{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import neo4j DB: 2/4\n",
    "\n",
    "Code to translate v2.7.4_PIS-model.xlsx to neo4j database. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check number families, nodes, clades"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import os\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from py2neo import Graph, Node, Relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to graph via docker-compose link. See http://localhost:7474/browser/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = Graph(host=\"neo4j\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = graph.run(\"MATCH (n) RETURN n LIMIT 10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "constraints_added: 0\n",
       "constraints_removed: 0\n",
       "contains_updates: False\n",
       "indexes_added: 0\n",
       "indexes_removed: 0\n",
       "labels_added: 0\n",
       "labels_removed: 0\n",
       "nodes_created: 0\n",
       "nodes_deleted: 0\n",
       "properties_set: 0\n",
       "relationships_created: 0\n",
       "relationships_deleted: 0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q.stats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_labels = [\n",
    "    'PlantCoding',\n",
    "    'PlantNonCoding',\n",
    "    'PlantAbstract',\n",
    "    'Complex',\n",
    "    'ExternalOrganism', \n",
    "    'ExternalCoding',\n",
    "    'ExternalNonCoding',\n",
    "    'Process', \n",
    "    'MetaboliteFamily',\n",
    "    'Metabolite',\n",
    "    'GeneExpression', \n",
    "    'PseudoNode'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# start from scratch\n",
    "graph.delete_all()\n",
    "\n",
    "for l in node_labels:\n",
    "    try:\n",
    "        graph.schema.drop_uniqueness_constraint(l, \"name\")\n",
    "    except:\n",
    "        pass    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Constraints on node names\n",
    "\n",
    "Constraint automatically adds an index. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PlantCoding 1 0\n",
      "PlantNonCoding 1 0\n",
      "PlantAbstract 1 0\n",
      "Complex 1 0\n",
      "ExternalOrganism 1 0\n",
      "ExternalCoding 1 0\n",
      "ExternalNonCoding 1 0\n",
      "Process 1 0\n",
      "MetaboliteFamily 1 0\n",
      "Metabolite 1 0\n",
      "GeneExpression 1 0\n",
      "PseudoNode 1 0\n"
     ]
    }
   ],
   "source": [
    "for label in node_labels:\n",
    "    q = f'''CREATE CONSTRAINT unique_name_{label.lower()}\n",
    "        ON (node:{label}) ASSERT node.name IS UNIQUE'''\n",
    "    qr = graph.run(q)\n",
    "    print(label, qr.stats()['constraints_added'], qr.stats()['indexes_added'])   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = os.path.join(\"..\", \"data\", \"raw\")\n",
    "path = os.path.join(input_path, 'parsed-components.tsv')\n",
    "\n",
    "df_components = pd.read_csv(path, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "label = 'Metabolite'\n",
    "df_metabolites = df_components[df_components['NodeLabel'] == label].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AddedBy', 'Species', 'NodeType', 'Family', 'Clade', 'NodeID',\n",
       "       'NodeName', 'ModelStatus', 'NodeDescription', 'AdditionalInfo',\n",
       "       'ExtDBlink', 'Process', 'ExternalDB', 'ModelV', 'GMM_OCD1', 'GMM_OCD',\n",
       "       'GMM_Description', 'GMM_ShortName', 'GMM_Synonyms', 'NodeLabel',\n",
       "       'observed_species', 'also_observed_in', 'chebi_identifier',\n",
       "       'pubmed_identifier'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metabolites.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>chebi_identifier</th>\n",
       "      <th>pubmed_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [AddedBy, Species, NodeType, Family, Clade, NodeID, NodeName, ModelStatus, NodeDescription, AdditionalInfo, ExtDBlink, Process, ExternalDB, ModelV, GMM_OCD1, GMM_OCD, GMM_Description, GMM_ShortName, GMM_Synonyms, NodeLabel, observed_species, also_observed_in, chebi_identifier, pubmed_identifier]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metabolites[df_metabolites['NodeName'].duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "want_columns = ['AddedBy', 'Family',\n",
    "       'NodeID', 'NodeName', 'NodeDescription',\n",
    "       'AdditionalInfo', 'Process', 'ExternalDB', 'ModelV',\n",
    "       'GMM_OCD1', 'GMM_OCD', 'GMM_Description', 'GMM_ShortName',\n",
    "       'GMM_Synonyms', 'chebi_identifier', 'pubmed_identifier']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = '%s-components.tsv'%label\n",
    "df_metabolites[want_columns].to_csv(\"../data/import/\" + f, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = helpers.component_node_query(f, label, \n",
    "                     n_name=\"line.NodeName\"\n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>chebi_identifier</th>\n",
       "      <th>pubmed_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [AddedBy, Species, NodeType, Family, Clade, NodeID, NodeName, ModelStatus, NodeDescription, AdditionalInfo, ExtDBlink, Process, ExternalDB, ModelV, GMM_OCD1, GMM_OCD, GMM_Description, GMM_ShortName, GMM_Synonyms, NodeLabel, observed_species, also_observed_in, chebi_identifier, pubmed_identifier]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metabolites[df_metabolites['NodeName'].duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "qr = graph.run(q)\n",
    "if not df_metabolites.shape[0] == qr.stats()['nodes_created']:\n",
    "    raise Exception"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>chebi_identifier</th>\n",
       "      <th>pubmed_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [AddedBy, Species, NodeType, Family, Clade, NodeID, NodeName, ModelStatus, NodeDescription, AdditionalInfo, ExtDBlink, Process, ExternalDB, ModelV, GMM_OCD1, GMM_OCD, GMM_Description, GMM_ShortName, GMM_Synonyms, NodeLabel, observed_species, also_observed_in, chebi_identifier, pubmed_identifier]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metabolites[df_metabolites['Clade'] != df_metabolites['NodeID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "label = \"MetaboliteFamily\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>chebi_identifier</th>\n",
       "      <th>pubmed_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>use</td>\n",
       "      <td>L-methionine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:16643</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>use</td>\n",
       "      <td>S-adenosyl-L-methionine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:15414</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACC</td>\n",
       "      <td>ACC</td>\n",
       "      <td>ACC</td>\n",
       "      <td>ACC</td>\n",
       "      <td>use</td>\n",
       "      <td>1-aminocyclopropane-1-carboxylate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:30526</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ET</td>\n",
       "      <td>ET</td>\n",
       "      <td>ET</td>\n",
       "      <td>ET</td>\n",
       "      <td>use</td>\n",
       "      <td>ethylene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:18153</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>use</td>\n",
       "      <td>copper</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:29036</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>PUFA:LA</td>\n",
       "      <td>PUFA:LA</td>\n",
       "      <td>PUFA:LA</td>\n",
       "      <td>PUFA:LA</td>\n",
       "      <td>use</td>\n",
       "      <td>polyunsaturated fatty acids: linoleic acid</td>\n",
       "      <td>HA | polyunsaturated fatty acids: hexadecatrie...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:17351</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>13-HPOT</td>\n",
       "      <td>13-HPOT</td>\n",
       "      <td>13-HPOT</td>\n",
       "      <td>13-HPOT</td>\n",
       "      <td>use</td>\n",
       "      <td>13-hydroperoxy-9,11,15-octadecatrienoic acid</td>\n",
       "      <td>11-HPHT | 11(S)-hydroperoxy-hexadecatrienoic a...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:48905</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>12,13-EOT</td>\n",
       "      <td>12,13-EOT</td>\n",
       "      <td>12,13-EOT</td>\n",
       "      <td>12,13-EOT</td>\n",
       "      <td>use</td>\n",
       "      <td>12,13(S)-epoxylinolenic acid</td>\n",
       "      <td>10,11-EHT | parallel path to OPDA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:15653</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>OPDA</td>\n",
       "      <td>OPDA</td>\n",
       "      <td>OPDA</td>\n",
       "      <td>OPDA</td>\n",
       "      <td>use</td>\n",
       "      <td>12-oxophytodienoic acid in the peroxisome</td>\n",
       "      <td>dnOPDA | dinorOPDA | parallel path to OPDA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:34005</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>OPC8</td>\n",
       "      <td>OPC8</td>\n",
       "      <td>OPC8</td>\n",
       "      <td>OPC8</td>\n",
       "      <td>use</td>\n",
       "      <td>oxopentenyl-cyclopentane-octanoic acid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:25244083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>OPC8-CoA</td>\n",
       "      <td>OPC8-CoA</td>\n",
       "      <td>OPC8-CoA</td>\n",
       "      <td>OPC8-CoA</td>\n",
       "      <td>use</td>\n",
       "      <td>oxopentenyl-cyclopentane-octanoyl-CoA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:44237325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>OPC6-CoA</td>\n",
       "      <td>OPC6-CoA</td>\n",
       "      <td>OPC6-CoA</td>\n",
       "      <td>OPC6-CoA</td>\n",
       "      <td>use</td>\n",
       "      <td>OPC6-CoA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:44237133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>OPC4-CoA</td>\n",
       "      <td>OPC4-CoA</td>\n",
       "      <td>OPC4-CoA</td>\n",
       "      <td>OPC4-CoA</td>\n",
       "      <td>use</td>\n",
       "      <td>OPC4-CoA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:44237211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>JA-CoA</td>\n",
       "      <td>JA-CoA</td>\n",
       "      <td>JA-CoA</td>\n",
       "      <td>JA-CoA</td>\n",
       "      <td>use</td>\n",
       "      <td>jasmonoyl-CoA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:44237154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>JA</td>\n",
       "      <td>JA</td>\n",
       "      <td>JA</td>\n",
       "      <td>JA</td>\n",
       "      <td>use</td>\n",
       "      <td>jasmonic acid</td>\n",
       "      <td>(+)-7-iso-jasmonate (interconverted through sp...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:25245750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>MeJA</td>\n",
       "      <td>MeJA</td>\n",
       "      <td>MeJA</td>\n",
       "      <td>MeJA</td>\n",
       "      <td>use</td>\n",
       "      <td>methyl jasmonic acid: (-)-jasmonic acid methyl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:5367719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>JA-Ile</td>\n",
       "      <td>JA-Ile</td>\n",
       "      <td>JA-Ile</td>\n",
       "      <td>JA-Ile</td>\n",
       "      <td>use</td>\n",
       "      <td>jasmonoyl-isoleucine: (+)-7-iso-jasmonoyl-L-is...</td>\n",
       "      <td>(-)-jasmonoyl-L-isoleucine (the non-active for...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PubChem:54758681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>12-OH-JA-Ile</td>\n",
       "      <td>12-OH-JA-Ile</td>\n",
       "      <td>12-OH-JA-Ile</td>\n",
       "      <td>12-OH-JA-Ile</td>\n",
       "      <td>use</td>\n",
       "      <td>12-hydroxyjasmonic acid 12-O-beta-D-glucoside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:34005</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Thr</td>\n",
       "      <td>Thr</td>\n",
       "      <td>Thr</td>\n",
       "      <td>Thr</td>\n",
       "      <td>use</td>\n",
       "      <td>amino acid Threonine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:57926</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Ile</td>\n",
       "      <td>Ile</td>\n",
       "      <td>Ile</td>\n",
       "      <td>Ile</td>\n",
       "      <td>use</td>\n",
       "      <td>amino acid Isoleucine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:58045</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Chor</td>\n",
       "      <td>Chor</td>\n",
       "      <td>Chor</td>\n",
       "      <td>Chor</td>\n",
       "      <td>use</td>\n",
       "      <td>chorismate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:29748</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>IsoChor</td>\n",
       "      <td>IsoChor</td>\n",
       "      <td>IsoChor</td>\n",
       "      <td>IsoChor</td>\n",
       "      <td>use</td>\n",
       "      <td>isochorismate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:29780</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Prep</td>\n",
       "      <td>Prep</td>\n",
       "      <td>Prep</td>\n",
       "      <td>Prep</td>\n",
       "      <td>use</td>\n",
       "      <td>prephenate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:57852</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>PhePyr</td>\n",
       "      <td>PhePyr</td>\n",
       "      <td>PhePyr</td>\n",
       "      <td>PhePyr</td>\n",
       "      <td>use</td>\n",
       "      <td>phenylpyruvate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:32815</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>PheAla</td>\n",
       "      <td>PheAla</td>\n",
       "      <td>PheAla</td>\n",
       "      <td>PheAla</td>\n",
       "      <td>use</td>\n",
       "      <td>phenylalanine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:28044</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>TransCA</td>\n",
       "      <td>TransCA</td>\n",
       "      <td>TransCA</td>\n",
       "      <td>TransCA</td>\n",
       "      <td>use</td>\n",
       "      <td>trans-cinnamic acid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:23252</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>OrthoCA</td>\n",
       "      <td>OrthoCA</td>\n",
       "      <td>OrthoCA</td>\n",
       "      <td>OrthoCA</td>\n",
       "      <td>use</td>\n",
       "      <td>ortho-coumaric acid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:23401</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>BA</td>\n",
       "      <td>BA</td>\n",
       "      <td>BA</td>\n",
       "      <td>BA</td>\n",
       "      <td>use</td>\n",
       "      <td>benzoic acid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:22723</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SA</td>\n",
       "      <td>SA</td>\n",
       "      <td>SA</td>\n",
       "      <td>SA</td>\n",
       "      <td>use</td>\n",
       "      <td>salicylate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:16914</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>KG</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>MeSA</td>\n",
       "      <td>MeSA</td>\n",
       "      <td>MeSA</td>\n",
       "      <td>MeSA</td>\n",
       "      <td>use</td>\n",
       "      <td>methyl salicylate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:31832</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>936</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ATP</td>\n",
       "      <td>ATP</td>\n",
       "      <td>ATP</td>\n",
       "      <td>ATP</td>\n",
       "      <td>use</td>\n",
       "      <td>adenosine 5'-triphosphate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>937</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>UDP glucose</td>\n",
       "      <td>UDP glucose</td>\n",
       "      <td>UDP glucose</td>\n",
       "      <td>UDP glucose</td>\n",
       "      <td>use</td>\n",
       "      <td>UDP-D-glucose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>938</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>NADPH</td>\n",
       "      <td>NADPH</td>\n",
       "      <td>NADPH</td>\n",
       "      <td>NADPH</td>\n",
       "      <td>use</td>\n",
       "      <td>nicotinamide adenine dinucleotide phosphate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>DHZ</td>\n",
       "      <td>DZ</td>\n",
       "      <td>DZ</td>\n",
       "      <td>DZ</td>\n",
       "      <td>use</td>\n",
       "      <td>dihydrozeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>DHZR</td>\n",
       "      <td>DZR</td>\n",
       "      <td>DZR</td>\n",
       "      <td>DZR</td>\n",
       "      <td>use</td>\n",
       "      <td>dihydrozeatin riboside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>941</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>DZRMP</td>\n",
       "      <td>DZRMP</td>\n",
       "      <td>DZRMP</td>\n",
       "      <td>DZRMP</td>\n",
       "      <td>use</td>\n",
       "      <td>dihydrozeatin riboside monophosphate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>942</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>tZRDP</td>\n",
       "      <td>tZRDP</td>\n",
       "      <td>tZRDP</td>\n",
       "      <td>tZRDP</td>\n",
       "      <td>use</td>\n",
       "      <td>9-ribosyl-trans-zeatin-5'-diphosphate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>943</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>tZRTP</td>\n",
       "      <td>tZRTP</td>\n",
       "      <td>tZRTP</td>\n",
       "      <td>tZRTP</td>\n",
       "      <td>use</td>\n",
       "      <td>9-ribosyl-trans-zeatin-5'-triphosphate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>cZR</td>\n",
       "      <td>cZR</td>\n",
       "      <td>cZR</td>\n",
       "      <td>cZR</td>\n",
       "      <td>use</td>\n",
       "      <td>cis-zeatin riboside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>cZRMP</td>\n",
       "      <td>cZRMP</td>\n",
       "      <td>cZRMP</td>\n",
       "      <td>cZRMP</td>\n",
       "      <td>use</td>\n",
       "      <td>cis-zeatin riboside monophosphate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>946</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>cis-prenyl-tRNA</td>\n",
       "      <td>cis-prenyl-tRNA</td>\n",
       "      <td>cis-prenyl-tRNA</td>\n",
       "      <td>cis-prenyl-tRNA</td>\n",
       "      <td>use</td>\n",
       "      <td>cis-prenyl-tRNA; N6-(cis-hydroxy)isopentenylad...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>947</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>prenyl-tRNA</td>\n",
       "      <td>prenyl-tRNA</td>\n",
       "      <td>prenyl-tRNA</td>\n",
       "      <td>prenyl-tRNA</td>\n",
       "      <td>use</td>\n",
       "      <td>prenyl-tRNA; N6-dimethylallyladenosine in tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>use</td>\n",
       "      <td>adenosine in tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>iP7G</td>\n",
       "      <td>iP7G</td>\n",
       "      <td>iP7G</td>\n",
       "      <td>iP7G</td>\n",
       "      <td>use</td>\n",
       "      <td>7-(alpha-D-glucosyl)-N6-isopentenyladenine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>950</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>iP9G</td>\n",
       "      <td>iP9G</td>\n",
       "      <td>iP9G</td>\n",
       "      <td>iP9G</td>\n",
       "      <td>use</td>\n",
       "      <td>9-(alpha-D-glucosyl)-N6-isopentenyladenine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>951</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Ade</td>\n",
       "      <td>Ade</td>\n",
       "      <td>Ade</td>\n",
       "      <td>Ade</td>\n",
       "      <td>use</td>\n",
       "      <td>adenine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Ado</td>\n",
       "      <td>Ado</td>\n",
       "      <td>Ado</td>\n",
       "      <td>Ado</td>\n",
       "      <td>use</td>\n",
       "      <td>adenosine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>953</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>tZOG</td>\n",
       "      <td>tZOG</td>\n",
       "      <td>tZOG</td>\n",
       "      <td>tZOG</td>\n",
       "      <td>use</td>\n",
       "      <td>O-beta-D-glucosyl-trans-zeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>tZROG</td>\n",
       "      <td>tZROG</td>\n",
       "      <td>tZROG</td>\n",
       "      <td>tZROG</td>\n",
       "      <td>use</td>\n",
       "      <td>O-glucosyl-trans-zeatin riboside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>tZ7G</td>\n",
       "      <td>tZ7G</td>\n",
       "      <td>tZ7G</td>\n",
       "      <td>tZ7G</td>\n",
       "      <td>use</td>\n",
       "      <td>trans-zeatin-7-beta-D-glucoside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>tZ9G</td>\n",
       "      <td>tZ9G</td>\n",
       "      <td>tZ9G</td>\n",
       "      <td>tZ9G</td>\n",
       "      <td>use</td>\n",
       "      <td>9-(alpha-D-glucosyl)-trans-zeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>DHZOG</td>\n",
       "      <td>DZOG</td>\n",
       "      <td>DZOG</td>\n",
       "      <td>DZOG</td>\n",
       "      <td>use</td>\n",
       "      <td>dihydrozeatin-O-glucoside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>DHZROG</td>\n",
       "      <td>DZROG</td>\n",
       "      <td>DZROG</td>\n",
       "      <td>DZROG</td>\n",
       "      <td>use</td>\n",
       "      <td>dihydrozeatin riboside-O-glucoside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>DHZ7G</td>\n",
       "      <td>DZ7G</td>\n",
       "      <td>DZ7G</td>\n",
       "      <td>DZ7G</td>\n",
       "      <td>use</td>\n",
       "      <td>7-(alpha-D-glucosyl)dihydrozeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>DHZ9G</td>\n",
       "      <td>DZ9G</td>\n",
       "      <td>DZ9G</td>\n",
       "      <td>DZ9G</td>\n",
       "      <td>use</td>\n",
       "      <td>9-(alpha-D-glucosyl)dihydrozeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>961</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>cZOG</td>\n",
       "      <td>cZOG</td>\n",
       "      <td>cZOG</td>\n",
       "      <td>cZOG</td>\n",
       "      <td>use</td>\n",
       "      <td>O-beta-D-glucosyl-cis-zeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>cZROG</td>\n",
       "      <td>cZROG</td>\n",
       "      <td>cZROG</td>\n",
       "      <td>cZROG</td>\n",
       "      <td>use</td>\n",
       "      <td>O-glucosyl-cis-zeatin riboside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>963</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>cZ7G</td>\n",
       "      <td>cZ7G</td>\n",
       "      <td>cZ7G</td>\n",
       "      <td>cZ7G</td>\n",
       "      <td>use</td>\n",
       "      <td>7-(beta-D-glucosyl)-cis-zeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>964</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>cZ9G</td>\n",
       "      <td>cZ9G</td>\n",
       "      <td>cZ9G</td>\n",
       "      <td>cZ9G</td>\n",
       "      <td>use</td>\n",
       "      <td>9-(beta-D-glucosyl)-cis-zeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>965</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>Glu</td>\n",
       "      <td>Glu</td>\n",
       "      <td>Glu</td>\n",
       "      <td>Glu</td>\n",
       "      <td>use</td>\n",
       "      <td>UDP-alpha-D-glucose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy    Species    NodeType           Family            Clade  \\\n",
       "20       KG  plant_all  metabolite            L-Met            L-Met   \n",
       "21       KG  plant_all  metabolite             SAMe             SAMe   \n",
       "22       KG  plant_all  metabolite              ACC              ACC   \n",
       "23       KG  plant_all  metabolite               ET               ET   \n",
       "24       KG  plant_all  metabolite             Cu2+             Cu2+   \n",
       "165      KG  plant_all  metabolite          PUFA:LA          PUFA:LA   \n",
       "166      KG  plant_all  metabolite          13-HPOT          13-HPOT   \n",
       "167      KG  plant_all  metabolite        12,13-EOT        12,13-EOT   \n",
       "168      KG  plant_all  metabolite             OPDA             OPDA   \n",
       "169      KG  plant_all  metabolite             OPC8             OPC8   \n",
       "170      KG  plant_all  metabolite         OPC8-CoA         OPC8-CoA   \n",
       "171      KG  plant_all  metabolite         OPC6-CoA         OPC6-CoA   \n",
       "172      KG  plant_all  metabolite         OPC4-CoA         OPC4-CoA   \n",
       "173      KG  plant_all  metabolite           JA-CoA           JA-CoA   \n",
       "174      KG  plant_all  metabolite               JA               JA   \n",
       "175      KG  plant_all  metabolite             MeJA             MeJA   \n",
       "176      KG  plant_all  metabolite           JA-Ile           JA-Ile   \n",
       "177      KG  plant_all  metabolite     12-OH-JA-Ile     12-OH-JA-Ile   \n",
       "179      KG  plant_all  metabolite              Thr              Thr   \n",
       "180      KG  plant_all  metabolite              Ile              Ile   \n",
       "278      KG  plant_all  metabolite             Chor             Chor   \n",
       "279      KG  plant_all  metabolite          IsoChor          IsoChor   \n",
       "280      KG  plant_all  metabolite             Prep             Prep   \n",
       "281      KG  plant_all  metabolite           PhePyr           PhePyr   \n",
       "282      KG  plant_all  metabolite           PheAla           PheAla   \n",
       "283      KG  plant_all  metabolite          TransCA          TransCA   \n",
       "284      KG  plant_all  metabolite          OrthoCA          OrthoCA   \n",
       "285      KG  plant_all  metabolite               BA               BA   \n",
       "286      KG  plant_all  metabolite               SA               SA   \n",
       "287      KG  plant_all  metabolite             MeSA             MeSA   \n",
       "..      ...        ...         ...              ...              ...   \n",
       "936      AG        all  metabolite              ATP              ATP   \n",
       "937      AG        all  metabolite      UDP glucose      UDP glucose   \n",
       "938      AG        all  metabolite            NADPH            NADPH   \n",
       "939      AG        all  metabolite              DHZ               DZ   \n",
       "940      AG        all  metabolite             DHZR              DZR   \n",
       "941      AG        all  metabolite            DZRMP            DZRMP   \n",
       "942      AG        all  metabolite            tZRDP            tZRDP   \n",
       "943      AG        all  metabolite            tZRTP            tZRTP   \n",
       "944      AG        all  metabolite              cZR              cZR   \n",
       "945      AG        all  metabolite            cZRMP            cZRMP   \n",
       "946      AG        all  metabolite  cis-prenyl-tRNA  cis-prenyl-tRNA   \n",
       "947      AG        all  metabolite      prenyl-tRNA      prenyl-tRNA   \n",
       "948      AG        all  metabolite             tRNA             tRNA   \n",
       "949      AG        all  metabolite             iP7G             iP7G   \n",
       "950      AG        all  metabolite             iP9G             iP9G   \n",
       "951      AG        all  metabolite              Ade              Ade   \n",
       "952      AG        all  metabolite              Ado              Ado   \n",
       "953      AG        all  metabolite             tZOG             tZOG   \n",
       "954      AG        all  metabolite            tZROG            tZROG   \n",
       "955      AG        all  metabolite             tZ7G             tZ7G   \n",
       "956      AG        all  metabolite             tZ9G             tZ9G   \n",
       "957      AG        all  metabolite            DHZOG             DZOG   \n",
       "958      AG        all  metabolite           DHZROG            DZROG   \n",
       "959      AG        all  metabolite            DHZ7G             DZ7G   \n",
       "960      AG        all  metabolite            DHZ9G             DZ9G   \n",
       "961      AG        all  metabolite             cZOG             cZOG   \n",
       "962      AG        all  metabolite            cZROG            cZROG   \n",
       "963      AG        all  metabolite             cZ7G             cZ7G   \n",
       "964      AG        all  metabolite             cZ9G             cZ9G   \n",
       "965      AG        all  metabolite              Glu              Glu   \n",
       "\n",
       "              NodeID         NodeName ModelStatus  \\\n",
       "20             L-Met            L-Met         use   \n",
       "21              SAMe             SAMe         use   \n",
       "22               ACC              ACC         use   \n",
       "23                ET               ET         use   \n",
       "24              Cu2+             Cu2+         use   \n",
       "165          PUFA:LA          PUFA:LA         use   \n",
       "166          13-HPOT          13-HPOT         use   \n",
       "167        12,13-EOT        12,13-EOT         use   \n",
       "168             OPDA             OPDA         use   \n",
       "169             OPC8             OPC8         use   \n",
       "170         OPC8-CoA         OPC8-CoA         use   \n",
       "171         OPC6-CoA         OPC6-CoA         use   \n",
       "172         OPC4-CoA         OPC4-CoA         use   \n",
       "173           JA-CoA           JA-CoA         use   \n",
       "174               JA               JA         use   \n",
       "175             MeJA             MeJA         use   \n",
       "176           JA-Ile           JA-Ile         use   \n",
       "177     12-OH-JA-Ile     12-OH-JA-Ile         use   \n",
       "179              Thr              Thr         use   \n",
       "180              Ile              Ile         use   \n",
       "278             Chor             Chor         use   \n",
       "279          IsoChor          IsoChor         use   \n",
       "280             Prep             Prep         use   \n",
       "281           PhePyr           PhePyr         use   \n",
       "282           PheAla           PheAla         use   \n",
       "283          TransCA          TransCA         use   \n",
       "284          OrthoCA          OrthoCA         use   \n",
       "285               BA               BA         use   \n",
       "286               SA               SA         use   \n",
       "287             MeSA             MeSA         use   \n",
       "..               ...              ...         ...   \n",
       "936              ATP              ATP         use   \n",
       "937      UDP glucose      UDP glucose         use   \n",
       "938            NADPH            NADPH         use   \n",
       "939               DZ               DZ         use   \n",
       "940              DZR              DZR         use   \n",
       "941            DZRMP            DZRMP         use   \n",
       "942            tZRDP            tZRDP         use   \n",
       "943            tZRTP            tZRTP         use   \n",
       "944              cZR              cZR         use   \n",
       "945            cZRMP            cZRMP         use   \n",
       "946  cis-prenyl-tRNA  cis-prenyl-tRNA         use   \n",
       "947      prenyl-tRNA      prenyl-tRNA         use   \n",
       "948             tRNA             tRNA         use   \n",
       "949             iP7G             iP7G         use   \n",
       "950             iP9G             iP9G         use   \n",
       "951              Ade              Ade         use   \n",
       "952              Ado              Ado         use   \n",
       "953             tZOG             tZOG         use   \n",
       "954            tZROG            tZROG         use   \n",
       "955             tZ7G             tZ7G         use   \n",
       "956             tZ9G             tZ9G         use   \n",
       "957             DZOG             DZOG         use   \n",
       "958            DZROG            DZROG         use   \n",
       "959             DZ7G             DZ7G         use   \n",
       "960             DZ9G             DZ9G         use   \n",
       "961             cZOG             cZOG         use   \n",
       "962            cZROG            cZROG         use   \n",
       "963             cZ7G             cZ7G         use   \n",
       "964             cZ9G             cZ9G         use   \n",
       "965              Glu              Glu         use   \n",
       "\n",
       "                                       NodeDescription  \\\n",
       "20                                        L-methionine   \n",
       "21                             S-adenosyl-L-methionine   \n",
       "22                   1-aminocyclopropane-1-carboxylate   \n",
       "23                                            ethylene   \n",
       "24                                              copper   \n",
       "165         polyunsaturated fatty acids: linoleic acid   \n",
       "166       13-hydroperoxy-9,11,15-octadecatrienoic acid   \n",
       "167                       12,13(S)-epoxylinolenic acid   \n",
       "168          12-oxophytodienoic acid in the peroxisome   \n",
       "169             oxopentenyl-cyclopentane-octanoic acid   \n",
       "170              oxopentenyl-cyclopentane-octanoyl-CoA   \n",
       "171                                           OPC6-CoA   \n",
       "172                                           OPC4-CoA   \n",
       "173                                      jasmonoyl-CoA   \n",
       "174                                      jasmonic acid   \n",
       "175  methyl jasmonic acid: (-)-jasmonic acid methyl...   \n",
       "176  jasmonoyl-isoleucine: (+)-7-iso-jasmonoyl-L-is...   \n",
       "177      12-hydroxyjasmonic acid 12-O-beta-D-glucoside   \n",
       "179                               amino acid Threonine   \n",
       "180                              amino acid Isoleucine   \n",
       "278                                         chorismate   \n",
       "279                                      isochorismate   \n",
       "280                                         prephenate   \n",
       "281                                     phenylpyruvate   \n",
       "282                                      phenylalanine   \n",
       "283                                trans-cinnamic acid   \n",
       "284                                ortho-coumaric acid   \n",
       "285                                       benzoic acid   \n",
       "286                                         salicylate   \n",
       "287                                  methyl salicylate   \n",
       "..                                                 ...   \n",
       "936                          adenosine 5'-triphosphate   \n",
       "937                                      UDP-D-glucose   \n",
       "938        nicotinamide adenine dinucleotide phosphate   \n",
       "939                                      dihydrozeatin   \n",
       "940                             dihydrozeatin riboside   \n",
       "941               dihydrozeatin riboside monophosphate   \n",
       "942              9-ribosyl-trans-zeatin-5'-diphosphate   \n",
       "943             9-ribosyl-trans-zeatin-5'-triphosphate   \n",
       "944                               cis-zeatin riboside    \n",
       "945                 cis-zeatin riboside monophosphate    \n",
       "946  cis-prenyl-tRNA; N6-(cis-hydroxy)isopentenylad...   \n",
       "947     prenyl-tRNA; N6-dimethylallyladenosine in tRNA   \n",
       "948                                  adenosine in tRNA   \n",
       "949        7-(alpha-D-glucosyl)-N6-isopentenyladenine    \n",
       "950        9-(alpha-D-glucosyl)-N6-isopentenyladenine    \n",
       "951                                            adenine   \n",
       "952                                          adenosine   \n",
       "953                    O-beta-D-glucosyl-trans-zeatin    \n",
       "954                   O-glucosyl-trans-zeatin riboside   \n",
       "955                    trans-zeatin-7-beta-D-glucoside   \n",
       "956                  9-(alpha-D-glucosyl)-trans-zeatin   \n",
       "957                          dihydrozeatin-O-glucoside   \n",
       "958                 dihydrozeatin riboside-O-glucoside   \n",
       "959                 7-(alpha-D-glucosyl)dihydrozeatin    \n",
       "960                 9-(alpha-D-glucosyl)dihydrozeatin    \n",
       "961                      O-beta-D-glucosyl-cis-zeatin    \n",
       "962                     O-glucosyl-cis-zeatin riboside   \n",
       "963                    7-(beta-D-glucosyl)-cis-zeatin    \n",
       "964                    9-(beta-D-glucosyl)-cis-zeatin    \n",
       "965                                UDP-alpha-D-glucose   \n",
       "\n",
       "                                        AdditionalInfo  ... GMM_OCD1 GMM_OCD  \\\n",
       "20                                                 NaN  ...      NaN     NaN   \n",
       "21                                                 NaN  ...      NaN     NaN   \n",
       "22                                                 NaN  ...      NaN     NaN   \n",
       "23                                                 NaN  ...      NaN     NaN   \n",
       "24                                                 NaN  ...      NaN     NaN   \n",
       "165  HA | polyunsaturated fatty acids: hexadecatrie...  ...      NaN     NaN   \n",
       "166  11-HPHT | 11(S)-hydroperoxy-hexadecatrienoic a...  ...      NaN     NaN   \n",
       "167                  10,11-EHT | parallel path to OPDA  ...      NaN     NaN   \n",
       "168         dnOPDA | dinorOPDA | parallel path to OPDA  ...      NaN     NaN   \n",
       "169                                                NaN  ...      NaN     NaN   \n",
       "170                                                NaN  ...      NaN     NaN   \n",
       "171                                                NaN  ...      NaN     NaN   \n",
       "172                                                NaN  ...      NaN     NaN   \n",
       "173                                                NaN  ...      NaN     NaN   \n",
       "174  (+)-7-iso-jasmonate (interconverted through sp...  ...      NaN     NaN   \n",
       "175                                                NaN  ...      NaN     NaN   \n",
       "176  (-)-jasmonoyl-L-isoleucine (the non-active for...  ...      NaN     NaN   \n",
       "177                                                NaN  ...      NaN     NaN   \n",
       "179                                                NaN  ...      NaN     NaN   \n",
       "180                                                NaN  ...      NaN     NaN   \n",
       "278                                                NaN  ...      NaN     NaN   \n",
       "279                                                NaN  ...      NaN     NaN   \n",
       "280                                                NaN  ...      NaN     NaN   \n",
       "281                                                NaN  ...      NaN     NaN   \n",
       "282                                                NaN  ...      NaN     NaN   \n",
       "283                                                NaN  ...      NaN     NaN   \n",
       "284                                                NaN  ...      NaN     NaN   \n",
       "285                                                NaN  ...      NaN     NaN   \n",
       "286                                                NaN  ...      NaN     NaN   \n",
       "287                                                NaN  ...      NaN     NaN   \n",
       "..                                                 ...  ...      ...     ...   \n",
       "936                                                NaN  ...      NaN     NaN   \n",
       "937                                                NaN  ...      NaN     NaN   \n",
       "938                                                NaN  ...      NaN     NaN   \n",
       "939                                                NaN  ...      NaN     NaN   \n",
       "940                                                NaN  ...      NaN     NaN   \n",
       "941                                                NaN  ...      NaN     NaN   \n",
       "942                                                NaN  ...      NaN     NaN   \n",
       "943                                                NaN  ...      NaN     NaN   \n",
       "944                                                NaN  ...      NaN     NaN   \n",
       "945                                                NaN  ...      NaN     NaN   \n",
       "946                                                NaN  ...      NaN     NaN   \n",
       "947                                                NaN  ...      NaN     NaN   \n",
       "948                                                NaN  ...      NaN     NaN   \n",
       "949                                                NaN  ...      NaN     NaN   \n",
       "950                                                NaN  ...      NaN     NaN   \n",
       "951                                                NaN  ...      NaN     NaN   \n",
       "952                                                NaN  ...      NaN     NaN   \n",
       "953                                                NaN  ...      NaN     NaN   \n",
       "954                                                NaN  ...      NaN     NaN   \n",
       "955                                                NaN  ...      NaN     NaN   \n",
       "956                                                NaN  ...      NaN     NaN   \n",
       "957                                                NaN  ...      NaN     NaN   \n",
       "958                                                NaN  ...      NaN     NaN   \n",
       "959                                                NaN  ...      NaN     NaN   \n",
       "960                                                NaN  ...      NaN     NaN   \n",
       "961                                                NaN  ...      NaN     NaN   \n",
       "962                                                NaN  ...      NaN     NaN   \n",
       "963                                                NaN  ...      NaN     NaN   \n",
       "964                                                NaN  ...      NaN     NaN   \n",
       "965                                                NaN  ...      NaN     NaN   \n",
       "\n",
       "    GMM_Description GMM_ShortName GMM_Synonyms   NodeLabel observed_species  \\\n",
       "20              NaN           NaN          NaN  Metabolite        plant_all   \n",
       "21              NaN           NaN          NaN  Metabolite        plant_all   \n",
       "22              NaN           NaN          NaN  Metabolite        plant_all   \n",
       "23              NaN           NaN          NaN  Metabolite        plant_all   \n",
       "24              NaN           NaN          NaN  Metabolite        plant_all   \n",
       "165             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "166             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "167             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "168             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "169             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "170             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "171             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "172             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "173             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "174             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "175             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "176             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "177             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "179             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "180             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "278             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "279             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "280             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "281             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "282             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "283             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "284             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "285             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "286             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "287             NaN           NaN          NaN  Metabolite        plant_all   \n",
       "..              ...           ...          ...         ...              ...   \n",
       "936             NaN           NaN          NaN  Metabolite              all   \n",
       "937             NaN           NaN          NaN  Metabolite              all   \n",
       "938             NaN           NaN          NaN  Metabolite              all   \n",
       "939             NaN           NaN          NaN  Metabolite              all   \n",
       "940             NaN           NaN          NaN  Metabolite              all   \n",
       "941             NaN           NaN          NaN  Metabolite              all   \n",
       "942             NaN           NaN          NaN  Metabolite              all   \n",
       "943             NaN           NaN          NaN  Metabolite              all   \n",
       "944             NaN           NaN          NaN  Metabolite              all   \n",
       "945             NaN           NaN          NaN  Metabolite              all   \n",
       "946             NaN           NaN          NaN  Metabolite              all   \n",
       "947             NaN           NaN          NaN  Metabolite              all   \n",
       "948             NaN           NaN          NaN  Metabolite              all   \n",
       "949             NaN           NaN          NaN  Metabolite              all   \n",
       "950             NaN           NaN          NaN  Metabolite              all   \n",
       "951             NaN           NaN          NaN  Metabolite              all   \n",
       "952             NaN           NaN          NaN  Metabolite              all   \n",
       "953             NaN           NaN          NaN  Metabolite              all   \n",
       "954             NaN           NaN          NaN  Metabolite              all   \n",
       "955             NaN           NaN          NaN  Metabolite              all   \n",
       "956             NaN           NaN          NaN  Metabolite              all   \n",
       "957             NaN           NaN          NaN  Metabolite              all   \n",
       "958             NaN           NaN          NaN  Metabolite              all   \n",
       "959             NaN           NaN          NaN  Metabolite              all   \n",
       "960             NaN           NaN          NaN  Metabolite              all   \n",
       "961             NaN           NaN          NaN  Metabolite              all   \n",
       "962             NaN           NaN          NaN  Metabolite              all   \n",
       "963             NaN           NaN          NaN  Metabolite              all   \n",
       "964             NaN           NaN          NaN  Metabolite              all   \n",
       "965             NaN           NaN          NaN  Metabolite              all   \n",
       "\n",
       "    also_observed_in chebi_identifier pubmed_identifier  \n",
       "20               NaN      CHEBI:16643               NaN  \n",
       "21               NaN      CHEBI:15414               NaN  \n",
       "22               NaN      CHEBI:30526               NaN  \n",
       "23               NaN      CHEBI:18153               NaN  \n",
       "24               NaN      CHEBI:29036               NaN  \n",
       "165              NaN      CHEBI:17351               NaN  \n",
       "166              NaN      CHEBI:48905               NaN  \n",
       "167              NaN      CHEBI:15653               NaN  \n",
       "168              NaN      CHEBI:34005               NaN  \n",
       "169              NaN              NaN  PubChem:25244083  \n",
       "170              NaN              NaN  PubChem:44237325  \n",
       "171              NaN              NaN  PubChem:44237133  \n",
       "172              NaN              NaN  PubChem:44237211  \n",
       "173              NaN              NaN  PubChem:44237154  \n",
       "174              NaN              NaN  PubChem:25245750  \n",
       "175              NaN              NaN   PubChem:5367719  \n",
       "176              NaN              NaN  PubChem:54758681  \n",
       "177              NaN      CHEBI:34005               NaN  \n",
       "179              NaN      CHEBI:57926               NaN  \n",
       "180              NaN      CHEBI:58045               NaN  \n",
       "278              NaN      CHEBI:29748               NaN  \n",
       "279              NaN      CHEBI:29780               NaN  \n",
       "280              NaN      CHEBI:57852               NaN  \n",
       "281              NaN      CHEBI:32815               NaN  \n",
       "282              NaN      CHEBI:28044               NaN  \n",
       "283              NaN      CHEBI:23252               NaN  \n",
       "284              NaN      CHEBI:23401               NaN  \n",
       "285              NaN      CHEBI:22723               NaN  \n",
       "286              NaN      CHEBI:16914               NaN  \n",
       "287              NaN      CHEBI:31832               NaN  \n",
       "..               ...              ...               ...  \n",
       "936              NaN              NaN               NaN  \n",
       "937              NaN              NaN               NaN  \n",
       "938              NaN              NaN               NaN  \n",
       "939              NaN              NaN               NaN  \n",
       "940              NaN              NaN               NaN  \n",
       "941              NaN              NaN               NaN  \n",
       "942              NaN              NaN               NaN  \n",
       "943              NaN              NaN               NaN  \n",
       "944              NaN              NaN               NaN  \n",
       "945              NaN              NaN               NaN  \n",
       "946              NaN              NaN               NaN  \n",
       "947              NaN              NaN               NaN  \n",
       "948              NaN              NaN               NaN  \n",
       "949              NaN              NaN               NaN  \n",
       "950              NaN              NaN               NaN  \n",
       "951              NaN              NaN               NaN  \n",
       "952              NaN              NaN               NaN  \n",
       "953              NaN              NaN               NaN  \n",
       "954              NaN              NaN               NaN  \n",
       "955              NaN              NaN               NaN  \n",
       "956              NaN              NaN               NaN  \n",
       "957              NaN              NaN               NaN  \n",
       "958              NaN              NaN               NaN  \n",
       "959              NaN              NaN               NaN  \n",
       "960              NaN              NaN               NaN  \n",
       "961              NaN              NaN               NaN  \n",
       "962              NaN              NaN               NaN  \n",
       "963              NaN              NaN               NaN  \n",
       "964              NaN              NaN               NaN  \n",
       "965              NaN              NaN               NaN  \n",
       "\n",
       "[102 rows x 24 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metabolites_has_family = df_metabolites[df_metabolites['Family'] != df_metabolites['NodeID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "want_columns = ['AddedBy', 'Family',\n",
    "        'NodeDescription',\n",
    "       'AdditionalInfo', 'Process', 'ModelV']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Family</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Process</th>\n",
       "      <th>ModelV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>MZ</td>\n",
       "      <td>PreROS</td>\n",
       "      <td>ROS precursors; oxygen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>MZ</td>\n",
       "      <td>ROS</td>\n",
       "      <td>hydrogen peroxide (reactive oxygen species)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>MZ</td>\n",
       "      <td>ROS</td>\n",
       "      <td>superoxide anion (reactive oxygen species)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>MZ</td>\n",
       "      <td>ROS</td>\n",
       "      <td>hydroxyl radical (reactive oxygen species)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>MZ</td>\n",
       "      <td>ROS</td>\n",
       "      <td>hydroperoxyl radical (reactive oxygen species)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>MZ</td>\n",
       "      <td>ROS</td>\n",
       "      <td>ozone (reactive oxygen species)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>MZ</td>\n",
       "      <td>PostROS</td>\n",
       "      <td>water</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>KG</td>\n",
       "      <td>DAMP/HAMP</td>\n",
       "      <td>oligogalacturonides</td>\n",
       "      <td>damage/host associated molecular patterns</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>933</th>\n",
       "      <td>AG</td>\n",
       "      <td>tZRMP</td>\n",
       "      <td>9-ribosyl-trans-zeatin-5'-monophosphate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>vNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>AG</td>\n",
       "      <td>DHZ</td>\n",
       "      <td>dihydrozeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>vNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>AG</td>\n",
       "      <td>DHZR</td>\n",
       "      <td>dihydrozeatin riboside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>vNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>AG</td>\n",
       "      <td>DHZOG</td>\n",
       "      <td>dihydrozeatin-O-glucoside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>vNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>AG</td>\n",
       "      <td>DHZROG</td>\n",
       "      <td>dihydrozeatin riboside-O-glucoside</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>vNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>AG</td>\n",
       "      <td>DHZ7G</td>\n",
       "      <td>7-(alpha-D-glucosyl)dihydrozeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>vNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>AG</td>\n",
       "      <td>DHZ9G</td>\n",
       "      <td>9-(alpha-D-glucosyl)dihydrozeatin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>vNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy     Family                                 NodeDescription  \\\n",
       "523      MZ     PreROS                          ROS precursors; oxygen   \n",
       "524      MZ        ROS     hydrogen peroxide (reactive oxygen species)   \n",
       "525      MZ        ROS      superoxide anion (reactive oxygen species)   \n",
       "526      MZ        ROS      hydroxyl radical (reactive oxygen species)   \n",
       "527      MZ        ROS  hydroperoxyl radical (reactive oxygen species)   \n",
       "528      MZ        ROS                 ozone (reactive oxygen species)   \n",
       "529      MZ    PostROS                                           water   \n",
       "752      KG  DAMP/HAMP                             oligogalacturonides   \n",
       "933      AG      tZRMP         9-ribosyl-trans-zeatin-5'-monophosphate   \n",
       "939      AG        DHZ                                   dihydrozeatin   \n",
       "940      AG       DHZR                          dihydrozeatin riboside   \n",
       "957      AG      DHZOG                       dihydrozeatin-O-glucoside   \n",
       "958      AG     DHZROG              dihydrozeatin riboside-O-glucoside   \n",
       "959      AG      DHZ7G              7-(alpha-D-glucosyl)dihydrozeatin    \n",
       "960      AG      DHZ9G              9-(alpha-D-glucosyl)dihydrozeatin    \n",
       "\n",
       "                                AdditionalInfo     Process ModelV  \n",
       "523                                        NaN       S:ROS   v2.6  \n",
       "524                                        NaN       S:ROS   v2.6  \n",
       "525                                        NaN       S:ROS   v2.6  \n",
       "526                                        NaN       S:ROS   v1.0  \n",
       "527                                        NaN       S:ROS   v2.6  \n",
       "528                                        NaN       S:ROS   v2.6  \n",
       "529                                        NaN       S:ROS   v2.7  \n",
       "752  damage/host associated molecular patterns         NaN   v2.6  \n",
       "933                                        NaN  Hormone:CK    vNA  \n",
       "939                                        NaN  Hormone:CK    vNA  \n",
       "940                                        NaN  Hormone:CK    vNA  \n",
       "957                                        NaN  Hormone:CK    vNA  \n",
       "958                                        NaN  Hormone:CK    vNA  \n",
       "959                                        NaN  Hormone:CK    vNA  \n",
       "960                                        NaN  Hormone:CK    vNA  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metabolites_has_family[want_columns].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_metabolites_family = df_metabolites_has_family.groupby('Family').agg({\n",
    "                                                                  'Family':lambda x:list(x)[0],\n",
    "                                                                  'AddedBy':lambda x:list(x)[0], \n",
    "                                                                  'NodeDescription':lambda x: ', '.join(x), \n",
    "                                                                  'AdditionalInfo':lambda x: helpers.list_to_string(x), \n",
    "                                                                  'Process':lambda x:list(x)[0], \n",
    "                                                                  'ModelV':helpers.get_latest_model})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Family</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Process</th>\n",
       "      <th>ModelV</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Family</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DAMP/HAMP</th>\n",
       "      <td>DAMP/HAMP</td>\n",
       "      <td>KG</td>\n",
       "      <td>oligogalacturonides</td>\n",
       "      <td>damage/host associated molecular patterns</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHZ</th>\n",
       "      <td>DHZ</td>\n",
       "      <td>AG</td>\n",
       "      <td>dihydrozeatin</td>\n",
       "      <td></td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHZ7G</th>\n",
       "      <td>DHZ7G</td>\n",
       "      <td>AG</td>\n",
       "      <td>7-(alpha-D-glucosyl)dihydrozeatin</td>\n",
       "      <td></td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHZ9G</th>\n",
       "      <td>DHZ9G</td>\n",
       "      <td>AG</td>\n",
       "      <td>9-(alpha-D-glucosyl)dihydrozeatin</td>\n",
       "      <td></td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHZOG</th>\n",
       "      <td>DHZOG</td>\n",
       "      <td>AG</td>\n",
       "      <td>dihydrozeatin-O-glucoside</td>\n",
       "      <td></td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHZR</th>\n",
       "      <td>DHZR</td>\n",
       "      <td>AG</td>\n",
       "      <td>dihydrozeatin riboside</td>\n",
       "      <td></td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DHZROG</th>\n",
       "      <td>DHZROG</td>\n",
       "      <td>AG</td>\n",
       "      <td>dihydrozeatin riboside-O-glucoside</td>\n",
       "      <td></td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PostROS</th>\n",
       "      <td>PostROS</td>\n",
       "      <td>MZ</td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PreROS</th>\n",
       "      <td>PreROS</td>\n",
       "      <td>MZ</td>\n",
       "      <td>ROS precursors; oxygen</td>\n",
       "      <td></td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ROS</th>\n",
       "      <td>ROS</td>\n",
       "      <td>MZ</td>\n",
       "      <td>hydrogen peroxide (reactive oxygen species), s...</td>\n",
       "      <td></td>\n",
       "      <td>S:ROS</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tZRMP</th>\n",
       "      <td>tZRMP</td>\n",
       "      <td>AG</td>\n",
       "      <td>9-ribosyl-trans-zeatin-5'-monophosphate</td>\n",
       "      <td></td>\n",
       "      <td>Hormone:CK</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Family AddedBy  \\\n",
       "Family                         \n",
       "DAMP/HAMP  DAMP/HAMP      KG   \n",
       "DHZ              DHZ      AG   \n",
       "DHZ7G          DHZ7G      AG   \n",
       "DHZ9G          DHZ9G      AG   \n",
       "DHZOG          DHZOG      AG   \n",
       "DHZR            DHZR      AG   \n",
       "DHZROG        DHZROG      AG   \n",
       "PostROS      PostROS      MZ   \n",
       "PreROS        PreROS      MZ   \n",
       "ROS              ROS      MZ   \n",
       "tZRMP          tZRMP      AG   \n",
       "\n",
       "                                             NodeDescription  \\\n",
       "Family                                                         \n",
       "DAMP/HAMP                                oligogalacturonides   \n",
       "DHZ                                            dihydrozeatin   \n",
       "DHZ7G                     7-(alpha-D-glucosyl)dihydrozeatin    \n",
       "DHZ9G                     9-(alpha-D-glucosyl)dihydrozeatin    \n",
       "DHZOG                              dihydrozeatin-O-glucoside   \n",
       "DHZR                                  dihydrozeatin riboside   \n",
       "DHZROG                    dihydrozeatin riboside-O-glucoside   \n",
       "PostROS                                                water   \n",
       "PreROS                                ROS precursors; oxygen   \n",
       "ROS        hydrogen peroxide (reactive oxygen species), s...   \n",
       "tZRMP                9-ribosyl-trans-zeatin-5'-monophosphate   \n",
       "\n",
       "                                      AdditionalInfo     Process ModelV  \n",
       "Family                                                                   \n",
       "DAMP/HAMP  damage/host associated molecular patterns         NaN   v2.6  \n",
       "DHZ                                                   Hormone:CK   v0.0  \n",
       "DHZ7G                                                 Hormone:CK   v0.0  \n",
       "DHZ9G                                                 Hormone:CK   v0.0  \n",
       "DHZOG                                                 Hormone:CK   v0.0  \n",
       "DHZR                                                  Hormone:CK   v0.0  \n",
       "DHZROG                                                Hormone:CK   v0.0  \n",
       "PostROS                                                    S:ROS   v2.7  \n",
       "PreROS                                                     S:ROS   v2.6  \n",
       "ROS                                                        S:ROS   v2.6  \n",
       "tZRMP                                                 Hormone:CK   v0.0  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metabolites_family"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = '%s-components.tsv'%label\n",
    "df_metabolites_family.to_csv(\"../data/import/\" + f, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = helpers.component_node_query(f, label, \n",
    "                     n_name=\"line.Family\"\n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "qr = graph.run(q)\n",
    "if not df_metabolites_family.shape[0] == qr.stats()['nodes_created']:\n",
    "    raise Exception "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Metabolite to MetaboliteFamily edges\n",
    "want_columns = ['AddedBy', 'Family', 'NodeName', 'ModelV']\n",
    "edge_type = 'TYPE_OF'\n",
    "f = '%s-edges.tsv'%edge_type\n",
    "df_metabolites_has_family[want_columns].to_csv(\"../data/import/\" + f, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = helpers.make_create_edge_query(f, edge_type, \n",
    "                           source_label=\"Metabolite\", target_label=\"MetaboliteFamily\",\n",
    "                           source_name=\"line.NodeName\", target_name=\"line.Family\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "qr = graph.run(q)\n",
    "if not df_metabolites_has_family.shape[0] == qr.stats()['relationships_created']:\n",
    "    raise Exception"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pathogens (External)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_external = df_components[df_components[\"NodeLabel\"].isin([\"ExternalCoding\", \"ExternalNonCoding\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_external.columns = ['AddedBy', 'O-Species', 'NodeType', 'classification', 'species',\n",
    "       'NodeID', 'NodeName', 'ModelStatus', 'NodeDescription',\n",
    "       'AdditionalInfo', 'ExtDBlink', 'Process', 'ExternalDB', 'ModelV',\n",
    "       'GMM_OCD1', 'GMM_OCD', 'GMM_Description', 'GMM_ShortName',\n",
    "       'GMM_Synonyms', 'NodeLabel', 'observed_species',\n",
    "       'also_observed_in', 'chebi_identifier', 'pubmed_identifier']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "want_columns = ['AddedBy',\n",
    "       'NodeID', 'NodeName', 'NodeDescription',\n",
    "       'AdditionalInfo', 'Process', 'ExternalDB', 'ModelV',\n",
    "       'GMM_OCD1', 'GMM_OCD', 'GMM_Description', 'GMM_ShortName',\n",
    "       'GMM_Synonyms', 'chebi_identifier', 'pubmed_identifier']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>O-Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>classification</th>\n",
       "      <th>species</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>observed_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>6K1</td>\n",
       "      <td>6K1</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>6K2</td>\n",
       "      <td>6K2</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>CI</td>\n",
       "      <td>CI</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>CP</td>\n",
       "      <td>CP</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>HC-Pro</td>\n",
       "      <td>HC-Pro</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>NIa-Pro</td>\n",
       "      <td>NIa-Pro</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>NIb</td>\n",
       "      <td>NIb</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>P1</td>\n",
       "      <td>P1</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>P3</td>\n",
       "      <td>P3</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>P3N-PIPO</td>\n",
       "      <td>P3N-PIPO</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>virus</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>VPg</td>\n",
       "      <td>VPg</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>trichous-bacteria</td>\n",
       "      <td>flg22</td>\n",
       "      <td>flg22</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>trichous-bacteria</td>\n",
       "      <td>elf18</td>\n",
       "      <td>elf18</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>external</td>\n",
       "      <td>pathogen_coding</td>\n",
       "      <td>fungi</td>\n",
       "      <td>oomycete</td>\n",
       "      <td>ch</td>\n",
       "      <td>ch</td>\n",
       "      <td>external</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    O-Species         NodeType classification            species    NodeID  \\\n",
       "451  external  pathogen_coding          virus          potyvirus       6K1   \n",
       "452  external  pathogen_coding          virus          potyvirus       6K2   \n",
       "453  external  pathogen_coding          virus          potyvirus        CI   \n",
       "454  external  pathogen_coding          virus          potyvirus        CP   \n",
       "455  external  pathogen_coding          virus          potyvirus    HC-Pro   \n",
       "456  external  pathogen_coding          virus          potyvirus   NIa-Pro   \n",
       "457  external  pathogen_coding          virus          potyvirus       NIb   \n",
       "458  external  pathogen_coding          virus          potyvirus        P1   \n",
       "459  external  pathogen_coding          virus          potyvirus        P3   \n",
       "460  external  pathogen_coding          virus          potyvirus  P3N-PIPO   \n",
       "461  external  pathogen_coding          virus          potyvirus       VPg   \n",
       "462  external  pathogen_coding       bacteria  trichous-bacteria     flg22   \n",
       "463  external  pathogen_coding       bacteria  trichous-bacteria     elf18   \n",
       "464  external  pathogen_coding          fungi           oomycete        ch   \n",
       "\n",
       "     NodeName observed_species  \n",
       "451       6K1         external  \n",
       "452       6K2         external  \n",
       "453        CI         external  \n",
       "454        CP         external  \n",
       "455    HC-Pro         external  \n",
       "456   NIa-Pro         external  \n",
       "457       NIb         external  \n",
       "458        P1         external  \n",
       "459        P3         external  \n",
       "460  P3N-PIPO         external  \n",
       "461       VPg         external  \n",
       "462     flg22         external  \n",
       "463     elf18         external  \n",
       "464        ch         external  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_external[['O-Species', 'NodeType', 'classification', 'species', 'NodeID', 'NodeName', 'observed_species']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ExternalCoding\t"
     ]
    }
   ],
   "source": [
    "for label, subdf in df_external.groupby('NodeLabel'):\n",
    "    print(label, end='\\t')\n",
    "    f = '%s-components.tsv'%label\n",
    "    subdf[want_columns].to_csv(\"../data/import/\" + f, sep=\"\\t\", index=None)\n",
    "    \n",
    "    q = helpers.component_node_query(f, label, \n",
    "                     n_name=\"line.NodeName\"\n",
    "                    )\n",
    "    \n",
    "    qr = graph.run(q)\n",
    "    if not subdf.shape[0] == qr.stats()['nodes_created']:\n",
    "        raise Exception"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "external_organisms = df_external.groupby('species').agg({\n",
    "                                                      'classification':lambda x:list(x)[0],\n",
    "                                                      'AddedBy':lambda x:list(x)[0], \n",
    "                                                      #'NodeDescription':lambda x: ', '.join(x), \n",
    "                                                      #'AdditionalInfo':lambda x: ', '.join(x), \n",
    "                                                      #'Process':lambda x:list(x)[0], \n",
    "                                                      'ModelV':helpers.get_latest_model}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>classification</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ModelV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>oomycete</td>\n",
       "      <td>fungi</td>\n",
       "      <td>KG</td>\n",
       "      <td>v0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>potyvirus</td>\n",
       "      <td>virus</td>\n",
       "      <td>KG</td>\n",
       "      <td>v2.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>trichous-bacteria</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>KG</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             species classification AddedBy ModelV\n",
       "0           oomycete          fungi      KG   v0.0\n",
       "1          potyvirus          virus      KG   v2.6\n",
       "2  trichous-bacteria       bacteria      KG       "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "external_organisms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "label = 'ExternalOrganism'\n",
    "f = '%s-components.tsv'%label\n",
    "external_organisms.to_csv(\"../data/import/\" + f, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = helpers.component_node_query(f, label, \n",
    "                     n_name=\"line.species\", \n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "qr = graph.run(q)\n",
    "if not external_organisms.shape[0] == qr.stats()['nodes_created']:\n",
    "    raise Exception"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# External to ExternalOrganism edges\n",
    "want_columns = ['NodeName', 'species', 'ModelV']\n",
    "edge_type = 'AGENT_OF'\n",
    "f = '%s-edges.tsv'%edge_type\n",
    "df_external[want_columns].to_csv(\"../data/import/\" + f, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = helpers.make_create_edge_query(f, edge_type, \n",
    "                           source_label=\"\", target_label=\"ExternalOrganism\",\n",
    "                           source_name=\"line.NodeName\", target_name=\"line.species\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "qr = graph.run(q)\n",
    "if not df_external.shape[0] == qr.stats()['relationships_created']:\n",
    "    raise Exception"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bio-elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bioelements = df_components[df_components['NodeLabel'].isin(['PlantCoding', \n",
    "                                                                 'PlantNonCoding',\n",
    "                                                                 'PlantAbstract', \n",
    "                                                                 'Complex'\n",
    "                                                                ])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ath', 'plant_all', 'stu', 'osa'], dtype=object)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bioelements['Species'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PlantCoding       825\n",
       "PlantAbstract      12\n",
       "PlantNonCoding     12\n",
       "Complex             1\n",
       "Name: NodeLabel, dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_bioelements['NodeLabel'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ath', 'plant_all', 'stu', 'osa']"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species_listed = list(df_bioelements['Species'].unique()); all_species_listed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_species = all_species_listed.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "list.remove(x): x not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-52-d3eecc67a504>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mall_species\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'plant_all'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mall_species\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'all'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m: list.remove(x): x not in list"
     ]
    }
   ],
   "source": [
    "all_species.remove('plant_all')\n",
    "all_species.remove('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ath', 'stu', 'osa']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pick_the_set(x):\n",
    "    for v in x:\n",
    "        if type(v) == set:\n",
    "            return v\n",
    "    return {}\n",
    "\n",
    "\n",
    "def get_species_homologues(level):\n",
    "    df_level_species  = df_bioelements.groupby([level, 'Species']).agg({\n",
    "      'NodeID':lambda x: set(x), \n",
    "    })\n",
    "    df_level_species.reset_index(1, inplace=True)\n",
    "\n",
    "    cols = []\n",
    "    for specie in all_species:\n",
    "        col = specie + '_homologues'\n",
    "        cols.append(col)\n",
    "        df_level_species.loc[df_level_species['Species'] == specie, col] =\\\n",
    "         df_level_species[df_level_species['Species'] == specie]['NodeID']\n",
    "    \n",
    "    df_level_species = df_level_species[cols].groupby(level).agg({\n",
    "        y:pick_the_set for y in cols\n",
    "    })\n",
    "    \n",
    "    df_level = df_bioelements.groupby(level).agg({\n",
    "          'AddedBy':lambda x:list(x)[0], \n",
    "          'NodeLabel':lambda x:list(x)[0], \n",
    "          'NodeDescription':lambda x: ', '.join(list(set(x))), \n",
    "          'AdditionalInfo':lambda x: helpers.list_to_string(x), \n",
    "          'Process':lambda x:list(x)[0], \n",
    "          'ModelV':helpers.get_latest_model, \n",
    "          'Species':lambda x: set(x), \n",
    "    })    \n",
    "\n",
    "    df_level = df_level.join(df_level_species[cols])\n",
    "    df_level.reset_index(inplace=True)\n",
    "    \n",
    "    return df_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_families = get_species_homologues(\"Family\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Complex \t 1\n",
      "PlantAbstract \t 7\n",
      "PlantCoding \t 161\n",
      "PlantNonCoding \t 9\n"
     ]
    }
   ],
   "source": [
    "# save node types \n",
    "family_node_labels = []\n",
    "for t, subdf in df_families.groupby(\"NodeLabel\"):\n",
    "    print(t, \"\\t\", subdf.shape[0])\n",
    "    subdf.to_csv(\"../data/import/%s-components.tsv\"%t, sep=\"\\t\", index=None)\n",
    "    family_node_labels.append(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Complex \t 1\n",
      "PlantAbstract \t 7\n",
      "PlantCoding \t 161\n",
      "PlantNonCoding \t 9\n"
     ]
    }
   ],
   "source": [
    "for t in family_node_labels:\n",
    "    query = helpers.component_node_query(\"%s-components.tsv\"%t, n_name=\"line.Family\", labels=[t, 'Family'])\n",
    "    #query = f\"MATCH (n:{t}) DELETE n\"\n",
    "    qr = graph.run(query)\n",
    "    print(t, \"\\t\", qr.stats()['nodes_created'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = os.path.join(input_path, \"bio_elements.tsv\")\n",
    "df_bioelements.to_csv(file_name, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_process = df_components[df_components['NodeLabel'].isin(['Process'\n",
    "                                                                ])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "want_columns = ['AddedBy',\n",
    "       'NodeName', 'NodeDescription', 'Process', \n",
    "       'AdditionalInfo', 'Process', 'ExternalDB', 'ModelV',\n",
    "       'GMM_OCD1', 'GMM_OCD', 'GMM_Description', 'GMM_ShortName',\n",
    "       'GMM_Synonyms', 'chebi_identifier', 'pubmed_identifier', 'observed_species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "label = \"Process\"\n",
    "f = \"%s-components.tsv\"%label\n",
    "df_process[want_columns].to_csv(\"../data/import/\" + f, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = helpers.component_node_query(f, label, \n",
    "                     n_name=\"line.NodeName\", \n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "qr = graph.run(q)\n",
    "if not df_process.shape[0] == qr.stats()['nodes_created']:\n",
    "    raise Exception"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "constraints_added: 0\n",
       "constraints_removed: 0\n",
       "contains_updates: True\n",
       "indexes_added: 0\n",
       "indexes_removed: 0\n",
       "labels_added: 6\n",
       "labels_removed: 0\n",
       "nodes_created: 6\n",
       "nodes_deleted: 0\n",
       "properties_set: 29\n",
       "relationships_created: 0\n",
       "relationships_deleted: 0"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qr.stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# END "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein, Clade, Family keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/raw'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = os.path.join(input_path, \"bio_elements.tsv\")\n",
    "df_bioelements.to_csv(file_name, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_to_name = df_bioelements[['NodeID', 'NodeName']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CPS', 'GA20ox3', 'GA20ox1', 'GA20ox4', 'MYB33'], dtype=object)"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_to_name[id_to_name['NodeName'].duplicated()]['NodeName'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_nodes = get_species_homologues('NodeName')\n",
    "df_nodes.set_index('NodeName', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# node name to node IDs\n",
    "node_ids_key = {}\n",
    "for species in all_species:\n",
    "    col = species + '_homologues'\n",
    "    node_ids_key[col] = df_nodes[col].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clades = get_species_homologues('Clade')\n",
    "df_clades.set_index('Clade', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clade name to node IDs\n",
    "clade_ids_key = {}\n",
    "for species in all_species:\n",
    "    col = species + '_homologues'\n",
    "    clade_ids_key[col] = df_clades[col].to_dict() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# family name to node IDs\n",
    "df_families.set_index(\"Family\", inplace=True)\n",
    "family_ids_key = {}\n",
    "for species in all_species:\n",
    "    col = species + '_homologues'\n",
    "    family_ids_key[col] = df_families[col].to_dict() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_to_family = df_bioelements[[\"NodeName\", \"Family\"]].set_index(\"NodeName\")[\"Family\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_to_family = df_bioelements[[\"Clade\", \"Family\"]].set_index(\"Clade\")[\"Family\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "780    Anthocyanin-accumulation\n",
       "781         Trichome-initiation\n",
       "782               RNA-silencing\n",
       "783              ROS-production\n",
       "784              AUX-signalling\n",
       "785              CO2-deficiency\n",
       "Name: NodeName, dtype: object"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_process['NodeName']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Hormone:ET', 'Hormone:JA', 'Hormone:SA', 'Hormone:AUX',\n",
       "       'Hormone:BS', 'Hormone:GA', 'S:Ca', 'P:Stress_HSPs', 'S:MAPKs',\n",
       "       'Pathogen', 'Pathogen_Effector', 'S:ROS', 'S:TFs',\n",
       "       'P:Primary-metabolism_Photosynthesis',\n",
       "       'P:Primary-metabolism_Respiration',\n",
       "       'P:Secondary-metabolism_Anthocyanins',\n",
       "       'P:Secondary-metabolism_Isoprenoids',\n",
       "       'P:Secondary-metabolism_Terpenoids', 'P:Ubiquitination',\n",
       "       'P:Silencing', 'P:Translation', nan, 'S:Rgenes', 'Hormone:CK',\n",
       "       'Hormone:SLs'], dtype=object)"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_components[\"Process\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Components summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = '''MATCH (n) RETURN DISTINCT n.name AS name, n.level AS level'''\n",
    "nodes = graph.run(q).data()\n",
    "all_nodes_in_components = set([(d[\"name\"], d[\"level\"]) for d in nodes])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "313"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_nodes_in_components)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Complex',\n",
       " 'ExternalCoding',\n",
       " 'Metabolite',\n",
       " 'PlantAbstract',\n",
       " 'PlantCoding',\n",
       " 'PlantNonCoding',\n",
       " 'Process',\n",
       " 'Undefined']"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(df_components['NodeLabel'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PlantCoding 142\n",
      "PlantNonCoding 9\n",
      "PlantAbstract 7\n",
      "Complex 1\n",
      "ExternalOrganism 3\n",
      "ExternalCoding 14\n",
      "ExternalNonCoding 0\n",
      "Process 6\n",
      "MetaboliteFamily 4\n",
      "Metabolite 59\n"
     ]
    }
   ],
   "source": [
    "node_dict = {}\n",
    "for label in node_labels:\n",
    "    q = '''MATCH (n:%s) RETURN DISTINCT n.name'''%label\n",
    "    s = set([d['n.name'] for d in graph.run(q).data()])\n",
    "    print(label, len(s))\n",
    "    node_dict[label] = s\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reactions sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges_orig = pd.read_excel(input_path, \n",
    "                         sheet_name=\"Reactions\", \n",
    "                         header=[1], \n",
    "                         dtype=str, \n",
    "                         na_values=helpers.empty_strings)\n",
    "df_edges_orig = df_edges_orig[df_edges_orig['Status'].isin([\"forCB\", \"forCB_INVENTED\"])]\n",
    "df_edges_orig.drop(['Status', 'FOXMES', 'Legacy:Process', 'Legacy:ReactionMode'], axis=1, inplace=True)\n",
    "df_edges_orig.columns = pd.Index(['AddedBy', 'ConnID', 'Species', \n",
    "          'input1_ID', 'input1_level', 'input1_localisation', 'input1_type', \n",
    "          'input2_ID', 'input2_level', 'input2_localisation', 'input2_type', \n",
    "          'input3_ID', 'input3_level', 'input3_localisation', 'input3_type', \n",
    "          'ReactionEffect', 'ReactionMode', 'Modifications',\n",
    "          'output1_ID', 'output1_level', 'output1_localisation', 'output1_type', \n",
    "          'TrustLevel', 'Literature', 'AdditionalInfo', 'Comment', 'ModelV', 'kinetics'],\n",
    "      dtype='object')\n",
    "\n",
    "df_edges_new = pd.read_excel(input_path, \n",
    "                         sheet_name=\"Reactions_New\", \n",
    "                         header=[1], \n",
    "                         dtype=str, \n",
    "                         na_values=helpers.empty_strings)\n",
    "df_edges_new = df_edges_new[~df_edges_new['AddedBy'].isin(['-'])]\n",
    "df_edges_new.drop(['Status'], axis=1, inplace=True)\n",
    "df_edges_new.columns = pd.Index(['AddedBy', 'ConnID', 'Species', \n",
    "          'input1_ID', 'input1_level', 'input1_type', \n",
    "          'input2_ID', 'input2_level', 'input2_type', \n",
    "          'input3_ID', 'input3_level', 'input3_type',                                  \n",
    "          'ReactionEffect', 'ReactionMode', \n",
    "          'output1_ID', 'output1_level', 'output1_type', \n",
    "          'TrustLevel', 'Literature', 'AdditionalInfo', 'Comment', 'ModelV'],\n",
    "      dtype='object')\n",
    "\n",
    "\n",
    "df_edges = pd.concat([df_edges_orig, df_edges_new], sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges = df_edges[~df_edges[\"AddedBy\"].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>output1_ID</th>\n",
       "      <th>output1_level</th>\n",
       "      <th>output1_localisation</th>\n",
       "      <th>output1_type</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>LSD1</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.113.225805</td>\n",
       "      <td>LSD1 interacted with all three catalases both ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>SA</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R4] indirect reaction</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>SA decreased CAT2 activity in a dose-dependent...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX3</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "46        sb  Conn310     ath       CAT       family                 NaN   \n",
       "47        sb  Conn310     ath        SA         node                 NaN   \n",
       "48        sb  Conn310     ath      CAT2         node                 NaN   \n",
       "49        sb  Conn310     ath      CAT2         node                 NaN   \n",
       "1737       x        x       x         x            x                 NaN   \n",
       "\n",
       "       input1_type input2_ID input2_level input2_localisation  ... output1_ID  \\\n",
       "46    plant_coding      LSD1         node                 NaN  ...        NaN   \n",
       "47      metabolite      CAT2         node                 NaN  ...        NaN   \n",
       "48    plant_coding      ACX2         node                 NaN  ...        NaN   \n",
       "49    plant_coding      ACX3         node                 NaN  ...        NaN   \n",
       "1737             x       NaN            x                 NaN  ...          x   \n",
       "\n",
       "     output1_level output1_localisation output1_type  \\\n",
       "46             NaN                  NaN          NaN   \n",
       "47             NaN                  NaN          NaN   \n",
       "48             NaN                  NaN          NaN   \n",
       "49             NaN                  NaN          NaN   \n",
       "1737             x                  NaN            x   \n",
       "\n",
       "                                       TrustLevel  \\\n",
       "46    [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "47                         [R4] indirect reaction   \n",
       "48    [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "49    [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1737                                            x   \n",
       "\n",
       "                          Literature  \\\n",
       "46         DOI:10.1104/pp.113.225805   \n",
       "47    DOI:10.1016/j.chom.2017.01.007   \n",
       "48    DOI:10.1016/j.chom.2017.01.007   \n",
       "49    DOI:10.1016/j.chom.2017.01.007   \n",
       "1737                               x   \n",
       "\n",
       "                                         AdditionalInfo Comment ModelV  \\\n",
       "46    LSD1 interacted with all three catalases both ...     NaN    vNA   \n",
       "47    SA decreased CAT2 activity in a dose-dependent...     NaN    vNA   \n",
       "48    CAT2 Promotes the Activityof ACX2/ACX3 ( (test...     NaN    vNA   \n",
       "49    CAT2 Promotes the Activityof ACX2/ACX3 ( (test...     NaN    vNA   \n",
       "1737                                                  x       x      x   \n",
       "\n",
       "     kinetics  \n",
       "46        NaN  \n",
       "47        NaN  \n",
       "48        NaN  \n",
       "49        NaN  \n",
       "1737      NaN  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Int64Index([1737], dtype='int64')\n"
     ]
    }
   ],
   "source": [
    "x = df_edges[df_edges['AddedBy']=='x'].index\n",
    "print(x)\n",
    "df_edges.drop(x, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[R1] targetted experiments (e.g. Y2H, BIFC)',\n",
       "       '[Ry] invented reaction', '[Rx] incomplete/unspecific reaction',\n",
       "       '[R2] high-throughput experiment (e.g. ChIP-seq)',\n",
       "       '[R3] in-silico prediction', '[R4] indirect reaction'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[\"TrustLevel\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['trust_level']  = df_edges[\"TrustLevel\"].apply(lambda x: re.search( r\"(R[1|2|3|4|x|y]|undefined)\", x).groups()[0])\n",
    "df_edges['observed_species'] = df_edges[\"Species\"].apply(helpers.get_second_item)\n",
    "df_edges['also_observed_in'] = df_edges[\"Species\"].apply(helpers.rest_of_items)\n",
    "df_edges['Comment'] = df_edges['Comment'].fillna(\"\")\n",
    "df_edges['AdditionalInfo'] = df_edges['AdditionalInfo'].fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['KG', 'MZ', 'ZR', 'MPE', 'ACR', 'MAK', 'Å T', 'SB'], dtype=object)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['AddedBy'] = df_edges['AddedBy'].apply(lambda x: x.upper())\n",
    "df_edges[\"AddedBy\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['v1.0', 'v2.5', 'v2.7', 'v2.6', 'vNA'], dtype=object)"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.loc[df_edges['ModelV'].isna(), 'ModelV'] = 'vNA'\n",
    "df_edges['ModelV'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def only_asci(x):\n",
    "    return \"\".join([character for character in x if character.isascii()])\n",
    "\n",
    "def doi_list(x):\n",
    "    x = only_asci(x.lower())\n",
    "    match = re.findall(\"(?:doi)(?:\\:|\\/)\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"doi:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "def pubmed_list(x):\n",
    "    x = only_asci(x.lower())\n",
    "    match = re.findall(\"(?:pmid)\\:\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"pmid:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "def list_to_string(x):\n",
    "    return \",\".join([str(i) for i in x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no/bad reference Conn040 \n",
      "no/bad reference Conn118 \n",
      "no/bad reference Conn120 \n",
      "no/bad reference Conn122 \n",
      "no/bad reference Conn182 \n",
      "no/bad reference Conn183 \n",
      "no/bad reference Conn199 ? Kg need to find reference\n",
      "no/bad reference Conn273 KEGG \n",
      "no/bad reference Conn274 KEGG \n",
      "no/bad reference Conn275 KEGG \n",
      "no/bad reference Conn276 KEGG \n",
      "no/bad reference Conn276  10.1073/pnas.98.4.2065\n",
      "no/bad reference Conn277 KEGG \n",
      "no/bad reference Conn278 KEGG \n",
      "no/bad reference Conn279 KEGG \n",
      "no/bad reference Conn280 KEGG \n",
      "no/bad reference Conn281 KEGG \n",
      "no/bad reference Conn282 KEGG \n",
      "no/bad reference Conn283 KEGG \n",
      "no/bad reference Conn309 \n"
     ]
    }
   ],
   "source": [
    "# format literature sources\n",
    "df_edges[\"Literature\"].fillna(value=\"\", inplace=True)\n",
    "for i, row in df_edges.iterrows():\n",
    "    s = row['Literature']\n",
    "    source = doi_list(s)\n",
    "    source += pubmed_list(s)\n",
    "    for z in s.split(\"|\"):\n",
    "        key = z.lower()\n",
    "        if \":\" in key:\n",
    "            if \"aracyc\" in key:\n",
    "                aracyc_string = \"aracyc:\" + z.split(\":\")[1].strip()\n",
    "                source.append(aracyc_string)\n",
    "            elif \"kegg\" in key:\n",
    "                kegg_string = \"kegg:\" + z.split(\":\")[1].strip()\n",
    "                source.append(kegg_string)\n",
    "            elif \"doi\" in key:\n",
    "                # already fetched\n",
    "                continue\n",
    "            elif (\"pubmed\" in key) or (\"pmid\" in key):\n",
    "                pmid_string = \"pubmed:\" + z.split(\":\")[1].strip()\n",
    "        elif \"invented\" in key:\n",
    "            source.append(\"invented\")\n",
    "        else:\n",
    "            print(\"no/bad reference\", row[\"ConnID\"], z)\n",
    "            source.append(\"other:\" + only_asci(z.strip()))\n",
    "    if len(source) > 0:\n",
    "        df_edges.loc[i, \"literature_sources\"] = list_to_string(source)\n",
    "    else:\n",
    "        print(row[\"ConnID\"], z)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges[[\"ConnID\", \"Literature\", \"literature_sources\"]].to_csv(\"lit-check.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_df = df_edges.copy()\n",
    "#df_edges = save_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn080</td>\n",
       "      <td>ath</td>\n",
       "      <td>Trichome-initiation</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>process</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[Ry] invented reaction</td>\n",
       "      <td>DOI:10.1105/tpc.111.083261 (Qi, 2011: The jasm...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>v2.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ry</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.111.083261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GID1a,b,c</td>\n",
       "      <td>clade/orthologue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA3</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.106.047415</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.106.047415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GID1a,b,c</td>\n",
       "      <td>clade/orthologue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA4</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.106.047415</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.106.047415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA1-GID1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA3-GID1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA4-GID1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>SLY1</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SCF</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.010827</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.010827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>osa</td>\n",
       "      <td>GID2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SCF</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1126/science.1081077</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>osa</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1126/science.1081077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA1-GID1-DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>SCF-SLY1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956 | DOI:10.1105/tpc.01...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956,doi:10.1105/tpc.010827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>osa</td>\n",
       "      <td>GA1-GID1-SLR1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>SCF-GID2</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956 | https://doi.org/10...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>osa</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>SB</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>SA</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R4] indirect reaction</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>SA decreased CAT2 activity in a dose-dependent...</td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R4</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1016/j.chom.2017.01.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>SB</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1016/j.chom.2017.01.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>SB</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX3</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1016/j.chom.2017.01.007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy   ConnID Species            input1_ID      input1_level  \\\n",
       "74       KG  Conn080     ath  Trichome-initiation            family   \n",
       "260     MAK      NaN     ath            GID1a,b,c  clade/orthologue   \n",
       "261     MAK      NaN     ath            GID1a,b,c  clade/orthologue   \n",
       "268     MAK      NaN     ath             GA1-GID1            family   \n",
       "269     MAK      NaN     ath             GA3-GID1            family   \n",
       "270     MAK      NaN     ath             GA4-GID1            family   \n",
       "271     MAK      NaN     ath                 SLY1              node   \n",
       "272     MAK      NaN     osa                 GID2              node   \n",
       "273     MAK      NaN     ath       GA1-GID1-DELLA            family   \n",
       "274     MAK      NaN     osa        GA1-GID1-SLR1            family   \n",
       "287      SB  Conn310     ath                   SA              node   \n",
       "288      SB  Conn310     ath                 CAT2              node   \n",
       "289      SB  Conn310     ath                 CAT2              node   \n",
       "\n",
       "    input1_localisation    input1_type  input2_ID input2_level  \\\n",
       "74                  NaN        process  potyvirus       family   \n",
       "260                 NaN   plant_coding        GA3       family   \n",
       "261                 NaN   plant_coding        GA4       family   \n",
       "268                 NaN  plant_complex      DELLA       family   \n",
       "269                 NaN  plant_complex      DELLA       family   \n",
       "270                 NaN  plant_complex      DELLA       family   \n",
       "271                 NaN   plant_coding        SCF       family   \n",
       "272                 NaN   plant_coding        SCF       family   \n",
       "273                 NaN  plant_complex   SCF-SLY1       family   \n",
       "274                 NaN  plant_complex   SCF-GID2       family   \n",
       "287                 NaN     metabolite       CAT2         node   \n",
       "288                 NaN   plant_coding       ACX2         node   \n",
       "289                 NaN   plant_coding       ACX3         node   \n",
       "\n",
       "    input2_localisation  ...                                   TrustLevel  \\\n",
       "74                  NaN  ...                       [Ry] invented reaction   \n",
       "260                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "261                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "268                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "269                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "270                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "271                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "272                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "273                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "274                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "287                 NaN  ...                       [R4] indirect reaction   \n",
       "288                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "289                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                            Literature  \\\n",
       "74   DOI:10.1105/tpc.111.083261 (Qi, 2011: The jasm...   \n",
       "260                         DOI:10.1105/tpc.106.047415   \n",
       "261                         DOI:10.1105/tpc.106.047415   \n",
       "268                          DOI:10.1104/pp.112.200956   \n",
       "269                          DOI:10.1104/pp.112.200956   \n",
       "270                          DOI:10.1104/pp.112.200956   \n",
       "271                             DOI:10.1105/tpc.010827   \n",
       "272                        DOI:10.1126/science.1081077   \n",
       "273  DOI:10.1104/pp.112.200956 | DOI:10.1105/tpc.01...   \n",
       "274  DOI:10.1104/pp.112.200956 | https://doi.org/10...   \n",
       "287                     DOI:10.1016/j.chom.2017.01.007   \n",
       "288                     DOI:10.1016/j.chom.2017.01.007   \n",
       "289                     DOI:10.1016/j.chom.2017.01.007   \n",
       "\n",
       "                                        AdditionalInfo Comment ModelV  \\\n",
       "74                                                               v2.7   \n",
       "260                                                               vNA   \n",
       "261                                                               vNA   \n",
       "268                                                               vNA   \n",
       "269                                                               vNA   \n",
       "270                                                               vNA   \n",
       "271                                                               vNA   \n",
       "272                                                               vNA   \n",
       "273                                                               vNA   \n",
       "274                                                               vNA   \n",
       "287  SA decreased CAT2 activity in a dose-dependent...            vNA   \n",
       "288  CAT2 Promotes the Activityof ACX2/ACX3 ( (test...            vNA   \n",
       "289  CAT2 Promotes the Activityof ACX2/ACX3 ( (test...            vNA   \n",
       "\n",
       "    kinetics trust_level observed_species also_observed_in  \\\n",
       "74       NaN          Ry              ath                    \n",
       "260      NaN          R1              ath                    \n",
       "261      NaN          R1              ath                    \n",
       "268      NaN          R1              ath                    \n",
       "269      NaN          R1              ath                    \n",
       "270      NaN          R1              ath                    \n",
       "271      NaN          R1              ath                    \n",
       "272      NaN          R1              osa                    \n",
       "273      NaN          R1              ath                    \n",
       "274      NaN          R1              osa                    \n",
       "287      NaN          R4              ath                    \n",
       "288      NaN          R1              ath                    \n",
       "289      NaN          R1              ath                    \n",
       "\n",
       "                                   literature_sources  \n",
       "74                         doi:10.1105/tpc.111.083261  \n",
       "260                        doi:10.1105/tpc.106.047415  \n",
       "261                        doi:10.1105/tpc.106.047415  \n",
       "268                         doi:10.1104/pp.112.200956  \n",
       "269                         doi:10.1104/pp.112.200956  \n",
       "270                         doi:10.1104/pp.112.200956  \n",
       "271                            doi:10.1105/tpc.010827  \n",
       "272                       doi:10.1126/science.1081077  \n",
       "273  doi:10.1104/pp.112.200956,doi:10.1105/tpc.010827  \n",
       "274                         doi:10.1104/pp.112.200956  \n",
       "287                    doi:10.1016/j.chom.2017.01.007  \n",
       "288                    doi:10.1016/j.chom.2017.01.007  \n",
       "289                    doi:10.1016/j.chom.2017.01.007  \n",
       "\n",
       "[13 rows x 32 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[df_edges['ConnID'].duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG,other:10.107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "                                    TrustLevel  \\\n",
       "0  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo Comment ModelV kinetics  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.           v1.0      NaN   \n",
       "1                 ACS catalyse ACC to SAMe reaction.           v1.0      NaN   \n",
       "2                   ACO catalyse ACC to ET reaction.           v1.0      NaN   \n",
       "3  Copper gets transported from the cytoplasm to ...           v1.0      NaN   \n",
       "4  Copper activates the membrane bound ethylene r...           v1.0      NaN   \n",
       "\n",
       "  trust_level observed_species also_observed_in  \\\n",
       "0          R1              ath                    \n",
       "1          R1              ath                    \n",
       "2          R1              ath                    \n",
       "3          R1              ath                    \n",
       "4          R1              ath                    \n",
       "\n",
       "                                  literature_sources  \n",
       "0                  doi:10.1042/bj20120245,other:KEGG  \n",
       "1                  doi:10.1042/bj20120245,other:KEGG  \n",
       "2                  doi:10.1042/bj20120245,other:KEGG  \n",
       "3  doi:10.1042/bj20120245,other:KEGG,other:10.107...  \n",
       "4                  doi:10.1042/bj20120245,other:KEGG  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in ['input1', 'input2', 'input3', 'output1']:\n",
    "    df_edges.loc[:, x + \"_ID\"] = df_edges[x + \"_ID\"].apply(helpers.reorder_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG,other:10.107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "                                    TrustLevel  \\\n",
       "0  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo Comment ModelV kinetics  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.           v1.0      NaN   \n",
       "1                 ACS catalyse ACC to SAMe reaction.           v1.0      NaN   \n",
       "2                   ACO catalyse ACC to ET reaction.           v1.0      NaN   \n",
       "3  Copper gets transported from the cytoplasm to ...           v1.0      NaN   \n",
       "4  Copper activates the membrane bound ethylene r...           v1.0      NaN   \n",
       "\n",
       "  trust_level observed_species also_observed_in  \\\n",
       "0          R1              ath                    \n",
       "1          R1              ath                    \n",
       "2          R1              ath                    \n",
       "3          R1              ath                    \n",
       "4          R1              ath                    \n",
       "\n",
       "                                  literature_sources  \n",
       "0                  doi:10.1042/bj20120245,other:KEGG  \n",
       "1                  doi:10.1042/bj20120245,other:KEGG  \n",
       "2                  doi:10.1042/bj20120245,other:KEGG  \n",
       "3  doi:10.1042/bj20120245,other:KEGG,other:10.107...  \n",
       "4                  doi:10.1042/bj20120245,other:KEGG  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_node_to_family(x):\n",
    "    # pathogen proteins are listed as proteins, so cannot use dict\n",
    "    id_, type_, level_ = x.values\n",
    "      \n",
    "    if id_ is np.nan:\n",
    "        return np.nan, np.nan\n",
    "\n",
    "    new_label = None\n",
    "    family_id = None\n",
    "    \n",
    "    ########################\n",
    "    # Simple Cases\n",
    "    ########################\n",
    "    if type_ in ['complex', 'complex [active]', 'complex [activated]', 'complex [inactive]', 'plant_complex']:\n",
    "        if not (id_ in node_dict[\"Complex\"]):\n",
    "            complexes_to_add.append(id_)\n",
    "        new_label = 'Complex'\n",
    "        family_id = id_\n",
    "    \n",
    "    elif type_ in [\"metabolite\"]:\n",
    "        for label in [\"Metabolite\", \"MetaboliteFamily\"]:\n",
    "            if id_ in node_dict[label]:\n",
    "                new_label = label\n",
    "                break\n",
    "        if not new_label:\n",
    "            missing_in_components.update([id_])        \n",
    "            print(id_, type_, level_, \" | (label) not a listed metabolite\")\n",
    "        family_id = id_\n",
    "\n",
    "    elif type_ in ['process']:\n",
    "        if (id_ in node_dict[\"Process\"]):\n",
    "            family_id = id_\n",
    "            new_label = \"Process\"\n",
    "        else:\n",
    "            print(id_, type_, level_, \" | (label) process not a listed process\")\n",
    "        \n",
    "    else:\n",
    "        ########################\n",
    "        # family ID\n",
    "        ########################\n",
    "        check_external = False\n",
    "        if level_ == \"family\":    \n",
    "            family_id = id_\n",
    "        elif level_ in [\"clade\", \"clade/orthologue\"]:\n",
    "            try:\n",
    "                family_id = clade_to_family[id_]\n",
    "                replace_w_family.update([id_])\n",
    "            except KeyError:\n",
    "                check_external = True\n",
    "        elif level_ == \"node\":\n",
    "            try:\n",
    "                family_id = node_to_family[id_]\n",
    "                replace_w_family.update([id_])\n",
    "            except KeyError:\n",
    "                check_external = True\n",
    "        \n",
    "        if check_external:\n",
    "            for label in [\"ExternalOrganism\", \"ExternalCoding\", \"ExternalNonCoding\"]:\n",
    "                if id_ in node_dict[label]:\n",
    "                    new_label = label\n",
    "                    family_id = id_\n",
    "                    break\n",
    "                    \n",
    "        if not family_id:\n",
    "            print(id_, type_, level_, \" | (family id) could not convert to family/external\")\n",
    "                \n",
    "        ########################\n",
    "        # Label\n",
    "        ########################\n",
    "        if (family_id) and (not new_label):\n",
    "            id_labels = [] #looping just in case an id occurs mutiple times\n",
    "            for label in node_labels:\n",
    "                if family_id in node_dict[label]:\n",
    "                    id_labels.append(label)\n",
    "\n",
    "            if len(id_labels) == 1:\n",
    "                new_label = id_labels[0]\n",
    "            elif len(id_labels) > 1:\n",
    "                print(id_, type_, level_, \" | (label) many labels fit\")\n",
    "                missing_in_components.update([id_])       \n",
    "\n",
    "            else:\n",
    "                print(id_, type_, level_, \" | (label) could not find label\")\n",
    "                missing_in_components.update([id_])        \n",
    "\n",
    "    return family_id, new_label\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miR6022 plant_ncRNA node  | (family id) could not convert to family/external\n",
      "LSD1 plant_coding node  | (family id) could not convert to family/external\n"
     ]
    }
   ],
   "source": [
    "complexes_to_add = []\n",
    "missing_in_components = set()\n",
    "replace_w_family = set()\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, level_col, new_id, new_label_col  =\\\n",
    "            [prefix + x for x in ('_ID',  '_type',  '_level',  '_newID', '_label')]    \n",
    "    \n",
    "    df_edges[[new_id, new_label_col]] = df_edges[[id_col, type_col, level_col]].apply(convert_node_to_family, axis=1, result_type='expand')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "      <th>input1_newID</th>\n",
       "      <th>input1_label</th>\n",
       "      <th>input2_newID</th>\n",
       "      <th>input2_label</th>\n",
       "      <th>input3_newID</th>\n",
       "      <th>input3_label</th>\n",
       "      <th>output1_newID</th>\n",
       "      <th>output1_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>SAM</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ET</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG,other:10.107...</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>ETR</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETR</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  also_observed_in                                 literature_sources  \\\n",
       "0                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "1                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "2                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "3                   doi:10.1042/bj20120245,other:KEGG,other:10.107...   \n",
       "4                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "\n",
       "  input1_newID input1_label input2_newID input2_label input3_newID  \\\n",
       "0        L-Met   Metabolite          SAM  PlantCoding          NaN   \n",
       "1         SAMe   Metabolite          ACS  PlantCoding          NaN   \n",
       "2          ACC   Metabolite          ACO  PlantCoding          NaN   \n",
       "3         Cu2+   Metabolite          HMA  PlantCoding          NaN   \n",
       "4          ETR  PlantCoding         Cu2+   Metabolite          NaN   \n",
       "\n",
       "  input3_label output1_newID output1_label  \n",
       "0          NaN          SAMe    Metabolite  \n",
       "1          NaN           ACC    Metabolite  \n",
       "2          NaN            ET    Metabolite  \n",
       "3          NaN          Cu2+    Metabolite  \n",
       "4          NaN           ETR   PlantCoding  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_orthologues(x, prefix=\"\"):\n",
    "    id_, level_, label_ = x.values\n",
    "    ########################\n",
    "    # Specie specific homologues\n",
    "    ########################\n",
    "    \n",
    "    return_D = {f\"{prefix}_{specie}_homologues\":\"\" for specie in all_species}\n",
    "    \n",
    "    if label_ in ['PlantCoding', 'PlantNonCoding',  'PlantAbstract']:\n",
    "\n",
    "        for species in all_species:\n",
    "            species = f\"{species}_homologues\"\n",
    "            k = f\"{prefix}_{species}\"\n",
    "            if level_ == 'node':\n",
    "                return_D[k] = node_ids_key[species][id_]\n",
    "            elif level_ == 'clade':\n",
    "                return_D[k] =  clade_ids_key[species][id_]\n",
    "            elif level_ == 'family':\n",
    "                return_D[k] =  family_ids_key[species][id_]\n",
    "\n",
    "    return_D = {x:helpers.list_to_string(list(return_D[x])) for x in return_D}\n",
    "    return return_D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "new_dfs = []\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    print(prefix)\n",
    "    id_col, level_col, new_label_col  =\\\n",
    "                [prefix + x for x in ('_ID',  '_level',  '_label')]        \n",
    "    \n",
    "    new_df = df_edges[[id_col, level_col, new_label_col ]].apply(get_orthologues, axis=1, result_type='expand', prefix=prefix)\n",
    "    new_dfs.append(new_df)\n",
    "    #df_edges = df_edges.join(new_df, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "homologues_df = pd.concat(new_dfs, sort=False, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "input1_ath_homologues                                            \n",
       "input1_osa_homologues                                            \n",
       "input1_stu_homologues                                            \n",
       "input2_ath_homologues     AT2G36880,AT1G02500,AT4G01850,AT3G17390\n",
       "input2_osa_homologues                                            \n",
       "input2_stu_homologues                                            \n",
       "input3_ath_homologues                                            \n",
       "input3_osa_homologues                                            \n",
       "input3_stu_homologues                                            \n",
       "output1_ath_homologues                                           \n",
       "output1_osa_homologues                                           \n",
       "output1_stu_homologues                                           \n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homologues_df.loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges = df_edges.join(homologues_df, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>input1_stu_homologues</th>\n",
       "      <th>input2_ath_homologues</th>\n",
       "      <th>input2_osa_homologues</th>\n",
       "      <th>input2_stu_homologues</th>\n",
       "      <th>input3_ath_homologues</th>\n",
       "      <th>input3_osa_homologues</th>\n",
       "      <th>input3_stu_homologues</th>\n",
       "      <th>output1_ath_homologues</th>\n",
       "      <th>output1_osa_homologues</th>\n",
       "      <th>output1_stu_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT2G36880,AT1G02500,AT4G01850,AT3G17390</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT5G65800,AT5G51690,AT4G26200,AT3G49700,AT1G01...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT1G62380,AT1G12010,AT1G77330,AT1G05010,AT2G19590</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT5G44790,AT5G21930,AT4G33520,AT1G63440</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  input1_stu_homologues                              input2_ath_homologues  \\\n",
       "0                                  AT2G36880,AT1G02500,AT4G01850,AT3G17390   \n",
       "1                        AT5G65800,AT5G51690,AT4G26200,AT3G49700,AT1G01...   \n",
       "2                        AT1G62380,AT1G12010,AT1G77330,AT1G05010,AT2G19590   \n",
       "3                                  AT5G44790,AT5G21930,AT4G33520,AT1G63440   \n",
       "4                                                                            \n",
       "\n",
       "  input2_osa_homologues input2_stu_homologues input3_ath_homologues  \\\n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "\n",
       "  input3_osa_homologues input3_stu_homologues  \\\n",
       "0                                               \n",
       "1                                               \n",
       "2                                               \n",
       "3                                               \n",
       "4                                               \n",
       "\n",
       "                              output1_ath_homologues output1_osa_homologues  \\\n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                                                             \n",
       "4  AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580                          \n",
       "\n",
       "  output1_stu_homologues  \n",
       "0                         \n",
       "1                         \n",
       "2                         \n",
       "3                         \n",
       "4                         \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "node_type_to_node_form_dict = {\n",
    "    \"gene\":\"gene\",\n",
    "    \n",
    "    \"protein\":\"protein\",\n",
    "    \"protein [activated]\":\"protein_active\",\n",
    "    'protein [active]': \"protein_active\",\n",
    "    \n",
    "    \"ncRNA\":\"ncRNA\",\n",
    "    \"plant_ncRNA\":\"ncRNA\",\n",
    "    'ta-siRNA':\"ta-siRNA\", \n",
    "        \n",
    "    \"complex\":\"complex\", \n",
    "    \"plant_complex\":\"complex\",\n",
    "    'complex [active]': \"complex_active\",\n",
    "    \n",
    "    \"metabolite\":\"metabolite\",\n",
    "    \n",
    "    \"process\":\"process\", \n",
    "    'process [active]':\"process_active\",\n",
    "\n",
    "    np.nan:\"\", \n",
    "    \"plant_coding\":\"unknown\"\n",
    "}\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, new_form_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_form')]\n",
    "    \n",
    "    x = df_edges[[id_col, type_col]].dropna(how='all')\n",
    "    \n",
    "    print(prefix)\n",
    "    df_edges[new_form_col] = df_edges[type_col].apply(lambda x: node_type_to_node_form_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>input3_ath_homologues</th>\n",
       "      <th>input3_osa_homologues</th>\n",
       "      <th>input3_stu_homologues</th>\n",
       "      <th>output1_ath_homologues</th>\n",
       "      <th>output1_osa_homologues</th>\n",
       "      <th>output1_stu_homologues</th>\n",
       "      <th>input1_form</th>\n",
       "      <th>input2_form</th>\n",
       "      <th>input3_form</th>\n",
       "      <th>output1_form</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>protein</td>\n",
       "      <td>metabolite</td>\n",
       "      <td></td>\n",
       "      <td>protein_active</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  input3_ath_homologues input3_osa_homologues input3_stu_homologues  \\\n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "\n",
       "                              output1_ath_homologues output1_osa_homologues  \\\n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                                                             \n",
       "4  AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580                          \n",
       "\n",
       "  output1_stu_homologues input1_form     input2_form input3_form  \\\n",
       "0                         metabolite  protein_active               \n",
       "1                         metabolite  protein_active               \n",
       "2                         metabolite  protein_active               \n",
       "3                         metabolite         protein               \n",
       "4                            protein      metabolite               \n",
       "\n",
       "     output1_form  \n",
       "0      metabolite  \n",
       "1      metabolite  \n",
       "2      metabolite  \n",
       "3      metabolite  \n",
       "4  protein_active  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_localisation_dict = {\n",
    "    'nuc':'nucleus',\n",
    "    'er':'endoplasmic reticulum',\n",
    "    'golgi':'golgi apparatus'\n",
    "}\n",
    "\n",
    "\n",
    "good_localisations = set([\n",
    "     'nucleus',\n",
    "     'nucleolus',\n",
    "     'cytoplasm',\n",
    "     'vacuole',\n",
    "     'endoplasmic reticulum',\n",
    "     'chloroplast',\n",
    "     'mitochondrion',\n",
    "     'golgi apparatus',\n",
    "     'peroxisome',\n",
    "     'apoplast',\n",
    "     'extracellular'\n",
    "])\n",
    "\n",
    "\n",
    "def node_localisation_std(x):\n",
    "    if not type(x) == str:\n",
    "        return \"\"\n",
    "    \n",
    "    x = x.lower()\n",
    "    if x in node_localisation_dict.keys():\n",
    "        x = node_localisation_dict[x]\n",
    "    \n",
    "    if x in good_localisations:\n",
    "        return x\n",
    "    else:\n",
    "        return \"\"\n",
    "\n",
    "node_localisations = set()\n",
    "new_localisation = set()\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, localisation_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_localisation')]\n",
    "    \n",
    "    x = df_edges[['ConnID', id_col, localisation_col]].dropna(how='all', subset=[ id_col, localisation_col])\n",
    "    \n",
    "    #for _, y in x.iterrows():\n",
    "    #    if y[localisation_col] in ['mitochondria?', np.nan]:\n",
    "    #        print(y['ConnID'], \"\\t\", y[id_col], \"\\t\", y[localisation_col])\n",
    "    \n",
    "    \n",
    "    node_localisations.update(x[localisation_col])\n",
    "    \n",
    "    #print(prefix)\n",
    "    df_edges[localisation_col] = df_edges[localisation_col].apply(node_localisation_std)\n",
    "    \n",
    "    new_localisation.update(df_edges[localisation_col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ER',\n",
       " 'Golgi',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'extracellular',\n",
       " 'mitochondria?',\n",
       " nan,\n",
       " 'nuc',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_localisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'endoplasmic reticulum',\n",
       " 'extracellular',\n",
       " 'golgi apparatus',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_localisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.to_csv(os.path.join(\"..\", \"data\", \"raw\", \"edges-sheet.tsv\"), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "homologue_cols = [f\"{x}_homologues\" for x in all_species]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ath', 'stu', 'osa']"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(\"..\", \"data\", \"raw\", \"complexes_to_add.tsv\"), \"w\") as out:\n",
    "    for c in complexes_to_add:\n",
    "        out.write(f\"{c}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
