{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import neo4j DB: 3/4\n",
    "\n",
    "Code to translate v2.7.4_PIS-model.xlsx to neo4j database. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import os\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "base_path = Path(\"..\")\n",
    "input_path = base_path / \"data\" / \"raw\"\n",
    "output_path = base_path / \"data\" / \"parsed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to graph via docker-compose link. See http://localhost:7474/browser/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from py2neo import Graph, Node, Relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = Graph(host=\"neo4j\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "sheets = [#(file, sheet_name)\n",
    "    (\"v2.7.9_PIS-model.xlsx\", \"Reactions\"),\n",
    "#     (\"v2.7.5_PIS-model.xlsx\", \"Reactions\"),\n",
    "#     (\"v2.7.5_PIS-model.xlsx\", \"Reactions_New\"), \n",
    "#     (\"Model_CK.xlsx\", \"Reactions_new\"), \n",
    "#     (\"v2.7.2_PIS-model-JALR.xlsx\", \"Reactions_New\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.9_PIS-model.xlsx Reactions\n"
     ]
    }
   ],
   "source": [
    "# resave xlsx as tsv\n",
    "drops = ['FOXMES', 'Legacy:Process', 'Legacy:ReactionMode', \"Comment\"]#, 'ConnID']\n",
    "col_rename = {\n",
    "    'Status':'Status',\n",
    "    'AddedBy':'AddedBy',\n",
    "    'Species':'Species',\n",
    "    'ID':'input1_ID',\n",
    "    'level':'input1_level',\n",
    "    'localisation':'input1_localisation',\n",
    "    'type':'input1_type',\n",
    "    'ID.1':'input2_ID',\n",
    "    'level.1':'input2_level',\n",
    "    'localisation.1':'input2_localisation',\n",
    "    'type.1':'input2_type',\n",
    "    'ID.2':'input3_ID',\n",
    "    'level.2':'input3_level',\n",
    "    'localisation.2':'input3_localisation',\n",
    "    'type.2':'input3_type',\n",
    "    'ReactionEffect':'ReactionEffect',\n",
    "    'ReactionMode':'ReactionMode',\n",
    "    'Modifications':'Modifications',\n",
    "    'ID.3':'output1_ID',\n",
    "    'level.3':'output1_level',\n",
    "    'localisation.3':'output1_localisation',\n",
    "    'type.3':'output1_type',\n",
    "    'TrustLevel':'TrustLevel',\n",
    "    'Literature':'Literature',\n",
    "    'AdditionalInfo':'AdditionalInfo',\n",
    "    'Comment':'Comment',\n",
    "    'Model-v':'ModelV',\n",
    "    'KINETICS':'kinetics', \n",
    "    'ConnID': 'ConnID'\n",
    "}\n",
    "\n",
    "for file_name, sheet_name in sheets:\n",
    "    \n",
    "    print(file_name, sheet_name)\n",
    "    \n",
    "    file_path = input_path / file_name\n",
    "    \n",
    "    base_name, extension = os.path.splitext(file_name)\n",
    "    new_file_path = output_path / f'{base_name}-{sheet_name}.tsv'\n",
    "    \n",
    "    #if os.path.exists(new_file_path):\n",
    "    #    continue\n",
    "        \n",
    "    df = pd.read_excel(file_path, \n",
    "                    sheet_name=sheet_name, \n",
    "                    header=[1], \n",
    "                    dtype=str, \n",
    "                    na_values=helpers.empty_strings)\n",
    "\n",
    "    #df = df[~df[\"AddedBy\"].isna()]\n",
    "    #if 'Status' in df.columns:\n",
    "    #    df = df[df['Status'].isin([\"forCB\", \"forCB_INVENTED\", np.nan])]\n",
    "\n",
    "    to_drop = list(set(drops) & set(df.columns)) + list(df.filter(regex=(\"Unnamed.*\")).columns)\n",
    "    df.drop(to_drop, axis=1, inplace=True)\n",
    "    \n",
    "    new_cols = [col_rename[x] for x in df.columns]\n",
    "    df.columns = new_cols\n",
    "    \n",
    "    df['origin'] = df['Status'] + f'-{base_name}-{sheet_name}'\n",
    "    \n",
    "    df.to_csv(new_file_path, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.9_PIS-model.xlsx Reactions\n"
     ]
    }
   ],
   "source": [
    "dfs = []\n",
    "\n",
    "for file_name, sheet_name in sheets:\n",
    "    \n",
    "    print(file_name, sheet_name)\n",
    "    base_name, extension = os.path.splitext(file_name)\n",
    "\n",
    "    file_path = output_path / f'{base_name}-{sheet_name}.tsv'\n",
    "    \n",
    "    df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "    dfs.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges = pd.concat(dfs, sort=False)\n",
    "df_edges.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>...</th>\n",
       "      <th>Modifications</th>\n",
       "      <th>output1_ID</th>\n",
       "      <th>output1_level</th>\n",
       "      <th>output1_localisation</th>\n",
       "      <th>output1_type</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ET</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>translocation</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein [active]</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Status AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0  forCB      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1  forCB      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2  forCB      KG  Conn003     ath       ACC       family                  ER   \n",
       "3  forCB      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4  forCB      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level  ...  Modifications output1_ID  \\\n",
       "0  metabolite      SAMS        clade  ...            NaN       SAMe   \n",
       "1  metabolite       ACS       family  ...            NaN        ACC   \n",
       "2  metabolite       ACO       family  ...            NaN         ET   \n",
       "3  metabolite       HMA       family  ...  translocation       Cu2+   \n",
       "4     protein      Cu2+       family  ...            NaN        ETR   \n",
       "\n",
       "  output1_level output1_localisation      output1_type  \\\n",
       "0        family                   ER        metabolite   \n",
       "1        family                   ER        metabolite   \n",
       "2        family                   ER        metabolite   \n",
       "3        family                   ER        metabolite   \n",
       "4        family                   ER  protein [active]   \n",
       "\n",
       "                                    TrustLevel  \\\n",
       "0  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo ModelV  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.   v1.0   \n",
       "1                 ACS catalyse ACC to SAMe reaction.   v1.0   \n",
       "2                   ACO catalyse ACC to ET reaction.   v1.0   \n",
       "3  Copper gets transported from the cytoplasm to ...   v1.0   \n",
       "4  Copper activates the membrane bound ethylene r...   v1.0   \n",
       "\n",
       "                             origin  \n",
       "0  forCB-v2.7.9_PIS-model-Reactions  \n",
       "1  forCB-v2.7.9_PIS-model-Reactions  \n",
       "2  forCB-v2.7.9_PIS-model-Reactions  \n",
       "3  forCB-v2.7.9_PIS-model-Reactions  \n",
       "4  forCB-v2.7.9_PIS-model-Reactions  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "binding / oligomerisation                     104\n",
       "catalysis / auto-catalysis                     79\n",
       "transcriptional / translational induction      60\n",
       "protein activation                             52\n",
       "degradation / secretion                        32\n",
       "cleavage / auto-cleavage                       29\n",
       "transcriptional / translational repression     19\n",
       "translocation                                   5\n",
       "protein deactivation                            4\n",
       "dissociation                                    1\n",
       "Name: ReactionMode, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['ReactionMode'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in df_edges.columns:\n",
    "    df_edges[c] = df_edges[c].str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Int64Index([], dtype='int64')\n"
     ]
    }
   ],
   "source": [
    "x = df_edges[df_edges['AddedBy']=='x'].index\n",
    "print(x)\n",
    "#df_edges.drop(x, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[R1] targetted experiments (e.g. Y2H, BIFC)\n",
      "[Ry] invented reaction\n",
      "[R4] indirect reaction\n",
      "[Rx] incomplete/unspecific reaction\n",
      "[R2] high-throughput experiment (e.g. ChIP-seq)\n",
      "[R3] in-silico prediction\n"
     ]
    }
   ],
   "source": [
    "for x in df_edges[\"TrustLevel\"].unique():\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['trust_level']  = df_edges[\"TrustLevel\"].apply(lambda x: re.search( r\"(R[1|2|3|4|x|y]|undefined)\", x).groups()[0])\n",
    "#df_edges['observed_species'] = df_edges['Species'].apply(helpers.lower_string)\n",
    "#df_edges['species_also_observed_in'] = df_edges[\"Species\"].apply(helpers.rest_of_items)\n",
    "#df_edges['Comment'] = df_edges['Comment'].fillna(\"\")\n",
    "df_edges['AdditionalInfo'] = df_edges['AdditionalInfo'].fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['KG', 'MZ', 'ZR', 'MPE', 'ACR', 'MAK', 'ST', 'SB', 'JALR', 'AG'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['AddedBy'] = df_edges['AddedBy'].apply(lambda x: x.upper())\n",
    "df_edges[\"AddedBy\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['v1.0', 'v2.5', 'v2.7', 'v2.6', 'vNA'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.loc[df_edges['ModelV'].isna(), 'ModelV'] = 'vNA'\n",
    "df_edges['ModelV'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# See here for keys https://www.utf8-chartable.de/unicode-utf8-table.pl \n",
    "\n",
    "def only_asci(x):\n",
    "    return \"\".join([character for character in x if character.isascii()])\n",
    "\n",
    "def find_non_ascii(x):\n",
    "    x = str(x)\n",
    "    has_nonascii = False\n",
    "    for character in x:\n",
    "        if not character.isascii():\n",
    "            has_nonascii = True\n",
    "            print('    ', character, ord(character), character.encode())\n",
    "    if has_nonascii:\n",
    "        print(x)\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "ascii_replacers = {\n",
    "    b'\\xc2\\xa0'         : b' ',            # funky WIN whitespace\n",
    "    b'\\xe2\\x80\\xa6'     : b'...',          # …\n",
    "    b'\\xe2\\x80\\x8b'     : b'',             # have no figging clue\n",
    "    b'\\xe2\\x80\\x93'     : b'-',            # –\n",
    "    \n",
    "    b'\\xce\\xb1'         : b'&alpha;',      # α\n",
    "    b'\\xc3\\x9f'         : b'&beta;',       # ß\n",
    "    b'\\xce\\xb2'         : b'&beta;',       # β\n",
    "    \n",
    "    # some \"prime\" symbols...\n",
    "    b'\\xe2\\x80\\x98'     : b'&prime;',      # ‘ Left Single Quotation Mark\n",
    "    b'\\xe2\\x80\\x99'     : b'&prime;',      # ’ Right Single Quotation Mark\n",
    "    b'\\xc2\\xb4'         : b'&prime;',      # ´ Acute Accent\n",
    "    # actual prime\n",
    "    b'\\xe2\\x80\\xb2'     : b'&prime;',      # ′ Prime\n",
    "    \n",
    "    # Sorry accents :(\n",
    "    b'\\xc5\\xa0'         : b'S',            # Š\n",
    "    b'\\xc5\\xa1'         : b's',            # š\n",
    "    b'\\xc5\\xbd'         : b'Z',            # Ž\n",
    "    b'\\xc4\\x8d'         : b'c'             # č\n",
    "}\n",
    "\n",
    "def replacer(x, verbose=False):\n",
    "    if type(x) == float:\n",
    "        return x\n",
    "    y = x.encode('utf-8')\n",
    "    for old, new in ascii_replacers.items():\n",
    "        y = y.replace(old, new)\n",
    "    y = y.decode('utf-8')\n",
    "    \n",
    "    if verbose and ( y != x):\n",
    "        print(f\"'{x}' : '{y}'\")\n",
    "    \n",
    "    return y.strip()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Status \n",
      "-------------\n",
      "\n",
      "AddedBy \n",
      "-------------\n",
      "\n",
      "ConnID \n",
      "-------------\n",
      "\n",
      "Species \n",
      "-------------\n",
      "\n",
      "input1_ID \n",
      "-------------\n",
      "     β 946 b'\\xce\\xb2'\n",
      "all-trans-β-carotene\n",
      "     β 946 b'\\xce\\xb2'\n",
      "9-cis-β-carotene\n",
      "     ′ 8242 b'\\xe2\\x80\\xb2'\n",
      "     β 946 b'\\xce\\xb2'\n",
      "9-cis-10′-apo-β-carotenal\n",
      "\n",
      "input1_level \n",
      "-------------\n",
      "\n",
      "input1_localisation \n",
      "-------------\n",
      "\n",
      "input1_type \n",
      "-------------\n",
      "\n",
      "input2_ID \n",
      "-------------\n",
      "\n",
      "input2_level \n",
      "-------------\n",
      "\n",
      "input2_localisation \n",
      "-------------\n",
      "\n",
      "input2_type \n",
      "-------------\n",
      "\n",
      "input3_ID \n",
      "-------------\n",
      "\n",
      "input3_level \n",
      "-------------\n",
      "\n",
      "input3_localisation \n",
      "-------------\n",
      "\n",
      "input3_type \n",
      "-------------\n",
      "\n",
      "ReactionEffect \n",
      "-------------\n",
      "\n",
      "ReactionMode \n",
      "-------------\n",
      "\n",
      "Modifications \n",
      "-------------\n",
      "\n",
      "output1_ID \n",
      "-------------\n",
      "     β 946 b'\\xce\\xb2'\n",
      "9-cis-β-carotene\n",
      "     ′ 8242 b'\\xe2\\x80\\xb2'\n",
      "     β 946 b'\\xce\\xb2'\n",
      "9-cis-10′-apo-β-carotenal\n",
      "\n",
      "output1_level \n",
      "-------------\n",
      "\n",
      "output1_localisation \n",
      "-------------\n",
      "\n",
      "output1_type \n",
      "-------------\n",
      "\n",
      "TrustLevel \n",
      "-------------\n",
      "\n",
      "Literature \n",
      "-------------\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.​1104/​pp.​107.​104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent…)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.​1104/​pp.​107.​104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent…)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1104/pp.106.095588 (Chen, 2007: Stability of Plant Defense Proteins...); DOI:10.1105/tpc.106.041103 Plant Cell. 2006 Nov; 18(11): 3303–3320.\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1105/tpc.017954 Plant Cell. 2004 Jan; 16(1): 126–143\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1126/scisignal.3109cm4 (Arabidopsis Jasmonate Signaling Pathway) | DOI:10.​1105/​tpc.​110.​080788 (The Arabidopsis bHLH Transcription Factors...) | DOI:10.1016/j.jtbi.2010.12.037 (Transient pulse formation...) | DOI:10.1146/annurev.arplant.043008.092007 (Jasmonate Passes Muster...) | DOI:10.1007/978-0-387-85498-4_8 (Regulation of Secondary Metabolism...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1094/MPMI-23-9-1151 (Manosalva, 2010; : Methyl esterase 1 (StMES1) is required...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "AraCyc:PWY-981 | DOI:​10.​1104/​pp.​103.​2.​315 (Pathway of Salicylic Acid Biosynthesis...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1365-313X.2006.02806.x (Arabidopsis MAP kinase 4 regulates...) | DOI:10.​1104/​pp.​112.​194647 (Dissecting Phosphite-Induced Priming...)\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1105/tpc.010376 (EDS5, an Essential Component of Salicylic Acid–Dependent Signaling...) | DOI:10.1104/pp.010862 (Identification of Arabidopsis Ethylene-Responsive Element...) | DOI:10.1104/pp.105.071662 (Role of Salicylic Acid and Fatty Acid Desaturation...) | DOI:10.1104/pp.106.090035 (Silverleaf Whitefly Induces Salicylic Acid Defenses...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:​10.​1105/​tpc.​015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:​10.​1105/​tpc.​015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "EMBO J., 20 (2001), pp. 5400–5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "EMBO J., 20 (2001), pp. 5400–5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "EMBO J., 20 (2001), pp. 5400–5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126–42\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126–42\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...) | PMID:24989234 (Zander, 2014: TGA Transcription Factors Activate the SA-Suppressible Branch ...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.​1105/​tpc.​108.​058974 (Fode, 2008: The Arabidopsis GRAS Protein SCL14 Interacts...)\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:​10.1105/tpc.113.121731 (Qi, 2014: Arabidopsis DELLA and JAZ Proteins...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI:10.​1105/​tpc.​112.​101428 (Wild, 2012: The Arabidopsis DELLA RGA-LIKE3...)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases… ker R-gene unknown also this interaction postulated\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.​1104/​pp.​111.​192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "     ​ 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.​1104/​pp.​111.​192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "       160 b'\\xc2\\xa0'\n",
      "DOI: 10.1105/tpc.112.098335\n",
      "\n",
      "AdditionalInfo \n",
      "-------------\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "     β 946 b'\\xce\\xb2'\n",
      "Acyl-CoA oxidase: In Arabidopsis, five genes appear to en-code acyl-CoA oxidases (ACX1–ACX5, with varying chain-length specificities). Mutant analysis indicated that ACX1 and ACX5 act redundantly in initiating β-oxidation of OPC-8:0-CoA. Beta-oxidation itself involves three core enzymes, acyl-CoA oxi-dase (ACX), multifunctional protein (MFP; comprising enoyl-CoA hydratase and b-hydroxy-acyl-CoA dehydrogenase activities), and 3-ketoacyl-CoA thiolase (KAT).\n",
      "       160 b'\\xc2\\xa0'\n",
      "bHLH transcription factors EGL3, GL3, and TT8 ; MYB factors MYB75 and GL1;WD-repeat protein TTG1\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "Targets to proteasome (er ligase); HECT domain E3 ubiquitin ligase UPL5 … UPL5 regulates leaf senescence in Arabidopsis through degradation of WRKY53 and ensures that senescence is executed in the correct time frame.\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "     – 8211 b'\\xe2\\x80\\x93'\n",
      "...Upon pathogen challenge, the defense repressor AtSR1 is degradedthrough the SR1IP1–CUL3-mediated ubiquitin/proteasome pathway torelieve the repression of EDS1 transcription exerted by Ca2+/CaM signaling...These results indicatethat SR1IP1–CUL3 E3/proteasome-mediated degradation of AtSR1 is a critical mechanism to relieve the suppression of immune responses exerted by Ca2+/CaM/AtSR1. '|'...The negative regulation of immune response by Ca2+-CaM-AtSR1 complex is proposed to be released through AtSR1-Interacting-Protein-1 (SR1-1P1) that binds to AtSR1 and facilitates ubiquitination and degradation of the latter upon pathogen challenge (Zhang et al., 2014)...'\n",
      "     … 8230 b'\\xe2\\x80\\xa6'\n",
      "This inactivation of tga256 diables binding on the ORA59promotor and its induction by EIN3/ERF+TGA256… Glutaredoxins catalyze the reduction of target proteins through a redox-active site...ROXY19/GRX480-mediated repression depended on the GSH binding site, suggesting that redox modification of either TGA factors or as yet unknown target proteins is important for the suppression of ORA59 promoter activity.\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "       160 b'\\xc2\\xa0'\n",
      "based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)\n",
      "     β 946 b'\\xce\\xb2'\n",
      "Original reaction had ntaCF1-beta (P00826 in SwissProt; Z00044.2 in GenBank), closest match in TAIR10 is ATCG00480 (ATPB). NtCF1-beta is a subunit of chloroplast ATP synthase. HC-Pro may interfere with the assembly of NtCF1β -subunit into the whole enzyme. HC-Pro from PVY with Y1H.\n",
      "     ´ 180 b'\\xc2\\xb4'\n",
      "CCD7 cleaves 9-cis-b-carotene into 9-cis-b-apo-10´-carotenal (Output_1) and b-ionone (Output_2)\n",
      "     ´ 180 b'\\xc2\\xb4'\n",
      "cis-Zeatin riboside + H2O <=> cis-Zeatin + D-Ribose; cZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08080;\n",
      "     ´ 180 b'\\xc2\\xb4'\n",
      "Dihydrozeatin riboside + H2O <=> Dihydrozeatin + D-Ribose; DHZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08074;\n",
      "     ´ 180 b'\\xc2\\xb4'\n",
      "trans-Zeatin riboside + H2O <=> Zeatin + D-Ribose; tZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified;https://www.genome.jp/entry/R08073\n",
      "     ´ 180 b'\\xc2\\xb4'\n",
      "Isopentenyl adenosine + H2O <=> N6-(Delta2-Isopentenyl)-adenine + D-Ribose; iP synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08072\n",
      "\n",
      "ModelV \n",
      "-------------\n",
      "\n",
      "origin \n",
      "-------------\n",
      "\n",
      "trust_level \n",
      "-------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bad_cols = []\n",
    "for c in df_edges.columns:\n",
    "    print(c, \"\\n-------------\")\n",
    "    if any(df_edges[c].apply(find_non_ascii)):\n",
    "        bad_cols.append(c)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1_ID \n",
      "-------------\n",
      "'all-trans-β-carotene' : 'all-trans-&beta;-carotene'\n",
      "'9-cis-β-carotene' : '9-cis-&beta;-carotene'\n",
      "'9-cis-10′-apo-β-carotenal' : '9-cis-10&prime;-apo-&beta;-carotenal'\n",
      "\n",
      "output1_ID \n",
      "-------------\n",
      "'9-cis-β-carotene' : '9-cis-&beta;-carotene'\n",
      "'9-cis-10′-apo-β-carotenal' : '9-cis-10&prime;-apo-&beta;-carotenal'\n",
      "\n",
      "Literature \n",
      "-------------\n",
      "'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.​1104/​pp.​107.​104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent…)' : 'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes...) | DOI:10.1104/pp.107.104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent...)'\n",
      "'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.​1104/​pp.​107.​104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent…)' : 'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes...) | DOI:10.1104/pp.107.104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent...)'\n",
      "'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)' : 'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes...) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)'\n",
      "'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)' : 'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes...) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)'\n",
      "'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)' : 'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes...) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)'\n",
      "'DOI:10.1104/pp.106.095588 (Chen, 2007: Stability of Plant Defense Proteins...); DOI:10.1105/tpc.106.041103 Plant Cell. 2006 Nov; 18(11): 3303–3320.' : 'DOI:10.1104/pp.106.095588 (Chen, 2007: Stability of Plant Defense Proteins...); DOI:10.1105/tpc.106.041103 Plant Cell. 2006 Nov; 18(11): 3303-3320.'\n",
      "'DOI:10.1105/tpc.017954 Plant Cell. 2004 Jan; 16(1): 126–143' : 'DOI:10.1105/tpc.017954 Plant Cell. 2004 Jan; 16(1): 126-143'\n",
      "'DOI:10.1126/scisignal.3109cm4 (Arabidopsis Jasmonate Signaling Pathway) | DOI:10.​1105/​tpc.​110.​080788 (The Arabidopsis bHLH Transcription Factors...) | DOI:10.1016/j.jtbi.2010.12.037 (Transient pulse formation...) | DOI:10.1146/annurev.arplant.043008.092007 (Jasmonate Passes Muster...) | DOI:10.1007/978-0-387-85498-4_8 (Regulation of Secondary Metabolism...)' : 'DOI:10.1126/scisignal.3109cm4 (Arabidopsis Jasmonate Signaling Pathway) | DOI:10.1105/tpc.110.080788 (The Arabidopsis bHLH Transcription Factors...) | DOI:10.1016/j.jtbi.2010.12.037 (Transient pulse formation...) | DOI:10.1146/annurev.arplant.043008.092007 (Jasmonate Passes Muster...) | DOI:10.1007/978-0-387-85498-4_8 (Regulation of Secondary Metabolism...)'\n",
      "'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)' : 'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.1105/tpc.106.048017 (MYC2 Differentially Modulates Diverse...)'\n",
      "'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)' : 'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.1105/tpc.106.048017 (MYC2 Differentially Modulates Diverse...)'\n",
      "'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)' : 'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.1105/tpc.106.048017 (MYC2 Differentially Modulates Diverse...)'\n",
      "'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)' : 'DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.1105/tpc.106.048017 (MYC2 Differentially Modulates Diverse...)'\n",
      "'DOI:10.1094/MPMI-23-9-1151 (Manosalva, 2010; : Methyl esterase 1 (StMES1) is required...)' : 'DOI:10.1094/MPMI-23-9-1151 (Manosalva, 2010; : Methyl esterase 1 (StMES1) is required...)'\n",
      "'AraCyc:PWY-981 | DOI:​10.​1104/​pp.​103.​2.​315 (Pathway of Salicylic Acid Biosynthesis...)' : 'AraCyc:PWY-981 | DOI:10.1104/pp.103.2.315 (Pathway of Salicylic Acid Biosynthesis...)'\n",
      "'DOI:10.1111/j.1365-313X.2006.02806.x (Arabidopsis MAP kinase 4 regulates...) | DOI:10.​1104/​pp.​112.​194647 (Dissecting Phosphite-Induced Priming...)' : 'DOI:10.1111/j.1365-313X.2006.02806.x (Arabidopsis MAP kinase 4 regulates...) | DOI:10.1104/pp.112.194647 (Dissecting Phosphite-Induced Priming...)'\n",
      "'DOI:10.1105/tpc.010376 (EDS5, an Essential Component of Salicylic Acid–Dependent Signaling...) | DOI:10.1104/pp.010862 (Identification of Arabidopsis Ethylene-Responsive Element...) | DOI:10.1104/pp.105.071662 (Role of Salicylic Acid and Fatty Acid Desaturation...) | DOI:10.1104/pp.106.090035 (Silverleaf Whitefly Induces Salicylic Acid Defenses...)' : 'DOI:10.1105/tpc.010376 (EDS5, an Essential Component of Salicylic Acid-Dependent Signaling...) | DOI:10.1104/pp.010862 (Identification of Arabidopsis Ethylene-Responsive Element...) | DOI:10.1104/pp.105.071662 (Role of Salicylic Acid and Fatty Acid Desaturation...) | DOI:10.1104/pp.106.090035 (Silverleaf Whitefly Induces Salicylic Acid Defenses...)'\n",
      "'DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)' : 'DOI:10.1105/tpc.104.027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)'\n",
      "'DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)' : 'DOI:10.1105/tpc.104.027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)'\n",
      "'DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)' : 'DOI:10.1105/tpc.104.027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)'\n",
      "'DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)' : 'DOI:10.1105/tpc.104.027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)'\n",
      "'DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:​10.​1105/​tpc.​015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)' : 'DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:10.1105/tpc.015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)'\n",
      "'DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:​10.​1105/​tpc.​015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)' : 'DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:10.1105/tpc.015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)'\n",
      "'EMBO J., 20 (2001), pp. 5400–5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472' : 'EMBO J., 20 (2001), pp. 5400-5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472'\n",
      "'EMBO J., 20 (2001), pp. 5400–5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472' : 'EMBO J., 20 (2001), pp. 5400-5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472'\n",
      "'EMBO J., 20 (2001), pp. 5400–5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472' : 'EMBO J., 20 (2001), pp. 5400-5411 Direct interaction between the Arabidopsis disease resistance signaling proteins, EDS1 and PAD4 | doi:10.1093/emboj/20.19.5400 | PMID:11574472'\n",
      "'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)' : 'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)'\n",
      "'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)' : 'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)'\n",
      "'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)' : 'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)'\n",
      "'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)' : 'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)'\n",
      "'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)' : 'DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)'\n",
      "'DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126–42' : 'DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126-42'\n",
      "'DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126–42' : 'DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126-42'\n",
      "'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)' : 'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)'\n",
      "'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...) | PMID:24989234 (Zander, 2014: TGA Transcription Factors Activate the SA-Suppressible Branch ...)' : 'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...) | PMID:24989234 (Zander, 2014: TGA Transcription Factors Activate the SA-Suppressible Branch ...)'\n",
      "'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)' : 'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)'\n",
      "'DOI:10.​1105/​tpc.​108.​058974 (Fode, 2008: The Arabidopsis GRAS Protein SCL14 Interacts...)' : 'DOI:10.1105/tpc.108.058974 (Fode, 2008: The Arabidopsis GRAS Protein SCL14 Interacts...)'\n",
      "'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)' : 'DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)'\n",
      "'DOI:​10.1105/tpc.113.121731 (Qi, 2014: Arabidopsis DELLA and JAZ Proteins...)' : 'DOI:10.1105/tpc.113.121731 (Qi, 2014: Arabidopsis DELLA and JAZ Proteins...)'\n",
      "'DOI:10.​1105/​tpc.​112.​101428 (Wild, 2012: The Arabidopsis DELLA RGA-LIKE3...)' : 'DOI:10.1105/tpc.112.101428 (Wild, 2012: The Arabidopsis DELLA RGA-LIKE3...)'\n",
      "'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…' : 'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases...'\n",
      "'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases… ker R-gene unknown also this interaction postulated' : 'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases... ker R-gene unknown also this interaction postulated'\n",
      "'DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.​1104/​pp.​111.​192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)' : 'DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.1104/pp.111.192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)'\n",
      "'DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.​1104/​pp.​111.​192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)' : 'DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.1104/pp.111.192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)'\n",
      "'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…' : 'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases...'\n",
      "'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…' : 'DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases...'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)' : 'DOI:10.1016/j.chom.2012.04.014. (coronatine promotes....)'\n",
      "'DOI: 10.1105/tpc.112.098335' : 'DOI: 10.1105/tpc.112.098335'\n",
      "\n",
      "AdditionalInfo \n",
      "-------------\n",
      "'Acyl-CoA oxidase: In Arabidopsis, five genes appear to en-code acyl-CoA oxidases (ACX1–ACX5, with varying chain-length specificities). Mutant analysis indicated that ACX1 and ACX5 act redundantly in initiating β-oxidation of OPC-8:0-CoA. Beta-oxidation itself involves three core enzymes, acyl-CoA oxi-dase (ACX), multifunctional protein (MFP; comprising enoyl-CoA hydratase and b-hydroxy-acyl-CoA dehydrogenase activities), and 3-ketoacyl-CoA thiolase (KAT).' : 'Acyl-CoA oxidase: In Arabidopsis, five genes appear to en-code acyl-CoA oxidases (ACX1-ACX5, with varying chain-length specificities). Mutant analysis indicated that ACX1 and ACX5 act redundantly in initiating &beta;-oxidation of OPC-8:0-CoA. Beta-oxidation itself involves three core enzymes, acyl-CoA oxi-dase (ACX), multifunctional protein (MFP; comprising enoyl-CoA hydratase and b-hydroxy-acyl-CoA dehydrogenase activities), and 3-ketoacyl-CoA thiolase (KAT).'\n",
      "'bHLH transcription factors EGL3, GL3, and TT8 ; MYB factors MYB75 and GL1;WD-repeat protein TTG1' : 'bHLH transcription factors EGL3, GL3, and TT8 ; MYB factors MYB75 and GL1;WD-repeat protein TTG1'\n",
      "'Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).' : 'Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).'\n",
      "'Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).' : 'Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).'\n",
      "'Targets to proteasome (er ligase); HECT domain E3 ubiquitin ligase UPL5 … UPL5 regulates leaf senescence in Arabidopsis through degradation of WRKY53 and ensures that senescence is executed in the correct time frame.' : 'Targets to proteasome (er ligase); HECT domain E3 ubiquitin ligase UPL5 ... UPL5 regulates leaf senescence in Arabidopsis through degradation of WRKY53 and ensures that senescence is executed in the correct time frame.'\n",
      "'...Upon pathogen challenge, the defense repressor AtSR1 is degradedthrough the SR1IP1–CUL3-mediated ubiquitin/proteasome pathway torelieve the repression of EDS1 transcription exerted by Ca2+/CaM signaling...These results indicatethat SR1IP1–CUL3 E3/proteasome-mediated degradation of AtSR1 is a critical mechanism to relieve the suppression of immune responses exerted by Ca2+/CaM/AtSR1. '|'...The negative regulation of immune response by Ca2+-CaM-AtSR1 complex is proposed to be released through AtSR1-Interacting-Protein-1 (SR1-1P1) that binds to AtSR1 and facilitates ubiquitination and degradation of the latter upon pathogen challenge (Zhang et al., 2014)...'' : '...Upon pathogen challenge, the defense repressor AtSR1 is degradedthrough the SR1IP1-CUL3-mediated ubiquitin/proteasome pathway torelieve the repression of EDS1 transcription exerted by Ca2+/CaM signaling...These results indicatethat SR1IP1-CUL3 E3/proteasome-mediated degradation of AtSR1 is a critical mechanism to relieve the suppression of immune responses exerted by Ca2+/CaM/AtSR1. '|'...The negative regulation of immune response by Ca2+-CaM-AtSR1 complex is proposed to be released through AtSR1-Interacting-Protein-1 (SR1-1P1) that binds to AtSR1 and facilitates ubiquitination and degradation of the latter upon pathogen challenge (Zhang et al., 2014)...''\n",
      "'This inactivation of tga256 diables binding on the ORA59promotor and its induction by EIN3/ERF+TGA256… Glutaredoxins catalyze the reduction of target proteins through a redox-active site...ROXY19/GRX480-mediated repression depended on the GSH binding site, suggesting that redox modification of either TGA factors or as yet unknown target proteins is important for the suppression of ORA59 promoter activity.' : 'This inactivation of tga256 diables binding on the ORA59promotor and its induction by EIN3/ERF+TGA256... Glutaredoxins catalyze the reduction of target proteins through a redox-active site...ROXY19/GRX480-mediated repression depended on the GSH binding site, suggesting that redox modification of either TGA factors or as yet unknown target proteins is important for the suppression of ORA59 promoter activity.'\n",
      "'based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)' : 'based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)'\n",
      "'based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)' : 'based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)'\n",
      "'Original reaction had ntaCF1-beta (P00826 in SwissProt; Z00044.2 in GenBank), closest match in TAIR10 is ATCG00480 (ATPB). NtCF1-beta is a subunit of chloroplast ATP synthase. HC-Pro may interfere with the assembly of NtCF1β -subunit into the whole enzyme. HC-Pro from PVY with Y1H.' : 'Original reaction had ntaCF1-beta (P00826 in SwissProt; Z00044.2 in GenBank), closest match in TAIR10 is ATCG00480 (ATPB). NtCF1-beta is a subunit of chloroplast ATP synthase. HC-Pro may interfere with the assembly of NtCF1&beta; -subunit into the whole enzyme. HC-Pro from PVY with Y1H.'\n",
      "'CCD7 cleaves 9-cis-b-carotene into 9-cis-b-apo-10´-carotenal (Output_1) and b-ionone (Output_2)' : 'CCD7 cleaves 9-cis-b-carotene into 9-cis-b-apo-10&prime;-carotenal (Output_1) and b-ionone (Output_2)'\n",
      "'cis-Zeatin riboside + H2O <=> cis-Zeatin + D-Ribose; cZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08080;' : 'cis-Zeatin riboside + H2O <=> cis-Zeatin + D-Ribose; cZ synthesis; enzyme 5&prime;- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08080;'\n",
      "'Dihydrozeatin riboside + H2O <=> Dihydrozeatin + D-Ribose; DHZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08074;' : 'Dihydrozeatin riboside + H2O <=> Dihydrozeatin + D-Ribose; DHZ synthesis; enzyme 5&prime;- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08074;'\n",
      "'trans-Zeatin riboside + H2O <=> Zeatin + D-Ribose; tZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified;https://www.genome.jp/entry/R08073' : 'trans-Zeatin riboside + H2O <=> Zeatin + D-Ribose; tZ synthesis; enzyme 5&prime;- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified;https://www.genome.jp/entry/R08073'\n",
      "'Isopentenyl adenosine + H2O <=> N6-(Delta2-Isopentenyl)-adenine + D-Ribose; iP synthesis; enzyme 5´- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08072' : 'Isopentenyl adenosine + H2O <=> N6-(Delta2-Isopentenyl)-adenine + D-Ribose; iP synthesis; enzyme 5&prime;- ribonucleotide phosphohydrolase; reactions whose genes for the enzyme are not identified; https://www.genome.jp/entry/R08072'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for c in bad_cols:\n",
    "    print(c, \"\\n-------------\")\n",
    "    df_edges[c] = df_edges[c].apply(replacer, verbose=True)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1_ID \n",
      "-------------\n",
      "output1_ID \n",
      "-------------\n",
      "Literature \n",
      "-------------\n",
      "AdditionalInfo \n",
      "-------------\n"
     ]
    }
   ],
   "source": [
    "for c in bad_cols:\n",
    "    print(c, \"\\n-------------\")\n",
    "    if any(df_edges[c].apply(find_non_ascii)):\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# also need to look for ' (quote) and \" (doube; quote)\n",
    "# which are used instead of prime, (may cause string issues??)\n",
    "\n",
    "# replacing 5' with 5&prime; and 3' with 3&prime;\n",
    "def find_quotes(x):\n",
    "    x = str(x)\n",
    "    if x.find('\"') != -1: \n",
    "        print(x)\n",
    "        return True\n",
    "    elif x.find(\"'\") != -1:\n",
    "        print(x)\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "        \n",
    "\n",
    "def quote_replacer(x, verbose=False):\n",
    "    if type(x) == float:\n",
    "        x = \"\"\n",
    "    y = x.replace(\"5'\", \"5&prime;\")\n",
    "    y = y.replace(\"3'\", \"3&prime;\")\n",
    "\n",
    "    if verbose and ( y != x):\n",
    "        print(f\"'{x}' : '{y}'\")\n",
    "        \n",
    "    return y\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Status \n",
      "-------------\n",
      "\n",
      "AddedBy \n",
      "-------------\n",
      "\n",
      "ConnID \n",
      "-------------\n",
      "\n",
      "Species \n",
      "-------------\n",
      "\n",
      "input1_ID \n",
      "-------------\n",
      "\n",
      "input1_level \n",
      "-------------\n",
      "\n",
      "input1_localisation \n",
      "-------------\n",
      "\n",
      "input1_type \n",
      "-------------\n",
      "\n",
      "input2_ID \n",
      "-------------\n",
      "\n",
      "input2_level \n",
      "-------------\n",
      "\n",
      "input2_localisation \n",
      "-------------\n",
      "\n",
      "input2_type \n",
      "-------------\n",
      "\n",
      "input3_ID \n",
      "-------------\n",
      "\n",
      "input3_level \n",
      "-------------\n",
      "\n",
      "input3_localisation \n",
      "-------------\n",
      "\n",
      "input3_type \n",
      "-------------\n",
      "\n",
      "ReactionEffect \n",
      "-------------\n",
      "\n",
      "ReactionMode \n",
      "-------------\n",
      "\n",
      "Modifications \n",
      "-------------\n",
      "\n",
      "output1_ID \n",
      "-------------\n",
      "\n",
      "output1_level \n",
      "-------------\n",
      "\n",
      "output1_localisation \n",
      "-------------\n",
      "\n",
      "output1_type \n",
      "-------------\n",
      "\n",
      "TrustLevel \n",
      "-------------\n",
      "\n",
      "Literature \n",
      "-------------\n",
      "\n",
      "AdditionalInfo \n",
      "-------------\n",
      "actually binding - acts as inhibition as it outcompetes bdingin to promotor, can be extended to include binding to HISTONE DEACETYLASE 6 (HDA6) 'interacts with JAZs and EIN3/EIL1 as a co-repressor'\n",
      "RTM proteins block the long distance transport of plant viruses. They don't directly interact with CP, but the process of RTM resistance involves at least 5 different proteins.\n",
      "RTM proteins block the long distance transport of plant viruses. They don't directly interact with CP, but the process of RTM resistance involves at least 5 different proteins.\n",
      "...Upon pathogen challenge, the defense repressor AtSR1 is degradedthrough the SR1IP1-CUL3-mediated ubiquitin/proteasome pathway torelieve the repression of EDS1 transcription exerted by Ca2+/CaM signaling...These results indicatethat SR1IP1-CUL3 E3/proteasome-mediated degradation of AtSR1 is a critical mechanism to relieve the suppression of immune responses exerted by Ca2+/CaM/AtSR1. '|'...The negative regulation of immune response by Ca2+-CaM-AtSR1 complex is proposed to be released through AtSR1-Interacting-Protein-1 (SR1-1P1) that binds to AtSR1 and facilitates ubiquitination and degradation of the latter upon pathogen challenge (Zhang et al., 2014)...'\n",
      "HC-Pro from PVY with Y2H, BiFC. Arabidopsis 20S proteasome subunits. Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "The canonical GA signaling model illustrating GA-dependent GID1-DELLA complex formation resulting in DELLA recognition and ubiquitylation by the SCFSLY1 E3. Polyubiquitylation leads to DELLA proteolysis by the 26S proteasome, thereby lifting DELLA repression of GA responses.'\n",
      "...phosphorylated SLR1 is caught by the SCFGID2 complex through an interacting affinity between GID2 and phosphorylated SLR1, triggering the ubiquitin-mediated degradation of SLR1.'\n",
      "D53 binds to the complex S-D14-MAX2 to be degraded by ubiquitination;  'A model of the SL signaling complex in Arabidopsis that includes SL-dependent interaction of Arabidopsis D14 with both MAX2 and SMXL proteins, although the sequence in which these interactions occur is not known. It is not known if BRC1 is a direct or indirect target of this SL signaling mechanism. ASK, CUL1, RBX, and E2 are components of the ubiquitination complex. U, Ubiquitin; TFs, transcription factors (unidentified)....It is also unknown if TPR2 remains associated with SMXL proteins when they associated with D14 and MAX2.'\n",
      "tZ synthesis; 'The reduced recovery of radiolabeled nucleotides ZRMP and iPRMP in sADK indicated that CK ribosides are indeed substrates for ADK in planta and that ADK has an impact in CK riboside phosphorylation.'\n",
      "iP synthesis; 'Feeding studies using chloronemal tissues of Physcomitrella revealed that exogenous iPR is converted into iPR monophosphate (iPRMP) via an ADK-dependent pathway (von Schwartzenberg et al., 1998, 2003)....The reduced recovery of radiolabeled nucleotides ZRMP and iPRMP in sADK indicated that CK ribosides are indeed substrates for ADK in planta and that ADK has an impact in CK riboside phosphorylation.'\n",
      "Isopentenyladenosine-5'-triphosphate + H2O <=> Isopentenyladenosine-5'-diphosphate + Orthophosphate; https://www.genome.jp/entry/R08061;\n",
      "Isopentenyladenosine-5'-diphosphate + H2O <=> N6-(delta2-Isopentenyl)-adenosine 5'-monophosphate + Orthophosphate; https://www.genome.jp/entry/R08062;\n",
      "Histological analysis of root vasculature stained with toluidine blue (top) and fuchsin red (bottom, where provided) indicates that protoxylem is more abundant in genetic backgrounds with reduced cytokinin signaling, and ahp6 is able to suppress wol and cre1 ahk3....We also demonstrated that ahp6-1 could not suppress the cytokinin-insensitive cre1 ahk2 ahk3 phenotype, indicating that the suppressor phenotype in the wol background is due to residual cytokinin signaling'\n",
      "The auxin response negatively controls the cytokinin response by the activation of ARR-A genes. The cytokinin-mediated phosphorylation of ARR-B transcription factors enhances Aux/IAA transcription, and so balances the auxin response. '\n",
      "\n",
      "ModelV \n",
      "-------------\n",
      "\n",
      "origin \n",
      "-------------\n",
      "\n",
      "trust_level \n",
      "-------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bad_cols = []\n",
    "for c in df_edges.columns:\n",
    "    print(c, \"\\n-------------\")\n",
    "    if any(df_edges[c].apply(find_quotes)):\n",
    "        bad_cols.append(c)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AdditionalInfo \n",
      "-------------\n",
      "'Isopentenyladenosine-5'-triphosphate + H2O <=> Isopentenyladenosine-5'-diphosphate + Orthophosphate; https://www.genome.jp/entry/R08061;' : 'Isopentenyladenosine-5&prime;-triphosphate + H2O <=> Isopentenyladenosine-5&prime;-diphosphate + Orthophosphate; https://www.genome.jp/entry/R08061;'\n",
      "'Isopentenyladenosine-5'-diphosphate + H2O <=> N6-(delta2-Isopentenyl)-adenosine 5'-monophosphate + Orthophosphate; https://www.genome.jp/entry/R08062;' : 'Isopentenyladenosine-5&prime;-diphosphate + H2O <=> N6-(delta2-Isopentenyl)-adenosine 5&prime;-monophosphate + Orthophosphate; https://www.genome.jp/entry/R08062;'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for c in bad_cols:\n",
    "    print(c, \"\\n-------------\")\n",
    "    df_edges[c] = df_edges[c].apply(quote_replacer, verbose=True)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AdditionalInfo \n",
      "-------------\n",
      "actually binding - acts as inhibition as it outcompetes bdingin to promotor, can be extended to include binding to HISTONE DEACETYLASE 6 (HDA6) 'interacts with JAZs and EIN3/EIL1 as a co-repressor'\n",
      "RTM proteins block the long distance transport of plant viruses. They don't directly interact with CP, but the process of RTM resistance involves at least 5 different proteins.\n",
      "RTM proteins block the long distance transport of plant viruses. They don't directly interact with CP, but the process of RTM resistance involves at least 5 different proteins.\n",
      "...Upon pathogen challenge, the defense repressor AtSR1 is degradedthrough the SR1IP1-CUL3-mediated ubiquitin/proteasome pathway torelieve the repression of EDS1 transcription exerted by Ca2+/CaM signaling...These results indicatethat SR1IP1-CUL3 E3/proteasome-mediated degradation of AtSR1 is a critical mechanism to relieve the suppression of immune responses exerted by Ca2+/CaM/AtSR1. '|'...The negative regulation of immune response by Ca2+-CaM-AtSR1 complex is proposed to be released through AtSR1-Interacting-Protein-1 (SR1-1P1) that binds to AtSR1 and facilitates ubiquitination and degradation of the latter upon pathogen challenge (Zhang et al., 2014)...'\n",
      "HC-Pro from PVY with Y2H, BiFC. Arabidopsis 20S proteasome subunits. Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "Elena's gathered literature and PPIs (Y2H).\n",
      "The canonical GA signaling model illustrating GA-dependent GID1-DELLA complex formation resulting in DELLA recognition and ubiquitylation by the SCFSLY1 E3. Polyubiquitylation leads to DELLA proteolysis by the 26S proteasome, thereby lifting DELLA repression of GA responses.'\n",
      "...phosphorylated SLR1 is caught by the SCFGID2 complex through an interacting affinity between GID2 and phosphorylated SLR1, triggering the ubiquitin-mediated degradation of SLR1.'\n",
      "D53 binds to the complex S-D14-MAX2 to be degraded by ubiquitination;  'A model of the SL signaling complex in Arabidopsis that includes SL-dependent interaction of Arabidopsis D14 with both MAX2 and SMXL proteins, although the sequence in which these interactions occur is not known. It is not known if BRC1 is a direct or indirect target of this SL signaling mechanism. ASK, CUL1, RBX, and E2 are components of the ubiquitination complex. U, Ubiquitin; TFs, transcription factors (unidentified)....It is also unknown if TPR2 remains associated with SMXL proteins when they associated with D14 and MAX2.'\n",
      "tZ synthesis; 'The reduced recovery of radiolabeled nucleotides ZRMP and iPRMP in sADK indicated that CK ribosides are indeed substrates for ADK in planta and that ADK has an impact in CK riboside phosphorylation.'\n",
      "iP synthesis; 'Feeding studies using chloronemal tissues of Physcomitrella revealed that exogenous iPR is converted into iPR monophosphate (iPRMP) via an ADK-dependent pathway (von Schwartzenberg et al., 1998, 2003)....The reduced recovery of radiolabeled nucleotides ZRMP and iPRMP in sADK indicated that CK ribosides are indeed substrates for ADK in planta and that ADK has an impact in CK riboside phosphorylation.'\n",
      "Histological analysis of root vasculature stained with toluidine blue (top) and fuchsin red (bottom, where provided) indicates that protoxylem is more abundant in genetic backgrounds with reduced cytokinin signaling, and ahp6 is able to suppress wol and cre1 ahk3....We also demonstrated that ahp6-1 could not suppress the cytokinin-insensitive cre1 ahk2 ahk3 phenotype, indicating that the suppressor phenotype in the wol background is due to residual cytokinin signaling'\n",
      "The auxin response negatively controls the cytokinin response by the activation of ARR-A genes. The cytokinin-mediated phosphorylation of ARR-B transcription factors enhances Aux/IAA transcription, and so balances the auxin response. '\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for c in bad_cols:\n",
    "    print(c, \"\\n-------------\")\n",
    "    if any(df_edges[c].apply(find_quotes)):\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges[\"Literature\"].fillna(value=\"\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format literature sources\n",
    "\n",
    "re_ec = \"ec(?:\\:|\\s)?(\\d+(?:\\.(?:\\-|\\d+)){1,3}(?:\\.n\\d+)?)(?:\\s|$|\\]|,|\\.)\"\n",
    "\n",
    "\n",
    "def doi_list(x):\n",
    "    x = x.lower()\n",
    "    match = re.findall(\"(?:doi)(?:\\:|\\/)\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"doi:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "def pubmed_list(x):\n",
    "    x =x.lower()\n",
    "    match = re.findall(\"(?:pmid)\\:\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"pmid:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "\n",
    "def format_literature(row):\n",
    "    issued = False\n",
    "    dbs_list = []\n",
    "    \n",
    "    x = row['Literature'].lower().strip()\n",
    "    doi_match = re.findall(\"(?:doi)(?:\\:|\\/)\\s*(.+?)(?:\\s|$)\", x)\n",
    "    dbs_list += [f\"doi:{idf.strip().rstrip('.')}\" for idf in doi_match]    \n",
    "\n",
    "    pubmed_match = re.findall(\"pmid(?:\\:|)\\s*(\\d+)\", x)\n",
    "    dbs_list += [f\"pmid:{idf}\" for idf in pubmed_match]\n",
    "    \n",
    "    pubmedc_match =  re.findall(\"pmcid(?:\\:|)\\s*(pmc\\d+)\", x)\n",
    "    dbs_list += [f\"pmcid:{idf}\" for idf in pubmedc_match]\n",
    "                 \n",
    "    for key in x.split(\"|\"):\n",
    "        if key ==\"\":\n",
    "            print(f\"BLANK\\\\{row['ConnID']}\")\n",
    "            issued = True\n",
    "\n",
    "        elif \":\" in key:\n",
    "            if \"aracyc\" in key:\n",
    "                aracyc_string = \"aracyc:\" + key.split(\":\")[1].strip()\n",
    "                dbs_list.append(aracyc_string)\n",
    "            elif \"kegg\" in key:\n",
    "                kegg_string = \"kegg:\" + key.split(\":\")[1].strip()\n",
    "                dbs_list.append(kegg_string)\n",
    "            elif \"doi\" in key:\n",
    "                # already fetched\n",
    "                continue\n",
    "            elif (\"pmcid\" in key) or (\"pmid\" in key):\n",
    "                continue\n",
    "            else:\n",
    "                print(f\"UNKOWN DB\\\\{row['ConnID']}\\\\{key}\")                 \n",
    "                issued = True\n",
    "        elif \"invented\" in key:\n",
    "            dbs_list.append(\"invented:reason\")\n",
    "        else:\n",
    "            print(f\"NOVALUE\\\\{row['ConnID']}\\\\{key}\")\n",
    "            issued = True\n",
    "\n",
    "    if (len(dbs_list)==0) and not issued:\n",
    "        print(f\"BADorMISSING\\\\{row['ConnID']}\\\\{x}\")\n",
    "\n",
    "    ################\n",
    "    x = row['AdditionalInfo']\n",
    "    x = x.lower()\n",
    "    \n",
    "    # EC:3.3.3.-\n",
    "    ec_match = re.findall(re_ec, x)\n",
    "    dbs_list += [f\"ec:{idf}\" for idf in ec_match]\n",
    "    \n",
    "    pubmed_match = re.findall(\"pmid(?:\\:|)\\s*(\\d+)\", x)\n",
    "    dbs_list += [f\"pmid:{idf}\" for idf in pubmed_match]\n",
    "    \n",
    "    pubmedc_match =  re.findall(\"pmcid(?:\\:|)\\s*(pmc\\d+)\", x)\n",
    "    dbs_list += [f\"pmcid:{idf}\" for idf in pubmedc_match]\n",
    "    \n",
    "    doi_match = re.findall(\"doi(?:\\:|\\/)\\s*(.+?)(?:\\s|$)\", x)\n",
    "    dbs_list += [f\"doi:{idf.strip().rstrip('.')}\" for idf in doi_match]    \n",
    "    \n",
    "    kegg_match = re.findall(r\"((?:k|map|ko|ec|rn)\\d{5})\", x)\n",
    "    dbs_list += [f\"kegg:{idf}\" for idf in kegg_match]              \n",
    "    \n",
    "    ncbi_nuccore_match = re.findall(\"NCBI ID: (.+)\", x)\n",
    "    dbs_list += [f\"ncbi_nuccore:{idf}\" for idf in ncbi_nuccore_match]              \n",
    "         \n",
    "    return ','.join(list(set(dbs_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLANK\\Conn039\n",
      "BLANK\\Conn115\n",
      "NOVALUE\\Conn116\\embo j., 20 (2001), pp. 5400-5411 direct interaction between the arabidopsis disease resistance signaling proteins, eds1 and pad4 \n",
      "BLANK\\Conn117\n",
      "NOVALUE\\Conn118\\embo j., 20 (2001), pp. 5400-5411 direct interaction between the arabidopsis disease resistance signaling proteins, eds1 and pad4 \n",
      "BLANK\\Conn119\n",
      "NOVALUE\\Conn120\\embo j., 20 (2001), pp. 5400-5411 direct interaction between the arabidopsis disease resistance signaling proteins, eds1 and pad4 \n",
      "BLANK\\Conn178\n",
      "BLANK\\Conn179\n",
      "NOVALUE\\Conn194\\? kg need to find reference\n",
      "NOVALUE\\Conn267\\ 10.1073/pnas.98.4.2065\n",
      "BLANK\\Conn309\n"
     ]
    }
   ],
   "source": [
    "df_edges['external_links'] = df_edges.apply(format_literature, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ConnID</th>\n",
       "      <th>origin</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Conn039</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>Conn115</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td></td>\n",
       "      <td>This is viral PTI.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Conn117</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>Conn119</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td></td>\n",
       "      <td>RNA silencing inhibits virus, part of PTI.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Conn178</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td></td>\n",
       "      <td>have to check which one, not sure that this mp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>Conn179</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td></td>\n",
       "      <td>have to check which one, not sure that this mp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>Conn194</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>? Kg need to find reference</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>Conn309</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ConnID                            origin                   Literature  \\\n",
       "38   Conn039  forCB-v2.7.9_PIS-model-Reactions                                \n",
       "114  Conn115  forCB-v2.7.9_PIS-model-Reactions                                \n",
       "116  Conn117  forCB-v2.7.9_PIS-model-Reactions                                \n",
       "118  Conn119  forCB-v2.7.9_PIS-model-Reactions                                \n",
       "177  Conn178  forCB-v2.7.9_PIS-model-Reactions                                \n",
       "178  Conn179  forCB-v2.7.9_PIS-model-Reactions                                \n",
       "193  Conn194  forCB-v2.7.9_PIS-model-Reactions  ? Kg need to find reference   \n",
       "308  Conn309  forCB-v2.7.9_PIS-model-Reactions                                \n",
       "\n",
       "                                        AdditionalInfo  \n",
       "38                                                      \n",
       "114                                 This is viral PTI.  \n",
       "116                                                     \n",
       "118         RNA silencing inhibits virus, part of PTI.  \n",
       "177  have to check which one, not sure that this mp...  \n",
       "178  have to check which one, not sure that this mp...  \n",
       "193                                                     \n",
       "308                                                     "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[df_edges['external_links']==''][['ConnID', 'origin', 'Literature', 'AdditionalInfo']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges[[\"ConnID\",  \"external_links\", \"Literature\", \"AdditionalInfo\"]].to_csv(output_path / \"reactions-lit-check.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_df = df_edges.copy()\n",
    "#df_edges = save_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>...</th>\n",
       "      <th>output1_level</th>\n",
       "      <th>output1_localisation</th>\n",
       "      <th>output1_type</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>origin</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>external_links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>aracyc:ethyl-pwy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>aracyc:ethyl-pwy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>aracyc:ethyl-pwy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein [active]</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Status AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0  forCB      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1  forCB      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2  forCB      KG  Conn003     ath       ACC       family                  ER   \n",
       "3  forCB      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4  forCB      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level  ... output1_level output1_localisation  \\\n",
       "0  metabolite      SAMS        clade  ...        family                   ER   \n",
       "1  metabolite       ACS       family  ...        family                   ER   \n",
       "2  metabolite       ACO       family  ...        family                   ER   \n",
       "3  metabolite       HMA       family  ...        family                   ER   \n",
       "4     protein      Cu2+       family  ...        family                   ER   \n",
       "\n",
       "       output1_type                                   TrustLevel  \\\n",
       "0        metabolite  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1        metabolite  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2        metabolite  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3        metabolite  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  protein [active]  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo ModelV  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.   v1.0   \n",
       "1                 ACS catalyse ACC to SAMe reaction.   v1.0   \n",
       "2                   ACO catalyse ACC to ET reaction.   v1.0   \n",
       "3  Copper gets transported from the cytoplasm to ...   v1.0   \n",
       "4  Copper activates the membrane bound ethylene r...   v1.0   \n",
       "\n",
       "                             origin trust_level          external_links  \n",
       "0  forCB-v2.7.9_PIS-model-Reactions          R1        aracyc:ethyl-pwy  \n",
       "1  forCB-v2.7.9_PIS-model-Reactions          R1        aracyc:ethyl-pwy  \n",
       "2  forCB-v2.7.9_PIS-model-Reactions          R1        aracyc:ethyl-pwy  \n",
       "3  forCB-v2.7.9_PIS-model-Reactions          R1  doi:10.1105/tpc.001768  \n",
       "4  forCB-v2.7.9_PIS-model-Reactions          R1  doi:10.1105/tpc.001768  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in ['input1', 'input2', 'input3', 'output1']:\n",
    "    df_edges.loc[:, x + \"_ID\"] = df_edges[x + \"_ID\"].apply(helpers.reorder_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "ARR-A(a)(p) protein --> protein_active\n",
      "ARR-B(a)(p) protein --> protein_active\n",
      "ARR-B(a)(p) protein --> protein_active\n",
      "input2\n",
      "AHK2,3,4(a) protein [active] --> protein_active\n",
      "AHK2,3,4(a) protein [active] --> protein_active\n",
      "ARR-B(a)(p) protein [active] --> protein_active\n",
      "AHP1,2,3,4,5(a) protein [active] --> protein_active\n",
      "AHP1,2,3,4,5(a) protein [active] --> protein_active\n",
      "input3\n",
      "output1\n",
      "AHK2,3,4(a) protein [active] --> protein_active\n",
      "AHK2,3,4(a) protein [active] --> protein_active\n",
      "AHK2,3,4(a) protein [active] --> protein_active\n",
      "AHK2,3,4(a) protein [active] --> protein_active\n",
      "AHP1,2,3,4,5(a) protein [active] --> protein_active\n",
      "ARR-A(a)(p) protein --> protein_active\n",
      "ARR-B(a)(p) protein --> protein_active\n"
     ]
    }
   ],
   "source": [
    "node_type_to_node_form_dict = {\n",
    "    \"gene\":\"gene\",\n",
    "    \n",
    "    \"protein\":\"protein\",\n",
    "    \"protein [inactivated]\":\"protein\", \n",
    "    \"protein [activated]\":\"protein_active\",\n",
    "    'protein [active]': \"protein_active\",\n",
    "    'protein_active': 'protein_active',\n",
    "    \n",
    "    \"ncRNA\":\"ncRNA\",\n",
    "    \"plant_ncRNA\":\"ncRNA\",\n",
    "    'ta-siRNA':\"ta-siRNA\", \n",
    "        \n",
    "    \"complex\":\"complex\", \n",
    "    \"plant_complex\":\"complex\",\n",
    "    'complex [active]': \"complex_active\",\n",
    "    'complex_active': \"complex_active\",\n",
    "    \n",
    "    \"metabolite\":\"metabolite\",\n",
    "    \n",
    "    \"process\":\"process\", \n",
    "    \"process_active\":\"process_active\", \n",
    "    'process [active]':\"process_active\",\n",
    "\n",
    "    np.nan:\"\", \n",
    "}\n",
    "\n",
    "node_type_to_active_form_dict = {\n",
    "    \"protein [active]\":\"protein_active\",\n",
    "    \"protein\":\"protein_active\",\n",
    "    \"protein_active\":\"protein_active\",\n",
    "    \n",
    "    \"complex\":\"complex_active\", \n",
    "    \"complex_active\": \"complex_active\",\n",
    "    \n",
    "    \"process\":\"process_active\", \n",
    "    'process_active':\"process_active\",\n",
    "}\n",
    "\n",
    "node_type_to_inactive_form_dict = {\n",
    "    \"protein\":\"protein\",\n",
    "    \"protein_active\":\"protein\",\n",
    "    \"protein [inactivated]\":\"protein\",\n",
    "    \n",
    "    \"complex\":\"complex\", \n",
    "    \"complex_active\": \"complex\",\n",
    "    \n",
    "    \"process\":\"process\", \n",
    "    'process_active':\"process\",\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "def get_node_form(row, prefix=\"input1\"):\n",
    "    id_ = row.iloc[0]\n",
    "    if type(id_) == float:\n",
    "        return np.nan\n",
    "    \n",
    "    type_ = row.iloc[1]\n",
    "    form = node_type_to_node_form_dict[type_]\n",
    "    if (  len(re.findall(\"\\(a\\)\", id_)) > 0  ) or (  len(re.findall(\"\\[active\\]\", id_)) > 0  ):\n",
    "        form = node_type_to_active_form_dict[type_]\n",
    "        print(id_, type_, \"-->\", form)\n",
    "        \n",
    "    return form\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, new_form_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_form')]\n",
    "    \n",
    "    #x = df_edges[[id_col, type_col]].dropna(how='all')\n",
    "    \n",
    "    print(prefix)\n",
    "    df_edges[new_form_col] = df_edges[[id_col, type_col]].apply(get_node_form, prefix=prefix, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_brackets(x):\n",
    "    # reorder ids for complexes\n",
    "    if type(x) == np.float:\n",
    "        return np.nan\n",
    "    else:\n",
    "        s = re.sub(\"\\(a\\)\", \"\", x)\n",
    "        s = re.sub(\"\\(p\\)\", \"\", s)\n",
    "        if s != x:\n",
    "            print(x, s)\n",
    "        return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARR-A(a)(p) ARR-A\n",
      "ARR-B(a)(p) ARR-B\n",
      "ARR-B(a)(p) ARR-B\n",
      "AHK2,3,4(a) AHK2,3,4\n",
      "AHK2,3,4(a) AHK2,3,4\n",
      "ARR-B(a)(p) ARR-B\n",
      "AHP1,2,3,4,5(a) AHP1,2,3,4,5\n",
      "AHP1,2,3,4,5(a) AHP1,2,3,4,5\n",
      "AHK2,3,4(a) AHK2,3,4\n",
      "AHK2,3,4(a) AHK2,3,4\n",
      "AHK2,3,4(a) AHK2,3,4\n",
      "AHK2,3,4(a) AHK2,3,4\n",
      "AHP1,2,3,4,5(a) AHP1,2,3,4,5\n",
      "ARR-A(a)(p) ARR-A\n",
      "ARR-B(a)(p) ARR-B\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-38-6a8595da6a17>:3: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  if type(x) == np.float:\n"
     ]
    }
   ],
   "source": [
    "for x in ['input1', 'input2', 'input3', 'output1']:\n",
    "    df_edges.loc[:, x + \"_ID\"] = df_edges[x + \"_ID\"].apply(remove_brackets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['species'] = df_edges['Species'].apply(lambda x: \",\".join(x.lower().split('/')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ath', 'ath,nbe', 'ath,stu', 'all', 'ath,nta', 'stu', 'osa',\n",
       "       'ath,osa', 'ath,osa,psa', 'ath,osa,sly,zma', 'ath,osa,phy'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['species'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_localisation_dict = {\n",
    "    'nuc':'nucleus',\n",
    "    'er':'endoplasmic reticulum',\n",
    "    'golgi':'golgi apparatus', \n",
    "    'mitochondria?': 'putative:mitochondrion', \n",
    "    'cytoplasm?': 'putative:cytoplasm', \n",
    "}\n",
    "\n",
    "\n",
    "good_localisations = set([\n",
    "     'nucleus',\n",
    "     'nucleolus',\n",
    "     'cytoplasm',\n",
    "     'vacuole',\n",
    "     'endoplasmic reticulum',\n",
    "     'chloroplast',\n",
    "     'mitochondrion',\n",
    "     'golgi apparatus',\n",
    "     'peroxisome',\n",
    "     'apoplast',\n",
    "     'extracellular'\n",
    "])\n",
    "\n",
    "good_localisations.update(['putative:' + s for s in good_localisations])\n",
    "\n",
    "\n",
    "def node_localisation_std(x):\n",
    "    if not type(x) == str:\n",
    "        return \"putative:cytoplasm\"\n",
    "    \n",
    "    x = x.lower()\n",
    "    if x in node_localisation_dict.keys():\n",
    "        x = node_localisation_dict[x]\n",
    "    \n",
    "    if x in good_localisations:\n",
    "        return x\n",
    "    else:\n",
    "        print(x)\n",
    "        return \"\"\n",
    "\n",
    "node_localisations = set()\n",
    "new_localisation = set()\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, localisation_col, location_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_localisation', '_location')]\n",
    "    \n",
    "    x = df_edges[['ConnID', 'origin', id_col, localisation_col]].dropna(how='all', subset=[ id_col, localisation_col])\n",
    "    for _, y in x.iterrows():\n",
    "        if (not (y[id_col] in helpers.empty_strings)) and (y[localisation_col] in helpers.empty_strings + ['']):\n",
    "            print(y['origin'], \"|\", y['ConnID'], \"|\", prefix, \"|\", y[id_col], \"|\", y[localisation_col])\n",
    "    \n",
    "    \n",
    "    node_localisations.update(x[localisation_col])\n",
    "    \n",
    "    #print(prefix)\n",
    "    df_edges[location_col] = df_edges[localisation_col].apply(node_localisation_std)\n",
    "    \n",
    "    new_localisation.update(df_edges[location_col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ER',\n",
       " 'Golgi',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'cytoplasm?',\n",
       " 'extracellular',\n",
       " 'mitochondria?',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_localisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'endoplasmic reticulum',\n",
       " 'extracellular',\n",
       " 'golgi apparatus',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'putative:cytoplasm',\n",
       " 'putative:mitochondrion',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_localisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'apoplast',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'endoplasmic reticulum',\n",
       " 'extracellular',\n",
       " 'golgi apparatus',\n",
       " 'mitochondrion',\n",
       " 'nucleolus',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'putative:apoplast',\n",
       " 'putative:chloroplast',\n",
       " 'putative:cytoplasm',\n",
       " 'putative:endoplasmic reticulum',\n",
       " 'putative:extracellular',\n",
       " 'putative:golgi apparatus',\n",
       " 'putative:mitochondrion',\n",
       " 'putative:nucleolus',\n",
       " 'putative:nucleus',\n",
       " 'putative:peroxisome',\n",
       " 'putative:vacuole',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "good_localisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reaction_mode_dict = {\n",
      "\t'catalysis / auto-catalysis':'catalysis/auto-catalysis',\n",
      "\t'translocation':'translocation',\n",
      "\t'protein activation':'protein activation',\n",
      "\t'binding / oligomerisation':'binding/oligomerisation',\n",
      "\t'protein deactivation':'protein deactivation',\n",
      "\t'transcriptional / translational repression':'transcriptional/translational repression',\n",
      "\t'transcriptional / translational induction':'transcriptional/translational induction',\n",
      "\t'degradation / secretion':'degradation/secretion',\n",
      "\t'nan':'nan',\n",
      "\t'dissociation':'dissociation',\n",
      "\t'cleavage / auto-cleavage':'cleavage/auto-cleavage',\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print('reaction_mode_dict = {')\n",
    "for s in df_edges['ReactionMode'].unique():\n",
    "    print(f\"\\t'{s}':'{'/'.join([x.lower().strip() for x in str(s).split('/')])}',\")\n",
    "print('}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Old version before edit\n",
    "# reaction_mode_dict = {\n",
    "# \t'catalysis / auto-catalysis':'catalysis/auto-catalysis',\n",
    "# \t'translocation':'translocation',\n",
    "# \t'protein activation':'protein activation',\n",
    "# \t'binding / oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein deactivation':'protein deactivation',\n",
    "# \t'transcription/translation repression':'transcription/translation repression',\n",
    "# \t'transcription/translation induction':'transcription/translation induction',\n",
    "# \t'degradation / secretion':'degradation/secretion',\n",
    "# \t'dissociation':'dissociation',\n",
    "# \t'translation':'translation',\n",
    "# \t'transcription':'transcription',\n",
    "# \t'by binding':'by binding',\n",
    "# \t'nan':'nan',\n",
    "# \t'cleavage / auto-cleavage':'cleavage/auto-cleavage',\n",
    "# \t'binding/oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein phosphorylation':'protein phosphorylation',\n",
    "# }\n",
    "\n",
    "# Old version after edit\n",
    "# reaction_mode_dict = {\n",
    "# \t'catalysis / auto-catalysis':'catalysis/auto-catalysis',\n",
    "# \t'translocation':'translocation',\n",
    "# \t'protein activation':'protein activation',\n",
    "# \t'binding / oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein deactivation':'protein deactivation',\n",
    "# \t'transcription/translation repression':'transcription/translation repression',\n",
    "# \t'transcription/translation induction':'transcription/translation induction',\n",
    "# \t'degradation / secretion':'degradation/secretion',\n",
    "# \t'dissociation':'dissociation',\n",
    "# \t'translation':'bad',\n",
    "# \t'transcription':'bad',\n",
    "# \t'by binding':'bad',\n",
    "# \tnp.nan:'bad',\n",
    "# \t'cleavage / auto-cleavage':'bad',\n",
    "# \t'binding/oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein phosphorylation':'bad',\n",
    "# }\n",
    "\n",
    "reaction_mode_dict = {\n",
    "\t'catalysis / auto-catalysis':'catalysis',\n",
    "\t'translocation':'translocation',\n",
    "\t'protein activation':'protein activation',\n",
    "\t'binding / oligomerisation':'binding/oligomerisation',\n",
    "\t'protein deactivation':'protein deactivation',\n",
    "\t'transcriptional / translational repression':'transcriptional/translational repression',\n",
    "\t'transcriptional / translational induction':'transcriptional/translational activation',\n",
    "\t'degradation / secretion':'degradation/secretion',\n",
    "\tnp.nan:'undefined',\n",
    "\t'dissociation':'dissociation',\n",
    "\t'cleavage / auto-cleavage':'cleavage/auto-cleavage',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['reaction_type'] = df_edges['ReactionMode'].apply(lambda x: reaction_mode_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>ReactionMode</th>\n",
       "      <th>ReactionEffect</th>\n",
       "      <th>reaction_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>activation</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn121</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>_TBD-v2.7.9_PIS-model-Reactions</td>\n",
       "      <td>Conn311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              origin   ConnID ReactionMode ReactionEffect  \\\n",
       "31   _TBD-v2.7.9_PIS-model-Reactions  Conn032          NaN     inhibition   \n",
       "60   _TBD-v2.7.9_PIS-model-Reactions  Conn061          NaN     inhibition   \n",
       "87   _TBD-v2.7.9_PIS-model-Reactions  Conn088          NaN     inhibition   \n",
       "88   _TBD-v2.7.9_PIS-model-Reactions  Conn089          NaN     inhibition   \n",
       "91   _TBD-v2.7.9_PIS-model-Reactions  Conn092          NaN     activation   \n",
       "120  _TBD-v2.7.9_PIS-model-Reactions  Conn121          NaN     inhibition   \n",
       "150  _TBD-v2.7.9_PIS-model-Reactions  Conn151          NaN     inhibition   \n",
       "209  _TBD-v2.7.9_PIS-model-Reactions  Conn210          NaN     inhibition   \n",
       "310  _TBD-v2.7.9_PIS-model-Reactions  Conn311          NaN     inhibition   \n",
       "\n",
       "    reaction_type  \n",
       "31      undefined  \n",
       "60      undefined  \n",
       "87      undefined  \n",
       "88      undefined  \n",
       "91      undefined  \n",
       "120     undefined  \n",
       "150     undefined  \n",
       "209     undefined  \n",
       "310     undefined  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[df_edges['reaction_type']=='undefined'][['origin', 'ConnID', 'ReactionMode', 'ReactionEffect', 'reaction_type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.to_csv(\"parsed_reactions.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# END"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
