{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import neo4j DB: 3/4\n",
    "\n",
    "Code to translate v2.7.4_PIS-model.xlsx to neo4j database. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import os\n",
    "from IPython.display import Image\n",
    "\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "base_path = Path(\"..\")\n",
    "input_path = base_path / \"data\" / \"raw\"\n",
    "output_path = base_path / \"data\" / \"parsed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to graph via docker-compose link. See http://localhost:7474/browser/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from py2neo import Graph, Node, Relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = Graph(host=\"neo4j\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in Bio elements\n",
    "bioelements = PlantCoding/PlantNonCoding/PlantAbstract\n",
    "\n",
    "have family hierachy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>species</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>external_links</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Process</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>gmm_ocd</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>node0278</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>Benzoic acid 2-hydroxylase: Isolated from toba...</td>\n",
       "      <td>Hormone:SA</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BA2H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>node0255</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Arabidopsis contains two ICS genes but has no ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:SA</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IPL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>node0695</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>invented:unidentified</td>\n",
       "      <td>green peach aphid 2; resistance against PVX (stu)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:Rgenes</td>\n",
       "      <td>v2.6</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>node0692</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>HRT</td>\n",
       "      <td>HRT</td>\n",
       "      <td>HRT</td>\n",
       "      <td>invented:unidentified</td>\n",
       "      <td>HR against TCV (ath)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:Rgenes</td>\n",
       "      <td>v2.5</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>node0691</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>Ny</td>\n",
       "      <td>Ny</td>\n",
       "      <td>Ny</td>\n",
       "      <td>invented:unidentified</td>\n",
       "      <td>HR against PVY (stu)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:Rgenes</td>\n",
       "      <td>v2.5</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R-gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier AddedBy species      NodeLabel        NodeType  Family     Clade  \\\n",
       "0   node0278      KG     ath  PlantAbstract  plant_abstract    BA2H      BA2H   \n",
       "1   node0255      KG     ath  PlantAbstract  plant_abstract     IPL       IPL   \n",
       "2   node0695      KG     ath  PlantAbstract  plant_abstract  R-gene  GPAphid2   \n",
       "3   node0692      KG     ath  PlantAbstract  plant_abstract  R-gene       HRT   \n",
       "4   node0691      KG     ath  PlantAbstract  plant_abstract  R-gene        Ny   \n",
       "\n",
       "     NodeID  NodeName         external_links  \\\n",
       "0      BA2H      BA2H                    NaN   \n",
       "1       IPL       IPL                    NaN   \n",
       "2  GPAphid2  GPAphid2  invented:unidentified   \n",
       "3       HRT       HRT  invented:unidentified   \n",
       "4        Ny        Ny  invented:unidentified   \n",
       "\n",
       "                                     NodeDescription  \\\n",
       "0                                               BA2H   \n",
       "1  Arabidopsis contains two ICS genes but has no ...   \n",
       "2  green peach aphid 2; resistance against PVX (stu)   \n",
       "3                               HR against TCV (ath)   \n",
       "4                               HR against PVY (stu)   \n",
       "\n",
       "                                      AdditionalInfo     Process ModelV  \\\n",
       "0  Benzoic acid 2-hydroxylase: Isolated from toba...  Hormone:SA   v1.0   \n",
       "1                                                NaN  Hormone:SA   v1.0   \n",
       "2                                                NaN    S:Rgenes   v2.6   \n",
       "3                                                NaN    S:Rgenes   v2.5   \n",
       "4                                                NaN    S:Rgenes   v2.5   \n",
       "\n",
       "  ModelStatus gmm_ocd GMM_Description GMM_ShortName synonyms  family  \n",
       "0      ignore     NaN             NaN           NaN      NaN    BA2H  \n",
       "1      ignore     NaN             NaN           NaN      NaN     IPL  \n",
       "2      ignore     NaN             NaN           NaN      NaN  R-gene  \n",
       "3      ignore     NaN             NaN           NaN      NaN  R-gene  \n",
       "4      ignore     NaN             NaN           NaN      NaN  R-gene  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_name = output_path / \"bio_elements.tsv\"\n",
    "df_bioelements = pd.read_csv(file_name, sep=\"\\t\")\n",
    "df_bioelements.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_bioelements[df_bioelements['NodeName'].duplicated(keep=False)].sort_values('NodeName')[['identifier', 'species', 'Family', 'Clade', 'NodeID', 'NodeName']]\n",
    "repeated_node_level = df_bioelements[df_bioelements['NodeName'].duplicated(keep='first')]['NodeName'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th>level</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;alpha;/&amp;beta; hydroxylase</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;beta;-carotene isomerase</th>\n",
       "      <th>clade</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4CLL</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT1G20490,AT1G20510,AT4G05160,AT1G20500,AT1G20...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4CLL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                      ath_homologues  \\\n",
       "name                       level                                                       \n",
       "&alpha;/&beta; hydroxylase clade                                           AT3G03990   \n",
       "                           family                                          AT3G03990   \n",
       "&beta;-carotene isomerase  clade                                                       \n",
       "                           family                                                      \n",
       "4CLL                       clade   AT1G20490,AT1G20510,AT4G05160,AT1G20500,AT1G20...   \n",
       "\n",
       "                                  stu_homologues sly_homologues  \\\n",
       "name                       level                                  \n",
       "&alpha;/&beta; hydroxylase clade                                  \n",
       "                           family                                 \n",
       "&beta;-carotene isomerase  clade                                  \n",
       "                           family                                 \n",
       "4CLL                       clade                                  \n",
       "\n",
       "                                  osa_homologues                      family  \n",
       "name                       level                                              \n",
       "&alpha;/&beta; hydroxylase clade                  &alpha;/&beta; hydroxylase  \n",
       "                           family                 &alpha;/&beta; hydroxylase  \n",
       "&beta;-carotene isomerase  clade    OS11G0587000   &beta;-carotene isomerase  \n",
       "                           family   OS11G0587000   &beta;-carotene isomerase  \n",
       "4CLL                       clade                                        4CLL  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_name = output_path / \"level_translation.tsv\"\n",
    "translate_df = pd.read_csv(file_name, sep=\"\\t\", index_col=[0, 1])\n",
    "translate_df.fillna('', inplace=True)\n",
    "translate_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in parsed reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ath', 'stu', 'sly', 'osa']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species = list(df_bioelements['species'].unique())\n",
    "try: all_species.remove('plant_all')\n",
    "except ValueError: pass\n",
    "try: all_species.remove('all')\n",
    "except ValueError: pass    \n",
    "all_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>...</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>origin</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>external_links</th>\n",
       "      <th>input1_form</th>\n",
       "      <th>input2_form</th>\n",
       "      <th>input3_form</th>\n",
       "      <th>output1_form</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>...</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>aracyc:ethyl-pwy</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>aracyc:ethyl-pwy</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>aracyc:ethyl-pwy</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>forCB-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "      <td>protein</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>ath</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Status AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0  forCB      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1  forCB      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2  forCB      KG  Conn003     ath       ACC       family                  ER   \n",
       "3  forCB      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4  forCB      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level  ...  \\\n",
       "0  metabolite      SAMS        clade  ...   \n",
       "1  metabolite       ACS       family  ...   \n",
       "2  metabolite       ACO       family  ...   \n",
       "3  metabolite       HMA       family  ...   \n",
       "4     protein      Cu2+       family  ...   \n",
       "\n",
       "                                      AdditionalInfo ModelV  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.   v1.0   \n",
       "1                 ACS catalyse ACC to SAMe reaction.   v1.0   \n",
       "2                   ACO catalyse ACC to ET reaction.   v1.0   \n",
       "3  Copper gets transported from the cytoplasm to ...   v1.0   \n",
       "4  Copper activates the membrane bound ethylene r...   v1.0   \n",
       "\n",
       "                             origin trust_level          external_links  \\\n",
       "0  forCB-v2.7.8_PIS-model-Reactions          R1        aracyc:ethyl-pwy   \n",
       "1  forCB-v2.7.8_PIS-model-Reactions          R1        aracyc:ethyl-pwy   \n",
       "2  forCB-v2.7.8_PIS-model-Reactions          R1        aracyc:ethyl-pwy   \n",
       "3  forCB-v2.7.8_PIS-model-Reactions          R1  doi:10.1105/tpc.001768   \n",
       "4  forCB-v2.7.8_PIS-model-Reactions          R1  doi:10.1105/tpc.001768   \n",
       "\n",
       "  input1_form     input2_form input3_form    output1_form species  \n",
       "0  metabolite  protein_active         NaN      metabolite     ath  \n",
       "1  metabolite  protein_active         NaN      metabolite     ath  \n",
       "2  metabolite  protein_active         NaN      metabolite     ath  \n",
       "3  metabolite         protein         NaN      metabolite     ath  \n",
       "4     protein      metabolite         NaN  protein_active     ath  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges = pd.read_csv(\"parsed_reactions.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_edges['ConnID'].unique()) == df_edges.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    rx00001\n",
       "1    rx00002\n",
       "2    rx00003\n",
       "3    rx00004\n",
       "4    rx00005\n",
       "Name: reaction_id, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['reaction_id'] = df_edges.index.map(lambda x: f\"rx{x+1:05}\")\n",
    "df_edges['reaction_id'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FunctionalCluster 0\n",
      "PlantCoding 162\n",
      "PlantNonCoding 9\n",
      "PlantAbstract 7\n",
      "ForeignEntity 3\n",
      "ForeignCoding 14\n",
      "ForeignNonCoding 0\n",
      "ForeignAbstract 0\n",
      "Complex 3\n",
      "Process 6\n",
      "MetaboliteFamily 6\n",
      "Metabolite 110\n",
      "Reaction 0\n"
     ]
    }
   ],
   "source": [
    "node_dict = {}\n",
    "for label in helpers.node_labels:\n",
    "    if label == \"Metabolite\":\n",
    "        q = '''MATCH (n:%s) WHERE NOT n:MetaboliteFamily RETURN DISTINCT n.name'''%label\n",
    "    else:\n",
    "        q = '''MATCH (n:%s) RETURN DISTINCT n.name'''%label\n",
    "    s = set([d['n.name'] for d in graph.run(q).data()])\n",
    "    print(label, len(s))\n",
    "    node_dict[label] = s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_label(x):\n",
    "    # pathogen proteins are listed as proteins, so cannot use dict\n",
    "    id_, form_, level_, ConnID, origin = x.values\n",
    "      \n",
    "    if id_ in helpers.empty_strings:\n",
    "        return np.nan\n",
    "\n",
    "    new_label = None\n",
    "    \n",
    "    ########################\n",
    "    # Simple Cases\n",
    "    ########################\n",
    "    if form_ in ['complex', 'complex_active']:\n",
    "        if not (id_ in node_dict[\"Complex\"]):\n",
    "            complexes_to_add.append(id_)\n",
    "        new_label = 'Complex'\n",
    "    \n",
    "    elif form_ in [\"metabolite\"]:\n",
    "        for label in [\"Metabolite\", \"MetaboliteFamily\"]:\n",
    "            if id_ in node_dict[label]:\n",
    "                new_label = label\n",
    "                break\n",
    "        if not new_label:\n",
    "            missing_in_components.update([id_])        \n",
    "            print(origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_, \" |a (label) not a listed Metabolite/MetaboliteFamily\")\n",
    "\n",
    "    elif form_ in ['process', 'process_active']:\n",
    "        if (id_ in node_dict[\"Process\"]):\n",
    "            new_label = \"Process\"\n",
    "        else:\n",
    "            print(origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |b (label) process not a listed process\")\n",
    "    \n",
    "    else:\n",
    "        # somethng plant, something foreign\n",
    "        family_id = get_family_id(id_, level_)\n",
    "                \n",
    "        if (family_id):\n",
    "            id_labels = [] #looping just in case an id occurs mutiple times\n",
    "            for label in helpers.plant_node_labels:\n",
    "                if family_id in node_dict[label]:\n",
    "                    id_labels.append(label)\n",
    "\n",
    "            if len(id_labels) == 1:\n",
    "                new_label = id_labels[0]\n",
    "            elif len(id_labels) > 1:\n",
    "                print(origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |d (label) many labels fit\")\n",
    "                missing_in_components.update([id_])       \n",
    "\n",
    "            else:\n",
    "                print(origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |e (label) could not find label\")\n",
    "                missing_in_components.update([id_]) \n",
    "        else:\n",
    "            # check if foreign\n",
    "            for label in helpers.foreign_node_labels:\n",
    "                if id_ in node_dict[label]:\n",
    "                    new_label = label\n",
    "                    break\n",
    "        \n",
    "    if not new_label:\n",
    "        print(origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |c (family id) could not convert to label\")\n",
    "        \n",
    "    return new_label\n",
    "\n",
    "\n",
    "    \n",
    "def get_family_id(id_, level_):\n",
    "    family_id = None\n",
    "    if id_ in repeated_node_level:\n",
    "        print(id_, level_, 'node level has multiple identifiers')\n",
    "    if level_ == \"family\":    \n",
    "        try:\n",
    "            family_id = translate_df.loc[(id_, 'family')]['family']\n",
    "        except KeyError:\n",
    "            pass\n",
    "    elif level_ in [\"clade\", \"clade/orthologue\"]:\n",
    "        try:\n",
    "            family_id = translate_df.loc[(id_, 'clade')]['family']\n",
    "        except KeyError:\n",
    "            pass\n",
    "    elif level_ == \"node\":\n",
    "        try:\n",
    "            family_id = translate_df.loc[(id_, 'node')]['family']\n",
    "        except KeyError:\n",
    "            pass\n",
    "\n",
    "    return family_id\n",
    "\n",
    "def apply_get_family_id(x):\n",
    "    id_, level_, label_, ConnID, origin = x.values\n",
    "      \n",
    "    if id_ in helpers.empty_strings:\n",
    "        return np.nan\n",
    "    \n",
    "    family_id = np.nan\n",
    "    if label_ in helpers.plant_node_labels:\n",
    "        family_id = get_family_id(id_, level_)\n",
    "    \n",
    "    return family_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "PR1 family node level has multiple identifiers\n",
      "PR1 family node level has multiple identifiers\n",
      "input2\n",
      "PR1 family node level has multiple identifiers\n",
      "PR1 family node level has multiple identifiers\n",
      "CPS node node level has multiple identifiers\n",
      "CPS node node level has multiple identifiers\n",
      "GA20ox3 node node level has multiple identifiers\n",
      "GA20ox3 node node level has multiple identifiers\n",
      "GA20ox1 node node level has multiple identifiers\n",
      "GA20ox4 node node level has multiple identifiers\n",
      "PR1 clade node level has multiple identifiers\n",
      "PR1 family node level has multiple identifiers\n",
      "PR1 family node level has multiple identifiers\n",
      "CPS node node level has multiple identifiers\n",
      "CPS node node level has multiple identifiers\n",
      "GA20ox3 node node level has multiple identifiers\n",
      "GA20ox3 node node level has multiple identifiers\n",
      "GA20ox1 node node level has multiple identifiers\n",
      "GA20ox4 node node level has multiple identifiers\n",
      "PR1 clade node level has multiple identifiers\n",
      "input3\n",
      "output1\n",
      "PR1 family node level has multiple identifiers\n",
      "PR1 family node level has multiple identifiers\n",
      "GA20ox3 node node level has multiple identifiers\n",
      "GA20ox1 node node level has multiple identifiers\n",
      "GA20ox4 node node level has multiple identifiers\n",
      "PR1 clade node level has multiple identifiers\n",
      "PR1 family node level has multiple identifiers\n",
      "PR1 family node level has multiple identifiers\n",
      "GA20ox3 node node level has multiple identifiers\n",
      "GA20ox1 node node level has multiple identifiers\n",
      "GA20ox4 node node level has multiple identifiers\n",
      "PR1 clade node level has multiple identifiers\n"
     ]
    }
   ],
   "source": [
    "complexes_to_add = []\n",
    "missing_in_components = set()\n",
    "replace_w_family = set()\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    print(prefix)\n",
    "    id_col, form_col, level_col  =\\\n",
    "            [prefix + x for x in ('_ID',  '_form',  '_level')]\n",
    "\n",
    "    label_col = prefix + '_label'\n",
    "    family_col = prefix + '_family'    \n",
    "    \n",
    "    df_edges[label_col] = df_edges[[id_col, form_col, level_col, 'ConnID', 'origin']].apply(get_label, axis=1, result_type='expand')\n",
    "    df_edges[family_col] = df_edges[[id_col, level_col, label_col, 'ConnID', 'origin']].apply(apply_get_family_id, axis=1, result_type='expand')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functional groups\n",
    "Node --> identifiers\n",
    "\n",
    "Clade --> identifiers\n",
    "\n",
    "Family --> identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_entry():\n",
    "    d = {'level':'-', 'members':'-'}      \n",
    "    for c in [f\"{specie}_homologues\" for specie in all_species]:\n",
    "        d[c] = '-'\n",
    "\n",
    "    return d\n",
    "\n",
    "def list_of_string_list_to_list(x, delim=\",\"):\n",
    "    '''e.g: ['AT5G26130,AT1G50060,AT2G14610,AT2G14580', '', '', '']\n",
    "    to: ['AT4G33710', 'AT4G33720', 'AT5G26130', 'AT1G50060', 'AT2G14610', 'AT2G14580']\n",
    "    '''\n",
    "    new_list = []\n",
    "    for sub_s in x:\n",
    "        if not sub_s in helpers.empty_strings:\n",
    "            for s in sub_s.split(delim):\n",
    "                if not s in helpers.empty_strings:\n",
    "                    new_list.append(s)\n",
    "    new_list = sorted(new_list)\n",
    "    return new_list\n",
    "\n",
    "reaction_players = defaultdict(lambda : defaultdict(create_entry))\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, from_col, level_col, new_id, label_col  =\\\n",
    "            [prefix + x for x in ('_ID',  '_form',  '_level',  '_newID', '_label')]    \n",
    "    homologue_cols = [f\"{prefix}_{specie}_homologues\" for specie in all_species]\n",
    "    \n",
    "    this_players = df_edges[['reaction_id', id_col, level_col, label_col]]# + homologue_cols]\n",
    "    \n",
    "    for _, row in this_players.iterrows():\n",
    "        if row[label_col] in helpers.plant_node_labels:\n",
    "            name = row[id_col]\n",
    "            level = row[level_col]\n",
    "            members = []\n",
    "            \n",
    "            r = translate_df.loc[(name, level)]\n",
    "            for specie in all_species:\n",
    "                c = f\"{specie}_homologues\"\n",
    "                members.append(r[c])\n",
    "                reaction_players[row[id_col]][row['reaction_id']][c] = r[c]\n",
    "            \n",
    "            reaction_players[row[id_col]][row['reaction_id']]['level'] =  level\n",
    "            reaction_players[row[id_col]][row['reaction_id']]['family'] =  r['family']\n",
    "            reaction_players[row[id_col]][row['reaction_id']]['members'] =  list_of_string_list_to_list(members)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['forCB', 'KG', 'Conn010', 'ath', 'CTR|ETR', 'family', 'ER',\n",
       "        'complex', 'EIN2', 'family', 'ER', 'protein [active]', nan, nan,\n",
       "        nan, nan, 'inhibition', 'protein deactivation',\n",
       "        'phosphorylation', 'EIN2', 'node', 'ER', 'protein',\n",
       "        '[R1] targetted experiments (e.g. Y2H, BIFC)',\n",
       "        'DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes...) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)',\n",
       "        'ETR/CTR domain keeps EIN2 phosphorylated = inactive. Modelled as ETR(a)/CTR(a) catalysing the phosphorylation reaction that turns EIN2 inactive.',\n",
       "        'v1.0', 'forCB-v2.7.8_PIS-model-Reactions', 'R1',\n",
       "        'doi:10.1073/pnas.1214848109,doi:10.1042/bj20091102,doi:10.1093/mp/ssq036,doi:10.1073/pnas.0605528103,doi:10.1016/j.pbi.2008.06.011,doi:10.1093/mp/ssr042',\n",
       "        'complex', 'protein_active', nan, 'protein', 'ath', 'rx00010',\n",
       "        'Complex', nan, 'PlantCoding', 'EIN2', nan, nan, 'PlantCoding',\n",
       "        'EIN2'],\n",
       "       ['forCB', 'KG', 'Conn015', 'ath', 'EIN3(like)', 'family',\n",
       "        'nucleus', 'protein', 'EIN2', 'node', 'ER', 'protein [active]',\n",
       "        nan, nan, nan, nan, 'activation', 'protein activation', nan,\n",
       "        'EIN3(like)', 'family', 'nucleus', 'protein [active]',\n",
       "        '[R1] targetted experiments (e.g. Y2H, BIFC)',\n",
       "        'DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1105/tpc.001768 (Ethylene Biosynthesis and Signaling Networks) | DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes...)',\n",
       "        'Deposphorylated (thus active) EIN2 activates EIN3(like).',\n",
       "        'v1.0', 'forCB-v2.7.8_PIS-model-Reactions', 'R1',\n",
       "        'doi:10.1016/j.pbi.2008.06.011,doi:10.1105/tpc.001768,doi:10.1073/pnas.0605528103',\n",
       "        'protein', 'protein_active', nan, 'protein_active', 'ath',\n",
       "        'rx00015', 'PlantCoding', 'EIN3(like)', 'PlantCoding', 'EIN2',\n",
       "        nan, nan, 'PlantCoding', 'EIN3(like)']], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[df_edges['input2_ID'] == 'EIN2'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"FunctionalClusters_first-definition.tsv\", \"w\") as out:\n",
    "    c = '\\t'.join([f'{specie}_homologues' for specie in all_species])\n",
    "    out.write(f\"node_name\\tfamily\\treaction_id\\tlevel\\tmembers\\t{c}\\n\")\n",
    "    for node in list(reaction_players):\n",
    "        for reaction in list(reaction_players[node]):\n",
    "            c =  '\\t'.join([reaction_players[node][reaction][f'{specie}_homologues'] for specie in all_species])\n",
    "            out.write(f\"{node}\\t{reaction_players[node][reaction]['family']}\\t{reaction}\\t{reaction_players[node][reaction]['level']}\\t{','.join(reaction_players[node][reaction]['members'])}\\t{c}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "node_name\tfamily\treaction_id\tlevel\tmembers\tath_homologues\tstu_homologues\tsly_homologues\tosa_homologues\r\n",
      "ETR\tETR\trx00005\tfamily\tAT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150\tAT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940\t\t\t\r\n",
      "ETR\tETR\trx00006\tfamily\tAT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150\tAT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940\t\t\t\r\n",
      "ETR\tETR\trx00008\tfamily\tAT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150\tAT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940\t\t\t\r\n",
      "CTR\tCTR\trx00006\tfamily\tAT4G24480,AT5G03730\tAT5G03730,AT4G24480\t\t\t\r\n",
      "EIN2\tEIN2\trx00009\tnode\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "EIN2\tEIN2\trx00210\tnode\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "EIN2\tEIN2\trx00010\tnode\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "EIN2\tEIN2\trx00015\tnode\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "ETP\tETP\trx00011\tfamily\tAT2G04920,AT2G07140,AT2G24510,AT2G27520,AT3G17265,AT3G17280,AT3G17480,AT3G17490,AT3G17500,AT3G17530,AT3G17540,AT3G17560,AT3G17570,AT3G18320,AT3G18330,AT3G18900,AT3G18910,AT3G18980,AT3G19410,AT3G21120,AT3G21130,AT3G21170,AT3G22350,AT3G22700,AT3G22710,AT3G22720,AT3G22730,AT3G23260,AT3G24700,AT3G25460,AT3G44120,AT3G49980,AT4G09870,AT4G10190,AT5G47300,AT5G62830\tAT3G49980,AT3G19410,AT3G25460,AT5G47300,AT3G22720,AT3G18910,AT3G17530,AT3G17570,AT3G22700,AT3G21120,AT2G07140,AT2G27520,AT5G62830,AT3G21170,AT3G17490,AT3G17265,AT3G17480,AT2G24510,AT3G18320,AT3G23260,AT3G18330,AT3G17560,AT3G17280,AT3G21130,AT3G24700,AT3G17540,AT3G22350,AT3G18980,AT2G04920,AT4G10190,AT3G18900,AT3G17500,AT3G44120,AT4G09870,AT3G22730,AT3G22710\t\t\t\r\n"
     ]
    }
   ],
   "source": [
    "!head FunctionalClusters_first-definition.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_clusters_df = pd.read_csv(\"FunctionalClusters_first-definition.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00009</td>\n",
       "      <td>node</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00210</td>\n",
       "      <td>node</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00010</td>\n",
       "      <td>node</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00015</td>\n",
       "      <td>node</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EIN5</td>\n",
       "      <td>XRN</td>\n",
       "      <td>rx00012</td>\n",
       "      <td>node</td>\n",
       "      <td>AT1G54490</td>\n",
       "      <td>AT1G54490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>MAX2</td>\n",
       "      <td>F-box/LRR-repeat</td>\n",
       "      <td>rx00319</td>\n",
       "      <td>node</td>\n",
       "      <td>AT2G42620</td>\n",
       "      <td>AT2G42620</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>D53</td>\n",
       "      <td>Class I Clp ATPase</td>\n",
       "      <td>rx00320</td>\n",
       "      <td>node</td>\n",
       "      <td>OS11G0104300</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OS11G0104300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>RTM3</td>\n",
       "      <td>TRAF</td>\n",
       "      <td>rx00060</td>\n",
       "      <td>node</td>\n",
       "      <td>AT3G58350</td>\n",
       "      <td>AT3G58350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>NAC032</td>\n",
       "      <td>NAC</td>\n",
       "      <td>rx00157</td>\n",
       "      <td>node</td>\n",
       "      <td>AT1G77450</td>\n",
       "      <td>AT1G77450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>BAK1</td>\n",
       "      <td>LRR-BAK1</td>\n",
       "      <td>rx00237</td>\n",
       "      <td>node</td>\n",
       "      <td>AT4G33430</td>\n",
       "      <td>AT4G33430</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name              family reaction_id level       members  \\\n",
       "4        EIN2                EIN2     rx00009  node     AT5G03280   \n",
       "5        EIN2                EIN2     rx00210  node     AT5G03280   \n",
       "6        EIN2                EIN2     rx00010  node     AT5G03280   \n",
       "7        EIN2                EIN2     rx00015  node     AT5G03280   \n",
       "9        EIN5                 XRN     rx00012  node     AT1G54490   \n",
       "..        ...                 ...         ...   ...           ...   \n",
       "462      MAX2    F-box/LRR-repeat     rx00319  node     AT2G42620   \n",
       "463       D53  Class I Clp ATPase     rx00320  node  OS11G0104300   \n",
       "518      RTM3                TRAF     rx00060  node     AT3G58350   \n",
       "521    NAC032                 NAC     rx00157  node     AT1G77450   \n",
       "522      BAK1            LRR-BAK1     rx00237  node     AT4G33430   \n",
       "\n",
       "    ath_homologues stu_homologues sly_homologues osa_homologues  \n",
       "4        AT5G03280            NaN            NaN            NaN  \n",
       "5        AT5G03280            NaN            NaN            NaN  \n",
       "6        AT5G03280            NaN            NaN            NaN  \n",
       "7        AT5G03280            NaN            NaN            NaN  \n",
       "9        AT1G54490            NaN            NaN            NaN  \n",
       "..             ...            ...            ...            ...  \n",
       "462      AT2G42620            NaN            NaN            NaN  \n",
       "463            NaN            NaN            NaN   OS11G0104300  \n",
       "518      AT3G58350            NaN            NaN            NaN  \n",
       "521      AT1G77450            NaN            NaN            NaN  \n",
       "522      AT4G33430            NaN            NaN            NaN  \n",
       "\n",
       "[282 rows x 9 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['level']=='node']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "523\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00005</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00006</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00008</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTR</td>\n",
       "      <td>CTR</td>\n",
       "      <td>rx00006</td>\n",
       "      <td>family</td>\n",
       "      <td>AT4G24480,AT5G03730</td>\n",
       "      <td>AT5G03730,AT4G24480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00009</td>\n",
       "      <td>node</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_name family reaction_id   level  \\\n",
       "0       ETR    ETR     rx00005  family   \n",
       "1       ETR    ETR     rx00006  family   \n",
       "2       ETR    ETR     rx00008  family   \n",
       "3       CTR    CTR     rx00006  family   \n",
       "4      EIN2   EIN2     rx00009    node   \n",
       "\n",
       "                                             members  \\\n",
       "0  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "1  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "2  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "3                                AT4G24480,AT5G03730   \n",
       "4                                          AT5G03280   \n",
       "\n",
       "                                      ath_homologues stu_homologues  \\\n",
       "0  AT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940            NaN   \n",
       "1  AT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940            NaN   \n",
       "2  AT3G04580,AT1G04310,AT3G23150,AT1G66340,AT2G40940            NaN   \n",
       "3                                AT5G03730,AT4G24480            NaN   \n",
       "4                                          AT5G03280            NaN   \n",
       "\n",
       "  sly_homologues osa_homologues  \n",
       "0            NaN            NaN  \n",
       "1            NaN            NaN  \n",
       "2            NaN            NaN  \n",
       "3            NaN            NaN  \n",
       "4            NaN            NaN  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(functional_clusters_df.shape[0])\n",
    "functional_clusters_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_clusters_df.sort_values('node_name', inplace=True)\n",
    "functional_clusters_df = functional_clusters_df.drop_duplicates(keep='first', subset=['node_name', 'members', 'level'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>CPS</td>\n",
       "      <td>CPS</td>\n",
       "      <td>rx00275</td>\n",
       "      <td>node</td>\n",
       "      <td>AT4G02780,SOTUB06G034690.1.1</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td>SOTUB06G034690.1.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id level                       members  \\\n",
       "424       CPS    CPS     rx00275  node  AT4G02780,SOTUB06G034690.1.1   \n",
       "\n",
       "    ath_homologues      stu_homologues sly_homologues osa_homologues  \n",
       "424      AT4G02780  SOTUB06G034690.1.1            NaN            NaN  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name']=='CPS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00196</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00045</td>\n",
       "      <td>node</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>EDS5</td>\n",
       "      <td>EDS5</td>\n",
       "      <td>rx00116</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G21340,AT4G39030</td>\n",
       "      <td>AT4G39030,AT2G21340</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>EDS5</td>\n",
       "      <td>EDS5</td>\n",
       "      <td>rx00070</td>\n",
       "      <td>node</td>\n",
       "      <td>AT4G39030</td>\n",
       "      <td>AT4G39030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>GA3ox</td>\n",
       "      <td>GA3ox</td>\n",
       "      <td>rx00282</td>\n",
       "      <td>clade</td>\n",
       "      <td>AT1G15550,AT1G80330,AT1G80340,AT4G21690,SOLYC0...</td>\n",
       "      <td>AT4G21690,AT1G80330,AT1G15550,AT1G80340</td>\n",
       "      <td>SOTUB03G033480.1.1,SOTUB06G023360.1.1</td>\n",
       "      <td>SOLYC03G119910.3.1,SOLYC06G066820.3.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>GA3ox</td>\n",
       "      <td>GA3ox</td>\n",
       "      <td>rx00272</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G15550,AT1G80330,AT1G80340,AT4G21690,SOLYC0...</td>\n",
       "      <td>AT4G21690,AT1G80330,AT1G15550,AT1G80340</td>\n",
       "      <td>SOTUB03G033480.1.1,SOTUB06G023360.1.1</td>\n",
       "      <td>SOLYC03G119910.3.1,SOLYC06G066820.3.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>GID1</td>\n",
       "      <td>GID</td>\n",
       "      <td>rx00288</td>\n",
       "      <td>node</td>\n",
       "      <td>LOC_OS05G33730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC_OS05G33730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>GID1</td>\n",
       "      <td>GID</td>\n",
       "      <td>rx00285</td>\n",
       "      <td>clade</td>\n",
       "      <td>AT3G05120,AT3G63010,AT5G27320,LOC_OS05G33730,S...</td>\n",
       "      <td>AT3G05120,AT5G27320,AT3G63010</td>\n",
       "      <td>SOTUB06G008680.1.1,SOTUB09G021750.1.1,SOTUB01G...</td>\n",
       "      <td>SOLYC01G098390.3.1,SOLYC06G008870.2.1,SOLYC09G...</td>\n",
       "      <td>LOC_OS05G33730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>rx00179</td>\n",
       "      <td>clade</td>\n",
       "      <td>AT1G09090,AT1G19230,AT1G64060,AT3G45810,AT4G11...</td>\n",
       "      <td>AT1G09090,AT3G45810,AT5G07390,AT5G47910,AT5G51...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>rx00132</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G09090,AT1G19230,AT1G64060,AT3G45810,AT4G11...</td>\n",
       "      <td>AT1G09090,AT3G45810,AT5G07390,AT5G47910,AT5G51...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>PAD4</td>\n",
       "      <td>PAD4</td>\n",
       "      <td>rx00087</td>\n",
       "      <td>node</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>PAD4</td>\n",
       "      <td>PAD4</td>\n",
       "      <td>rx00120</td>\n",
       "      <td>family</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PR1</td>\n",
       "      <td>PR1</td>\n",
       "      <td>rx00104</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G50060,AT2G14580,AT2G14610,AT4G33710,AT4G33...</td>\n",
       "      <td>AT2G14580,AT5G26130,AT2G14610,AT1G50060,AT4G33...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PR1</td>\n",
       "      <td>PR1</td>\n",
       "      <td>rx00394</td>\n",
       "      <td>clade</td>\n",
       "      <td>AT1G50060,AT2G14580,AT2G14610,AT4G33710,AT4G33...</td>\n",
       "      <td>AT2G14580,AT5G26130,AT2G14610,AT1G50060,AT4G33...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00259</td>\n",
       "      <td>clade</td>\n",
       "      <td>AT2G43820,AT2G43840,AT4G31570</td>\n",
       "      <td>AT4G31570,AT2G43820,AT2G43840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00085</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G43820,AT2G43840,AT4G31570</td>\n",
       "      <td>AT4G31570,AT2G43820,AT2G43840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         node_name         family reaction_id   level  \\\n",
       "45            COI1           COI1     rx00196  family   \n",
       "43            COI1           COI1     rx00045    node   \n",
       "333           EDS5           EDS5     rx00116  family   \n",
       "330           EDS5           EDS5     rx00070    node   \n",
       "433          GA3ox          GA3ox     rx00282   clade   \n",
       "432          GA3ox          GA3ox     rx00272  family   \n",
       "269           GID1            GID     rx00288    node   \n",
       "266           GID1            GID     rx00285   clade   \n",
       "213  NADPH-oxidase  NADPH-oxidase     rx00179   clade   \n",
       "215  NADPH-oxidase  NADPH-oxidase     rx00132  family   \n",
       "350           PAD4           PAD4     rx00087    node   \n",
       "352           PAD4           PAD4     rx00120  family   \n",
       "106            PR1            PR1     rx00104  family   \n",
       "109            PR1            PR1     rx00394   clade   \n",
       "349           SAGT           SAGT     rx00259   clade   \n",
       "345           SAGT           SAGT     rx00085  family   \n",
       "\n",
       "                                               members  \\\n",
       "45                                           AT2G39940   \n",
       "43                                           AT2G39940   \n",
       "333                                AT2G21340,AT4G39030   \n",
       "330                                          AT4G39030   \n",
       "433  AT1G15550,AT1G80330,AT1G80340,AT4G21690,SOLYC0...   \n",
       "432  AT1G15550,AT1G80330,AT1G80340,AT4G21690,SOLYC0...   \n",
       "269                                     LOC_OS05G33730   \n",
       "266  AT3G05120,AT3G63010,AT5G27320,LOC_OS05G33730,S...   \n",
       "213  AT1G09090,AT1G19230,AT1G64060,AT3G45810,AT4G11...   \n",
       "215  AT1G09090,AT1G19230,AT1G64060,AT3G45810,AT4G11...   \n",
       "350                                          AT3G52430   \n",
       "352                                          AT3G52430   \n",
       "106  AT1G50060,AT2G14580,AT2G14610,AT4G33710,AT4G33...   \n",
       "109  AT1G50060,AT2G14580,AT2G14610,AT4G33710,AT4G33...   \n",
       "349                      AT2G43820,AT2G43840,AT4G31570   \n",
       "345                      AT2G43820,AT2G43840,AT4G31570   \n",
       "\n",
       "                                        ath_homologues  \\\n",
       "45                                           AT2G39940   \n",
       "43                                           AT2G39940   \n",
       "333                                AT4G39030,AT2G21340   \n",
       "330                                          AT4G39030   \n",
       "433            AT4G21690,AT1G80330,AT1G15550,AT1G80340   \n",
       "432            AT4G21690,AT1G80330,AT1G15550,AT1G80340   \n",
       "269                                                NaN   \n",
       "266                      AT3G05120,AT5G27320,AT3G63010   \n",
       "213  AT1G09090,AT3G45810,AT5G07390,AT5G47910,AT5G51...   \n",
       "215  AT1G09090,AT3G45810,AT5G07390,AT5G47910,AT5G51...   \n",
       "350                                          AT3G52430   \n",
       "352                                          AT3G52430   \n",
       "106  AT2G14580,AT5G26130,AT2G14610,AT1G50060,AT4G33...   \n",
       "109  AT2G14580,AT5G26130,AT2G14610,AT1G50060,AT4G33...   \n",
       "349                      AT4G31570,AT2G43820,AT2G43840   \n",
       "345                      AT4G31570,AT2G43820,AT2G43840   \n",
       "\n",
       "                                        stu_homologues  \\\n",
       "45                                                 NaN   \n",
       "43                                                 NaN   \n",
       "333                                                NaN   \n",
       "330                                                NaN   \n",
       "433              SOTUB03G033480.1.1,SOTUB06G023360.1.1   \n",
       "432              SOTUB03G033480.1.1,SOTUB06G023360.1.1   \n",
       "269                                                NaN   \n",
       "266  SOTUB06G008680.1.1,SOTUB09G021750.1.1,SOTUB01G...   \n",
       "213                                                NaN   \n",
       "215                                                NaN   \n",
       "350                                                NaN   \n",
       "352                                                NaN   \n",
       "106                                                NaN   \n",
       "109                                                NaN   \n",
       "349                                                NaN   \n",
       "345                                                NaN   \n",
       "\n",
       "                                        sly_homologues  osa_homologues  \n",
       "45                                                 NaN             NaN  \n",
       "43                                                 NaN             NaN  \n",
       "333                                                NaN             NaN  \n",
       "330                                                NaN             NaN  \n",
       "433              SOLYC03G119910.3.1,SOLYC06G066820.3.1             NaN  \n",
       "432              SOLYC03G119910.3.1,SOLYC06G066820.3.1             NaN  \n",
       "269                                                NaN  LOC_OS05G33730  \n",
       "266  SOLYC01G098390.3.1,SOLYC06G008870.2.1,SOLYC09G...  LOC_OS05G33730  \n",
       "213                                                NaN             NaN  \n",
       "215                                                NaN             NaN  \n",
       "350                                                NaN             NaN  \n",
       "352                                                NaN             NaN  \n",
       "106                                                NaN             NaN  \n",
       "109                                                NaN             NaN  \n",
       "349                                                NaN             NaN  \n",
       "345                                                NaN             NaN  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = functional_clusters_df[['node_name']].value_counts()\n",
    "x = x[x>1].index[:]\n",
    "x = [z[0] for z in x]\n",
    "functional_clusters_df[functional_clusters_df['node_name'].isin(x)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_clusters_df.loc[:, \"functional_cluster_name\"] = functional_clusters_df.apply(lambda row: f\"{row['node_name']}[{row['members']}]\", axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...</td>\n",
       "      <td>AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...</td>\n",
       "      <td>AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level  \\\n",
       "340       AAO    AAO     rx00081  family   \n",
       "318       ACH    ACH     rx00038  family   \n",
       "301       ACO    ACO     rx00003  family   \n",
       "300       ACS    ACS     rx00002  family   \n",
       "148      ACS2    ACS     rx00248    node   \n",
       "\n",
       "                                               members  \\\n",
       "340  AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...   \n",
       "148                                          AT1G01480   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340  AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...            NaN   \n",
       "318                                AT2G30720,AT5G48370            NaN   \n",
       "301  AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330            NaN   \n",
       "300  AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...            NaN   \n",
       "148                                          AT1G01480            NaN   \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340            NaN            NaN   \n",
       "318            NaN            NaN   \n",
       "301            NaN            NaN   \n",
       "300            NaN            NaN   \n",
       "148            NaN            NaN   \n",
       "\n",
       "                               functional_cluster_name  \n",
       "340  AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...  \n",
       "318                           ACH[AT2G30720,AT5G48370]  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  \n",
       "300  ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...  \n",
       "148                                    ACS2[AT1G01480]  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00259</td>\n",
       "      <td>clade</td>\n",
       "      <td>AT2G43820,AT2G43840,AT4G31570</td>\n",
       "      <td>AT4G31570,AT2G43820,AT2G43840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAGT[AT2G43820,AT2G43840,AT4G31570]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00085</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G43820,AT2G43840,AT4G31570</td>\n",
       "      <td>AT4G31570,AT2G43820,AT2G43840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAGT[AT2G43820,AT2G43840,AT4G31570]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level                        members  \\\n",
       "349      SAGT   SAGT     rx00259   clade  AT2G43820,AT2G43840,AT4G31570   \n",
       "345      SAGT   SAGT     rx00085  family  AT2G43820,AT2G43840,AT4G31570   \n",
       "\n",
       "                    ath_homologues stu_homologues sly_homologues  \\\n",
       "349  AT4G31570,AT2G43820,AT2G43840            NaN            NaN   \n",
       "345  AT4G31570,AT2G43820,AT2G43840            NaN            NaN   \n",
       "\n",
       "    osa_homologues              functional_cluster_name  \n",
       "349            NaN  SAGT[AT2G43820,AT2G43840,AT4G31570]  \n",
       "345            NaN  SAGT[AT2G43820,AT2G43840,AT4G31570]  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'SAGT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00196</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COI1[AT2G39940]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00045</td>\n",
       "      <td>node</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COI1[AT2G39940]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   node_name family reaction_id   level    members ath_homologues  \\\n",
       "45      COI1   COI1     rx00196  family  AT2G39940      AT2G39940   \n",
       "43      COI1   COI1     rx00045    node  AT2G39940      AT2G39940   \n",
       "\n",
       "   stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "45            NaN            NaN            NaN         COI1[AT2G39940]  \n",
       "43            NaN            NaN            NaN         COI1[AT2G39940]  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'COI1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00210</td>\n",
       "      <td>node</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIN2[AT5G03280]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_name family reaction_id level    members ath_homologues stu_homologues  \\\n",
       "5      EIN2   EIN2     rx00210  node  AT5G03280      AT5G03280            NaN   \n",
       "\n",
       "  sly_homologues osa_homologues functional_cluster_name  \n",
       "5            NaN            NaN         EIN2[AT5G03280]  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'EIN2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create functional cluster nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_clusters_import = functional_clusters_df.drop_duplicates(keep='first', subset=['node_name', 'members'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_labels(labels):\n",
    "\tfor x in ['Family', 'Plant', 'Foreign', 'Node']:\n",
    "\t\tif x in labels:\n",
    "\t\t\tlabels.remove(x)\n",
    "\treturn labels[0]\n",
    "\n",
    "def functional_cluster_query(file_name, \n",
    "                         labels, \n",
    "                         n_name=\"line.functional_cluster_name\"\n",
    "                        ):\n",
    "    \n",
    "    if type(labels) == list:\n",
    "        node_label = ':' + ':'.join(labels)\n",
    "    else:\n",
    "        node_label = ':' + labels\n",
    "    \n",
    "    key = {\"file_name\":file_name, \n",
    "           \"node_label\":node_label, \n",
    "           \"name\":n_name}\n",
    "\n",
    "    species_str = \"\"\n",
    "    for specie in all_species:\n",
    "        species_str += f\"{specie}_homologues:split(line.{specie}_homologues, ','),\\n                \"\n",
    "\n",
    "    \n",
    "    q = '''USING PERIODIC COMMIT 500\n",
    "           LOAD CSV WITH HEADERS FROM  'file:///{file_name}' AS line FIELDTERMINATOR '\\t'\n",
    "           CREATE (p{node_label}   {{ \n",
    "                name:{name}, \n",
    "                                \n",
    "                {species_str}\n",
    "\n",
    "                family:line.family\n",
    "   \n",
    "            }})'''.format(**key, species_str=species_str)\n",
    "    \n",
    "    return q\n",
    "\n",
    "\n",
    "\n",
    "def make_create_type_of_edge_query(file_name, edge_type,\n",
    "                           source_label=\"\", target_label=\"\",\n",
    "                           source_name=\"line.source_name\", target_name=\"line.target_name\",\n",
    "                          ):\n",
    "\n",
    "    if not source_label == \"\":\n",
    "        source_label = ':' + source_label\n",
    "        \n",
    "    if not target_label == \"\":\n",
    "        target_label = ':' + target_label\n",
    "                \n",
    "    key ={\"file_name\":file_name, \"edge_type\":edge_type,\n",
    "          \"source_label\":source_label, \"target_label\":target_label,\n",
    "          \"source_name\":source_name, \"target_name\":target_name, \n",
    "         }\n",
    "    \n",
    "    q = '''USING PERIODIC COMMIT 500\n",
    "           LOAD CSV WITH HEADERS FROM  'file:///{file_name}' AS line FIELDTERMINATOR '\\t'\n",
    "           \n",
    "           MATCH (source{source_label} {{ name:{source_name}}}),\n",
    "                 (target{target_label} {{ name:{target_name}}})\n",
    "           \n",
    "           CREATE (source)-[:{edge_type}]->(target)'''.format(**key)\n",
    "\n",
    "    return q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOAD CSV WITH HEADERS FROM 'file:///FunctionalCluster-components.tsv' AS line  FIELDTERMINATOR '\t'\n",
      "       MATCH (parent:Family { name:line.family }) RETURN parent.name AS name, labels(parent)   AS labels\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...</td>\n",
       "      <td>AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...</td>\n",
       "      <td>AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level  \\\n",
       "340       AAO    AAO     rx00081  family   \n",
       "318       ACH    ACH     rx00038  family   \n",
       "301       ACO    ACO     rx00003  family   \n",
       "300       ACS    ACS     rx00002  family   \n",
       "148      ACS2    ACS     rx00248    node   \n",
       "\n",
       "                                               members  \\\n",
       "340  AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...   \n",
       "148                                          AT1G01480   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340  AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...            NaN   \n",
       "318                                AT2G30720,AT5G48370            NaN   \n",
       "301  AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330            NaN   \n",
       "300  AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...            NaN   \n",
       "148                                          AT1G01480            NaN   \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340            NaN            NaN   \n",
       "318            NaN            NaN   \n",
       "301            NaN            NaN   \n",
       "300            NaN            NaN   \n",
       "148            NaN            NaN   \n",
       "\n",
       "                               functional_cluster_name       labels  \n",
       "340  AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...  PlantCoding  \n",
       "318                           ACH[AT2G30720,AT5G48370]  PlantCoding  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  PlantCoding  \n",
       "300  ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...  PlantCoding  \n",
       "148                                    ACS2[AT1G01480]  PlantCoding  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label = \"FunctionalCluster\"\n",
    "f = f\"{label}-components.tsv\"\n",
    "functional_clusters_import.to_csv(f\"../data/import/{f}\", sep=\"\\t\", index=None)\n",
    "\n",
    "\n",
    "q = '''LOAD CSV WITH HEADERS FROM 'file:///{file_name}' AS line  FIELDTERMINATOR '\\t'\n",
    "       MATCH (parent:Family {{ name:line.family }}) RETURN parent.name AS name, labels(parent)   AS labels\n",
    "    '''.format(file_name=f)\n",
    "print(q)\n",
    "qr = graph.run(q)\n",
    "\n",
    "df = pd.DataFrame(qr.data())\n",
    "df['labels'] = df['labels'].apply(clean_labels)\n",
    "df = df.drop_duplicates(keep='first')\n",
    "df = functional_clusters_import.join(df.set_index('name'), on='family')\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PlantCoding       227\n",
       "PlantNonCoding     10\n",
       "PlantAbstract       8\n",
       "Name: labels, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['labels'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FunctionalCluster \t 245\n",
      "PlantAbstract \t 8 8\n",
      "PlantCoding \t 227 227\n",
      "PlantNonCoding \t 10 10\n"
     ]
    }
   ],
   "source": [
    "# save node types\n",
    "label = \"FunctionalCluster\"\n",
    "print(label, \"\\t\", df.shape[0])\n",
    "\n",
    "for t, subdf in df.groupby('labels'):\n",
    "    subdf.to_csv(f\"../data/import/{label}-{t}-components.tsv\", sep=\"\\t\", index=None)\n",
    "\n",
    "    query = functional_cluster_query(f\"{label}-{t}-components.tsv\", [label, t], n_name=\"line.functional_cluster_name\")\n",
    "    qr = graph.run(query)\n",
    "    print(t, \"\\t\", subdf.shape[0], qr.stats()['nodes_created'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...</td>\n",
       "      <td>AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...</td>\n",
       "      <td>AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>miR319a-3p</td>\n",
       "      <td>miR319</td>\n",
       "      <td>rx00307</td>\n",
       "      <td>node</td>\n",
       "      <td>miR319a-3p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR319a-3p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR319a-3p[miR319a-3p]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>miR6022</td>\n",
       "      <td>miR6022</td>\n",
       "      <td>rx00308</td>\n",
       "      <td>clade</td>\n",
       "      <td>miR6022-3p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR6022-3p</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR6022[miR6022-3p]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td>rx00282</td>\n",
       "      <td>node</td>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>phasiRNA931[phasiRNA931]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td>rx00276</td>\n",
       "      <td>node</td>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vsiRNA12986[vsiRNA12986]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td>rx00275</td>\n",
       "      <td>node</td>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vsiRNA34327[vsiRNA34327]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>245 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_name       family reaction_id   level  \\\n",
       "340          AAO          AAO     rx00081  family   \n",
       "318          ACH          ACH     rx00038  family   \n",
       "301          ACO          ACO     rx00003  family   \n",
       "300          ACS          ACS     rx00002  family   \n",
       "148         ACS2          ACS     rx00248    node   \n",
       "..           ...          ...         ...     ...   \n",
       "274   miR319a-3p       miR319     rx00307    node   \n",
       "275      miR6022      miR6022     rx00308   clade   \n",
       "263  phasiRNA931  phasiRNA931     rx00282    node   \n",
       "257  vsiRNA12986  vsiRNA12986     rx00276    node   \n",
       "256  vsiRNA34327  vsiRNA34327     rx00275    node   \n",
       "\n",
       "                                               members  \\\n",
       "340  AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...   \n",
       "148                                          AT1G01480   \n",
       "..                                                 ...   \n",
       "274                                         miR319a-3p   \n",
       "275                                         miR6022-3p   \n",
       "263                                        phasiRNA931   \n",
       "257                                        vsiRNA12986   \n",
       "256                                        vsiRNA34327   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340  AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...            NaN   \n",
       "318                                AT2G30720,AT5G48370            NaN   \n",
       "301  AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330            NaN   \n",
       "300  AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...            NaN   \n",
       "148                                          AT1G01480            NaN   \n",
       "..                                                 ...            ...   \n",
       "274                                                NaN     miR319a-3p   \n",
       "275                                                NaN     miR6022-3p   \n",
       "263                                                NaN    phasiRNA931   \n",
       "257                                                NaN    vsiRNA12986   \n",
       "256                                                NaN    vsiRNA34327   \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340            NaN            NaN   \n",
       "318            NaN            NaN   \n",
       "301            NaN            NaN   \n",
       "300            NaN            NaN   \n",
       "148            NaN            NaN   \n",
       "..             ...            ...   \n",
       "274            NaN            NaN   \n",
       "275            NaN            NaN   \n",
       "263            NaN            NaN   \n",
       "257            NaN            NaN   \n",
       "256            NaN            NaN   \n",
       "\n",
       "                               functional_cluster_name  \n",
       "340  AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...  \n",
       "318                           ACH[AT2G30720,AT5G48370]  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  \n",
       "300  ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...  \n",
       "148                                    ACS2[AT1G01480]  \n",
       "..                                                 ...  \n",
       "274                             miR319a-3p[miR319a-3p]  \n",
       "275                                miR6022[miR6022-3p]  \n",
       "263                           phasiRNA931[phasiRNA931]  \n",
       "257                           vsiRNA12986[vsiRNA12986]  \n",
       "256                           vsiRNA34327[vsiRNA34327]  \n",
       "\n",
       "[245 rows x 10 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USING PERIODIC COMMIT 500\n",
      "           LOAD CSV WITH HEADERS FROM  'file:///TAKES_PART-FunctionalCluster-edges.tsv' AS line FIELDTERMINATOR '\t'\n",
      "           \n",
      "           MATCH (source:Family { name:line.family}),\n",
      "                 (target:FunctionalCluster { name:line.functional_cluster_name})\n",
      "           \n",
      "           CREATE (source)-[:TAKES_PART]->(target)\n",
      "FunctionalCluster 251 245\n",
      "\tnot all edges created\n"
     ]
    }
   ],
   "source": [
    "# component to complex edges\n",
    "edge_type = 'TAKES_PART'\n",
    "want_cols = ['functional_cluster_name', 'family']\n",
    "\n",
    "f = f'{edge_type}-{label}-edges.tsv'  \n",
    "functional_clusters_import.to_csv(f\"../data/import/{f}\", index=None, sep=\"\\t\")\n",
    "\n",
    "query = make_create_type_of_edge_query(f, edge_type, \n",
    "                       source_label='Family', target_label=label,\n",
    "                       source_name=\"line.family\", target_name=\"line.functional_cluster_name\",\n",
    "                      )\n",
    "print(query)\n",
    "qr = graph.run(query)\n",
    "\n",
    "print(label, functional_clusters_df.shape[0], qr.stats()['relationships_created'])    \n",
    "if not functional_clusters_df.shape[0] == qr.stats()['relationships_created']:\n",
    "    print(\"\\tnot all edges created\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Replace ids in edges table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...</td>\n",
       "      <td>AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...</td>\n",
       "      <td>AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level  \\\n",
       "340       AAO    AAO     rx00081  family   \n",
       "318       ACH    ACH     rx00038  family   \n",
       "301       ACO    ACO     rx00003  family   \n",
       "300       ACS    ACS     rx00002  family   \n",
       "148      ACS2    ACS     rx00248    node   \n",
       "\n",
       "                                               members  \\\n",
       "340  AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT4G34...   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT3G61...   \n",
       "148                                          AT1G01480   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340  AT2G27150,AT5G20960,AT3G43600,AT4G34900,AT4G34...            NaN   \n",
       "318                                AT2G30720,AT5G48370            NaN   \n",
       "301  AT1G05010,AT1G62380,AT1G12010,AT2G19590,AT1G77330            NaN   \n",
       "300  AT4G37770,AT4G26200,AT3G61510,AT3G49700,AT1G01...            NaN   \n",
       "148                                          AT1G01480            NaN   \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340            NaN            NaN   \n",
       "318            NaN            NaN   \n",
       "301            NaN            NaN   \n",
       "300            NaN            NaN   \n",
       "148            NaN            NaN   \n",
       "\n",
       "                               functional_cluster_name  \n",
       "340  AAO[AT1G04580,AT2G27150,AT3G43600,AT4G34890,AT...  \n",
       "318                           ACH[AT2G30720,AT5G48370]  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  \n",
       "300  ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...  \n",
       "148                                    ACS2[AT1G01480]  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "functional_clusters_translate = functional_clusters_df[['node_name', 'level', 'functional_cluster_name']].set_index(['node_name', 'level']).to_dict('index')\n",
    "\n",
    "def get_new_id(x):\n",
    "    id_, level_, label_ = x.values\n",
    "    \n",
    "    if label_ in helpers.plant_node_labels:\n",
    "        try:\n",
    "            return functional_clusters_translate[(id_, level_)]['functional_cluster_name']\n",
    "        except KeyError:\n",
    "            pass\n",
    "        \n",
    "        # if reaction has same node on different levels, then one is removed\n",
    "        return functional_clusters_translate[(id_, 'node')]['functional_cluster_name']\n",
    "        print(id_, level_, label_)\n",
    "    else:\n",
    "        return id_\n",
    "\n",
    "\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    print(prefix)\n",
    "    id_col, form_col, level_col, label_col, new_id =\\\n",
    "            [prefix + x for x in ('_ID',  '_form',  '_level', '_label', '_newID')]\n",
    "\n",
    "    df_edges[new_id] = df_edges[[id_col, level_col, label_col]].apply(get_new_id, axis=1, result_type='expand')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>...</th>\n",
       "      <th>input2_label</th>\n",
       "      <th>input2_family</th>\n",
       "      <th>input3_label</th>\n",
       "      <th>input3_family</th>\n",
       "      <th>output1_label</th>\n",
       "      <th>output1_family</th>\n",
       "      <th>input1_newID</th>\n",
       "      <th>input2_newID</th>\n",
       "      <th>input3_newID</th>\n",
       "      <th>output1_newID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>...</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>SAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>SAMS[AT1G02500,AT2G36880,AT3G17390,AT4G01850]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ACS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ACO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>HMA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>HMA[AT1G63440,AT4G33520,AT5G21930,AT5G44790]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cu2+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Status AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0  forCB      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1  forCB      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2  forCB      KG  Conn003     ath       ACC       family                  ER   \n",
       "3  forCB      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4  forCB      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level  ... input2_label input2_family  \\\n",
       "0  metabolite      SAMS        clade  ...  PlantCoding           SAM   \n",
       "1  metabolite       ACS       family  ...  PlantCoding           ACS   \n",
       "2  metabolite       ACO       family  ...  PlantCoding           ACO   \n",
       "3  metabolite       HMA       family  ...  PlantCoding           HMA   \n",
       "4     protein      Cu2+       family  ...   Metabolite           NaN   \n",
       "\n",
       "  input3_label input3_family output1_label output1_family  \\\n",
       "0          NaN           NaN    Metabolite            NaN   \n",
       "1          NaN           NaN    Metabolite            NaN   \n",
       "2          NaN           NaN    Metabolite            NaN   \n",
       "3          NaN           NaN    Metabolite            NaN   \n",
       "4          NaN           NaN   PlantCoding            ETR   \n",
       "\n",
       "                                        input1_newID  \\\n",
       "0                                              L-Met   \n",
       "1                                               SAMe   \n",
       "2                                                ACC   \n",
       "3                                               Cu2+   \n",
       "4  ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...   \n",
       "\n",
       "                                        input2_newID input3_newID  \\\n",
       "0      SAMS[AT1G02500,AT2G36880,AT3G17390,AT4G01850]          NaN   \n",
       "1  ACS[AT1G01480,AT1G62960,AT2G22810,AT3G49700,AT...          NaN   \n",
       "2  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...          NaN   \n",
       "3       HMA[AT1G63440,AT4G33520,AT5G21930,AT5G44790]          NaN   \n",
       "4                                               Cu2+          NaN   \n",
       "\n",
       "                                       output1_newID  \n",
       "0                                               SAMe  \n",
       "1                                                ACC  \n",
       "2                                                 ET  \n",
       "3                                               Cu2+  \n",
       "4  ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_localisation_dict = {\n",
    "    'nuc':'nucleus',\n",
    "    'er':'endoplasmic reticulum',\n",
    "    'golgi':'golgi apparatus', \n",
    "    'mitochondria?': 'putative:mitochondrion', \n",
    "    'cytoplasm?': 'putative:cytoplasm', \n",
    "}\n",
    "\n",
    "\n",
    "good_localisations = set([\n",
    "     'nucleus',\n",
    "     'nucleolus',\n",
    "     'cytoplasm',\n",
    "     'vacuole',\n",
    "     'endoplasmic reticulum',\n",
    "     'chloroplast',\n",
    "     'mitochondrion',\n",
    "     'golgi apparatus',\n",
    "     'peroxisome',\n",
    "     'apoplast',\n",
    "     'extracellular'\n",
    "])\n",
    "\n",
    "good_localisations.update(['putative:' + s for s in good_localisations])\n",
    "\n",
    "\n",
    "def node_localisation_std(x):\n",
    "    if not type(x) == str:\n",
    "        return \"putative:cytoplasm\"\n",
    "    \n",
    "    x = x.lower()\n",
    "    if x in node_localisation_dict.keys():\n",
    "        x = node_localisation_dict[x]\n",
    "    \n",
    "    if x in good_localisations:\n",
    "        return x\n",
    "    else:\n",
    "        print(x)\n",
    "        return \"\"\n",
    "\n",
    "node_localisations = set()\n",
    "new_localisation = set()\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, localisation_col, location_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_localisation', '_location')]\n",
    "    \n",
    "    x = df_edges[['ConnID', 'origin', id_col, localisation_col]].dropna(how='all', subset=[ id_col, localisation_col])\n",
    "    for _, y in x.iterrows():\n",
    "        if (not (y[id_col] in helpers.empty_strings)) and (y[localisation_col] in helpers.empty_strings + ['']):\n",
    "            print(y['origin'], \"|\", y['ConnID'], \"|\", prefix, \"|\", y[id_col], \"|\", y[localisation_col])\n",
    "    \n",
    "    \n",
    "    node_localisations.update(x[localisation_col])\n",
    "    \n",
    "    #print(prefix)\n",
    "    df_edges[location_col] = df_edges[localisation_col].apply(node_localisation_std)\n",
    "    \n",
    "    new_localisation.update(df_edges[location_col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ER',\n",
       " 'Golgi',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'cytoplasm?',\n",
       " 'extracellular',\n",
       " 'mitochondria?',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_localisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'endoplasmic reticulum',\n",
       " 'extracellular',\n",
       " 'golgi apparatus',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'putative:cytoplasm',\n",
       " 'putative:mitochondrion',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_localisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'apoplast',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'endoplasmic reticulum',\n",
       " 'extracellular',\n",
       " 'golgi apparatus',\n",
       " 'mitochondrion',\n",
       " 'nucleolus',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'putative:apoplast',\n",
       " 'putative:chloroplast',\n",
       " 'putative:cytoplasm',\n",
       " 'putative:endoplasmic reticulum',\n",
       " 'putative:extracellular',\n",
       " 'putative:golgi apparatus',\n",
       " 'putative:mitochondrion',\n",
       " 'putative:nucleolus',\n",
       " 'putative:nucleus',\n",
       " 'putative:peroxisome',\n",
       " 'putative:vacuole',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "good_localisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reaction_mode_dict = {\n",
      "\t'catalysis / auto-catalysis':'catalysis/auto-catalysis',\n",
      "\t'translocation':'translocation',\n",
      "\t'protein activation':'protein activation',\n",
      "\t'binding / oligomerisation':'binding/oligomerisation',\n",
      "\t'protein deactivation':'protein deactivation',\n",
      "\t'transcriptional / translational repression':'transcriptional/translational repression',\n",
      "\t'transcriptional / translational induction':'transcriptional/translational induction',\n",
      "\t'degradation / secretion':'degradation/secretion',\n",
      "\t'nan':'nan',\n",
      "\t'dissociation':'dissociation',\n",
      "\t'cleavage / auto-cleavage':'cleavage/auto-cleavage',\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print('reaction_mode_dict = {')\n",
    "for s in df_edges['ReactionMode'].unique():\n",
    "    print(f\"\\t'{s}':'{'/'.join([x.lower().strip() for x in str(s).split('/')])}',\")\n",
    "print('}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Old version before edit\n",
    "# reaction_mode_dict = {\n",
    "# \t'catalysis / auto-catalysis':'catalysis/auto-catalysis',\n",
    "# \t'translocation':'translocation',\n",
    "# \t'protein activation':'protein activation',\n",
    "# \t'binding / oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein deactivation':'protein deactivation',\n",
    "# \t'transcription/translation repression':'transcription/translation repression',\n",
    "# \t'transcription/translation induction':'transcription/translation induction',\n",
    "# \t'degradation / secretion':'degradation/secretion',\n",
    "# \t'dissociation':'dissociation',\n",
    "# \t'translation':'translation',\n",
    "# \t'transcription':'transcription',\n",
    "# \t'by binding':'by binding',\n",
    "# \t'nan':'nan',\n",
    "# \t'cleavage / auto-cleavage':'cleavage/auto-cleavage',\n",
    "# \t'binding/oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein phosphorylation':'protein phosphorylation',\n",
    "# }\n",
    "\n",
    "# Old version after edit\n",
    "# reaction_mode_dict = {\n",
    "# \t'catalysis / auto-catalysis':'catalysis/auto-catalysis',\n",
    "# \t'translocation':'translocation',\n",
    "# \t'protein activation':'protein activation',\n",
    "# \t'binding / oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein deactivation':'protein deactivation',\n",
    "# \t'transcription/translation repression':'transcription/translation repression',\n",
    "# \t'transcription/translation induction':'transcription/translation induction',\n",
    "# \t'degradation / secretion':'degradation/secretion',\n",
    "# \t'dissociation':'dissociation',\n",
    "# \t'translation':'bad',\n",
    "# \t'transcription':'bad',\n",
    "# \t'by binding':'bad',\n",
    "# \tnp.nan:'bad',\n",
    "# \t'cleavage / auto-cleavage':'bad',\n",
    "# \t'binding/oligomerisation':'binding/oligomerisation',\n",
    "# \t'protein phosphorylation':'bad',\n",
    "# }\n",
    "\n",
    "reaction_mode_dict = {\n",
    "\t'catalysis / auto-catalysis':'catalysis/auto-catalysis',\n",
    "\t'translocation':'translocation',\n",
    "\t'protein activation':'protein activation',\n",
    "\t'binding / oligomerisation':'binding/oligomerisation',\n",
    "\t'protein deactivation':'protein deactivation',\n",
    "\t'transcriptional / translational repression':'transcriptional/translational repression',\n",
    "\t'transcriptional / translational induction':'transcriptional/translational induction',\n",
    "\t'degradation / secretion':'degradation/secretion',\n",
    "\tnp.nan:'undefined',\n",
    "\t'dissociation':'dissociation',\n",
    "\t'cleavage / auto-cleavage':'cleavage/auto-cleavage',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['reaction_type'] = df_edges['ReactionMode'].apply(lambda x: reaction_mode_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>origin</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>ReactionMode</th>\n",
       "      <th>ReactionEffect</th>\n",
       "      <th>reaction_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn061</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>activation</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn121</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>_TBD-v2.7.8_PIS-model-Reactions</td>\n",
       "      <td>Conn311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>inhibition</td>\n",
       "      <td>undefined</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              origin   ConnID ReactionMode ReactionEffect  \\\n",
       "31   _TBD-v2.7.8_PIS-model-Reactions  Conn032          NaN     inhibition   \n",
       "60   _TBD-v2.7.8_PIS-model-Reactions  Conn061          NaN     inhibition   \n",
       "87   _TBD-v2.7.8_PIS-model-Reactions  Conn088          NaN     inhibition   \n",
       "88   _TBD-v2.7.8_PIS-model-Reactions  Conn089          NaN     inhibition   \n",
       "91   _TBD-v2.7.8_PIS-model-Reactions  Conn092          NaN     activation   \n",
       "120  _TBD-v2.7.8_PIS-model-Reactions  Conn121          NaN     inhibition   \n",
       "150  _TBD-v2.7.8_PIS-model-Reactions  Conn151          NaN     inhibition   \n",
       "209  _TBD-v2.7.8_PIS-model-Reactions  Conn210          NaN     inhibition   \n",
       "310  _TBD-v2.7.8_PIS-model-Reactions  Conn311          NaN     inhibition   \n",
       "\n",
       "    reaction_type  \n",
       "31      undefined  \n",
       "60      undefined  \n",
       "87      undefined  \n",
       "88      undefined  \n",
       "91      undefined  \n",
       "120     undefined  \n",
       "150     undefined  \n",
       "209     undefined  \n",
       "310     undefined  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[df_edges['reaction_type']=='undefined'][['origin', 'ConnID', 'ReactionMode', 'ReactionEffect', 'reaction_type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_edges['reaction_id'].unique()) == df_edges.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.to_csv(output_path / \"edges-sheet.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(output_path / \"functional_clusters.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_path / \"complexes_to_add.tsv\", \"w\") as out:\n",
    "    for c in complexes_to_add:\n",
    "        out.write(f\"{c}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
