{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import neo4j DB: 3/4\n",
    "\n",
    "Code to translate v2.7.4_PIS-model.xlsx to neo4j database. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import os\n",
    "from IPython.display import Image\n",
    "\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "base_path = Path(\"..\")\n",
    "input_path = base_path / \"data\" / \"raw\"\n",
    "output_path = base_path / \"data\" / \"parsed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to graph via docker-compose link. See http://localhost:7474/browser/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from py2neo import Graph, Node, Relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = Graph(host=\"neo4j\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in Bio elements\n",
    "bioelements = PlantCoding/PlantNonCoding/PlantAbstract\n",
    "\n",
    "have family hierachy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>species</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>external_links</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Process</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>gmm_ocd</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>node0278</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>Benzoic acid 2-hydroxylase: Isolated from toba...</td>\n",
       "      <td>Hormone:SA</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>node0255</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Arabidopsis contains two ICS genes but has no ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone:SA</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>node0695</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>invented:unidentified</td>\n",
       "      <td>green peach aphid 2; resistance against PVX (stu)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:Rgenes</td>\n",
       "      <td>v2.6</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>node0692</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>HRT</td>\n",
       "      <td>HRT</td>\n",
       "      <td>HRT</td>\n",
       "      <td>invented:unidentified</td>\n",
       "      <td>HR against TCV (ath)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:Rgenes</td>\n",
       "      <td>v2.5</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>node0691</td>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>PlantAbstract</td>\n",
       "      <td>plant_abstract</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>Ny</td>\n",
       "      <td>Ny</td>\n",
       "      <td>Ny</td>\n",
       "      <td>invented:unidentified</td>\n",
       "      <td>HR against PVY (stu)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S:Rgenes</td>\n",
       "      <td>v2.5</td>\n",
       "      <td>ignore</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier AddedBy species      NodeLabel        NodeType  Family     Clade  \\\n",
       "0   node0278      KG     ath  PlantAbstract  plant_abstract    BA2H      BA2H   \n",
       "1   node0255      KG     ath  PlantAbstract  plant_abstract     IPL       IPL   \n",
       "2   node0695      KG     ath  PlantAbstract  plant_abstract  R-gene  GPAphid2   \n",
       "3   node0692      KG     ath  PlantAbstract  plant_abstract  R-gene       HRT   \n",
       "4   node0691      KG     ath  PlantAbstract  plant_abstract  R-gene        Ny   \n",
       "\n",
       "     NodeID  NodeName         external_links  \\\n",
       "0      BA2H      BA2H                    NaN   \n",
       "1       IPL       IPL                    NaN   \n",
       "2  GPAphid2  GPAphid2  invented:unidentified   \n",
       "3       HRT       HRT  invented:unidentified   \n",
       "4        Ny        Ny  invented:unidentified   \n",
       "\n",
       "                                     NodeDescription  \\\n",
       "0                                               BA2H   \n",
       "1  Arabidopsis contains two ICS genes but has no ...   \n",
       "2  green peach aphid 2; resistance against PVX (stu)   \n",
       "3                               HR against TCV (ath)   \n",
       "4                               HR against PVY (stu)   \n",
       "\n",
       "                                      AdditionalInfo     Process ModelV  \\\n",
       "0  Benzoic acid 2-hydroxylase: Isolated from toba...  Hormone:SA   v1.0   \n",
       "1                                                NaN  Hormone:SA   v1.0   \n",
       "2                                                NaN    S:Rgenes   v2.6   \n",
       "3                                                NaN    S:Rgenes   v2.5   \n",
       "4                                                NaN    S:Rgenes   v2.5   \n",
       "\n",
       "  ModelStatus gmm_ocd GMM_Description GMM_ShortName synonyms  \n",
       "0      ignore     NaN             NaN           NaN      NaN  \n",
       "1      ignore     NaN             NaN           NaN      NaN  \n",
       "2      ignore     NaN             NaN           NaN      NaN  \n",
       "3      ignore     NaN             NaN           NaN      NaN  \n",
       "4      ignore     NaN             NaN           NaN      NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_name = output_path / \"bio_elements.tsv\"\n",
    "df_bioelements = pd.read_csv(file_name, sep=\"\\t\")\n",
    "df_bioelements.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_bioelements[df_bioelements['NodeName'].duplicated(keep=False)].sort_values('NodeName')[['identifier', 'species', 'Family', 'Clade', 'NodeID', 'NodeName']]\n",
    "repeated_node_level = df_bioelements[df_bioelements['NodeName'].duplicated(keep='first')]['NodeName'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CPS', 'CPS.x1', 'CPS.x2', 'RGA2', 'RGA2', 'GA20ox.x1',\n",
       "       'GA20ox.x2', 'GA20ox1', 'GA20ox1', 'GA20ox2', 'GA20ox2', 'GA20ox3',\n",
       "       'GA20ox3', 'GA20ox4', 'GA20ox4', 'GA3ox.x1', 'GA3ox.x2', 'GID1.x1',\n",
       "       'GID1.x2', 'GID1.x3', 'MYB33-65.x1', 'MYB33-65.x2'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repeated_node_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th>level</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;alpha;/&amp;beta; hydroxylase</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;beta;-carotene isomerase</th>\n",
       "      <th>clade</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4CLL</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT1G20500,AT5G38120,AT1G20480,AT1G20510</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4CLL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            ath_homologues  \\\n",
       "name                       level                                             \n",
       "&alpha;/&beta; hydroxylase clade                                 AT3G03990   \n",
       "                           family                                AT3G03990   \n",
       "&beta;-carotene isomerase  clade                                             \n",
       "                           family                                            \n",
       "4CLL                       clade   AT1G20500,AT5G38120,AT1G20480,AT1G20510   \n",
       "\n",
       "                                  stu_homologues sly_homologues  \\\n",
       "name                       level                                  \n",
       "&alpha;/&beta; hydroxylase clade                                  \n",
       "                           family                                 \n",
       "&beta;-carotene isomerase  clade                                  \n",
       "                           family                                 \n",
       "4CLL                       clade                                  \n",
       "\n",
       "                                  osa_homologues                      family  \n",
       "name                       level                                              \n",
       "&alpha;/&beta; hydroxylase clade                  &alpha;/&beta; hydroxylase  \n",
       "                           family                 &alpha;/&beta; hydroxylase  \n",
       "&beta;-carotene isomerase  clade    OS11G0587000   &beta;-carotene isomerase  \n",
       "                           family   OS11G0587000   &beta;-carotene isomerase  \n",
       "4CLL                       clade                                        4CLL  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_name = output_path / \"level_translation.tsv\"\n",
    "translate_df = pd.read_csv(file_name, sep=\"\\t\", index_col=[0, 1])\n",
    "translate_df.fillna('', inplace=True)\n",
    "translate_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th>level</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;alpha;/&amp;beta; hydroxylase</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;beta;-carotene isomerase</th>\n",
       "      <th>clade</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4CLL</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT1G20500,AT5G38120,AT1G20480,AT1G20490,AT4G05...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4CLL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                      ath_homologues  \\\n",
       "name                       level                                                       \n",
       "&alpha;/&beta; hydroxylase clade                                           AT3G03990   \n",
       "                           family                                          AT3G03990   \n",
       "&beta;-carotene isomerase  clade                                                       \n",
       "                           family                                                      \n",
       "4CLL                       clade   AT1G20500,AT5G38120,AT1G20480,AT1G20490,AT4G05...   \n",
       "\n",
       "                                  stu_homologues sly_homologues  \\\n",
       "name                       level                                  \n",
       "&alpha;/&beta; hydroxylase clade                                  \n",
       "                           family                                 \n",
       "&beta;-carotene isomerase  clade                                  \n",
       "                           family                                 \n",
       "4CLL                       clade                                  \n",
       "\n",
       "                                  osa_homologues                      family  \n",
       "name                       level                                              \n",
       "&alpha;/&beta; hydroxylase clade                  &alpha;/&beta; hydroxylase  \n",
       "                           family                 &alpha;/&beta; hydroxylase  \n",
       "&beta;-carotene isomerase  clade    OS11G0587000   &beta;-carotene isomerase  \n",
       "                           family   OS11G0587000   &beta;-carotene isomerase  \n",
       "4CLL                       clade                                        4CLL  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_name = output_path / \"level_not_use_translation.tsv\"\n",
    "translate_ignore_etc_df = pd.read_csv(file_name, sep=\"\\t\", index_col=[0, 1])\n",
    "translate_ignore_etc_df.fillna('', inplace=True)\n",
    "translate_ignore_etc_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in parsed reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ath', 'stu', 'sly', 'osa']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species = list(df_bioelements['species'].unique())\n",
    "try: all_species.remove('plant_all')\n",
    "except ValueError: pass\n",
    "try: all_species.remove('all')\n",
    "except ValueError: pass    \n",
    "all_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>...</th>\n",
       "      <th>input1_form</th>\n",
       "      <th>input2_form</th>\n",
       "      <th>input3_form</th>\n",
       "      <th>output1_form</th>\n",
       "      <th>species</th>\n",
       "      <th>input1_location</th>\n",
       "      <th>input2_location</th>\n",
       "      <th>input3_location</th>\n",
       "      <th>output1_location</th>\n",
       "      <th>reaction_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>...</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>putative:cytoplasm</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>catalysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>putative:cytoplasm</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>catalysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>putative:cytoplasm</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>catalysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ath</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>putative:cytoplasm</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>translocation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>protein</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_active</td>\n",
       "      <td>ath</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>putative:cytoplasm</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "      <td>protein activation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Status AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0  forCB      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1  forCB      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2  forCB      KG  Conn003     ath       ACC       family                  ER   \n",
       "3  forCB      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4  forCB      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level  ... input1_form     input2_form  \\\n",
       "0  metabolite      SAMS        clade  ...  metabolite  protein_active   \n",
       "1  metabolite       ACS       family  ...  metabolite  protein_active   \n",
       "2  metabolite       ACO       family  ...  metabolite  protein_active   \n",
       "3  metabolite       HMA       family  ...  metabolite         protein   \n",
       "4     protein      Cu2+       family  ...     protein      metabolite   \n",
       "\n",
       "  input3_form    output1_form species        input1_location  \\\n",
       "0         NaN      metabolite     ath  endoplasmic reticulum   \n",
       "1         NaN      metabolite     ath  endoplasmic reticulum   \n",
       "2         NaN      metabolite     ath  endoplasmic reticulum   \n",
       "3         NaN      metabolite     ath              cytoplasm   \n",
       "4         NaN  protein_active     ath  endoplasmic reticulum   \n",
       "\n",
       "         input2_location     input3_location       output1_location  \\\n",
       "0  endoplasmic reticulum  putative:cytoplasm  endoplasmic reticulum   \n",
       "1  endoplasmic reticulum  putative:cytoplasm  endoplasmic reticulum   \n",
       "2  endoplasmic reticulum  putative:cytoplasm  endoplasmic reticulum   \n",
       "3  endoplasmic reticulum  putative:cytoplasm  endoplasmic reticulum   \n",
       "4  endoplasmic reticulum  putative:cytoplasm  endoplasmic reticulum   \n",
       "\n",
       "        reaction_type  \n",
       "0           catalysis  \n",
       "1           catalysis  \n",
       "2           catalysis  \n",
       "3       translocation  \n",
       "4  protein activation  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges = pd.read_csv(\"parsed_reactions.tsv\", sep=\"\\t\", index_col=0)\n",
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_edges['ConnID'].unique()) == df_edges.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    rx00001\n",
       "1    rx00002\n",
       "2    rx00003\n",
       "3    rx00004\n",
       "4    rx00005\n",
       "Name: reaction_id, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['reaction_id'] = df_edges.index.map(lambda x: f\"rx{x+1:05}\")\n",
    "df_edges['reaction_id'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FunctionalCluster 0\n",
      "PlantCoding 162\n",
      "PlantNonCoding 9\n",
      "PlantAbstract 7\n",
      "ForeignEntity 3\n",
      "ForeignCoding 14\n",
      "ForeignNonCoding 0\n",
      "ForeignAbstract 0\n",
      "Complex 3\n",
      "Process 6\n",
      "MetaboliteFamily 6\n",
      "Metabolite 110\n",
      "Reaction 0\n"
     ]
    }
   ],
   "source": [
    "node_dict = {}\n",
    "for label in helpers.node_labels:\n",
    "    if label == \"Metabolite\":\n",
    "        q = '''MATCH (n:%s) WHERE NOT n:MetaboliteFamily RETURN DISTINCT n.name'''%label\n",
    "    else:\n",
    "        q = '''MATCH (n:%s) RETURN DISTINCT n.name'''%label\n",
    "    s = set([d['n.name'] for d in graph.run(q).data()])\n",
    "    print(label, len(s))\n",
    "    node_dict[label] = s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th>level</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;alpha;/&amp;beta; hydroxylase</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td>AT3G03990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>&amp;alpha;/&amp;beta; hydroxylase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">&amp;beta;-carotene isomerase</th>\n",
       "      <th>clade</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>family</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0587000</td>\n",
       "      <td>&amp;beta;-carotene isomerase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4CLL</th>\n",
       "      <th>clade</th>\n",
       "      <td>AT1G20500,AT5G38120,AT1G20480,AT1G20510</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4CLL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            ath_homologues  \\\n",
       "name                       level                                             \n",
       "&alpha;/&beta; hydroxylase clade                                 AT3G03990   \n",
       "                           family                                AT3G03990   \n",
       "&beta;-carotene isomerase  clade                                             \n",
       "                           family                                            \n",
       "4CLL                       clade   AT1G20500,AT5G38120,AT1G20480,AT1G20510   \n",
       "\n",
       "                                  stu_homologues sly_homologues  \\\n",
       "name                       level                                  \n",
       "&alpha;/&beta; hydroxylase clade                                  \n",
       "                           family                                 \n",
       "&beta;-carotene isomerase  clade                                  \n",
       "                           family                                 \n",
       "4CLL                       clade                                  \n",
       "\n",
       "                                  osa_homologues                      family  \n",
       "name                       level                                              \n",
       "&alpha;/&beta; hydroxylase clade                  &alpha;/&beta; hydroxylase  \n",
       "                           family                 &alpha;/&beta; hydroxylase  \n",
       "&beta;-carotene isomerase  clade    OS11G0587000   &beta;-carotene isomerase  \n",
       "                           family   OS11G0587000   &beta;-carotene isomerase  \n",
       "4CLL                       clade                                        4CLL  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translate_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ath_homologues       Rx1\n",
       "stu_homologues          \n",
       "sly_homologues          \n",
       "osa_homologues          \n",
       "family            R-gene\n",
       "Name: (Rx1, node), dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translate_ignore_etc_df.loc[('Rx1', 'node')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_label(x):\n",
    "    # pathogen proteins are listed as proteins, so cannot use dict\n",
    "    id_, form_, level_, ConnID, origin = x.values\n",
    "      \n",
    "    if id_ in helpers.empty_strings:\n",
    "        return np.nan\n",
    "\n",
    "    new_label = None\n",
    "    \n",
    "    ########################\n",
    "    # Simple Cases\n",
    "    ########################\n",
    "    if form_ in ['complex', 'complex_active']:\n",
    "        if not (id_ in node_dict[\"Complex\"]):\n",
    "            complexes_to_add.append(id_)\n",
    "        new_label = 'Complex'\n",
    "    \n",
    "    elif form_ in [\"metabolite\"]:\n",
    "        for label in [\"Metabolite\", \"MetaboliteFamily\"]:\n",
    "            if id_ in node_dict[label]:\n",
    "                new_label = label\n",
    "                break\n",
    "        if not new_label:\n",
    "            missing_in_components.update([id_])        \n",
    "            print('ERROR', origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_, \" |a (label) not a listed Metabolite/MetaboliteFamily\")\n",
    "\n",
    "    elif form_ in ['process', 'process_active']:\n",
    "        if (id_ in node_dict[\"Process\"]):\n",
    "            new_label = \"Process\"\n",
    "        else:\n",
    "            print('ERROR', origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |b (label) process not a listed process\")\n",
    "    \n",
    "    else:\n",
    "        # somethng plant, something foreign\n",
    "        family_id = get_family_id(id_, level_, translate_df)\n",
    "        \n",
    "        if not family_id:\n",
    "            family_id = get_family_id(id_, level_, translate_ignore_etc_df)\n",
    "\n",
    "                \n",
    "        if (family_id):\n",
    "            id_labels = [] #looping just in case an id occurs mutiple times\n",
    "            for label in helpers.plant_node_labels:\n",
    "                if family_id in node_dict[label]:\n",
    "                    id_labels.append(label)\n",
    "\n",
    "            if len(id_labels) == 1:\n",
    "                new_label = id_labels[0]\n",
    "            elif len(id_labels) > 1:\n",
    "                print('ERROR', origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |d (label) many labels fit\")\n",
    "                missing_in_components.update([id_])       \n",
    "\n",
    "            else:\n",
    "                print('ERROR', origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |e (label) could not find label\")\n",
    "                missing_in_components.update([id_]) \n",
    "        else:\n",
    "            # check if foreign\n",
    "            for label in helpers.foreign_node_labels:\n",
    "                if id_ in node_dict[label]:\n",
    "                    new_label = label\n",
    "                    break\n",
    "        \n",
    "    if not new_label:\n",
    "        print('ERROR', origin, \"|\", ConnID, \"|\", id_, \"|\", form_, \"|\", level_,  \" |c (family id) could not convert to label\")\n",
    "        \n",
    "    return new_label\n",
    "\n",
    "\n",
    "    \n",
    "def get_family_id(id_, level_, df):\n",
    "    family_id = None\n",
    "    if id_ in repeated_node_level:\n",
    "        print('INFO', id_, level_, 'node level has multiple identifiers')\n",
    "    if level_ == \"family\":    \n",
    "        try:\n",
    "            family_id = df.loc[(id_, 'family')]['family']\n",
    "        except KeyError:\n",
    "            pass\n",
    "    elif level_ in [\"clade\", \"clade/orthologue\"]:\n",
    "        try:\n",
    "            family_id = df.loc[(id_, 'clade')]['family']\n",
    "        except KeyError:\n",
    "            pass\n",
    "    elif level_ == \"node\":\n",
    "        try:\n",
    "            family_id = df.loc[(id_, 'node')]['family']\n",
    "        except KeyError:\n",
    "            pass\n",
    "\n",
    "    return family_id\n",
    "\n",
    "def apply_get_family_id(x):\n",
    "    id_, level_, label_, ConnID, origin = x.values\n",
    "      \n",
    "    if id_ in helpers.empty_strings:\n",
    "        return np.nan\n",
    "    \n",
    "    family_id = np.nan\n",
    "    if label_ in helpers.plant_node_labels:\n",
    "        family_id = get_family_id(id_, level_)\n",
    "    \n",
    "    return family_id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "INFO CPS node node level has multiple identifiers\n",
      "INFO CPS node node level has multiple identifiers\n",
      "INFO GA20ox3 node node level has multiple identifiers\n",
      "INFO GA20ox3 node node level has multiple identifiers\n",
      "INFO GA20ox1 node node level has multiple identifiers\n",
      "INFO GA20ox4 node node level has multiple identifiers\n",
      "input3\n",
      "output1\n",
      "INFO GA20ox3 node node level has multiple identifiers\n",
      "INFO GA20ox1 node node level has multiple identifiers\n",
      "INFO GA20ox4 node node level has multiple identifiers\n"
     ]
    }
   ],
   "source": [
    "complexes_to_add = []\n",
    "missing_in_components = set()\n",
    "replace_w_family = set()\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    print(prefix)\n",
    "    id_col, form_col, level_col  =\\\n",
    "            [prefix + x for x in ('_ID',  '_form',  '_level')]\n",
    "\n",
    "    label_col = prefix + '_label'\n",
    "    family_col = prefix + '_family'    \n",
    "    \n",
    "    df_edges[label_col] = df_edges[[id_col, form_col, level_col, 'ConnID', 'origin']].apply(get_label, axis=1, result_type='expand')\n",
    "    #df_edges[family_col] = df_edges[[id_col, level_col, label_col, 'ConnID', 'origin']].apply(apply_get_family_id, axis=1, result_type='expand')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functional groups\n",
    "Node --> identifiers\n",
    "\n",
    "Clade --> identifiers\n",
    "\n",
    "Family --> identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PlantCoding', 'PlantNonCoding', 'PlantAbstract']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "helpers.plant_node_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "column\tname\trxid\tlevel\tspecies(as in excel)\tspecies(for identifiers)\n",
      "input1_ID\tCOI1\trx00045\tnode\tath,nbe\t['ath', 'nbe']\t['AT2G39940']\n",
      "input1_ID\tRx1\trx00057\tnode\tath,stu\t['ath', 'stu']\t[]\n",
      "input1_ID\tRx2\trx00058\tnode\tath,stu\t['ath', 'stu']\t[]\n",
      "input1_ID\tGPAphid2\trx00059\tnode\tath,stu\t['ath', 'stu']\t[]\n",
      "input1_ID\tR-gene\trx00108\tfamily\tath\t['ath']\t[]\n",
      "input1_ID\tR-gene\trx00109\tfamily\tath\t['ath']\t[]\n",
      "input1_ID\tAOX\trx00133\tfamily\tath,nbe\t['ath', 'nbe']\t[]\n",
      "input1_ID\tR-gene\trx00168\tfamily\tath\t['ath']\t[]\n",
      "input1_ID\tRBCS-3B\trx00232\tnode\tath\t['ath']\t[]\n",
      "input1_ID\tRBCS-2B\trx00233\tnode\tath\t['ath']\t[]\n",
      "input1_ID\tNAC055\trx00255\tnode\tath\t['ath']\t[]\n",
      "input1_ID\tNAC072\trx00256\tnode\tath\t['ath']\t[]\n",
      "input1_ID\tNAC055\trx00258\tnode\tath\t['ath']\t[]\n",
      "input1_ID\tNAC072\trx00259\tnode\tath\t['ath']\t[]\n",
      "input1_ID\tGPX\trx00262\tfamily\tall\t['ath', 'stu', 'sly', 'osa']\t[]\n",
      "input2_ID\tOMR1\trx00020\tnode\tath,nbe\t['ath', 'nbe']\t['AT3G10050']\n",
      "input2_ID\tX4\trx00039\tnode\tath\t['ath']\t[]\n",
      "input2_ID\tOMR1\trx00045\tnode\tath,nbe\t['ath', 'nbe']\t['AT3G10050']\n",
      "input2_ID\tIPL\trx00071\tfamily\tath\t['ath']\t[]\n",
      "input2_ID\tBA2H\trx00078\tfamily\tath\t['ath']\t[]\n",
      "input2_ID\tX1\trx00082\tfamily\tath\t['ath']\t[]\n",
      "input2_ID\tR-gene\trx00107\tfamily\tath\t['ath']\t[]\n",
      "input2_ID\tSOD\trx00131\tclade\tath,nbe\t['ath', 'nbe']\t['AT1G08830', 'AT1G12520', 'AT2G28190', 'AT3G56350', 'AT5G18100']\n",
      "input2_ID\tNADPH-oxidase\trx00132\tfamily\tath,nbe\t['ath', 'nbe']\t['AT5G47910']\n",
      "input2_ID\tSAM3\trx00197\tnode\tath,nta\t['ath', 'nta']\t['AT2G36880']\n",
      "input2_ID\tSAHH\trx00198\tfamily\tath,nta\t['ath', 'nta']\t['AT3G23810']\n",
      "input2_ID\tAGO1,5,7,10\trx00199\tclade\tath,nta\t['ath', 'nta']\t['AT1G48410']\n",
      "input2_ID\tAGO1,5,7,10\trx00200\tclade\tath,nta\t['ath', 'nta']\t['AT1G48410']\n",
      "input2_ID\tCPIP1,2b\trx00201\tnode\tath,nta\t['ath', 'nta']\t['AT1G10350']\n",
      "input2_ID\tCPIP2a\trx00202\tnode\tath,nta\t['ath', 'nta']\t['AT3G08910']\n",
      "input2_ID\tCPIP1,2b\trx00203\tnode\tath,nta\t['ath', 'nta']\t['AT1G10350']\n",
      "input2_ID\tDXPS2\trx00204\tnode\tath,nta\t['ath', 'nta']\t['AT4G15560']\n",
      "input2_ID\tATPB\trx00205\tnode\tath,nta\t['ath', 'nta']\t['ATCG00480']\n",
      "input2_ID\tMIND1\trx00207\tnode\tath,nta\t['ath', 'nta']\t['AT5G24020']\n",
      "input2_ID\tRBCS-3B\trx00222\tnode\tath\t['ath']\t[]\n",
      "input2_ID\tRBCS-2B\trx00223\tnode\tath\t['ath']\t[]\n",
      "input2_ID\tNAC055\trx00252\tnode\tath\t['ath']\t[]\n",
      "input2_ID\tNAC072\trx00253\tnode\tath\t['ath']\t[]\n",
      "input2_ID\tMYB33\trx00307\tnode\tstu\t['stu']\t[]\n",
      "input2_ID\tMYB33\trx00308\tnode\tstu\t['stu']\t[]\n",
      "input2_ID\tACX2\trx00312\tnode\tath\t['ath']\t[]\n",
      "input2_ID\tCCD7\trx00315\tnode\tath,osa,psa\t['ath', 'osa', 'psa']\t['AT2G44990']\n",
      "input2_ID\tCCD8\trx00316\tnode\tath,osa,psa\t['ath', 'osa', 'psa']\t['AT4G32810']\n",
      "input2_ID\tMAX1\trx00317\tnode\tath,osa,sly,zma\t['ath', 'osa', 'sly', 'zma']\t['AT2G26170']\n",
      "input2_ID\tD14\trx00318\tnode\tath,osa,phy\t['ath', 'osa', 'phy']\t['AT3G03990']\n",
      "output1_ID\tR-gene\trx00108\tfamily\tath\t['ath']\t[]\n",
      "output1_ID\tR-gene\trx00109\tfamily\tath\t['ath']\t[]\n",
      "output1_ID\tNAC055\trx00252\tnode\tath\t['ath']\t[]\n",
      "output1_ID\tNAC072\trx00253\tnode\tath\t['ath']\t[]\n",
      "output1_ID\tMYB33\trx00308\tnode\tstu\t['stu']\t[]\n",
      "output1_ID\tD14\trx00318\tnode\tath,osa,phy\t['ath', 'osa', 'phy']\t['AT3G03990']\n"
     ]
    }
   ],
   "source": [
    "def create_entry():\n",
    "    d = {'level':'', 'members':'', 'species':'', \"family\":''}      \n",
    "    for c in [f\"{specie}_homologues\" for specie in all_species]:\n",
    "        d[c] = '-'\n",
    "\n",
    "    return d\n",
    "\n",
    "def list_of_string_list_to_list(x, delim=\",\"):\n",
    "    '''e.g: ['AT5G26130,AT1G50060,AT2G14610,AT2G14580', '', '', '']\n",
    "    to: ['AT4G33710', 'AT4G33720', 'AT5G26130', 'AT1G50060', 'AT2G14610', 'AT2G14580']\n",
    "    '''\n",
    "    new_list = []\n",
    "    for sub_s in x:\n",
    "        if not sub_s in helpers.empty_strings:\n",
    "            for s in sub_s.split(delim):\n",
    "                if not s in helpers.empty_strings:\n",
    "                    new_list.append(s)\n",
    "    new_list = sorted(new_list)\n",
    "    return new_list\n",
    "\n",
    "print(\"column\\tname\\trxid\\tlevel\\tspecies(as in excel)\\tspecies(for identifiers)\")\n",
    "reaction_players = defaultdict(lambda : defaultdict(dict))  #lambda : defaultdict(create_entry))\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, from_col, level_col, new_id, label_col  =\\\n",
    "            [prefix + x for x in ('_ID',  '_form',  '_level',  '_newID', '_label')]    \n",
    "    homologue_cols = [f\"{prefix}_{specie}_homologues\" for specie in all_species]\n",
    "    \n",
    "    this_players = df_edges[['reaction_id', 'species', id_col, level_col, label_col]]# + homologue_cols]\n",
    "    \n",
    "    for _, row in this_players.iterrows():\n",
    "        if row[label_col] in helpers.plant_node_labels:\n",
    "            name = row[id_col]\n",
    "            level = row[level_col]\n",
    "            \n",
    "            members = []\n",
    "            \n",
    "            reaction_species = row['species']\n",
    "            \n",
    "            if reaction_species == 'all':\n",
    "                tmp_reaction_species = all_species\n",
    "            else:\n",
    "                tmp_reaction_species = row['species'].split(',')\n",
    "            \n",
    "            try:\n",
    "                r = translate_df.loc[(name, level)]\n",
    "                no_members = False\n",
    "            except KeyError:\n",
    "                r = translate_ignore_etc_df.loc[(name, level)]\n",
    "                no_members = True\n",
    "                #print(name, '- no members')\n",
    "            \n",
    "\n",
    "            was_in_species = []\n",
    "            for specie in all_species:\n",
    "                c = f\"{specie}_homologues\"\n",
    "                if (not no_members) and (specie in tmp_reaction_species):\n",
    "                    reaction_players[row[id_col]][row['reaction_id']][c] = r[c]\n",
    "                    members.append(r[c])\n",
    "                    was_in_species.append(specie)\n",
    "                else:\n",
    "                    reaction_players[row[id_col]][row['reaction_id']][c] = ''\n",
    "            \n",
    "            members = list_of_string_list_to_list(members)\n",
    "            \n",
    "            if ( not set(was_in_species) == set(tmp_reaction_species) ) or ( len(members) == 0 ):\n",
    "                print(f\"{id_col}\\t{row[id_col]}\\t{row['reaction_id']}\\t{level}\\t{reaction_species}\\t{tmp_reaction_species}\\t{members}\")\n",
    "                \n",
    "            reaction_players[row[id_col]][row['reaction_id']]['level'] =  level\n",
    "            reaction_players[row[id_col]][row['reaction_id']]['family'] =  r['family']\n",
    "            reaction_players[row[id_col]][row['reaction_id']]['members'] =  members\n",
    "            reaction_players[row[id_col]][row['reaction_id']]['species'] =  reaction_species\n",
    "        #else:\n",
    "            #print('---', id_col, row[label_col], row['reaction_id'])\n",
    "reaction_players = dict(reaction_players)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"FunctionalClusters_first-definition.tsv\", \"w\") as out:\n",
    "    c = '\\t'.join([f'{specie}_homologues' for specie in all_species])\n",
    "    out.write(f\"node_name\\tfamily\\treaction_id\\tlevel\\tspecies\\tmembers\\t{c}\\n\")\n",
    "    for node in list(reaction_players):\n",
    "        for reaction in list(reaction_players[node]):\n",
    "            members = reaction_players[node][reaction]['members']\n",
    "            c =  '\\t'.join([reaction_players[node][reaction][f'{specie}_homologues'] for specie in all_species])\n",
    "            out.write(f\"{node}\\t{reaction_players[node][reaction]['family']}\\t{reaction}\\t{reaction_players[node][reaction]['level']}\\t{reaction_players[node][reaction]['species']}\\t{','.join(members)}\\t{c}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "node_name\tfamily\treaction_id\tlevel\tspecies\tmembers\tath_homologues\tstu_homologues\tsly_homologues\tosa_homologues\r\n",
      "ETR\tETR\trx00005\tfamily\tath\tAT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150\tAT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340\t\t\t\r\n",
      "ETR\tETR\trx00006\tfamily\tath\tAT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150\tAT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340\t\t\t\r\n",
      "ETR\tETR\trx00008\tfamily\tath\tAT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150\tAT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340\t\t\t\r\n",
      "CTR\tCTR\trx00006\tfamily\tath\tAT5G03730\tAT5G03730\t\t\t\r\n",
      "EIN2\tEIN2\trx00009\tnode\tath\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "EIN2\tEIN2\trx00210\tnode\tath\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "EIN2\tEIN2\trx00010\tnode\tath\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "EIN2\tEIN2\trx00015\tnode\tath\tAT5G03280\tAT5G03280\t\t\t\r\n",
      "ETP\tETP\trx00011\tfamily\tath\tAT3G18910,AT3G18980\tAT3G18910,AT3G18980\t\t\t\r\n"
     ]
    }
   ],
   "source": [
    "!head FunctionalClusters_first-definition.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_clusters_df = pd.read_csv(\"FunctionalClusters_first-definition.tsv\", sep=\"\\t\")\n",
    "functional_clusters_df = functional_clusters_df.fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00009</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00210</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00010</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00015</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EIN5</td>\n",
       "      <td>XRN</td>\n",
       "      <td>rx00012</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G54490</td>\n",
       "      <td>AT1G54490</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>MAX2</td>\n",
       "      <td>F-box/LRR-repeat</td>\n",
       "      <td>rx00319</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,osa</td>\n",
       "      <td>AT2G42620</td>\n",
       "      <td>AT2G42620</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>D53</td>\n",
       "      <td>Class I Clp ATPase</td>\n",
       "      <td>rx00320</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,osa</td>\n",
       "      <td>OS11G0104300</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>OS11G0104300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>RTM3</td>\n",
       "      <td>TRAF</td>\n",
       "      <td>rx00060</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT3G58350</td>\n",
       "      <td>AT3G58350</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>NAC032</td>\n",
       "      <td>NAC</td>\n",
       "      <td>rx00157</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G77450</td>\n",
       "      <td>AT1G77450</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>BAK1</td>\n",
       "      <td>LRR-BAK1</td>\n",
       "      <td>rx00237</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G33430</td>\n",
       "      <td>AT4G33430</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name              family reaction_id level  species       members  \\\n",
       "4        EIN2                EIN2     rx00009  node      ath     AT5G03280   \n",
       "5        EIN2                EIN2     rx00210  node      ath     AT5G03280   \n",
       "6        EIN2                EIN2     rx00010  node      ath     AT5G03280   \n",
       "7        EIN2                EIN2     rx00015  node      ath     AT5G03280   \n",
       "9        EIN5                 XRN     rx00012  node      ath     AT1G54490   \n",
       "..        ...                 ...         ...   ...      ...           ...   \n",
       "462      MAX2    F-box/LRR-repeat     rx00319  node  ath,osa     AT2G42620   \n",
       "463       D53  Class I Clp ATPase     rx00320  node  ath,osa  OS11G0104300   \n",
       "518      RTM3                TRAF     rx00060  node      ath     AT3G58350   \n",
       "521    NAC032                 NAC     rx00157  node      ath     AT1G77450   \n",
       "522      BAK1            LRR-BAK1     rx00237  node      ath     AT4G33430   \n",
       "\n",
       "    ath_homologues stu_homologues sly_homologues osa_homologues  \n",
       "4        AT5G03280                                               \n",
       "5        AT5G03280                                               \n",
       "6        AT5G03280                                               \n",
       "7        AT5G03280                                               \n",
       "9        AT1G54490                                               \n",
       "..             ...            ...            ...            ...  \n",
       "462      AT2G42620                                               \n",
       "463                                                OS11G0104300  \n",
       "518      AT3G58350                                               \n",
       "521      AT1G77450                                               \n",
       "522      AT4G33430                                               \n",
       "\n",
       "[282 rows x 10 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['level']=='node']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "523\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00005</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00006</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00008</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTR</td>\n",
       "      <td>CTR</td>\n",
       "      <td>rx00006</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03730</td>\n",
       "      <td>AT5G03730</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00009</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_name family reaction_id   level species  \\\n",
       "0       ETR    ETR     rx00005  family     ath   \n",
       "1       ETR    ETR     rx00006  family     ath   \n",
       "2       ETR    ETR     rx00008  family     ath   \n",
       "3       CTR    CTR     rx00006  family     ath   \n",
       "4      EIN2   EIN2     rx00009    node     ath   \n",
       "\n",
       "                                             members  \\\n",
       "0  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "1  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "2  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "3                                          AT5G03730   \n",
       "4                                          AT5G03280   \n",
       "\n",
       "                                      ath_homologues stu_homologues  \\\n",
       "0  AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340                  \n",
       "1  AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340                  \n",
       "2  AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340                  \n",
       "3                                          AT5G03730                  \n",
       "4                                          AT5G03280                  \n",
       "\n",
       "  sly_homologues osa_homologues  \n",
       "0                                \n",
       "1                                \n",
       "2                                \n",
       "3                                \n",
       "4                                "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(functional_clusters_df.shape[0])\n",
    "functional_clusters_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "523\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00005</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00006</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ETR</td>\n",
       "      <td>ETR</td>\n",
       "      <td>rx00008</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150</td>\n",
       "      <td>AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTR</td>\n",
       "      <td>CTR</td>\n",
       "      <td>rx00006</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03730</td>\n",
       "      <td>AT5G03730</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00009</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_name family reaction_id   level species  \\\n",
       "0       ETR    ETR     rx00005  family     ath   \n",
       "1       ETR    ETR     rx00006  family     ath   \n",
       "2       ETR    ETR     rx00008  family     ath   \n",
       "3       CTR    CTR     rx00006  family     ath   \n",
       "4      EIN2   EIN2     rx00009    node     ath   \n",
       "\n",
       "                                             members  \\\n",
       "0  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "1  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "2  AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT3G23150   \n",
       "3                                          AT5G03730   \n",
       "4                                          AT5G03280   \n",
       "\n",
       "                                      ath_homologues stu_homologues  \\\n",
       "0  AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340                  \n",
       "1  AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340                  \n",
       "2  AT2G40940,AT3G04580,AT3G23150,AT1G04310,AT1G66340                  \n",
       "3                                          AT5G03730                  \n",
       "4                                          AT5G03280                  \n",
       "\n",
       "  sly_homologues osa_homologues  \n",
       "0                                \n",
       "1                                \n",
       "2                                \n",
       "3                                \n",
       "4                                "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(functional_clusters_df.shape[0])\n",
    "functional_clusters_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_clusters_df.sort_values('node_name', inplace=True)\n",
    "functional_clusters_df = functional_clusters_df.drop_duplicates(keep='first', subset=['node_name', 'members', 'level'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>CPS</td>\n",
       "      <td>CPS</td>\n",
       "      <td>rx00275</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td>SOTUB06G034690.1.1</td>\n",
       "      <td></td>\n",
       "      <td>SOTUB06G034690.1.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>CPS</td>\n",
       "      <td>CPS</td>\n",
       "      <td>rx00264</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id level species             members  \\\n",
       "424       CPS    CPS     rx00275  node     stu  SOTUB06G034690.1.1   \n",
       "423       CPS    CPS     rx00264  node     ath           AT4G02780   \n",
       "\n",
       "    ath_homologues      stu_homologues sly_homologues osa_homologues  \n",
       "424                 SOTUB06G034690.1.1                                \n",
       "423      AT4G02780                                                    "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name']=='CPS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00196</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00045</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,nbe</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>CPS</td>\n",
       "      <td>CPS</td>\n",
       "      <td>rx00275</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td>SOTUB06G034690.1.1</td>\n",
       "      <td></td>\n",
       "      <td>SOTUB06G034690.1.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>CPS</td>\n",
       "      <td>CPS</td>\n",
       "      <td>rx00264</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>EDS5</td>\n",
       "      <td>EDS5</td>\n",
       "      <td>rx00116</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G39030</td>\n",
       "      <td>AT4G39030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>EDS5</td>\n",
       "      <td>EDS5</td>\n",
       "      <td>rx00070</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT4G39030</td>\n",
       "      <td>AT4G39030</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>GA3ox</td>\n",
       "      <td>GA3ox</td>\n",
       "      <td>rx00282</td>\n",
       "      <td>clade</td>\n",
       "      <td>stu</td>\n",
       "      <td>SOTUB06G023360.1.1</td>\n",
       "      <td></td>\n",
       "      <td>SOTUB06G023360.1.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>GA3ox</td>\n",
       "      <td>GA3ox</td>\n",
       "      <td>rx00272</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G15550,AT1G80330,AT1G80340,AT4G21690</td>\n",
       "      <td>AT1G80340,AT1G80330,AT1G15550,AT4G21690</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>GID1</td>\n",
       "      <td>GID</td>\n",
       "      <td>rx00288</td>\n",
       "      <td>node</td>\n",
       "      <td>osa</td>\n",
       "      <td>LOC_OS05G33730</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LOC_OS05G33730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>GID1</td>\n",
       "      <td>GID</td>\n",
       "      <td>rx00285</td>\n",
       "      <td>clade</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT3G05120,AT3G63010,AT5G27320</td>\n",
       "      <td>AT3G63010,AT3G05120,AT5G27320</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>MYB33</td>\n",
       "      <td>MYB</td>\n",
       "      <td>rx00308</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>MYB33</td>\n",
       "      <td>MYB</td>\n",
       "      <td>rx00305</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G06100</td>\n",
       "      <td>AT5G06100</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>rx00179</td>\n",
       "      <td>clade</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G47910</td>\n",
       "      <td>AT5G47910</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>NADPH-oxidase</td>\n",
       "      <td>rx00132</td>\n",
       "      <td>family</td>\n",
       "      <td>ath,nbe</td>\n",
       "      <td>AT5G47910</td>\n",
       "      <td>AT5G47910</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>PAD4</td>\n",
       "      <td>PAD4</td>\n",
       "      <td>rx00087</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>PAD4</td>\n",
       "      <td>PAD4</td>\n",
       "      <td>rx00120</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td>AT3G52430</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PR1</td>\n",
       "      <td>PR1</td>\n",
       "      <td>rx00104</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G14580,AT2G14610</td>\n",
       "      <td>AT2G14580,AT2G14610</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PR1</td>\n",
       "      <td>PR1</td>\n",
       "      <td>rx00394</td>\n",
       "      <td>clade</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G14580,AT2G14610</td>\n",
       "      <td>AT2G14580,AT2G14610</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00259</td>\n",
       "      <td>clade</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00085</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         node_name         family reaction_id   level  species  \\\n",
       "45            COI1           COI1     rx00196  family      ath   \n",
       "43            COI1           COI1     rx00045    node  ath,nbe   \n",
       "424            CPS            CPS     rx00275    node      stu   \n",
       "423            CPS            CPS     rx00264    node      ath   \n",
       "333           EDS5           EDS5     rx00116  family      ath   \n",
       "330           EDS5           EDS5     rx00070    node      ath   \n",
       "433          GA3ox          GA3ox     rx00282   clade      stu   \n",
       "432          GA3ox          GA3ox     rx00272  family      ath   \n",
       "269           GID1            GID     rx00288    node      osa   \n",
       "266           GID1            GID     rx00285   clade      ath   \n",
       "453          MYB33            MYB     rx00308    node      stu   \n",
       "451          MYB33            MYB     rx00305    node      ath   \n",
       "213  NADPH-oxidase  NADPH-oxidase     rx00179   clade      ath   \n",
       "215  NADPH-oxidase  NADPH-oxidase     rx00132  family  ath,nbe   \n",
       "350           PAD4           PAD4     rx00087    node      ath   \n",
       "352           PAD4           PAD4     rx00120  family      ath   \n",
       "106            PR1            PR1     rx00104  family      ath   \n",
       "109            PR1            PR1     rx00394   clade      ath   \n",
       "349           SAGT           SAGT     rx00259   clade      ath   \n",
       "345           SAGT           SAGT     rx00085  family      ath   \n",
       "\n",
       "                                     members  \\\n",
       "45                                 AT2G39940   \n",
       "43                                 AT2G39940   \n",
       "424                       SOTUB06G034690.1.1   \n",
       "423                                AT4G02780   \n",
       "333                                AT4G39030   \n",
       "330                                AT4G39030   \n",
       "433                       SOTUB06G023360.1.1   \n",
       "432  AT1G15550,AT1G80330,AT1G80340,AT4G21690   \n",
       "269                           LOC_OS05G33730   \n",
       "266            AT3G05120,AT3G63010,AT5G27320   \n",
       "453                                            \n",
       "451                                AT5G06100   \n",
       "213                                AT5G47910   \n",
       "215                                AT5G47910   \n",
       "350                                AT3G52430   \n",
       "352                                AT3G52430   \n",
       "106                      AT2G14580,AT2G14610   \n",
       "109                      AT2G14580,AT2G14610   \n",
       "349                      AT2G43820,AT2G43840   \n",
       "345                      AT2G43820,AT2G43840   \n",
       "\n",
       "                              ath_homologues      stu_homologues  \\\n",
       "45                                 AT2G39940                       \n",
       "43                                 AT2G39940                       \n",
       "424                                           SOTUB06G034690.1.1   \n",
       "423                                AT4G02780                       \n",
       "333                                AT4G39030                       \n",
       "330                                AT4G39030                       \n",
       "433                                           SOTUB06G023360.1.1   \n",
       "432  AT1G80340,AT1G80330,AT1G15550,AT4G21690                       \n",
       "269                                                                \n",
       "266            AT3G63010,AT3G05120,AT5G27320                       \n",
       "453                                                                \n",
       "451                                AT5G06100                       \n",
       "213                                AT5G47910                       \n",
       "215                                AT5G47910                       \n",
       "350                                AT3G52430                       \n",
       "352                                AT3G52430                       \n",
       "106                      AT2G14580,AT2G14610                       \n",
       "109                      AT2G14580,AT2G14610                       \n",
       "349                      AT2G43820,AT2G43840                       \n",
       "345                      AT2G43820,AT2G43840                       \n",
       "\n",
       "    sly_homologues  osa_homologues  \n",
       "45                                  \n",
       "43                                  \n",
       "424                                 \n",
       "423                                 \n",
       "333                                 \n",
       "330                                 \n",
       "433                                 \n",
       "432                                 \n",
       "269                 LOC_OS05G33730  \n",
       "266                                 \n",
       "453                                 \n",
       "451                                 \n",
       "213                                 \n",
       "215                                 \n",
       "350                                 \n",
       "352                                 \n",
       "106                                 \n",
       "109                                 \n",
       "349                                 \n",
       "345                                 "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = functional_clusters_df[['node_name']].value_counts()\n",
    "x = x[x>1].index[:]\n",
    "x = [z[0] for z in x]\n",
    "functional_clusters_df[functional_clusters_df['node_name'].isin(x)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>ACX2</td>\n",
       "      <td>ACX</td>\n",
       "      <td>rx00312</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>AOX</td>\n",
       "      <td>AOX</td>\n",
       "      <td>rx00133</td>\n",
       "      <td>family</td>\n",
       "      <td>ath,nbe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>rx00078</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00059</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>GPX</td>\n",
       "      <td>GPX</td>\n",
       "      <td>rx00262</td>\n",
       "      <td>family</td>\n",
       "      <td>all</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>rx00071</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>MYB33</td>\n",
       "      <td>MYB</td>\n",
       "      <td>rx00308</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>NAC055</td>\n",
       "      <td>NAC</td>\n",
       "      <td>rx00255</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>NAC072</td>\n",
       "      <td>NAC</td>\n",
       "      <td>rx00253</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>R-gene</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00108</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>RBCS-2B</td>\n",
       "      <td>RBCS</td>\n",
       "      <td>rx00223</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>RBCS-3B</td>\n",
       "      <td>RBCS</td>\n",
       "      <td>rx00232</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Rx1</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00057</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Rx2</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00058</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>X1</td>\n",
       "      <td>X1</td>\n",
       "      <td>rx00082</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>X4</td>\n",
       "      <td>X4</td>\n",
       "      <td>rx00039</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name  family reaction_id   level  species members ath_homologues  \\\n",
       "456      ACX2     ACX     rx00312    node      ath                          \n",
       "135       AOX     AOX     rx00133  family  ath,nbe                          \n",
       "341      BA2H    BA2H     rx00078  family      ath                          \n",
       "53   GPAphid2  R-gene     rx00059    node  ath,stu                          \n",
       "254       GPX     GPX     rx00262  family      all                          \n",
       "334       IPL     IPL     rx00071  family      ath                          \n",
       "453     MYB33     MYB     rx00308    node      stu                          \n",
       "245    NAC055     NAC     rx00255    node      ath                          \n",
       "250    NAC072     NAC     rx00253    node      ath                          \n",
       "114    R-gene  R-gene     rx00108  family      ath                          \n",
       "238   RBCS-2B    RBCS     rx00223    node      ath                          \n",
       "235   RBCS-3B    RBCS     rx00232    node      ath                          \n",
       "51        Rx1  R-gene     rx00057    node  ath,stu                          \n",
       "52        Rx2  R-gene     rx00058    node  ath,stu                          \n",
       "342        X1      X1     rx00082  family      ath                          \n",
       "319        X4      X4     rx00039    node      ath                          \n",
       "\n",
       "    stu_homologues sly_homologues osa_homologues  \n",
       "456                                               \n",
       "135                                               \n",
       "341                                               \n",
       "53                                                \n",
       "254                                               \n",
       "334                                               \n",
       "453                                               \n",
       "245                                               \n",
       "250                                               \n",
       "114                                               \n",
       "238                                               \n",
       "235                                               \n",
       "51                                                \n",
       "52                                                \n",
       "342                                               \n",
       "319                                               "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['members'] == '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04580,AT2G27150,AT5G20960</td>\n",
       "      <td>AT2G27150,AT5G20960,AT1G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT5G20960]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...</td>\n",
       "      <td>AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level species  \\\n",
       "340       AAO    AAO     rx00081  family     ath   \n",
       "318       ACH    ACH     rx00038  family     ath   \n",
       "301       ACO    ACO     rx00003  family     ath   \n",
       "300       ACS    ACS     rx00002  family     ath   \n",
       "148      ACS2    ACS     rx00248    node     ath   \n",
       "\n",
       "                                               members  \\\n",
       "340                      AT1G04580,AT2G27150,AT5G20960   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...   \n",
       "148                                          AT1G01480   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340                      AT2G27150,AT5G20960,AT1G04580                  \n",
       "318                                AT2G30720,AT5G48370                  \n",
       "301  AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330                  \n",
       "300  AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...                  \n",
       "148                                          AT1G01480                  \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340                                 \n",
       "318                                 \n",
       "301                                 \n",
       "300                                 \n",
       "148                                 \n",
       "\n",
       "                               functional_cluster_name  \n",
       "340                 AAO[AT1G04580,AT2G27150,AT5G20960]  \n",
       "318                           ACH[AT2G30720,AT5G48370]  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  \n",
       "300  ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...  \n",
       "148                                    ACS2[AT1G01480]  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df.loc[:, \"functional_cluster_name\"] = functional_clusters_df.apply(lambda row: f\"{row['node_name']}[{row['members']}]\", axis=1)\n",
    "functional_clusters_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>ACX2</td>\n",
       "      <td>ACX</td>\n",
       "      <td>rx00312</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACX2[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>AOX</td>\n",
       "      <td>AOX</td>\n",
       "      <td>rx00133</td>\n",
       "      <td>family</td>\n",
       "      <td>ath,nbe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AOX[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>BA2H</td>\n",
       "      <td>BA2H</td>\n",
       "      <td>rx00078</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>BA2H[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>GPAphid2</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00059</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GPAphid2[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>GPX</td>\n",
       "      <td>GPX</td>\n",
       "      <td>rx00262</td>\n",
       "      <td>family</td>\n",
       "      <td>all</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>GPX[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>IPL</td>\n",
       "      <td>IPL</td>\n",
       "      <td>rx00071</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>IPL[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>MYB33</td>\n",
       "      <td>MYB</td>\n",
       "      <td>rx00308</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MYB33[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>NAC055</td>\n",
       "      <td>NAC</td>\n",
       "      <td>rx00255</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NAC055[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>NAC072</td>\n",
       "      <td>NAC</td>\n",
       "      <td>rx00253</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NAC072[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>R-gene</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00108</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>R-gene[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>RBCS-2B</td>\n",
       "      <td>RBCS</td>\n",
       "      <td>rx00223</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RBCS-2B[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>RBCS-3B</td>\n",
       "      <td>RBCS</td>\n",
       "      <td>rx00232</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>RBCS-3B[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Rx1</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00057</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Rx1[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Rx2</td>\n",
       "      <td>R-gene</td>\n",
       "      <td>rx00058</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,stu</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Rx2[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>X1</td>\n",
       "      <td>X1</td>\n",
       "      <td>rx00082</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>X1[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>X4</td>\n",
       "      <td>X4</td>\n",
       "      <td>rx00039</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>X4[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name  family reaction_id   level  species members ath_homologues  \\\n",
       "456      ACX2     ACX     rx00312    node      ath                          \n",
       "135       AOX     AOX     rx00133  family  ath,nbe                          \n",
       "341      BA2H    BA2H     rx00078  family      ath                          \n",
       "53   GPAphid2  R-gene     rx00059    node  ath,stu                          \n",
       "254       GPX     GPX     rx00262  family      all                          \n",
       "334       IPL     IPL     rx00071  family      ath                          \n",
       "453     MYB33     MYB     rx00308    node      stu                          \n",
       "245    NAC055     NAC     rx00255    node      ath                          \n",
       "250    NAC072     NAC     rx00253    node      ath                          \n",
       "114    R-gene  R-gene     rx00108  family      ath                          \n",
       "238   RBCS-2B    RBCS     rx00223    node      ath                          \n",
       "235   RBCS-3B    RBCS     rx00232    node      ath                          \n",
       "51        Rx1  R-gene     rx00057    node  ath,stu                          \n",
       "52        Rx2  R-gene     rx00058    node  ath,stu                          \n",
       "342        X1      X1     rx00082  family      ath                          \n",
       "319        X4      X4     rx00039    node      ath                          \n",
       "\n",
       "    stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "456                                                               ACX2[]  \n",
       "135                                                                AOX[]  \n",
       "341                                                               BA2H[]  \n",
       "53                                                            GPAphid2[]  \n",
       "254                                                                GPX[]  \n",
       "334                                                                IPL[]  \n",
       "453                                                              MYB33[]  \n",
       "245                                                             NAC055[]  \n",
       "250                                                             NAC072[]  \n",
       "114                                                             R-gene[]  \n",
       "238                                                            RBCS-2B[]  \n",
       "235                                                            RBCS-3B[]  \n",
       "51                                                                 Rx1[]  \n",
       "52                                                                 Rx2[]  \n",
       "342                                                                 X1[]  \n",
       "319                                                                 X4[]  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['members'] =='']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04580,AT2G27150,AT5G20960</td>\n",
       "      <td>AT2G27150,AT5G20960,AT1G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT5G20960]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...</td>\n",
       "      <td>AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level species  \\\n",
       "340       AAO    AAO     rx00081  family     ath   \n",
       "318       ACH    ACH     rx00038  family     ath   \n",
       "301       ACO    ACO     rx00003  family     ath   \n",
       "300       ACS    ACS     rx00002  family     ath   \n",
       "148      ACS2    ACS     rx00248    node     ath   \n",
       "\n",
       "                                               members  \\\n",
       "340                      AT1G04580,AT2G27150,AT5G20960   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...   \n",
       "148                                          AT1G01480   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340                      AT2G27150,AT5G20960,AT1G04580                  \n",
       "318                                AT2G30720,AT5G48370                  \n",
       "301  AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330                  \n",
       "300  AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...                  \n",
       "148                                          AT1G01480                  \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340                                 \n",
       "318                                 \n",
       "301                                 \n",
       "300                                 \n",
       "148                                 \n",
       "\n",
       "                               functional_cluster_name  \n",
       "340                 AAO[AT1G04580,AT2G27150,AT5G20960]  \n",
       "318                           ACH[AT2G30720,AT5G48370]  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  \n",
       "300  ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...  \n",
       "148                                    ACS2[AT1G01480]  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00259</td>\n",
       "      <td>clade</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>SAGT[AT2G43820,AT2G43840]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>SAGT</td>\n",
       "      <td>SAGT</td>\n",
       "      <td>rx00085</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td>AT2G43820,AT2G43840</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>SAGT[AT2G43820,AT2G43840]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level species              members  \\\n",
       "349      SAGT   SAGT     rx00259   clade     ath  AT2G43820,AT2G43840   \n",
       "345      SAGT   SAGT     rx00085  family     ath  AT2G43820,AT2G43840   \n",
       "\n",
       "          ath_homologues stu_homologues sly_homologues osa_homologues  \\\n",
       "349  AT2G43820,AT2G43840                                                \n",
       "345  AT2G43820,AT2G43840                                                \n",
       "\n",
       "       functional_cluster_name  \n",
       "349  SAGT[AT2G43820,AT2G43840]  \n",
       "345  SAGT[AT2G43820,AT2G43840]  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'SAGT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00196</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>COI1[AT2G39940]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>COI1</td>\n",
       "      <td>COI1</td>\n",
       "      <td>rx00045</td>\n",
       "      <td>node</td>\n",
       "      <td>ath,nbe</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td>AT2G39940</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>COI1[AT2G39940]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   node_name family reaction_id   level  species    members ath_homologues  \\\n",
       "45      COI1   COI1     rx00196  family      ath  AT2G39940      AT2G39940   \n",
       "43      COI1   COI1     rx00045    node  ath,nbe  AT2G39940      AT2G39940   \n",
       "\n",
       "   stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "45                                                      COI1[AT2G39940]  \n",
       "43                                                      COI1[AT2G39940]  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'COI1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EIN2</td>\n",
       "      <td>EIN2</td>\n",
       "      <td>rx00210</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td>AT5G03280</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>EIN2[AT5G03280]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_name family reaction_id level species    members ath_homologues  \\\n",
       "5      EIN2   EIN2     rx00210  node     ath  AT5G03280      AT5G03280   \n",
       "\n",
       "  stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "5                                                      EIN2[AT5G03280]  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'EIN2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>miR6022</td>\n",
       "      <td>miR6022</td>\n",
       "      <td>rx00308</td>\n",
       "      <td>clade</td>\n",
       "      <td>stu</td>\n",
       "      <td>miR6022-3p</td>\n",
       "      <td></td>\n",
       "      <td>miR6022-3p</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miR6022[miR6022-3p]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name   family reaction_id  level species     members ath_homologues  \\\n",
       "275   miR6022  miR6022     rx00308  clade     stu  miR6022-3p                  \n",
       "\n",
       "    stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "275     miR6022-3p                                   miR6022[miR6022-3p]  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'miR6022']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>miR159a</td>\n",
       "      <td>miR159</td>\n",
       "      <td>rx00305</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>miR159a</td>\n",
       "      <td>miR159a</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miR159a[miR159a]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name  family reaction_id level species  members ath_homologues  \\\n",
       "272   miR159a  miR159     rx00305  node     ath  miR159a        miR159a   \n",
       "\n",
       "    stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "272                                                     miR159a[miR159a]  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'miR159a']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>X4</td>\n",
       "      <td>X4</td>\n",
       "      <td>rx00039</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>X4[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id level species members ath_homologues  \\\n",
       "319        X4     X4     rx00039  node     ath                          \n",
       "\n",
       "    stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "319                                                                 X4[]  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['node_name'] == 'X4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>X4</td>\n",
       "      <td>X4</td>\n",
       "      <td>rx00039</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>X4[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id level species members ath_homologues  \\\n",
       "319        X4     X4     rx00039  node     ath                          \n",
       "\n",
       "    stu_homologues sly_homologues osa_homologues functional_cluster_name  \n",
       "319                                                                 X4[]  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df[functional_clusters_df['reaction_id'] == 'rx00039']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create functional cluster nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_clusters_import = functional_clusters_df.drop_duplicates(keep='first', subset=['functional_cluster_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(247, 11)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_import.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_labels(labels):\n",
    "\tfor x in ['Family', 'Plant', 'Foreign', 'Node']:\n",
    "\t\tif x in labels:\n",
    "\t\t\tlabels.remove(x)\n",
    "\treturn labels[0]\n",
    "\n",
    "def functional_cluster_query(file_name, \n",
    "                         labels, \n",
    "                         n_name=\"line.functional_cluster_name\"\n",
    "                        ):\n",
    "    \n",
    "    if type(labels) == list:\n",
    "        node_label = ':' + ':'.join(labels)\n",
    "    else:\n",
    "        node_label = ':' + labels\n",
    "    \n",
    "    key = {\"file_name\":file_name, \n",
    "           \"node_label\":node_label, \n",
    "           \"name\":n_name}\n",
    "\n",
    "    species_str = \"\"\n",
    "    for specie in all_species:\n",
    "        species_str += f\"{specie}_homologues:split(line.{specie}_homologues, ','),\\n                \"\n",
    "\n",
    "    \n",
    "    q = '''USING PERIODIC COMMIT 500\n",
    "           LOAD CSV WITH HEADERS FROM  'file:///{file_name}' AS line FIELDTERMINATOR '\\t'\n",
    "           CREATE (p{node_label}   {{ \n",
    "                name:{name}, \n",
    "                                \n",
    "                {species_str}\n",
    "\n",
    "                family:line.family\n",
    "   \n",
    "            }})'''.format(**key, species_str=species_str)\n",
    "    \n",
    "    return q\n",
    "\n",
    "\n",
    "\n",
    "def make_create_type_of_edge_query(file_name, edge_type,\n",
    "                           source_label=\"\", target_label=\"\",\n",
    "                           source_name=\"line.source_name\", target_name=\"line.target_name\",\n",
    "                          ):\n",
    "\n",
    "    if not source_label == \"\":\n",
    "        source_label = ':' + source_label\n",
    "        \n",
    "    if not target_label == \"\":\n",
    "        target_label = ':' + target_label\n",
    "                \n",
    "    key ={\"file_name\":file_name, \"edge_type\":edge_type,\n",
    "          \"source_label\":source_label, \"target_label\":target_label,\n",
    "          \"source_name\":source_name, \"target_name\":target_name, \n",
    "         }\n",
    "    \n",
    "    q = '''USING PERIODIC COMMIT 500\n",
    "           LOAD CSV WITH HEADERS FROM  'file:///{file_name}' AS line FIELDTERMINATOR '\\t'\n",
    "           \n",
    "           MATCH (source{source_label} {{ name:{source_name}}}),\n",
    "                 (target{target_label} {{ name:{target_name}}})\n",
    "           \n",
    "           CREATE (source)-[:{edge_type}]->(target)'''.format(**key)\n",
    "\n",
    "    return q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOAD CSV WITH HEADERS FROM 'file:///FunctionalCluster-components.tsv' AS line  FIELDTERMINATOR '\t'\n",
      "       MATCH (parent:Family { name:line.family }) RETURN parent.name AS name, labels(parent)   AS labels\n",
      "    \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04580,AT2G27150,AT5G20960</td>\n",
       "      <td>AT2G27150,AT5G20960,AT1G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT5G20960]</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...</td>\n",
       "      <td>AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    node_name family reaction_id   level species  \\\n",
       "340       AAO    AAO     rx00081  family     ath   \n",
       "318       ACH    ACH     rx00038  family     ath   \n",
       "301       ACO    ACO     rx00003  family     ath   \n",
       "300       ACS    ACS     rx00002  family     ath   \n",
       "148      ACS2    ACS     rx00248    node     ath   \n",
       "\n",
       "                                               members  \\\n",
       "340                      AT1G04580,AT2G27150,AT5G20960   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...   \n",
       "148                                          AT1G01480   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340                      AT2G27150,AT5G20960,AT1G04580                  \n",
       "318                                AT2G30720,AT5G48370                  \n",
       "301  AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330                  \n",
       "300  AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...                  \n",
       "148                                          AT1G01480                  \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340                                 \n",
       "318                                 \n",
       "301                                 \n",
       "300                                 \n",
       "148                                 \n",
       "\n",
       "                               functional_cluster_name       labels  \n",
       "340                 AAO[AT1G04580,AT2G27150,AT5G20960]  PlantCoding  \n",
       "318                           ACH[AT2G30720,AT5G48370]  PlantCoding  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  PlantCoding  \n",
       "300  ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...  PlantCoding  \n",
       "148                                    ACS2[AT1G01480]  PlantCoding  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label = \"FunctionalCluster\"\n",
    "f = f\"{label}-components.tsv\"\n",
    "functional_clusters_import.to_csv(f\"../data/import/{f}\", sep=\"\\t\", index=None)\n",
    "\n",
    "\n",
    "q = '''LOAD CSV WITH HEADERS FROM 'file:///{file_name}' AS line  FIELDTERMINATOR '\\t'\n",
    "       MATCH (parent:Family {{ name:line.family }}) RETURN parent.name AS name, labels(parent)   AS labels\n",
    "    '''.format(file_name=f)\n",
    "print(q)\n",
    "qr = graph.run(q)\n",
    "\n",
    "df = pd.DataFrame(qr.data())\n",
    "df['labels'] = df['labels'].apply(clean_labels)\n",
    "df = df.drop_duplicates(keep='first')\n",
    "df = functional_clusters_import.join(df.set_index('name'), on='family')\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PlantCoding       229\n",
       "PlantNonCoding     10\n",
       "PlantAbstract       8\n",
       "Name: labels, dtype: int64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['labels'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FunctionalCluster \t 247\n",
      "PlantAbstract \t 8 8\n",
      "PlantCoding \t 229 229\n",
      "PlantNonCoding \t 10 10\n"
     ]
    }
   ],
   "source": [
    "# save node types\n",
    "label = \"FunctionalCluster\"\n",
    "print(label, \"\\t\", df.shape[0])\n",
    "\n",
    "for t, subdf in df.groupby('labels'):\n",
    "    subdf.to_csv(f\"../data/import/{label}-{t}-components.tsv\", sep=\"\\t\", index=None)\n",
    "\n",
    "    query = functional_cluster_query(f\"{label}-{t}-components.tsv\", [label, t], n_name=\"line.functional_cluster_name\")\n",
    "    qr = graph.run(query)\n",
    "    print(t, \"\\t\", subdf.shape[0], qr.stats()['nodes_created'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_name</th>\n",
       "      <th>family</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>level</th>\n",
       "      <th>species</th>\n",
       "      <th>members</th>\n",
       "      <th>ath_homologues</th>\n",
       "      <th>stu_homologues</th>\n",
       "      <th>sly_homologues</th>\n",
       "      <th>osa_homologues</th>\n",
       "      <th>functional_cluster_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>AAO</td>\n",
       "      <td>AAO</td>\n",
       "      <td>rx00081</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G04580,AT2G27150,AT5G20960</td>\n",
       "      <td>AT2G27150,AT5G20960,AT1G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AAO[AT1G04580,AT2G27150,AT5G20960]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>ACH</td>\n",
       "      <td>ACH</td>\n",
       "      <td>rx00038</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td>AT2G30720,AT5G48370</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACH[AT2G30720,AT5G48370]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>ACO</td>\n",
       "      <td>ACO</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590</td>\n",
       "      <td>AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>family</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...</td>\n",
       "      <td>AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ACS2</td>\n",
       "      <td>ACS</td>\n",
       "      <td>rx00248</td>\n",
       "      <td>node</td>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td>AT1G01480</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ACS2[AT1G01480]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>miR319a-3p</td>\n",
       "      <td>miR319</td>\n",
       "      <td>rx00307</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td>miR319a-3p</td>\n",
       "      <td></td>\n",
       "      <td>miR319a-3p</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miR319a-3p[miR319a-3p]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>miR6022</td>\n",
       "      <td>miR6022</td>\n",
       "      <td>rx00308</td>\n",
       "      <td>clade</td>\n",
       "      <td>stu</td>\n",
       "      <td>miR6022-3p</td>\n",
       "      <td></td>\n",
       "      <td>miR6022-3p</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>miR6022[miR6022-3p]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td>rx00282</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td></td>\n",
       "      <td>phasiRNA931</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>phasiRNA931[phasiRNA931]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td>rx00276</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td></td>\n",
       "      <td>vsiRNA12986</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vsiRNA12986[vsiRNA12986]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td>rx00275</td>\n",
       "      <td>node</td>\n",
       "      <td>stu</td>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td></td>\n",
       "      <td>vsiRNA34327</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vsiRNA34327[vsiRNA34327]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>247 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       node_name       family reaction_id   level species  \\\n",
       "340          AAO          AAO     rx00081  family     ath   \n",
       "318          ACH          ACH     rx00038  family     ath   \n",
       "301          ACO          ACO     rx00003  family     ath   \n",
       "300          ACS          ACS     rx00002  family     ath   \n",
       "148         ACS2          ACS     rx00248    node     ath   \n",
       "..           ...          ...         ...     ...     ...   \n",
       "274   miR319a-3p       miR319     rx00307    node     stu   \n",
       "275      miR6022      miR6022     rx00308   clade     stu   \n",
       "263  phasiRNA931  phasiRNA931     rx00282    node     stu   \n",
       "257  vsiRNA12986  vsiRNA12986     rx00276    node     stu   \n",
       "256  vsiRNA34327  vsiRNA34327     rx00275    node     stu   \n",
       "\n",
       "                                               members  \\\n",
       "340                      AT1G04580,AT2G27150,AT5G20960   \n",
       "318                                AT2G30720,AT5G48370   \n",
       "301  AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT2G19590   \n",
       "300  AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT4G08...   \n",
       "148                                          AT1G01480   \n",
       "..                                                 ...   \n",
       "274                                         miR319a-3p   \n",
       "275                                         miR6022-3p   \n",
       "263                                        phasiRNA931   \n",
       "257                                        vsiRNA12986   \n",
       "256                                        vsiRNA34327   \n",
       "\n",
       "                                        ath_homologues stu_homologues  \\\n",
       "340                      AT2G27150,AT5G20960,AT1G04580                  \n",
       "318                                AT2G30720,AT5G48370                  \n",
       "301  AT1G62380,AT1G12010,AT2G19590,AT1G05010,AT1G77330                  \n",
       "300  AT4G26200,AT4G11280,AT3G49700,AT3G61510,AT4G37...                  \n",
       "148                                          AT1G01480                  \n",
       "..                                                 ...            ...   \n",
       "274                                                        miR319a-3p   \n",
       "275                                                        miR6022-3p   \n",
       "263                                                       phasiRNA931   \n",
       "257                                                       vsiRNA12986   \n",
       "256                                                       vsiRNA34327   \n",
       "\n",
       "    sly_homologues osa_homologues  \\\n",
       "340                                 \n",
       "318                                 \n",
       "301                                 \n",
       "300                                 \n",
       "148                                 \n",
       "..             ...            ...   \n",
       "274                                 \n",
       "275                                 \n",
       "263                                 \n",
       "257                                 \n",
       "256                                 \n",
       "\n",
       "                               functional_cluster_name  \n",
       "340                 AAO[AT1G04580,AT2G27150,AT5G20960]  \n",
       "318                           ACH[AT2G30720,AT5G48370]  \n",
       "301  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...  \n",
       "300  ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...  \n",
       "148                                    ACS2[AT1G01480]  \n",
       "..                                                 ...  \n",
       "274                             miR319a-3p[miR319a-3p]  \n",
       "275                                miR6022[miR6022-3p]  \n",
       "263                           phasiRNA931[phasiRNA931]  \n",
       "257                           vsiRNA12986[vsiRNA12986]  \n",
       "256                           vsiRNA34327[vsiRNA34327]  \n",
       "\n",
       "[247 rows x 11 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USING PERIODIC COMMIT 500\n",
      "           LOAD CSV WITH HEADERS FROM  'file:///TAKES_PART-FunctionalCluster-edges.tsv' AS line FIELDTERMINATOR '\t'\n",
      "           \n",
      "           MATCH (source:Family { name:line.family}),\n",
      "                 (target:FunctionalCluster { name:line.functional_cluster_name})\n",
      "           \n",
      "           CREATE (source)-[:TAKES_PART]->(target)\n",
      "FunctionalCluster 253 247\n",
      "\tnot all edges created\n"
     ]
    }
   ],
   "source": [
    "# component to complex edges\n",
    "edge_type = 'TAKES_PART'\n",
    "want_cols = ['functional_cluster_name', 'family']\n",
    "\n",
    "f = f'{edge_type}-{label}-edges.tsv'  \n",
    "functional_clusters_import.to_csv(f\"../data/import/{f}\", index=None, sep=\"\\t\")\n",
    "\n",
    "query = make_create_type_of_edge_query(f, edge_type, \n",
    "                       source_label='Family', target_label=label,\n",
    "                       source_name=\"line.family\", target_name=\"line.functional_cluster_name\",\n",
    "                      )\n",
    "print(query)\n",
    "qr = graph.run(query)\n",
    "\n",
    "print(label, functional_clusters_df.shape[0], qr.stats()['relationships_created'])    \n",
    "if not functional_clusters_df.shape[0] == qr.stats()['relationships_created']:\n",
    "    print(\"\\tnot all edges created\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Replace ids in edges table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys([('ETR', 'family', 'ath'), ('CTR', 'family', 'ath'), ('EIN2', 'node', 'ath'), ('ETP', 'family', 'ath'), ('EIN5', 'node', 'ath'), ('EIN3(like)', 'family', 'ath'), ('EBF', 'family', 'ath'), ('JAZ', 'family', 'ath'), ('MYC', 'family', 'ath'), ('PR3', 'family', 'ath'), ('PR4', 'family', 'ath'), ('ERF/EDF', 'family', 'ath'), ('LOX', 'family', 'ath'), ('COI1', 'node', 'ath,nbe'), ('COI1', 'node', 'ath'), ('COI1', 'family', 'ath'), ('JAM', 'family', 'ath'), ('SGT1', 'clade', 'ath'), ('NDR1', 'node', 'ath'), ('Rx1', 'node', 'ath,stu'), ('Rx2', 'node', 'ath,stu'), ('GPAphid2', 'node', 'ath,stu'), ('RTM1', 'node', 'ath'), ('CLH', 'family', 'ath'), ('JR1', 'node', 'ath'), ('PR13', 'family', 'ath'), ('VSP', 'family', 'ath'), ('ICS', 'family', 'ath'), ('EDS1', 'node', 'ath'), ('NPR1', 'node', 'ath'), ('MPK4', 'node', 'ath'), ('MPK3', 'node', 'ath'), ('MPK6', 'node', 'ath'), ('TRX-H', 'family', 'ath'), ('NIMIN', 'family', 'ath'), ('PR1', 'family', 'ath'), ('PR1', 'clade', 'ath'), ('PR2', 'node', 'ath'), ('PR5', 'family', 'ath'), ('R-gene', 'family', 'ath'), ('RBX', 'clade', 'ath'), ('MIR390', 'family', 'ath'), ('TAS3', 'family', 'ath'), ('ARF2', 'node', 'ath'), ('WRKY53', 'node', 'ath'), ('UPL5', 'node', 'ath'), ('MAPKKK8', 'node', 'ath'), ('WRKY57', 'node', 'ath'), ('AOX', 'family', 'ath,nbe'), ('ERF6', 'node', 'ath'), ('WRKY33', 'node', 'ath'), ('ACS2', 'node', 'ath'), ('ACS6', 'node', 'ath'), ('ERF5', 'node', 'ath'), ('ERF104', 'node', 'ath'), ('ERF105', 'node', 'ath'), ('CAMTA3', 'node', 'ath'), ('GRX480', 'node', 'ath'), ('TGA2,5,6', 'clade', 'ath'), ('ORA59', 'node', 'ath'), ('EIN3', 'node', 'ath'), ('DELLA', 'family', 'ath'), ('MYC2', 'node', 'ath'), ('WAK1', 'node', 'ath'), ('BIK1', 'node', 'ath'), ('CDPK', 'family', 'ath'), ('CRT3', 'node', 'ath'), ('CRT2', 'node', 'ath'), ('CRT1', 'node', 'ath'), ('CAM', 'family', 'ath'), ('RBOHD', 'node', 'ath'), ('NADPH-oxidase', 'clade', 'ath'), ('NADPH-oxidase', 'family', 'ath,nbe'), ('WRKY8', 'node', 'ath'), ('WRKY28', 'node', 'ath'), ('WRKY48', 'node', 'ath'), ('CBP60G', 'node', 'ath'), ('ERF', 'family', 'ath'), ('CRT', 'family', 'ath'), ('CML42', 'node', 'ath'), ('OBE1', 'node', 'ath'), ('RBCS-3B', 'node', 'ath'), ('RBCS-2B', 'node', 'ath'), ('MKS1', 'node', 'ath'), ('MYB44', 'node', 'ath'), ('NAC019', 'node', 'ath'), ('NAC055', 'node', 'ath'), ('NAC072', 'node', 'ath'), ('CAT', 'family', 'all'), ('CAT', 'family', 'ath'), ('APX', 'family', 'all'), ('GPX', 'family', 'all'), ('NADPH-dehydrogenase', 'family', 'all'), ('vsiRNA34327', 'node', 'stu'), ('vsiRNA12986', 'node', 'stu'), ('miR167e-3p', 'node', 'stu'), ('phasiRNA931', 'node', 'stu'), ('GID1', 'clade', 'ath'), ('GID1', 'node', 'osa'), ('SLY1', 'node', 'ath'), ('GID2', 'node', 'osa'), ('miR159a', 'node', 'ath'), ('miR159b', 'node', 'ath'), ('miR319a-3p', 'node', 'stu'), ('miR6022', 'clade', 'stu'), ('WRKY6', 'node', 'ath'), ('CAT2', 'node', 'ath'), ('D14', 'node', 'ath,osa'), ('D14', 'node', 'ath,osa,phy'), ('AHK2,3,4', 'clade', 'ath'), ('AHP6', 'clade', 'ath'), ('AHP1,2,3,4,5', 'clade', 'ath'), ('ARR-A', 'clade', 'ath'), ('ARR-B', 'clade', 'ath'), ('SAMS', 'clade', 'ath'), ('ACS', 'family', 'ath'), ('ACO', 'family', 'ath'), ('HMA', 'family', 'ath'), ('RTE1', 'family', 'ath'), ('OMR1', 'node', 'ath,nbe'), ('PDF1.2', 'clade', 'ath'), ('AOS', 'family', 'ath'), ('AOC', 'family', 'ath'), ('CTS', 'family', 'ath'), ('OPR', 'family', 'ath'), ('OPCL1', 'node', 'ath'), ('ACX', 'family', 'ath'), ('MFP', 'family', 'ath'), ('KAT', 'family', 'ath'), ('ACH', 'family', 'ath'), ('X4', 'node', 'ath'), ('JMT', 'family', 'ath'), ('CYP94', 'family', 'ath'), ('RAR1', 'node', 'ath'), ('SAG101', 'node', 'ath'), ('RIN4', 'node', 'ath'), ('MES9', 'node', 'ath,stu'), ('RANGAP', 'family', 'ath,stu'), ('RTM2', 'node', 'ath'), ('EDS5', 'node', 'ath'), ('EDS5', 'family', 'ath'), ('IPL', 'family', 'ath'), ('CM', 'family', 'ath'), ('PA', 'family', 'ath'), ('ADT', 'family', 'ath'), ('PAL', 'family', 'ath'), ('AAO', 'family', 'ath'), ('BA2H', 'family', 'ath'), ('X1', 'family', 'ath'), ('SAMC', 'family', 'ath'), ('MES', 'family', 'ath'), ('SAGT', 'family', 'ath'), ('SAGT', 'clade', 'ath'), ('PAD4', 'node', 'ath'), ('PAD4', 'family', 'ath'), ('MOS', 'family', 'ath'), ('WRKY70', 'node', 'ath'), ('CUL', 'clade', 'ath'), ('ARF3', 'node', 'ath'), ('ARF4', 'node', 'ath'), ('ESP', 'node', 'ath'), ('WRKY30', 'node', 'ath'), ('SEN4', 'node', 'ath'), ('SAG12', 'node', 'ath'), ('JAZ4', 'node', 'ath'), ('JAZ8', 'node', 'ath'), ('IAA29', 'node', 'ath'), ('SOD', 'clade', 'ath,nbe'), ('GSTU24', 'node', 'ath'), ('SR1IP1', 'node', 'ath'), ('ERF1', 'clade', 'ath'), ('MYB113', 'node', 'ath'), ('ERF096', 'node', 'ath'), ('SCL14', 'node', 'ath'), ('CPK1', 'node', 'ath'), ('CPK2', 'node', 'ath'), ('CPK4', 'node', 'ath'), ('CPK11', 'node', 'ath'), ('CPK4,5,6,11', 'clade', 'ath'), ('ICS1', 'node', 'ath'), ('SAM3', 'node', 'ath,nta'), ('SAHH', 'family', 'ath,nta'), ('AGO1,5,7,10', 'clade', 'ath,nta'), ('CPIP1,2b', 'node', 'ath,nta'), ('CPIP2a', 'node', 'ath,nta'), ('DXPS2', 'node', 'ath,nta'), ('ATPB', 'node', 'ath,nta'), ('MIND1', 'node', 'ath,nta'), ('OPR1', 'node', 'ath'), ('PAA2', 'node', 'ath'), ('PBB2', 'node', 'ath'), ('PBE1', 'node', 'ath'), ('PSAK', 'node', 'ath'), ('EDF2', 'node', 'ath'), ('CML40', 'node', 'ath'), ('TDX', 'node', 'ath'), ('RBCS-1A', 'node', 'ath'), ('RBCS-1B', 'node', 'ath'), ('RBCL', 'node', 'ath'), ('RH8', 'node', 'ath'), ('ETR1', 'node', 'ath'), ('WRKY17', 'node', 'ath'), ('WRKY11', 'node', 'ath'), ('CO', 'node', 'ath'), ('JAZ3', 'node', 'ath'), ('GST', 'family', 'ath'), ('FLS2', 'node', 'ath'), ('WRKY46', 'node', 'ath'), ('CPS', 'node', 'ath'), ('CPS', 'node', 'stu'), ('KS', 'node', 'ath'), ('KO', 'node', 'ath'), ('KAO1,2', 'clade', 'ath'), ('GA20ox', 'family', 'ath'), ('GA3ox', 'family', 'ath'), ('GA3ox', 'clade', 'stu'), ('C19-GA2ox', 'clade', 'ath'), ('GA20ox3', 'node', 'stu'), ('GA20ox1', 'node', 'stu'), ('GA20ox4', 'node', 'stu'), ('GA20ox.x3', 'node', 'stu'), ('GA20ox.x5', 'node', 'stu'), ('SLR1', 'node', 'osa'), ('GAMT1,2', 'clade', 'ath'), ('MYB33,65', 'clade', 'ath'), ('MYB33', 'node', 'ath'), ('MYB33', 'node', 'stu'), ('MYB65', 'node', 'ath'), ('LSD1', 'node', 'ath'), ('ACX2', 'node', 'ath'), ('ACX3', 'node', 'ath'), ('D27', 'node', 'ath,osa'), ('CCD7', 'node', 'ath,osa,psa'), ('CCD8', 'node', 'ath,osa,psa'), ('MAX1', 'node', 'ath,osa,sly,zma'), ('MAX2', 'node', 'ath,osa'), ('D53', 'node', 'ath,osa'), ('ISPH', 'clade', 'ath'), ('IPT4,6', 'clade', 'ath'), ('IPT2,9', 'clade', 'ath'), ('IPT1,3,5,7,8', 'clade', 'ath'), ('APT', 'clade', 'ath'), ('LOG', 'clade', 'ath'), ('ADK', 'clade', 'ath'), ('CKX', 'clade', 'ath'), ('UGT76C1,C2', 'clade', 'ath'), ('UGT85A1', 'clade', 'ath'), ('UGT73C1,C5', 'clade', 'ath'), ('CYP735A1,A2', 'clade', 'ath'), ('JAR', 'family', 'ath'), ('HSP90', 'clade', 'ath'), ('RTM3', 'node', 'ath'), ('GSNO', 'family', 'ath'), ('ASK', 'clade', 'ath'), ('NAC032', 'node', 'ath'), ('BAK1', 'node', 'ath')])"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_clusters_df = pd.read_csv(\"FunctionalClusters_first-definition.tsv\", sep=\"\\t\")\n",
    "functional_clusters_df = functional_clusters_df.fillna('')\n",
    "functional_clusters_df.loc[:, \"functional_cluster_name\"] = functional_clusters_df.apply(lambda row: f\"{row['node_name']}[{row['members']}]\", axis=1)\n",
    "\n",
    "functional_clusters_df = functional_clusters_df.drop_duplicates(keep='first', subset=['node_name', 'level', 'species', 'functional_cluster_name'])\n",
    "\n",
    "functional_clusters_translate = functional_clusters_df[['node_name', 'level', 'species', 'functional_cluster_name']].set_index(['node_name', 'level', 'species']).to_dict('index')\n",
    "functional_clusters_translate.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "def get_new_id(x):\n",
    "    specie, id_, level_, label_ = x.values\n",
    "    \n",
    "    if label_ in helpers.plant_node_labels:\n",
    "        try:\n",
    "            return functional_clusters_translate[(id_, level_, specie)]['functional_cluster_name']\n",
    "        except KeyError:\n",
    "            pass\n",
    "        \n",
    "        # if reaction has same node on different levels, then one is removed\n",
    "        return functional_clusters_translate[(id_, 'node', specie)]['functional_cluster_name']\n",
    "        print(id_, level_, label_)\n",
    "    else:\n",
    "        return id_\n",
    "\n",
    "\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    print(prefix)\n",
    "    id_col, form_col, level_col, label_col, new_id =\\\n",
    "            [prefix + x for x in ('_ID',  '_form',  '_level', '_label', '_newID')]\n",
    "\n",
    "    df_edges[new_id] = df_edges[['species', id_col, level_col, label_col]].apply(get_new_id, axis=1, result_type='expand')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Status</th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>...</th>\n",
       "      <th>reaction_type</th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>input1_label</th>\n",
       "      <th>input2_label</th>\n",
       "      <th>input3_label</th>\n",
       "      <th>output1_label</th>\n",
       "      <th>input1_newID</th>\n",
       "      <th>input2_newID</th>\n",
       "      <th>input3_newID</th>\n",
       "      <th>output1_newID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>...</td>\n",
       "      <td>catalysis</td>\n",
       "      <td>rx00001</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>SAMS[AT1G02500,AT2G36880,AT3G17390,AT4G01850]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>catalysis</td>\n",
       "      <td>rx00002</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>catalysis</td>\n",
       "      <td>rx00003</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>ACC</td>\n",
       "      <td>ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>translocation</td>\n",
       "      <td>rx00004</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>HMA[AT1G63440,AT5G44790]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cu2+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>forCB</td>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>...</td>\n",
       "      <td>protein activation</td>\n",
       "      <td>rx00005</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Status AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0  forCB      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1  forCB      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2  forCB      KG  Conn003     ath       ACC       family                  ER   \n",
       "3  forCB      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4  forCB      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level  ...       reaction_type reaction_id  \\\n",
       "0  metabolite      SAMS        clade  ...           catalysis     rx00001   \n",
       "1  metabolite       ACS       family  ...           catalysis     rx00002   \n",
       "2  metabolite       ACO       family  ...           catalysis     rx00003   \n",
       "3  metabolite       HMA       family  ...       translocation     rx00004   \n",
       "4     protein      Cu2+       family  ...  protein activation     rx00005   \n",
       "\n",
       "  input1_label input2_label input3_label output1_label  \\\n",
       "0   Metabolite  PlantCoding          NaN    Metabolite   \n",
       "1   Metabolite  PlantCoding          NaN    Metabolite   \n",
       "2   Metabolite  PlantCoding          NaN    Metabolite   \n",
       "3   Metabolite  PlantCoding          NaN    Metabolite   \n",
       "4  PlantCoding   Metabolite          NaN   PlantCoding   \n",
       "\n",
       "                                        input1_newID  \\\n",
       "0                                              L-Met   \n",
       "1                                               SAMe   \n",
       "2                                                ACC   \n",
       "3                                               Cu2+   \n",
       "4  ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...   \n",
       "\n",
       "                                        input2_newID input3_newID  \\\n",
       "0      SAMS[AT1G02500,AT2G36880,AT3G17390,AT4G01850]          NaN   \n",
       "1  ACS[AT1G01480,AT2G22810,AT3G49700,AT3G61510,AT...          NaN   \n",
       "2  ACO[AT1G05010,AT1G12010,AT1G62380,AT1G77330,AT...          NaN   \n",
       "3                           HMA[AT1G63440,AT5G44790]          NaN   \n",
       "4                                               Cu2+          NaN   \n",
       "\n",
       "                                       output1_newID  \n",
       "0                                               SAMe  \n",
       "1                                                ACC  \n",
       "2                                                 ET  \n",
       "3                                               Cu2+  \n",
       "4  ETR[AT1G04310,AT1G66340,AT2G40940,AT3G04580,AT...  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_edges['reaction_id'].unique()) == df_edges.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.to_csv(output_path / \"edges-sheet.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(output_path / \"functional_clusters.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(output_path / \"complexes_to_add.tsv\", \"w\") as out:\n",
    "    for c in complexes_to_add:\n",
    "        out.write(f\"{c}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
