{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import neo4j DB: 3/4\n",
    "\n",
    "Code to translate v2.7.4_PIS-model.xlsx to neo4j database. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import os\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to graph via docker-compose link. See http://localhost:7474/browser/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from py2neo import Graph, Node, Relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = Graph(host=\"neo4j\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_labels = [\n",
    "    'PlantCoding',\n",
    "    'PlantNonCoding',\n",
    "    'PlantAbstract',\n",
    "    'Complex',\n",
    "    'ExternalOrganism', \n",
    "    'ExternalCoding',\n",
    "    'ExternalNonCoding',\n",
    "    'Process', \n",
    "    'MetaboliteFamily',\n",
    "    'Metabolite',\n",
    "    'Gene', \n",
    "    'Expression'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "reaction_relationships = [\n",
    "    'ACTIVATE',\n",
    "    'SUBSTRATE',\n",
    "    'INHIBIT',\n",
    "    'PRODUCT',\n",
    "    'TRANSLOCATE_FROM',\n",
    "    'TRANSLOCATE_TO'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = os.path.join(\"..\", \"data\", \"raw\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sheets = [#(file, sheet_name)]\n",
    "    (\"v2.7.5_PIS-model.xlsx\", \"Reactions\"),\n",
    "    (\"v2.7.5_PIS-model.xlsx\", \"Reactions_New\"), \n",
    "    (\"Model_CK.xlsx\", \"Reactions_new\"), \n",
    "    (\"v2.7.2_PIS-model-JALR.xlsx\", \"Reactions_New\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.5_PIS-model.xlsx Reactions\n",
      "v2.7.5_PIS-model.xlsx Reactions_New\n",
      "Model_CK.xlsx Reactions_new\n",
      "v2.7.2_PIS-model-JALR.xlsx Reactions_New\n"
     ]
    }
   ],
   "source": [
    "# resave xlsx as tsv\n",
    "drops = ['Status', 'FOXMES', 'Legacy:Process', 'Legacy:ReactionMode']#, 'ConnID']\n",
    "col_rename = {\n",
    "    'AddedBy':'AddedBy',\n",
    "    'Species':'Species',\n",
    "    'ID':'input1_ID',\n",
    "    'level':'input1_level',\n",
    "    'localisation':'input1_localisation',\n",
    "    'type':'input1_type',\n",
    "    'ID.1':'input2_ID',\n",
    "    'level.1':'input2_level',\n",
    "    'localisation.1':'input2_localisation',\n",
    "    'type.1':'input2_type',\n",
    "    'ID.2':'input3_ID',\n",
    "    'level.2':'input3_level',\n",
    "    'localisation.2':'input3_localisation',\n",
    "    'type.2':'input3_type',\n",
    "    'ReactionEffect':'ReactionEffect',\n",
    "    'ReactionMode':'ReactionMode',\n",
    "    'Modifications':'Modifications',\n",
    "    'ID.3':'output1_ID',\n",
    "    'level.3':'output1_level',\n",
    "    'localisation.3':'output1_localisation',\n",
    "    'type.3':'output1_type',\n",
    "    'TrustLevel':'TrustLevel',\n",
    "    'Literature':'Literature',\n",
    "    'AdditionalInfo':'AdditionalInfo',\n",
    "    'Comment':'Comment',\n",
    "    'Model-v':'ModelV',\n",
    "    'KINETICS':'kinetics', \n",
    "    'ConnID': 'ConnID'\n",
    "}\n",
    "\n",
    "for file_name, sheet_name in sheets:\n",
    "    \n",
    "    print(file_name, sheet_name)\n",
    "    \n",
    "    file_path = os.path.join(input_path, file_name)\n",
    "    \n",
    "    base_name, extension = os.path.splitext(file_name)\n",
    "    new_file_path = os.path.join(input_path, f'{base_name}-{sheet_name}.tsv')\n",
    "    \n",
    "     if os.path.exists(new_file_path):\n",
    "        continue\n",
    "        \n",
    "    df = pd.read_excel(file_path, \n",
    "                    sheet_name=sheet_name, \n",
    "                    header=[1], \n",
    "                    dtype=str, \n",
    "                    na_values=helpers.empty_strings)\n",
    "\n",
    "    df = df[~df[\"AddedBy\"].isna()]\n",
    "    \n",
    "    if 'Status' in df.columns:\n",
    "        df = df[df['Status'].isin([\"forCB\", \"forCB_INVENTED\", np.nan])]\n",
    "\n",
    "    to_drop = list(set(drops) & set(df.columns)) + list(df.filter(regex=(\"Unnamed.*\")).columns)\n",
    "    df.drop(to_drop, axis=1, inplace=True)\n",
    "    \n",
    "    new_cols = [col_rename[x] for x in df.columns]\n",
    "    df.columns = new_cols\n",
    "    \n",
    "    df['origin'] = f'{base_name}-{sheet_name}'\n",
    "    \n",
    "    df.to_csv(new_file_path, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.5_PIS-model.xlsx Reactions\n",
      "v2.7.5_PIS-model.xlsx Reactions_New\n",
      "Model_CK.xlsx Reactions_new\n",
      "v2.7.2_PIS-model-JALR.xlsx Reactions_New\n"
     ]
    }
   ],
   "source": [
    "dfs = []\n",
    "\n",
    "for file_name, sheet_name in sheets:\n",
    "    \n",
    "    print(file_name, sheet_name)\n",
    "    base_name, extension = os.path.splitext(file_name)\n",
    "\n",
    "    file_path = os.path.join(input_path, f'{base_name}-{sheet_name}.tsv')\n",
    "    \n",
    "    df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "\n",
    "    dfs.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges = pd.concat(dfs, sort=False)\n",
    "df_edges.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>output1_level</th>\n",
       "      <th>output1_localisation</th>\n",
       "      <th>output1_type</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein [active]</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ... output1_level  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...        family   \n",
       "1  metabolite       ACS       family                  ER  ...        family   \n",
       "2  metabolite       ACO       family                  ER  ...        family   \n",
       "3  metabolite       HMA       family                  ER  ...        family   \n",
       "4     protein      Cu2+       family                  ER  ...        family   \n",
       "\n",
       "  output1_localisation      output1_type  \\\n",
       "0                   ER        metabolite   \n",
       "1                   ER        metabolite   \n",
       "2                   ER        metabolite   \n",
       "3                   ER        metabolite   \n",
       "4                   ER  protein [active]   \n",
       "\n",
       "                                    TrustLevel  \\\n",
       "0  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo Comment ModelV kinetics  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.     NaN   v1.0      NaN   \n",
       "1                 ACS catalyse ACC to SAMe reaction.     NaN   v1.0      NaN   \n",
       "2                   ACO catalyse ACC to ET reaction.     NaN   v1.0      NaN   \n",
       "3  Copper gets transported from the cytoplasm to ...     NaN   v1.0      NaN   \n",
       "4  Copper activates the membrane bound ethylene r...     NaN   v1.0      NaN   \n",
       "\n",
       "                       origin  \n",
       "0  v2.7.5_PIS-model-Reactions  \n",
       "1  v2.7.5_PIS-model-Reactions  \n",
       "2  v2.7.5_PIS-model-Reactions  \n",
       "3  v2.7.5_PIS-model-Reactions  \n",
       "4  v2.7.5_PIS-model-Reactions  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Int64Index([], dtype='int64')\n"
     ]
    }
   ],
   "source": [
    "x = df_edges[df_edges['AddedBy']=='x'].index\n",
    "print(x)\n",
    "df_edges.drop(x, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[R1] targetted experiments (e.g. Y2H, BIFC)',\n",
       "       '[Ry] invented reaction', '[Rx] incomplete/unspecific reaction',\n",
       "       '[R2] high-throughput experiment (e.g. ChIP-seq)',\n",
       "       '[R3] in-silico prediction', '[R4] indirect reaction'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[\"TrustLevel\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['trust_level']  = df_edges[\"TrustLevel\"].apply(lambda x: re.search( r\"(R[1|2|3|4|x|y]|undefined)\", x).groups()[0])\n",
    "df_edges['observed_species'] = df_edges['Species'].apply(helpers.lower_string)\n",
    "df_edges['species_also_observed_in'] = df_edges[\"Species\"].apply(helpers.rest_of_items)\n",
    "df_edges['Comment'] = df_edges['Comment'].fillna(\"\")\n",
    "df_edges['AdditionalInfo'] = df_edges['AdditionalInfo'].fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['KG', 'MZ', 'ZR', 'MPE', 'ACR', 'MAK', 'ST', 'SB', 'AG', 'JALR'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['AddedBy'] = df_edges['AddedBy'].apply(lambda x: x.upper())\n",
    "df_edges[\"AddedBy\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['v1.0', 'v2.5', 'v2.7', 'v2.6', 'vNA'], dtype=object)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.loc[df_edges['ModelV'].isna(), 'ModelV'] = 'vNA'\n",
    "df_edges['ModelV'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_non_ascii(x):\n",
    "    x = str(x)\n",
    "    has_nonascii = False\n",
    "    for character in x:\n",
    "        if not character.isascii():\n",
    "            has_nonascii = True\n",
    "            print(f\"|{character}|\", ord(character), character.encode())\n",
    "    if has_nonascii:\n",
    "        print(x)\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "ascii_replacers = {\n",
    "    b'\\xc2\\xa0' : b\" \",           # funky WIN whitespace\n",
    "    b'\\xe2\\x80\\xa6': b'...',      # …\n",
    "    b'\\xe2\\x80\\x8b' : b'',        # have no figging clue\n",
    "    b'\\xe2\\x80\\x93' : b'-',       # –\n",
    "    b'\\xce\\xb1' : b\"alpha\",       # α\n",
    "    b'\\xc3\\x9f' : b\"beta\",        # ß\n",
    "    b'\\xce\\xb2' : b\"beta\",        # β\n",
    "    b'\\xe2\\x80\\x98' : b\"prime\",   # ‘\n",
    "    b'\\xe2\\x80\\x99' : b\"prime\",   # ’\n",
    "    b'\\xc2\\xb4': b'prime',        # ´\n",
    "    # Sorry accents :(\n",
    "    b'\\xc5\\xa0' : b\"S\",           # Š\n",
    "    b'\\xc5\\xa1' : b's',           # š\n",
    "    b'\\xc5\\xbd' : b'Z',           # Ž\n",
    "    b'\\xc4\\x8d' : b'c'            # č\n",
    "}\n",
    "\n",
    "\n",
    "def replacer(x):\n",
    "    if not x in [' ', '', np.nan]:\n",
    "        x = x.encode('utf-8')\n",
    "        for old, new in ascii_replacers.items():\n",
    "            x = x.replace(old, new)\n",
    "        return x.decode('utf-8')\n",
    "    return ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AddedBy \n",
      "-------------\n",
      "|Š| 352 b'\\xc5\\xa0'\n",
      "ŠT\n",
      "\n",
      "ConnID \n",
      "-------------\n",
      "\n",
      "Species \n",
      "-------------\n",
      "\n",
      "input1_ID \n",
      "-------------\n",
      "|´| 180 b'\\xc2\\xb4'\n",
      "9-cis-b-apo-10´-carotenal\n",
      "\n",
      "input1_level \n",
      "-------------\n",
      "\n",
      "input1_localisation \n",
      "-------------\n",
      "\n",
      "input1_type \n",
      "-------------\n",
      "\n",
      "input2_ID \n",
      "-------------\n",
      "\n",
      "input2_level \n",
      "-------------\n",
      "\n",
      "input2_localisation \n",
      "-------------\n",
      "\n",
      "input2_type \n",
      "-------------\n",
      "\n",
      "input3_ID \n",
      "-------------\n",
      "\n",
      "input3_level \n",
      "-------------\n",
      "\n",
      "input3_localisation \n",
      "-------------\n",
      "\n",
      "input3_type \n",
      "-------------\n",
      "\n",
      "ReactionEffect \n",
      "-------------\n",
      "\n",
      "ReactionMode \n",
      "-------------\n",
      "\n",
      "Modifications \n",
      "-------------\n",
      "\n",
      "output1_ID \n",
      "-------------\n",
      "|´| 180 b'\\xc2\\xb4'\n",
      "9-cis-b-apo-10´-carotenal\n",
      "\n",
      "output1_level \n",
      "-------------\n",
      "\n",
      "output1_localisation \n",
      "-------------\n",
      "\n",
      "output1_type \n",
      "-------------\n",
      "\n",
      "TrustLevel \n",
      "-------------\n",
      "\n",
      "Literature \n",
      "-------------\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.​1104/​pp.​107.​104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent…)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.​1104/​pp.​107.​104299 (RTE1 Is a Golgi-Associated and ETR1-Dependent…)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1073/pnas.0605528103 (ETHYLENE-INSENSITIVE5 encodes…) | DOI:10.1073/pnas.1214848109 (CTR1 phosphorylates the central regulator EIN2...) | DOI:10.1093/mp/ssq036 (New insight in ethylene signaling...) | DOI:10.1016/j.pbi.2008.06.011 (Ethylene signaling...) | DOI:10.1093/mp/ssr042 (Paradigms and paradox in the ethylene signaling...) | DOI:10.1042/BJ20091102 (EIN2, the central regulator of ethylene signalling...)\n",
      "|–| 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1104/pp.106.095588 (Chen, 2007: Stability of Plant Defense Proteins...); DOI:10.1105/tpc.106.041103 Plant Cell. 2006 Nov; 18(11): 3303–3320.\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1126/scisignal.3109cm4 (Arabidopsis Jasmonate Signaling Pathway) | DOI:10.​1105/​tpc.​110.​080788 (The Arabidopsis bHLH Transcription Factors...) | DOI:10.1016/j.jtbi.2010.12.037 (Transient pulse formation...) | DOI:10.1146/annurev.arplant.043008.092007 (Jasmonate Passes Muster...) | DOI:10.1007/978-0-387-85498-4_8 (Regulation of Secondary Metabolism...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1111/j.1399-3054.2004.00418.x (Jasmonate-regulated Arabidopsis stress signalling network) | DOI:10.​1105/​tpc.​106.​048017 (MYC2 Differentially Modulates Diverse...)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1094/MPMI-23-9-1151 (Manosalva, 2010; : Methyl esterase 1 (StMES1) is required...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "AraCyc:PWY-981 | DOI:​10.​1104/​pp.​103.​2.​315 (Pathway of Salicylic Acid Biosynthesis...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.​1105/​tpc.​104.​027441 (Interaction of NIMIN1 with NPR1...) | DOI:10.1016/S1369-5266(03)00058-X (The salicylic acid loop in plant defense)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:​10.​1105/​tpc.​015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/415977a (MAP kinase signalling cascade in...) | DOI:​10.​1105/​tpc.​015552 (Silencing of the Mitogen-Activated...) | DOI:10.1007/s00425-010-1308-5 (A broad-spectrum, efficient and nontransgenic...)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.113.117838 (Jiang, 2014: Arabidopsis WRKY57 functions as a node of convergence...)\n",
      "|–| 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126–42\n",
      "|–| 8211 b'\\xe2\\x80\\x93'\n",
      "DOI:10.1105/tpc.112.109074 (Meng, 2013: Phosphorylation of an ERF ...) Meng X, Xu J, He Y, Yang K-Y, Mordorski B., et al. 2013. Phosphorylation of an ERF transcription factor by Arabidopsis MPK3/MPK6 regulates plant defense gene induction and fungal resistance. Plant Cell 25(3):1126–42\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...) | PMID:24989234 (Zander, 2014: TGA Transcription Factors Activate the SA-Suppressible Branch ...)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1093/mp/ssr113 (Zander, 2012: Repression of the Arabidopsis thaliana JA/ET-induced defense...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.​1105/​tpc.​108.​058974 (Fode, 2008: The Arabidopsis GRAS Protein SCL14 Interacts...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:​10.1105/tpc.113.121731 (Qi, 2014: Arabidopsis DELLA and JAZ Proteins...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.​1105/​tpc.​112.​101428 (Wild, 2012: The Arabidopsis DELLA RGA-LIKE3...)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases… ker R-gene unknown also this interaction postulated\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.​1104/​pp.​111.​192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "|​| 8203 b'\\xe2\\x80\\x8b'\n",
      "DOI:10.1038/nature07612 (Du, 2009: Ca2+_calmodulin regulates ...) | DOI:10.​1104/​pp.​111.​192310 (Nie, 2012: SR1, a Calmodulin-Binding Transcription Factor ...)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.3390/plants3010160 Functions of Calcium-Dependent Protein Kinases…\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "DOI:10.1105/tpc.108.065193. \n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "DOI:10.1016/j.chom.2012.04.014. (coronatine promotes….)\n",
      "\n",
      "AdditionalInfo \n",
      "-------------\n",
      "|–| 8211 b'\\xe2\\x80\\x93'\n",
      "|β| 946 b'\\xce\\xb2'\n",
      "Acyl-CoA oxidase: In Arabidopsis, five genes appear to en-code acyl-CoA oxidases (ACX1–ACX5, with varying chain-length specificities). Mutant analysis indicated that ACX1 and ACX5 act redundantly in initiating β-oxidation of OPC-8:0-CoA. Beta-oxidation itself involves three core enzymes, acyl-CoA oxi-dase (ACX), multifunctional protein (MFP; comprising enoyl-CoA hydratase and b-hydroxy-acyl-CoA dehydrogenase activities), and 3-ketoacyl-CoA thiolase (KAT).\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "bHLH transcription factors EGL3, GL3, and TT8 ; MYB factors MYB75 and GL1;WD-repeat protein TTG1\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "Upregulating the expression of late anthocyanin biosynthetic genes: including NADPH-dependent dihydroflavonol reductase(DFR), leucoanthocyanidin dioxygenase (LDOX), and UDP-Glc:flavonoid 3-Oglucosyltransferase (UF3GT).\n",
      "|…| 8230 b'\\xe2\\x80\\xa6'\n",
      "Targets to proteasome (er ligase); HECT domain E3 ubiquitin ligase UPL5 … UPL5 regulates leaf senescence in Arabidopsis through degradation of WRKY53 and ensures that senescence is executed in the correct time frame.\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "MEKK1/MAPKKK8 (=MAP3K) AT4G08500 \n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "based on evidence that the activity of CBP60, WRKY8/28/48, and CAMTA3/SR1 factors are modulated by calcium dependent protein kinases (CDPKs) and calmodulin (CaM; Du et al., 2009; Gao et al., 2013; Truman et al., 2013)\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "Control of salicylic acid synthesis and systemic acquired \n",
      "|β| 946 b'\\xce\\xb2'\n",
      "Original reaction had ntaCF1-beta (P00826 in SwissProt; Z00044.2 in GenBank), closest match in TAIR10 is ATCG00480 (ATPB). NtCF1-beta is a subunit of chloroplast ATP synthase. HC-Pro may interfere with the assembly of NtCF1β -subunit into the whole enzyme. HC-Pro from PVY with Y1H.\n",
      "| | 160 b'\\xc2\\xa0'\n",
      "Control of salicylic acid synthesis and systemic acquired \n",
      "|´| 180 b'\\xc2\\xb4'\n",
      "cZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase\n",
      "|´| 180 b'\\xc2\\xb4'\n",
      "DHZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase\n",
      "|´| 180 b'\\xc2\\xb4'\n",
      "tZ synthesis; enzyme 5´- ribonucleotide phosphohydrolase\n",
      "|´| 180 b'\\xc2\\xb4'\n",
      "iP synthesis; enzyme 5´- ribonucleotide phosphohydrolase\n",
      "|´| 180 b'\\xc2\\xb4'\n",
      "CCD7 cleaves 9-cis-b-carotene into 9-cis-b-apo-10´-carotenal (Output_1) and b-ionone (Output_2)\n",
      "\n",
      "Comment \n",
      "-------------\n",
      "|š| 353 b'\\xc5\\xa1'\n",
      "|Ž| 381 b'\\xc5\\xbd'\n",
      "|č| 269 b'\\xc4\\x8d'\n",
      "deaktivacija aktivnih GA-vprašaj Živo če ok\n",
      "|š| 353 b'\\xc5\\xa1'\n",
      "|Ž| 381 b'\\xc5\\xbd'\n",
      "|č| 269 b'\\xc4\\x8d'\n",
      "deaktivacija aktivnih GA-vprašaj Živo če ok\n",
      "\n",
      "ModelV \n",
      "-------------\n",
      "\n",
      "kinetics \n",
      "-------------\n",
      "\n",
      "origin \n",
      "-------------\n",
      "\n",
      "trust_level \n",
      "-------------\n",
      "\n",
      "also_observed_in \n",
      "-------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bad_cols = []\n",
    "for c in df_edges.columns:\n",
    "    print(c, \"\\n-------------\")\n",
    "    if any(df_edges[c].apply(get_non_ascii)):\n",
    "        bad_cols.append(c)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in bad_cols:\n",
    "    df_edges[c] = df_edges[c].apply(replacer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def doi_list(x):\n",
    "    x = x.lower()\n",
    "    match = re.findall(\"(?:doi)(?:\\:|\\/)\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"doi:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "def pubmed_list(x):\n",
    "    x =x.lower()\n",
    "    match = re.findall(\"(?:pmid)\\:\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"pmid:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no/bad reference Conn040 v2.7.5_PIS-model-Reactions \n",
      "no/bad reference Conn118 v2.7.5_PIS-model-Reactions \n",
      "no/bad reference Conn120 v2.7.5_PIS-model-Reactions \n",
      "no/bad reference Conn122 v2.7.5_PIS-model-Reactions \n",
      "no/bad reference Conn182 v2.7.5_PIS-model-Reactions \n",
      "no/bad reference Conn183 v2.7.5_PIS-model-Reactions \n",
      "no/bad reference Conn199 v2.7.5_PIS-model-Reactions ? Kg need to find reference\n",
      "no/bad reference Conn273 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn274 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn275 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn276 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn276 v2.7.5_PIS-model-Reactions_New  10.1073/pnas.98.4.2065\n",
      "no/bad reference Conn277 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn278 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn279 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn280 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn281 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn282 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn283 v2.7.5_PIS-model-Reactions_New KEGG \n",
      "no/bad reference Conn309 v2.7.5_PIS-model-Reactions_New \n"
     ]
    }
   ],
   "source": [
    "# format literature sources\n",
    "df_edges[\"Literature\"].fillna(value=\"\", inplace=True)\n",
    "for i, row in df_edges.iterrows():\n",
    "    s = row['Literature']\n",
    "    source = doi_list(s)\n",
    "    source += pubmed_list(s)\n",
    "    for z in s.split(\"|\"):\n",
    "        key = z.lower()\n",
    "        if \":\" in key:\n",
    "            if \"aracyc\" in key:\n",
    "                aracyc_string = \"aracyc:\" + z.split(\":\")[1].strip()\n",
    "                source.append(aracyc_string)\n",
    "            elif \"kegg\" in key:\n",
    "                kegg_string = \"kegg:\" + z.split(\":\")[1].strip()\n",
    "                source.append(kegg_string)\n",
    "            elif \"doi\" in key:\n",
    "                # already fetched\n",
    "                continue\n",
    "            elif (\"pubmed\" in key) or (\"pmid\" in key):\n",
    "                pmid_string = \"pubmed:\" + z.split(\":\")[1].strip()\n",
    "        elif \"invented\" in key:\n",
    "            source.append(\"invented\")\n",
    "        else:\n",
    "            print(\"no/bad reference\", row['ConnID'], row['origin'], z)\n",
    "            source.append(\"other:\" + z.strip())\n",
    "    if len(source) > 0:\n",
    "        df_edges.loc[i, \"literature_sources\"] = helpers.list_to_string(source)\n",
    "    else:\n",
    "        print(row['ConnID'], row['origin'], z)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges[[\"origin\", \"ConnID\", \"Literature\", \"literature_sources\"]].to_csv(\"lit-check.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_df = df_edges.copy()\n",
    "#df_edges = save_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>origin</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td></td>\n",
       "      <td>aracyc:ETHYL-PWY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td></td>\n",
       "      <td>aracyc:ETHYL-PWY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td></td>\n",
       "      <td>aracyc:ETHYL-PWY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v2.7.5_PIS-model-Reactions</td>\n",
       "      <td>R1</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "                                    TrustLevel  \\\n",
       "0  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo Comment ModelV kinetics  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.           v1.0      NaN   \n",
       "1                 ACS catalyse ACC to SAMe reaction.           v1.0      NaN   \n",
       "2                   ACO catalyse ACC to ET reaction.           v1.0      NaN   \n",
       "3  Copper gets transported from the cytoplasm to ...           v1.0      NaN   \n",
       "4  Copper activates the membrane bound ethylene r...           v1.0      NaN   \n",
       "\n",
       "                       origin trust_level also_observed_in  \\\n",
       "0  v2.7.5_PIS-model-Reactions          R1                    \n",
       "1  v2.7.5_PIS-model-Reactions          R1                    \n",
       "2  v2.7.5_PIS-model-Reactions          R1                    \n",
       "3  v2.7.5_PIS-model-Reactions          R1                    \n",
       "4  v2.7.5_PIS-model-Reactions          R1                    \n",
       "\n",
       "       literature_sources  \n",
       "0        aracyc:ETHYL-PWY  \n",
       "1        aracyc:ETHYL-PWY  \n",
       "2        aracyc:ETHYL-PWY  \n",
       "3  doi:10.1105/tpc.001768  \n",
       "4  doi:10.1105/tpc.001768  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in ['input1', 'input2', 'input3', 'output1']:\n",
    "    df_edges.loc[:, x + \"_ID\"] = df_edges[x + \"_ID\"].apply(helpers.reorder_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein, Clade, Family keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PlantCoding 161\n",
      "PlantNonCoding 9\n",
      "PlantAbstract 7\n",
      "Complex 1\n",
      "ExternalOrganism 3\n",
      "ExternalCoding 14\n",
      "ExternalNonCoding 0\n",
      "Process 6\n",
      "MetaboliteFamily 11\n",
      "Metabolite 102\n",
      "Gene 0\n",
      "Expression 0\n"
     ]
    }
   ],
   "source": [
    "node_dict = {}\n",
    "for label in node_labels:\n",
    "    q = '''MATCH (n:%s) RETURN DISTINCT n.name'''%label\n",
    "    s = set([d['n.name'] for d in graph.run(q).data()])\n",
    "    print(label, len(s))\n",
    "    node_dict[label] = s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_species = ['ath', 'stu', 'osa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = os.path.join(input_path, \"bio_elements.tsv\")\n",
    "df_bioelements = pd.read_csv(file_name, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>chebi_identifier</th>\n",
       "      <th>pubmed_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SAM</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>AT1G02500</td>\n",
       "      <td>SAM1</td>\n",
       "      <td>use</td>\n",
       "      <td>SAM synthetase [EC:2.5.1.6]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_000621</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>S-adenosylmethionine synthetase 1</td>\n",
       "      <td>SAM1</td>\n",
       "      <td>AtSAM1,MAT1,METK1,SAM-1,SAM1</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SAM</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>AT4G01850</td>\n",
       "      <td>SAM2</td>\n",
       "      <td>use</td>\n",
       "      <td>SAM synthetase [EC:2.5.1.6]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_000621</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>S-adenosylmethionine synthetase 2</td>\n",
       "      <td>SAM2</td>\n",
       "      <td>AtSAM2,MAT2,SAM-2,SAM2</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SAM</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>AT2G36880</td>\n",
       "      <td>SAM3</td>\n",
       "      <td>use</td>\n",
       "      <td>SAM synthetase [EC:2.5.1.6]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_000621</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>methionine adenosyltransferase 3</td>\n",
       "      <td>METK3</td>\n",
       "      <td>MAT3,METK3</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SAM</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>AT3G17390</td>\n",
       "      <td>SAM4</td>\n",
       "      <td>use</td>\n",
       "      <td>SAM synthetase [EC:2.5.1.6]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_000621</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>S-adenosylmethionine synthetase family protein</td>\n",
       "      <td>METK4</td>\n",
       "      <td>MAT4,METK4,MTO3,SAMS3</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACS</td>\n",
       "      <td>ACS1,2,6</td>\n",
       "      <td>AT3G61510</td>\n",
       "      <td>ACS1</td>\n",
       "      <td>use</td>\n",
       "      <td>ACC synthase [EC:4.4.1.14]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_000133</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>ACC synthase 1</td>\n",
       "      <td>ACS1</td>\n",
       "      <td>ACC2,ACS1,AT-ACS1</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy Species      NodeType Family     Clade     NodeID NodeName  \\\n",
       "0      KG     ath  plant_coding    SAM      SAMS  AT1G02500     SAM1   \n",
       "1      KG     ath  plant_coding    SAM      SAMS  AT4G01850     SAM2   \n",
       "2      KG     ath  plant_coding    SAM      SAMS  AT2G36880     SAM3   \n",
       "3      KG     ath  plant_coding    SAM      SAMS  AT3G17390     SAM4   \n",
       "4      KG     ath  plant_coding    ACS  ACS1,2,6  AT3G61510     ACS1   \n",
       "\n",
       "  ModelStatus              NodeDescription AdditionalInfo  ...  \\\n",
       "0         use  SAM synthetase [EC:2.5.1.6]            NaN  ...   \n",
       "1         use  SAM synthetase [EC:2.5.1.6]            NaN  ...   \n",
       "2         use  SAM synthetase [EC:2.5.1.6]            NaN  ...   \n",
       "3         use  SAM synthetase [EC:2.5.1.6]            NaN  ...   \n",
       "4         use   ACC synthase [EC:4.4.1.14]            NaN  ...   \n",
       "\n",
       "         GMM_OCD1                                            GMM_OCD  \\\n",
       "0  OCD_all_000621  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "1  OCD_all_000621  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "2  OCD_all_000621  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "3  OCD_all_000621  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "4  OCD_all_000133  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "\n",
       "                                  GMM_Description GMM_ShortName  \\\n",
       "0               S-adenosylmethionine synthetase 1          SAM1   \n",
       "1               S-adenosylmethionine synthetase 2          SAM2   \n",
       "2                methionine adenosyltransferase 3         METK3   \n",
       "3  S-adenosylmethionine synthetase family protein         METK4   \n",
       "4                                  ACC synthase 1          ACS1   \n",
       "\n",
       "                   GMM_Synonyms    NodeLabel observed_species  \\\n",
       "0  AtSAM1,MAT1,METK1,SAM-1,SAM1  PlantCoding              ath   \n",
       "1        AtSAM2,MAT2,SAM-2,SAM2  PlantCoding              ath   \n",
       "2                    MAT3,METK3  PlantCoding              ath   \n",
       "3         MAT4,METK4,MTO3,SAMS3  PlantCoding              ath   \n",
       "4             ACC2,ACS1,AT-ACS1  PlantCoding              ath   \n",
       "\n",
       "  also_observed_in chebi_identifier pubmed_identifier  \n",
       "0              NaN              NaN               NaN  \n",
       "1              NaN              NaN               NaN  \n",
       "2              NaN              NaN               NaN  \n",
       "3              NaN              NaN               NaN  \n",
       "4              NaN              NaN               NaN  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bioelements.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pick_the_set(x):\n",
    "    for v in x:\n",
    "        if type(v) == set:\n",
    "            return v\n",
    "    return {}\n",
    "\n",
    "\n",
    "def get_species_homologues(level):\n",
    "    df_level_species  = df_bioelements.groupby([level, 'Species']).agg({\n",
    "      'NodeID':lambda x: set(x), \n",
    "    })\n",
    "    df_level_species.reset_index(1, inplace=True)\n",
    "\n",
    "    cols = []\n",
    "    for specie in all_species:\n",
    "        col = specie + '_homologues'\n",
    "        cols.append(col)\n",
    "        df_level_species.loc[df_level_species['Species'] == specie, col] =\\\n",
    "         df_level_species[df_level_species['Species'] == specie]['NodeID']\n",
    "    \n",
    "    df_level_species = df_level_species[cols].groupby(level).agg({\n",
    "        y:pick_the_set for y in cols\n",
    "    })\n",
    "    \n",
    "    df_level = df_bioelements.groupby(level).agg({\n",
    "          'AddedBy':lambda x:list(x)[0], \n",
    "          'NodeLabel':lambda x:list(x)[0], \n",
    "          'NodeDescription':lambda x: ', '.join(list(set(x))), \n",
    "          'AdditionalInfo':lambda x: helpers.list_to_string(x), \n",
    "          'Process':lambda x:list(x)[0], \n",
    "          'ModelV':helpers.get_latest_model, \n",
    "          'Species':lambda x: set(x), \n",
    "    })    \n",
    "\n",
    "    df_level = df_level.join(df_level_species[cols])\n",
    "    df_level.reset_index(inplace=True)\n",
    "    \n",
    "    return df_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CPS', 'GA20ox3', 'GA20ox1', 'GA20ox4', 'MYB33'], dtype=object)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_to_name = df_bioelements[['NodeID', 'NodeName']]\n",
    "id_to_name[id_to_name['NodeName'].duplicated()]['NodeName'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_nodes = get_species_homologues('NodeName')\n",
    "df_nodes.set_index('NodeName', inplace=True)\n",
    "\n",
    "df_families = get_species_homologues(\"Family\")\n",
    "\n",
    "# node name to node IDs\n",
    "node_ids_key = {}\n",
    "for species in all_species:\n",
    "    col = species + '_homologues'\n",
    "    node_ids_key[col] = df_nodes[col].to_dict()\n",
    "df_clades = get_species_homologues('Clade')\n",
    "df_clades.set_index('Clade', inplace=True)\n",
    "\n",
    "# clade name to node IDs\n",
    "clade_ids_key = {}\n",
    "for species in all_species:\n",
    "    col = species + '_homologues'\n",
    "    clade_ids_key[col] = df_clades[col].to_dict() \n",
    "\n",
    "# family name to node IDs\n",
    "df_families.set_index(\"Family\", inplace=True)\n",
    "family_ids_key = {}\n",
    "for species in all_species:\n",
    "    col = species + '_homologues'\n",
    "    family_ids_key[col] = df_families[col].to_dict() \n",
    "\n",
    "node_to_family = df_bioelements[[\"NodeName\", \"Family\"]].set_index(\"NodeName\")[\"Family\"].to_dict()\n",
    "clade_to_family = df_bioelements[[\"Clade\", \"Family\"]].set_index(\"Clade\")[\"Family\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_node_to_family(x):\n",
    "    # pathogen proteins are listed as proteins, so cannot use dict\n",
    "    id_, type_, level_, ConnID, origin = x.values\n",
    "      \n",
    "    if id_ in helpers.empty_strings:\n",
    "        return np.nan, np.nan\n",
    "\n",
    "    new_label = None\n",
    "    family_id = None\n",
    "    \n",
    "    ########################\n",
    "    # Simple Cases\n",
    "    ########################\n",
    "    if type_ in ['complex', 'complex [active]', 'complex [activated]', 'complex [inactive]', 'plant_complex']:\n",
    "        if not (id_ in node_dict[\"Complex\"]):\n",
    "            complexes_to_add.append(id_)\n",
    "        new_label = 'Complex'\n",
    "        family_id = id_\n",
    "    \n",
    "    elif type_ in [\"metabolite\"]:\n",
    "        for label in [\"Metabolite\", \"MetaboliteFamily\"]:\n",
    "            if id_ in node_dict[label]:\n",
    "                new_label = label\n",
    "                break\n",
    "        if not new_label:\n",
    "            missing_in_components.update([id_])        \n",
    "            print(origin, \"|\", ConnID, \"|\", id_, \"|\", type_, \"|\", level_, \" |a (label) not a listed metabolite\")\n",
    "        family_id = id_\n",
    "\n",
    "    elif type_ in ['process']:\n",
    "        if (id_ in node_dict[\"Process\"]):\n",
    "            family_id = id_\n",
    "            new_label = \"Process\"\n",
    "        else:\n",
    "            print(origin, \"|\", ConnID, \"|\", id_, \"|\", type_, \"|\", level_,  \" |b (label) process not a listed process\")\n",
    "        \n",
    "    else:\n",
    "        ########################\n",
    "        # family ID\n",
    "        ########################\n",
    "        check_external = False\n",
    "        if level_ == \"family\":    \n",
    "            family_id = id_\n",
    "        elif level_ in [\"clade\", \"clade/orthologue\"]:\n",
    "            try:\n",
    "                family_id = clade_to_family[id_]\n",
    "                replace_w_family.update([id_])\n",
    "            except KeyError:\n",
    "                check_external = True\n",
    "        elif level_ == \"node\":\n",
    "            try:\n",
    "                family_id = node_to_family[id_]\n",
    "                replace_w_family.update([id_])\n",
    "            except KeyError:\n",
    "                check_external = True\n",
    "        \n",
    "        if check_external:\n",
    "            for label in [\"ExternalOrganism\", \"ExternalCoding\", \"ExternalNonCoding\"]:\n",
    "                if id_ in node_dict[label]:\n",
    "                    new_label = label\n",
    "                    family_id = id_\n",
    "                    break\n",
    "                    \n",
    "        if not family_id:\n",
    "            print(origin, \"|\", ConnID, \"|\", id_, \"|\", type_, \"|\", level_,  \" |c (family id) could not convert to family/external\")\n",
    "                \n",
    "        ########################\n",
    "        # Label\n",
    "        ########################\n",
    "        if (family_id) and (not new_label):\n",
    "            id_labels = [] #looping just in case an id occurs mutiple times\n",
    "            for label in node_labels:\n",
    "                if family_id in node_dict[label]:\n",
    "                    id_labels.append(label)\n",
    "\n",
    "            if len(id_labels) == 1:\n",
    "                new_label = id_labels[0]\n",
    "            elif len(id_labels) > 1:\n",
    "                print(origin, \"|\", ConnID, \"|\", id_, \"|\", type_, \"|\", level_,  \" |d (label) many labels fit\")\n",
    "                missing_in_components.update([id_])       \n",
    "\n",
    "            else:\n",
    "                print(origin, \"|\", ConnID, \"|\", id_, \"|\", type_, \"|\", level_,  \" |e (label) could not find label\")\n",
    "                missing_in_components.update([id_])        \n",
    "\n",
    "    return family_id, new_label\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model_CK-Reactions_new | nan | DHZRMP | metabolite | node  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | DHZRMP | metabolite | node  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | metabolite | clade  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | metabolite | clade  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | metabolite | clade  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | ARR-B(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn273 | All-trans-b-carotene | metabolite | family  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn274 | 9-cis-b-carotene | metabolite | node  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn275 | 9-cis-b-apo-10prime-carotenal | metabolite | node  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn276 | CL | metabolite | node  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn277 | SLs | metabolite | family  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn278 | D14 | protein [activated] | family  |e (label) could not find label\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn279 | D53 | protein [activated] | family  |e (label) could not find label\n",
      "v2.7.5_PIS-model-Reactions_New | Conn308 | miR6022 | plant_ncRNA | node  |c (family id) could not convert to family/external\n",
      "v2.7.5_PIS-model-Reactions_New | Conn310 | LSD1 | plant_coding | node  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | ARR-B(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | ARR-A(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHP1,2,3,4,5(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHP1,2,3,4,5(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn276 | MAX1 | protein [activated] | family  |e (label) could not find label\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn277 | D14 | protein [inactivated] | family  |e (label) could not find label\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn278 | MAX2 | protein [activated] | family  |e (label) could not find label\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn279 | D14 | protein [activated] | family  |e (label) could not find label\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn279 | MAX2 | protein [activated] | nan  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | DHZRMP | metabolite | node  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | iPOG | metabolite | node  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | iPROG | metabolite | node  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | DHZRMP | metabolite | node  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | DHZRMP | metabolite | node  |a (label) not a listed metabolite\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHK2,3,4(a) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | AHP1,2,3,4,5(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | ARR-A(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | ABB-B | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | ARR-A(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "Model_CK-Reactions_new | nan | ARR-B(a)(p) | protein | clade  |c (family id) could not convert to family/external\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn273 | 9-cis-b-carotene | metabolite | node  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn274 | 9-cis-b-apo-10prime-carotenal | metabolite | node  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn275 | CL | metabolite | node  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn276 | CLA | metabolite | node  |a (label) not a listed metabolite\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn277 | D14 | protein [activated] | family  |e (label) could not find label\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn278 | MAX2 | protein [activated] | family  |e (label) could not find label\n",
      "v2.7.2_PIS-model-JALR-Reactions_New | Conn279 | D53 | protein [inactivated] | family  |e (label) could not find label\n"
     ]
    }
   ],
   "source": [
    "complexes_to_add = []\n",
    "missing_in_components = set()\n",
    "replace_w_family = set()\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, level_col, new_id, new_label_col  =\\\n",
    "            [prefix + x for x in ('_ID',  '_type',  '_level',  '_newID', '_label')]    \n",
    "    \n",
    "    df_edges[[new_id, new_label_col]] = df_edges[[id_col, type_col, level_col, 'ConnID', 'origin']].apply(convert_node_to_family, axis=1, result_type='expand')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "      <th>input1_newID</th>\n",
       "      <th>input1_label</th>\n",
       "      <th>input2_newID</th>\n",
       "      <th>input2_label</th>\n",
       "      <th>input3_newID</th>\n",
       "      <th>input3_label</th>\n",
       "      <th>output1_newID</th>\n",
       "      <th>output1_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>aracyc:ETHYL-PWY</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>SAM</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>aracyc:ETHYL-PWY</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>aracyc:ETHYL-PWY</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ET</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.001768</td>\n",
       "      <td>ETR</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETR</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  also_observed_in      literature_sources input1_newID input1_label  \\\n",
       "0                         aracyc:ETHYL-PWY        L-Met   Metabolite   \n",
       "1                         aracyc:ETHYL-PWY         SAMe   Metabolite   \n",
       "2                         aracyc:ETHYL-PWY          ACC   Metabolite   \n",
       "3                   doi:10.1105/tpc.001768         Cu2+   Metabolite   \n",
       "4                   doi:10.1105/tpc.001768          ETR  PlantCoding   \n",
       "\n",
       "  input2_newID input2_label input3_newID input3_label output1_newID  \\\n",
       "0          SAM  PlantCoding          NaN          NaN          SAMe   \n",
       "1          ACS  PlantCoding          NaN          NaN           ACC   \n",
       "2          ACO  PlantCoding          NaN          NaN            ET   \n",
       "3          HMA  PlantCoding          NaN          NaN          Cu2+   \n",
       "4         Cu2+   Metabolite          NaN          NaN           ETR   \n",
       "\n",
       "  output1_label  \n",
       "0    Metabolite  \n",
       "1    Metabolite  \n",
       "2    Metabolite  \n",
       "3    Metabolite  \n",
       "4   PlantCoding  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_orthologues(x, prefix=\"\"):\n",
    "    id_, level_, label_ = x.values\n",
    "    ########################\n",
    "    # Specie specific homologues\n",
    "    ########################\n",
    "    \n",
    "    return_D = {f\"{prefix}_{specie}_homologues\":\"\" for specie in all_species}\n",
    "    \n",
    "    if label_ in ['PlantCoding', 'PlantNonCoding',  'PlantAbstract']:\n",
    "\n",
    "        for species in all_species:\n",
    "            species = f\"{species}_homologues\"\n",
    "            k = f\"{prefix}_{species}\"\n",
    "            if level_ == 'node':\n",
    "                return_D[k] = node_ids_key[species][id_]\n",
    "            elif level_ == 'clade':\n",
    "                return_D[k] =  clade_ids_key[species][id_]\n",
    "            elif level_ == 'family':\n",
    "                return_D[k] =  family_ids_key[species][id_]\n",
    "\n",
    "    return_D = {x:helpers.list_to_string(list(return_D[x])) for x in return_D}\n",
    "    return return_D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "new_dfs = []\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    print(prefix)\n",
    "    id_col, level_col, new_label_col  =\\\n",
    "                [prefix + x for x in ('_ID',  '_level',  '_label')]        \n",
    "    \n",
    "    new_df = df_edges[[id_col, level_col, new_label_col ]].apply(get_orthologues, axis=1, result_type='expand', prefix=prefix)\n",
    "    new_dfs.append(new_df)\n",
    "    #df_edges = df_edges.join(new_df, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "homologues_df = pd.concat(new_dfs, sort=False, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "input1_ath_homologues                                            \n",
       "input1_osa_homologues                                            \n",
       "input1_stu_homologues                                            \n",
       "input2_ath_homologues     AT4G01850,AT3G17390,AT1G02500,AT2G36880\n",
       "input2_osa_homologues                                            \n",
       "input2_stu_homologues                                            \n",
       "input3_ath_homologues                                            \n",
       "input3_osa_homologues                                            \n",
       "input3_stu_homologues                                            \n",
       "output1_ath_homologues                                           \n",
       "output1_osa_homologues                                           \n",
       "output1_stu_homologues                                           \n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homologues_df.loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges = df_edges.join(homologues_df, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>input1_stu_homologues</th>\n",
       "      <th>input2_ath_homologues</th>\n",
       "      <th>input2_osa_homologues</th>\n",
       "      <th>input2_stu_homologues</th>\n",
       "      <th>input3_ath_homologues</th>\n",
       "      <th>input3_osa_homologues</th>\n",
       "      <th>input3_stu_homologues</th>\n",
       "      <th>output1_ath_homologues</th>\n",
       "      <th>output1_osa_homologues</th>\n",
       "      <th>output1_stu_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT4G01850,AT3G17390,AT1G02500,AT2G36880</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT3G61510,AT4G08040,AT2G22810,AT1G62960,AT5G51...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT2G19590,AT1G05010,AT1G77330,AT1G62380,AT1G12010</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT1G63440,AT4G33520,AT5G44790,AT5G21930</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AT1G66340,AT3G04580,AT1G04310,AT2G40940,AT3G23150</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  input1_stu_homologues                              input2_ath_homologues  \\\n",
       "0                                  AT4G01850,AT3G17390,AT1G02500,AT2G36880   \n",
       "1                        AT3G61510,AT4G08040,AT2G22810,AT1G62960,AT5G51...   \n",
       "2                        AT2G19590,AT1G05010,AT1G77330,AT1G62380,AT1G12010   \n",
       "3                                  AT1G63440,AT4G33520,AT5G44790,AT5G21930   \n",
       "4                                                                            \n",
       "\n",
       "  input2_osa_homologues input2_stu_homologues input3_ath_homologues  \\\n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "\n",
       "  input3_osa_homologues input3_stu_homologues  \\\n",
       "0                                               \n",
       "1                                               \n",
       "2                                               \n",
       "3                                               \n",
       "4                                               \n",
       "\n",
       "                              output1_ath_homologues output1_osa_homologues  \\\n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                                                             \n",
       "4  AT1G66340,AT3G04580,AT1G04310,AT2G40940,AT3G23150                          \n",
       "\n",
       "  output1_stu_homologues  \n",
       "0                         \n",
       "1                         \n",
       "2                         \n",
       "3                         \n",
       "4                         \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "node_type_to_node_form_dict = {\n",
    "    \"gene\":\"gene\",\n",
    "    \n",
    "    \"protein\":\"protein\",\n",
    "    \"protein [inactivated]\":\"protein\", \n",
    "    \"protein [activated]\":\"protein_active\",\n",
    "    'protein [active]': \"protein_active\",\n",
    "    \n",
    "    \"ncRNA\":\"ncRNA\",\n",
    "    \"plant_ncRNA\":\"ncRNA\",\n",
    "    'ta-siRNA':\"ta-siRNA\", \n",
    "        \n",
    "    \"complex\":\"complex\", \n",
    "    \"plant_complex\":\"complex\",\n",
    "    'complex [active]': \"complex_active\",\n",
    "    \n",
    "    \"metabolite\":\"metabolite\",\n",
    "    \n",
    "    \"process\":\"process\", \n",
    "    'process [active]':\"process_active\",\n",
    "\n",
    "    np.nan:\"\", \n",
    "    \"plant_coding\":\"unknown\"\n",
    "}\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, new_form_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_form')]\n",
    "    \n",
    "    x = df_edges[[id_col, type_col]].dropna(how='all')\n",
    "    \n",
    "    print(prefix)\n",
    "    df_edges[new_form_col] = df_edges[type_col].apply(lambda x: node_type_to_node_form_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>input3_ath_homologues</th>\n",
       "      <th>input3_osa_homologues</th>\n",
       "      <th>input3_stu_homologues</th>\n",
       "      <th>output1_ath_homologues</th>\n",
       "      <th>output1_osa_homologues</th>\n",
       "      <th>output1_stu_homologues</th>\n",
       "      <th>input1_form</th>\n",
       "      <th>input2_form</th>\n",
       "      <th>input3_form</th>\n",
       "      <th>output1_form</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AT1G66340,AT3G04580,AT1G04310,AT2G40940,AT3G23150</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>protein</td>\n",
       "      <td>metabolite</td>\n",
       "      <td></td>\n",
       "      <td>protein_active</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  input3_ath_homologues input3_osa_homologues input3_stu_homologues  \\\n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "\n",
       "                              output1_ath_homologues output1_osa_homologues  \\\n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                                                             \n",
       "4  AT1G66340,AT3G04580,AT1G04310,AT2G40940,AT3G23150                          \n",
       "\n",
       "  output1_stu_homologues input1_form     input2_form input3_form  \\\n",
       "0                         metabolite  protein_active               \n",
       "1                         metabolite  protein_active               \n",
       "2                         metabolite  protein_active               \n",
       "3                         metabolite         protein               \n",
       "4                            protein      metabolite               \n",
       "\n",
       "     output1_form  \n",
       "0      metabolite  \n",
       "1      metabolite  \n",
       "2      metabolite  \n",
       "3      metabolite  \n",
       "4  protein_active  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.5_PIS-model-Reactions Conn024 input1 PR3 nan\n",
      "v2.7.5_PIS-model-Reactions Conn025 input1 PR4 nan\n",
      "v2.7.5_PIS-model-Reactions Conn066 input1 CLH nan\n",
      "v2.7.5_PIS-model-Reactions Conn067 input1 JR1 nan\n",
      "v2.7.5_PIS-model-Reactions Conn068 input1 PR13 nan\n",
      "v2.7.5_PIS-model-Reactions Conn069 input1 VSP nan\n",
      "v2.7.5_PIS-model-Reactions Conn080 input1 Anthocyanin-accumulation nan\n",
      "v2.7.5_PIS-model-Reactions Conn080 input1 Trichome-initiation nan\n",
      "v2.7.5_PIS-model-Reactions Conn104 input1 PR1 nan\n",
      "v2.7.5_PIS-model-Reactions Conn105 input1 PR2 nan\n",
      "v2.7.5_PIS-model-Reactions Conn106 input1 PR5 nan\n",
      "v2.7.5_PIS-model-Reactions Conn126 input1 WRKY53 nan\n",
      "v2.7.5_PIS-model-Reactions Conn172 input1 R-gene nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn273 input1 Geranylgeranyl-PP  nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn274 input1 ent-Copalyl-PP nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn275 input1 ent-Kaurene nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn276 input1 ent-Kaurenoic acid nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn277 input1 GA12 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn278 input1 GA12 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn279 input1 GA9 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn280 input1 GA20 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn281 input1 GA20 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn282 input1 GA1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn283 input1 GA4 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn284 input1 vsiRNA34327 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn285 input1 vsiRNA12986 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn286 input1 miR167e-3p nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn287 input1 miR167e-3p nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn288 input1 miR167e-3p nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn289 input1 miR167e-3p nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn290 input1 miR167e-3p nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn291 input1 phasiRNA931 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GID1a,b,c nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GID1a,b,c nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GID1a,b,c nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn292 input1 GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn293 input1 GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn294 input1 GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn295 input1 GA1-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn296 input1 GA3-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn297 input1 GA4-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GA1-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GA3-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GA4-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 SLY1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GID2 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GA1-GID1-DELLA nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input1 GA1-GID1-SLR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn299 input1 GA4 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn300 input1 GA3 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn301 input1 GA1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn302 input1 GA4 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn303 input1 GA3 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn304 input1 GA1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn305 input1 miR159a nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn306 input1 miR159b nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn307 input1 miR319a-3p nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn308 input1 MYB33 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn309 input1 WRKY6 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input1 CAT nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input1 SA nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input1 CAT2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input1 CAT2 nan\n",
      "Model_CK-Reactions_new nan input1 HMBDP nan\n",
      "Model_CK-Reactions_new nan input1 DMAPP nan\n",
      "Model_CK-Reactions_new nan input1 DMAPP nan\n",
      "Model_CK-Reactions_new nan input1 DMAPP nan\n",
      "Model_CK-Reactions_new nan input1 DMAPP nan\n",
      "Model_CK-Reactions_new nan input1 DMAPP nan\n",
      "Model_CK-Reactions_new nan input1 DMAPP nan\n",
      "Model_CK-Reactions_new nan input1 prenyl-tRNA nan\n",
      "Model_CK-Reactions_new nan input1 cis-prenyl-tRNA nan\n",
      "Model_CK-Reactions_new nan input1 cZRMP nan\n",
      "Model_CK-Reactions_new nan input1 cZRMP nan\n",
      "Model_CK-Reactions_new nan input1 cZR nan\n",
      "Model_CK-Reactions_new nan input1 cZR nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 cZOG nan\n",
      "Model_CK-Reactions_new nan input1 cZROG nan\n",
      "Model_CK-Reactions_new nan input1 DHZRMP nan\n",
      "Model_CK-Reactions_new nan input1 DHZRMP nan\n",
      "Model_CK-Reactions_new nan input1 DHZR nan\n",
      "Model_CK-Reactions_new nan input1 DHZR nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZOG nan\n",
      "Model_CK-Reactions_new nan input1 DHZROG nan\n",
      "Model_CK-Reactions_new nan input1 tZRMP nan\n",
      "Model_CK-Reactions_new nan input1 tZRMP nan\n",
      "Model_CK-Reactions_new nan input1 tZR nan\n",
      "Model_CK-Reactions_new nan input1 tZR nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 tZOG nan\n",
      "Model_CK-Reactions_new nan input1 tZROG nan\n",
      "Model_CK-Reactions_new nan input1 iPRMP nan\n",
      "Model_CK-Reactions_new nan input1 iPRMP nan\n",
      "Model_CK-Reactions_new nan input1 iPR nan\n",
      "Model_CK-Reactions_new nan input1 iPR nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 iPRTP nan\n",
      "Model_CK-Reactions_new nan input1 iPRDP nan\n",
      "Model_CK-Reactions_new nan input1 iPRMP nan\n",
      "Model_CK-Reactions_new nan input1 tZRMP nan\n",
      "Model_CK-Reactions_new nan input1 iPRTP nan\n",
      "Model_CK-Reactions_new nan input1 iPRDP nan\n",
      "Model_CK-Reactions_new nan input1 tZRTP nan\n",
      "Model_CK-Reactions_new nan input1 tZRDP nan\n",
      "Model_CK-Reactions_new nan input1 iP nan\n",
      "Model_CK-Reactions_new nan input1 tZ nan\n",
      "Model_CK-Reactions_new nan input1 cZ nan\n",
      "Model_CK-Reactions_new nan input1 DHZ nan\n",
      "Model_CK-Reactions_new nan input1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan input1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan input1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan input1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan input1 AHP6 nan\n",
      "Model_CK-Reactions_new nan input1 AHP1,2,3,4,5 nan\n",
      "Model_CK-Reactions_new nan input1 ARR-A nan\n",
      "Model_CK-Reactions_new nan input1 ARR-B(a)(p) nan\n",
      "Model_CK-Reactions_new nan input1 ARR-A nan\n",
      "Model_CK-Reactions_new nan input1 ARR-B nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn273 input1 All-trans-b-carotene nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn274 input1 9-cis-b-carotene nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn275 input1 9-cis-b-apo-10prime-carotenal nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn276 input1 CL nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn277 input1 SLs nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn278 input1 D14 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn279 input1 D53 nan\n",
      "v2.7.5_PIS-model-Reactions Conn024 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn025 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn056 input2 SAG101 nan\n",
      "v2.7.5_PIS-model-Reactions Conn066 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn067 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn068 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn069 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn080 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn080 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn104 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn105 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn106 input2 potyvirus nan\n",
      "v2.7.5_PIS-model-Reactions Conn126 input2 UPL5 nan\n",
      "v2.7.5_PIS-model-Reactions Conn172 input2 Ca2+ nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn273 input2 CPS nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn274 input2 KS nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn275 input2 KO nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn276 input2 KAO1,2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn277 input2 GA20ox nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn278 input2 GA20ox nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn279 input2 GA3ox nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn280 input2 GA3ox nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn281 input2 GA3ox nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn282 input2 C19-GA2ox nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn283 input2 C19-GA2ox nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn284 input2 CPS nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn285 input2 GA20ox3 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn286 input2 GA20ox3 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn287 input2 GA20ox1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn288 input2 GA20ox4 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn289 input2 Sotub09g017710.1.1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn290 input2 Sotub10g011620.1.1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn291 input2 GA3ox nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 GA1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 GA3 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 GA4 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn292 input2 GA1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn293 input2 GA3 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn294 input2 GA4 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn295 input2 SLR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn296 input2 SLR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn297 input2 SLR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 DELLA nan\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.5_PIS-model-Reactions_New nan input2 DELLA nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 DELLA nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 SCF nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 SCF nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 SCF-SLY1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan input2 SCF-GID2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn299 input2 GAMT1,2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn300 input2 GAMT1,2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn301 input2 GAMT1,2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn302 input2 GAMYB nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn303 input2 GAMYB nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn304 input2 GAMYB nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn305 input2 MYB33 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn306 input2 MYB65 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn307 input2 MYB33 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn308 input2 miR6022 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn309 input2 NPR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input2 LSD1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input2 CAT2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input2 ACX2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn310 input2 ACX3 nan\n",
      "Model_CK-Reactions_new nan input2 ISPH nan\n",
      "Model_CK-Reactions_new nan input2 IPT4,6 nan\n",
      "Model_CK-Reactions_new nan input2 IPT2,9 nan\n",
      "Model_CK-Reactions_new nan input2 IPT1,3,5,7,8 nan\n",
      "Model_CK-Reactions_new nan input2 IPT4,6 nan\n",
      "Model_CK-Reactions_new nan input2 IPT1,3,5,7,8 nan\n",
      "Model_CK-Reactions_new nan input2 IPT1,3,5,7,8 nan\n",
      "Model_CK-Reactions_new nan input2 APT nan\n",
      "Model_CK-Reactions_new nan input2 LOG nan\n",
      "Model_CK-Reactions_new nan input2 ADK nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 UGT73C1,C5 nan\n",
      "Model_CK-Reactions_new nan input2 UGT73C1,C5 nan\n",
      "Model_CK-Reactions_new nan input2 APT nan\n",
      "Model_CK-Reactions_new nan input2 LOG nan\n",
      "Model_CK-Reactions_new nan input2 ADK nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 UGT73C1,C5 nan\n",
      "Model_CK-Reactions_new nan input2 UGT73C1,C5 nan\n",
      "Model_CK-Reactions_new nan input2 APT nan\n",
      "Model_CK-Reactions_new nan input2 LOG nan\n",
      "Model_CK-Reactions_new nan input2 ADK nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 UGT73C1,C5 nan\n",
      "Model_CK-Reactions_new nan input2 UGT73C1,C5 nan\n",
      "Model_CK-Reactions_new nan input2 APT nan\n",
      "Model_CK-Reactions_new nan input2 LOG nan\n",
      "Model_CK-Reactions_new nan input2 ADK nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 CKX nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT76C1,C2 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 UGT85A1 nan\n",
      "Model_CK-Reactions_new nan input2 CYP735A1,A2 nan\n",
      "Model_CK-Reactions_new nan input2 CYP735A1,A2 nan\n",
      "Model_CK-Reactions_new nan input2 CYP735A1,A2 nan\n",
      "Model_CK-Reactions_new nan input2 tZ nan\n",
      "Model_CK-Reactions_new nan input2 cZ nan\n",
      "Model_CK-Reactions_new nan input2 DZ nan\n",
      "Model_CK-Reactions_new nan input2 iP nan\n",
      "Model_CK-Reactions_new nan input2 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan input2 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan input2 ARR-B(a)(p) nan\n",
      "Model_CK-Reactions_new nan input2 ARR-A(a)(p) nan\n",
      "Model_CK-Reactions_new nan input2 AHP1,2,3,4,5(a)(p) nan\n",
      "Model_CK-Reactions_new nan input2 AHP1,2,3,4,5(a)(p) nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn273 input2 D27 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn274 input2 CCD7 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn275 input2 CCD8 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn276 input2 MAX1 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn277 input2 D14 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn278 input2 MAX2 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn279 input2 D14 nan\n",
      "v2.7.5_PIS-model-Reactions Conn042 input3 JAR nan\n",
      "Model_CK-Reactions_new nan input3 ATP nan\n",
      "Model_CK-Reactions_new nan input3 tRNA nan\n",
      "Model_CK-Reactions_new nan input3 ATP nan\n",
      "Model_CK-Reactions_new nan input3 ADP nan\n",
      "Model_CK-Reactions_new nan input3 ADP nan\n",
      "Model_CK-Reactions_new nan input3 AMP nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "Model_CK-Reactions_new nan input3 Glu nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn279 input3 MAX2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn273 output1 ent-Copalyl-PP nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn274 output1 ent-Kaurene nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn275 output1 ent-Kaurenoic acid nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn276 output1 GA12 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn277 output1 GA9 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn278 output1 GA20 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn279 output1 GA4 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn280 output1 GA3 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn281 output1 GA1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn282 output1 GA8 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn283 output1 GA34 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 GA1-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 GA3-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 GA4-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn292 output1 GA1-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn293 output1 GA3-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn294 output1 GA4-GID1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn295 output1 GA1-GID1-SLR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn296 output1 GA3-GID1-SLR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn297 output1 GA4-GID1-SLR1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 GA1-GID1-DELLA nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 GA3-GID1-DELLA nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 GA4-GID1-DELLA nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 SCF-SLY1 nan\n",
      "v2.7.5_PIS-model-Reactions_New nan output1 SCF-GID2 nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn299 output1 GA4-methylester nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn300 output1 GA3-methylester nan\n",
      "v2.7.5_PIS-model-Reactions_New Conn301 output1 GA1-methylester nan\n",
      "Model_CK-Reactions_new nan output1 DMAPP nan\n",
      "Model_CK-Reactions_new nan output1 iPRTP nan\n",
      "Model_CK-Reactions_new nan output1 prenyl-tRNA nan\n",
      "Model_CK-Reactions_new nan output1 iPRTP nan\n",
      "Model_CK-Reactions_new nan output1 iPRDP nan\n",
      "Model_CK-Reactions_new nan output1 iPRDP nan\n",
      "Model_CK-Reactions_new nan output1 iPRMP nan\n",
      "Model_CK-Reactions_new nan output1 cis-prenyl-tRNA nan\n",
      "Model_CK-Reactions_new nan output1 cZRMP nan\n",
      "Model_CK-Reactions_new nan output1 cZR nan\n",
      "Model_CK-Reactions_new nan output1 cZ nan\n",
      "Model_CK-Reactions_new nan output1 cZ nan\n",
      "Model_CK-Reactions_new nan output1 cZRMP nan\n",
      "Model_CK-Reactions_new nan output1 cZR nan\n",
      "Model_CK-Reactions_new nan output1 Ade nan\n",
      "Model_CK-Reactions_new nan output1 Ado nan\n",
      "Model_CK-Reactions_new nan output1 cZ7G nan\n",
      "Model_CK-Reactions_new nan output1 cZ9G nan\n",
      "Model_CK-Reactions_new nan output1 cZOG nan\n",
      "Model_CK-Reactions_new nan output1 cZROG nan\n",
      "Model_CK-Reactions_new nan output1 cZROG nan\n",
      "Model_CK-Reactions_new nan output1 cZOG nan\n",
      "Model_CK-Reactions_new nan output1 cZ nan\n",
      "Model_CK-Reactions_new nan output1 cZ nan\n",
      "Model_CK-Reactions_new nan output1 DHZR nan\n",
      "Model_CK-Reactions_new nan output1 DHZ nan\n",
      "Model_CK-Reactions_new nan output1 DHZ nan\n",
      "Model_CK-Reactions_new nan output1 DHZRMP nan\n",
      "Model_CK-Reactions_new nan output1 DHZR nan\n",
      "Model_CK-Reactions_new nan output1 Ade nan\n",
      "Model_CK-Reactions_new nan output1 Ado nan\n",
      "Model_CK-Reactions_new nan output1 DHZ7G nan\n",
      "Model_CK-Reactions_new nan output1 DHZ9G nan\n",
      "Model_CK-Reactions_new nan output1 DHZOG nan\n",
      "Model_CK-Reactions_new nan output1 DHZROG nan\n",
      "Model_CK-Reactions_new nan output1 DHZROG nan\n",
      "Model_CK-Reactions_new nan output1 DHZOG nan\n",
      "Model_CK-Reactions_new nan output1 DHZ nan\n",
      "Model_CK-Reactions_new nan output1 DHZ nan\n",
      "Model_CK-Reactions_new nan output1 tZR nan\n",
      "Model_CK-Reactions_new nan output1 tZ nan\n",
      "Model_CK-Reactions_new nan output1 tZ nan\n",
      "Model_CK-Reactions_new nan output1 tZRMP nan\n",
      "Model_CK-Reactions_new nan output1 tZR nan\n",
      "Model_CK-Reactions_new nan output1 Ade nan\n",
      "Model_CK-Reactions_new nan output1 Ado nan\n",
      "Model_CK-Reactions_new nan output1 tZ7G nan\n",
      "Model_CK-Reactions_new nan output1 tZ9G nan\n",
      "Model_CK-Reactions_new nan output1 tZOG nan\n",
      "Model_CK-Reactions_new nan output1 tZROG nan\n",
      "Model_CK-Reactions_new nan output1 tZROG nan\n",
      "Model_CK-Reactions_new nan output1 tZOG nan\n",
      "Model_CK-Reactions_new nan output1 tZ nan\n",
      "Model_CK-Reactions_new nan output1 tZ nan\n",
      "Model_CK-Reactions_new nan output1 iPR nan\n",
      "Model_CK-Reactions_new nan output1 iP nan\n",
      "Model_CK-Reactions_new nan output1 iP nan\n",
      "Model_CK-Reactions_new nan output1 iPRMP nan\n",
      "Model_CK-Reactions_new nan output1 iPR nan\n",
      "Model_CK-Reactions_new nan output1 Ade nan\n",
      "Model_CK-Reactions_new nan output1 Ado nan\n",
      "Model_CK-Reactions_new nan output1 iP7G nan\n",
      "Model_CK-Reactions_new nan output1 iP9G nan\n",
      "Model_CK-Reactions_new nan output1 iPOG nan\n",
      "Model_CK-Reactions_new nan output1 iPROG nan\n",
      "Model_CK-Reactions_new nan output1 tZRTP nan\n",
      "Model_CK-Reactions_new nan output1 tZRDP nan\n",
      "Model_CK-Reactions_new nan output1 tZRMP nan\n",
      "Model_CK-Reactions_new nan output1 DHZRMP nan\n",
      "Model_CK-Reactions_new nan output1 iPRDP nan\n",
      "Model_CK-Reactions_new nan output1 iPRMP nan\n",
      "Model_CK-Reactions_new nan output1 tZRDP nan\n",
      "Model_CK-Reactions_new nan output1 tZRMP nan\n",
      "Model_CK-Reactions_new nan output1 iPRMP nan\n",
      "Model_CK-Reactions_new nan output1 tZRMP nan\n",
      "Model_CK-Reactions_new nan output1 cZRMP nan\n",
      "Model_CK-Reactions_new nan output1 DHZRMP nan\n",
      "Model_CK-Reactions_new nan output1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan output1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan output1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan output1 AHK2,3,4(a) nan\n",
      "Model_CK-Reactions_new nan output1 AHP1,2,3,4,5(a)(p) nan\n",
      "Model_CK-Reactions_new nan output1 ARR-A(a)(p) nan\n",
      "Model_CK-Reactions_new nan output1 ABB-B nan\n",
      "Model_CK-Reactions_new nan output1 ARR-A(a)(p) nan\n",
      "Model_CK-Reactions_new nan output1 ARR-B(a)(p) nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn273 output1 9-cis-b-carotene nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn274 output1 9-cis-b-apo-10prime-carotenal nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn275 output1 CL nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn276 output1 CLA nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn277 output1 D14 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn278 output1 MAX2 nan\n",
      "v2.7.2_PIS-model-JALR-Reactions_New Conn279 output1 D53 nan\n"
     ]
    }
   ],
   "source": [
    "node_localisation_dict = {\n",
    "    'nuc':'nucleus',\n",
    "    'er':'endoplasmic reticulum',\n",
    "    'golgi':'golgi apparatus', \n",
    "    'mitochondria?': 'putative:mitochondrion'\n",
    "}\n",
    "\n",
    "\n",
    "good_localisations = set([\n",
    "     'nucleus',\n",
    "     'nucleolus',\n",
    "     'cytoplasm',\n",
    "     'vacuole',\n",
    "     'endoplasmic reticulum',\n",
    "     'chloroplast',\n",
    "     'mitochondrion',\n",
    "     'golgi apparatus',\n",
    "     'peroxisome',\n",
    "     'apoplast',\n",
    "     'extracellular'\n",
    "])\n",
    "\n",
    "good_localisations.update(['putative:' + s for s in good_localisations])\n",
    "\n",
    "\n",
    "def node_localisation_std(x):\n",
    "    if not type(x) == str:\n",
    "        return \"\"\n",
    "    \n",
    "    x = x.lower()\n",
    "    if x in node_localisation_dict.keys():\n",
    "        x = node_localisation_dict[x]\n",
    "    \n",
    "    if x in good_localisations:\n",
    "        return x\n",
    "    else:\n",
    "        return \"\"\n",
    "\n",
    "node_localisations = set()\n",
    "new_localisation = set()\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, localisation_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_localisation')]\n",
    "    \n",
    "    x = df_edges[['ConnID', 'origin', id_col, localisation_col]].dropna(how='all', subset=[ id_col, localisation_col])\n",
    "    \n",
    "    for _, y in x.iterrows():\n",
    "        if (not (y[id_col] in helpers.empty_strings)) and (y[localisation_col] in helpers.empty_strings):\n",
    "            print(y['origin'], y['ConnID'], prefix, y[id_col], y[localisation_col])\n",
    "    \n",
    "    \n",
    "    node_localisations.update(x[localisation_col])\n",
    "    \n",
    "    #print(prefix)\n",
    "    df_edges[localisation_col] = df_edges[localisation_col].apply(node_localisation_std)\n",
    "    \n",
    "    new_localisation.update(df_edges[localisation_col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ER',\n",
       " 'Golgi',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'extracellular',\n",
       " 'mitochondria?',\n",
       " nan,\n",
       " 'nuc',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_localisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'endoplasmic reticulum',\n",
       " 'extracellular',\n",
       " 'golgi apparatus',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'putative:mitochondrion',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_localisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_edges['ConnID'].unique()) == df_edges.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['RxID'] = df_edges.index.map(lambda x: \"RxID-\" + str(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_edges['RxID'].unique()) == df_edges.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.to_csv(os.path.join(\"..\", \"data\", \"raw\", \"edges-sheet.tsv\"), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "homologue_cols = [f\"{x}_homologues\" for x in all_species]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ath', 'stu', 'osa']"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(\"..\", \"data\", \"raw\", \"complexes_to_add.tsv\"), \"w\") as out:\n",
    "    for c in complexes_to_add:\n",
    "        out.write(f\"{c}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
