{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import neo4j DB: 1/4\n",
    "\n",
    "Code to translate v2.7.4_PIS-model.xlsx to neo4j database. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import os\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to graph via docker-compose link. See http://localhost:7474/browser/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_labels = [\n",
    "    'PlantCoding',\n",
    "    'PlantNonCoding',\n",
    "    'PlantAbstract',\n",
    "    'Complex',\n",
    "    'ExternalOrganism', \n",
    "    'ExternalCoding',\n",
    "    'ExternalNonCoding',\n",
    "    'Process', \n",
    "    'MetaboliteFamily',\n",
    "    'Metabolite',\n",
    "    'GeneExpression', \n",
    "    'PseudoNode'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in sheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = os.path.join(\"..\", \"data\", \"raw\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Components sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sheets = [#(file, sheet_name)]\n",
    "    (\"v2.7.5_PIS-model.xlsx\", \"Components\"),\n",
    "    (\"v2.7.5_PIS-model.xlsx\", \"Components_New\"), \n",
    "    (\"Model_CK.xlsx\", \"Components_new\"), \n",
    "    (\"v2.7.2_PIS-model-JALR.xlsx\", \"Components_New\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.5_PIS-model.xlsx Components\n",
      "v2.7.5_PIS-model.xlsx Components_New\n",
      "Model_CK.xlsx Components_new\n",
      "v2.7.2_PIS-model-JALR.xlsx Components_New\n"
     ]
    }
   ],
   "source": [
    "# resave xlsx as tsv\n",
    "drops = ['Legacy:Process', 'mID', 'Notes', 'Unnamed: 21']\n",
    "col_rename = {\n",
    "    'mID':'mID', \n",
    "    'Notes':'Notes', \n",
    "    'AddedBy':'AddedBy', \n",
    "    'Species':'Species', \n",
    "    'NodeType':'NodeType', \n",
    "    'Family':'Family', \n",
    "    'Clade':'Clade', \n",
    "    'NodeID':'NodeID', \n",
    "    'NodeName':'NodeName', \n",
    "    'ModelStatus':'ModelStatus', \n",
    "    'NodeDescription':'NodeDescription', \n",
    "    'AdditionalInfo':'AdditionalInfo', \n",
    "    'ExtDBlink':'ExtDBlink', \n",
    "    'Process':'Process', \n",
    "    'ExternalDB':'ExternalDB', \n",
    "    'ModelV':'ModelV', \n",
    "    'GMM_OCD1':'GMM_OCD1', \n",
    "    'GMM_OCD':'GMM_OCD', \n",
    "    'GMM:Description':'GMM_Description', \n",
    "    'GMM:ShortName':'GMM_ShortName', \n",
    "    'GMM:Synonyms':'GMM_Synonyms', \n",
    "\n",
    "    'Node':'NodeName'\n",
    "}\n",
    "\n",
    "for file_name, sheet_name in sheets:\n",
    "    \n",
    "    print(file_name, sheet_name)\n",
    "    \n",
    "    file_path = os.path.join(input_path, file_name)\n",
    "    \n",
    "    base_name, extension = os.path.splitext(file_name)\n",
    "    new_file_path = os.path.join(input_path, f'{base_name}-{sheet_name}.tsv')\n",
    "    \n",
    "    if os.path.exists(new_file_path):\n",
    "        continue\n",
    "        \n",
    "    df = pd.read_excel(file_path, \n",
    "                    sheet_name=sheet_name, \n",
    "                    header=[1], \n",
    "                    dtype=str, \n",
    "                    na_values=helpers.empty_strings)\n",
    "    df = df[~df[\"AddedBy\"].isna()]\n",
    "    \n",
    "    to_drop = list(set(drops) & set(df.columns)) + list(df.filter(regex=(\"Unnamed.*\")).columns)\n",
    "    df.drop(to_drop, axis=1, inplace=True)\n",
    "    \n",
    "    new_cols = [col_rename[x] for x in df.columns]\n",
    "    df.columns = new_cols\n",
    "    \n",
    "    df.to_csv(new_file_path, sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v2.7.5_PIS-model.xlsx Components\n",
      "v2.7.5_PIS-model.xlsx Components_New\n",
      "Model_CK.xlsx Components_new\n",
      "v2.7.2_PIS-model-JALR.xlsx Components_New\n"
     ]
    }
   ],
   "source": [
    "dfs = []\n",
    "\n",
    "for file_name, sheet_name in sheets:\n",
    "    \n",
    "    print(file_name, sheet_name)\n",
    "    base_name, extension = os.path.splitext(file_name)\n",
    "\n",
    "    file_path = os.path.join(input_path, f'{base_name}-{sheet_name}.tsv')\n",
    "    \n",
    "    df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "    \n",
    "    dfs.append(df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components = pd.concat(dfs, sort=False)\n",
    "df_components.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "plant_coding       756\n",
       "metabolite         104\n",
       "plant coding        69\n",
       "pathogen_coding     14\n",
       "plant_abstract      12\n",
       "plant_ncRNA          8\n",
       "process              6\n",
       "plant_noncoding      4\n",
       "x                    2\n",
       "plant_complex        1\n",
       "Name: NodeType, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_components['NodeType'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>ExtDBlink</th>\n",
       "      <th>Process</th>\n",
       "      <th>ExternalDB</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy Species NodeType Family Clade NodeID NodeName ModelStatus  \\\n",
       "854       x       x        x      x     x      x        x           x   \n",
       "976       x       x        x      x     x      x        x           x   \n",
       "\n",
       "    NodeDescription AdditionalInfo ExtDBlink Process ExternalDB ModelV  \\\n",
       "854               x              x         x     NaN          x      x   \n",
       "976               x              x         x     NaN          x      x   \n",
       "\n",
       "    GMM_OCD1 GMM_OCD GMM_Description GMM_ShortName GMM_Synonyms  \n",
       "854        x       x               x             x            x  \n",
       "976        x       x               x             x            x  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_components[df_components['NodeType']=='x']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([854, 976], dtype='int64')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = df_components[df_components['NodeType']=='x'].index; x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components.drop(x, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "830    NaN\n",
       "831    NaN\n",
       "Name: NodeName, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_components.loc[df_components[\"NodeName\"].isna(), 'NodeName']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components.loc[df_components[\"NodeName\"].isna(), 'NodeName'] =  df_components.loc[df_components[\"NodeName\"].isna(), 'NodeID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "components_node_type_to_node_label = {\n",
    "    \"plant coding\":\"PlantCoding\",\n",
    "    \"plant_coding\":\"PlantCoding\",\n",
    "    \"plant_noncoding\":\"PlantNonCoding\",\n",
    "    \"plant_ncRNA\":\"PlantNonCoding\",\n",
    "\n",
    "    \"plant_complex\":\"Complex\", \n",
    "\n",
    "    \"metabolite\":\"Metabolite\",\n",
    "\n",
    "    \"pathogen_coding\":\"ExternalCoding\",\n",
    "    \"pathogen_noncoding\":\"ExternalNonCoding\",\n",
    "    \n",
    "    \"plant_abstract\":\"PlantAbstract\",\n",
    "    \n",
    "    \"process\":\"Process\", \n",
    "\n",
    "    np.nan:\"Undefined\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PlantCoding       825\n",
       "Metabolite        104\n",
       "ExternalCoding     14\n",
       "PlantNonCoding     12\n",
       "PlantAbstract      12\n",
       "Process             6\n",
       "Undefined           1\n",
       "Complex             1\n",
       "Name: NodeLabel, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update node labels\n",
    "df_components[\"NodeLabel\"] = df_components[\"NodeType\"].apply(lambda x: components_node_type_to_node_label[x])\n",
    "pd.value_counts(df_components['NodeLabel'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>ExtDBlink</th>\n",
       "      <th>Process</th>\n",
       "      <th>ExternalDB</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>JALR</td>\n",
       "      <td>osa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Class I Clp ATPase</td>\n",
       "      <td>Class I Clp ATPase</td>\n",
       "      <td>Os11g0104300</td>\n",
       "      <td>D53</td>\n",
       "      <td>use</td>\n",
       "      <td>D53</td>\n",
       "      <td>D53 binds to the complex S-D14-MAX2 to be degr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GMM-OCDs</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Undefined</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy Species NodeType              Family               Clade  \\\n",
       "975    JALR     osa      NaN  Class I Clp ATPase  Class I Clp ATPase   \n",
       "\n",
       "            NodeID NodeName ModelStatus NodeDescription  \\\n",
       "975  Os11g0104300       D53         use             D53   \n",
       "\n",
       "                                        AdditionalInfo ExtDBlink Process  \\\n",
       "975  D53 binds to the complex S-D14-MAX2 to be degr...       NaN     NaN   \n",
       "\n",
       "    ExternalDB ModelV GMM_OCD1 GMM_OCD GMM_Description GMM_ShortName  \\\n",
       "975   GMM-OCDs    vNA      NaN     NaN             NaN           NaN   \n",
       "\n",
       "    GMM_Synonyms  NodeLabel  \n",
       "975          NaN  Undefined  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_components[df_components[\"NodeLabel\"] == \"Undefined\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def only_asci(x):\n",
    "    return \"\".join([character for character in x if character.isascii()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components['Species'] = df_components[\"Species\"].apply(lambda x: x.lower().strip())\n",
    "df_components['observed_species'] = df_components[\"Species\"].apply(helpers.get_second_item)\n",
    "df_components['also_observed_in'] = df_components[\"Species\"].apply(helpers.rest_of_items)\n",
    "\n",
    "\n",
    "df_components[\"GMM_Synonyms\"] = df_components[\"GMM_Synonyms\"].apply(helpers.string_to_nice_string)\n",
    "df_components['chebi_identifier'] = df_components[df_components['ExternalDB'] == \"ChEBI\"][\"ExtDBlink\"]\n",
    "df_components['pubmed_identifier'] = df_components[df_components['ExternalDB'] == \"PubChem\"][\"ExtDBlink\"]\n",
    "df_components['AdditionalInfo'].fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components.loc[df_components['ModelV'].isna(), 'ModelV'] = 'vNA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components.loc[df_components['AddedBy'].isna(), 'AddedBy'] = 'NAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components['AddedBy'] = df_components['AddedBy'].apply(lambda x: x.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_components.loc[df_components['AddedBy']=='ZR/MZ', 'AddedBy'] = 'MZ' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['KG', 'ZR', 'MZ', 'MAK', 'ŠT', 'AG', 'JALR'], dtype=object)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_components['AddedBy'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def only_asci(x):\n",
    "    return \"\".join([character for character in x if character.isascii()])\n",
    "\n",
    "def get_non_ascii(x):\n",
    "    x = str(x)\n",
    "    has_nonascii = False\n",
    "    for character in x:\n",
    "        if not character.isascii():\n",
    "            has_nonascii = True\n",
    "            print(character, ord(character), character.encode())\n",
    "    if has_nonascii:\n",
    "        print(x)\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "ascii_replacers = {\n",
    "    b'\\xc2\\xa0' : b\" \",           # funky WIN whitespace\n",
    "    b'\\xe2\\x80\\xa6': b'...',      # …\n",
    "    b'\\xe2\\x80\\x8b' : b'',        # have no figging clue\n",
    "    b'\\xe2\\x80\\x93' : b'-',       # –\n",
    "    b'\\xce\\xb1' : b\"alpha\",       # α\n",
    "    b'\\xc3\\x9f' : b\"beta\",        # ß\n",
    "    b'\\xce\\xb2' : b\"beta\",        # β\n",
    "    b'\\xe2\\x80\\x98' : b\"prime\",   # ‘\n",
    "    b'\\xe2\\x80\\x99' : b\"prime\",   # ’\n",
    "    b'\\xc2\\xb4': b'prime',        # ´\n",
    "    # Sorry accents :(\n",
    "    b'\\xc5\\xa0' : b\"S\",           # Š\n",
    "    b'\\xc5\\xa1' : b's',           # š\n",
    "    b'\\xc5\\xbd' : b'Z',           # Ž\n",
    "    b'\\xc4\\x8d' : b'c'             # č\n",
    "}\n",
    "\n",
    "\n",
    "def replacer(x):\n",
    "    x = x.encode('utf-8')\n",
    "    for old, new in ascii_replacers.items():\n",
    "        x = x.replace(old, new)\n",
    "    return x.decode('utf-8')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AddedBy \n",
      "-------------\n",
      "Š 352 b'\\xc5\\xa0'\n",
      "ŠT\n",
      "\n",
      "Species \n",
      "-------------\n",
      "\n",
      "NodeType \n",
      "-------------\n",
      "\n",
      "Family \n",
      "-------------\n",
      "\n",
      "Clade \n",
      "-------------\n",
      "\n",
      "NodeID \n",
      "-------------\n",
      "\n",
      "NodeName \n",
      "-------------\n",
      "\n",
      "ModelStatus \n",
      "-------------\n",
      "\n",
      "NodeDescription \n",
      "-------------\n",
      "ß 223 b'\\xc3\\x9f'\n",
      "12-hydroxyjasmonic acid 12-O-ß-D-glucoside\n",
      "  160 b'\\xc2\\xa0'\n",
      "gibberellin 3-oxidase \n",
      "‘ 8216 b'\\xe2\\x80\\x98'\n",
      "N6-(dimethyallyl)adenosine 5‘-monophosphate\n",
      "‘ 8216 b'\\xe2\\x80\\x98'\n",
      "N6-(dimethyallyl)adenosine 5‘-diphosphate\n",
      "‘ 8216 b'\\xe2\\x80\\x98'\n",
      "N6-(dimethyallyl)adenosine 5‘-triphosphate\n",
      "α 945 b'\\xce\\xb1'\n",
      "7-(α-D-glucosyl)-N6-isopentenyladenine \n",
      "α 945 b'\\xce\\xb1'\n",
      "9-(α-D-glucosyl)-N6-isopentenyladenine \n",
      "β 946 b'\\xce\\xb2'\n",
      "O-β-D-glucosyl-trans-zeatin \n",
      "α 945 b'\\xce\\xb1'\n",
      "9-(α-D-glucosyl)-trans-zeatin\n",
      "α 945 b'\\xce\\xb1'\n",
      "7-(α-D-glucosyl)dihydrozeatin \n",
      "α 945 b'\\xce\\xb1'\n",
      "9-(α-D-glucosyl)dihydrozeatin \n",
      "β 946 b'\\xce\\xb2'\n",
      "O-β-D-glucosyl-cis-zeatin \n",
      "β 946 b'\\xce\\xb2'\n",
      "7-(β-D-glucosyl)-cis-zeatin \n",
      "β 946 b'\\xce\\xb2'\n",
      "9-(β-D-glucosyl)-cis-zeatin \n",
      "α 945 b'\\xce\\xb1'\n",
      "UDP-α-D-glucose\n",
      "\n",
      "AdditionalInfo \n",
      "-------------\n",
      "  160 b'\\xc2\\xa0'\n",
      "Chitin is a polymer of N-acetyl-d-glucosamine that is a major component of fungal cell walls and has been recognized as a general elicitor of plant defense responses for many years.\n",
      "  160 b'\\xc2\\xa0'\n",
      "EC 2.2.1.7, ec00900\n",
      "  160 b'\\xc2\\xa0'\n",
      "EC 2.2.1.7, ec00900\n",
      "  160 b'\\xc2\\xa0'\n",
      "EC 2.2.1.7, ec00900\n",
      "’ 8217 b'\\xe2\\x80\\x99'\n",
      "Encodes a trans-acting siRNA (tasi-RNA) that regulates the expression of auxin response factor genes (ARF2, ARF4, ETT). One of 3 genomic loci that encode the TAS3 siRNA. | In contrast to the single mRNA binding family, TAS3 requires the guide mRNA miR390 bind the transcript at two sites. The transcript is then cleaved at the 3’ binding site only, by AGO7. As is the case for the TAS1, TAS2, and TAS3 families, RDR6 then synthesizes the dsRNA fragment which is further processed by DCL4.\n",
      "\n",
      "ExtDBlink \n",
      "-------------\n",
      "\n",
      "Process \n",
      "-------------\n",
      "\n",
      "ExternalDB \n",
      "-------------\n",
      "\n",
      "ModelV \n",
      "-------------\n",
      "\n",
      "GMM_OCD1 \n",
      "-------------\n",
      "\n",
      "GMM_OCD \n",
      "-------------\n",
      "\n",
      "GMM_Description \n",
      "-------------\n",
      "\n",
      "GMM_ShortName \n",
      "-------------\n",
      "\n",
      "GMM_Synonyms \n",
      "-------------\n",
      "\n",
      "NodeLabel \n",
      "-------------\n",
      "\n",
      "observed_species \n",
      "-------------\n",
      "\n",
      "also_observed_in \n",
      "-------------\n",
      "\n",
      "chebi_identifier \n",
      "-------------\n",
      "\n",
      "pubmed_identifier \n",
      "-------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bad_cols = []\n",
    "for c in df_components.columns:\n",
    "    print(c, \"\\n-------------\")\n",
    "    if any(df_components[c].apply(get_non_ascii)):\n",
    "        bad_cols.append(c)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in bad_cols:\n",
    "    df_components[c] = df_components[c].apply(replacer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metabolite\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>chebi_identifier</th>\n",
       "      <th>pubmed_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>MZ</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>PostROS</td>\n",
       "      <td>H2O</td>\n",
       "      <td>H2O</td>\n",
       "      <td>H2O</td>\n",
       "      <td>ignore</td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>H2O</td>\n",
       "      <td>H2O</td>\n",
       "      <td>H2O</td>\n",
       "      <td>H2O</td>\n",
       "      <td>use</td>\n",
       "      <td>water</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>MZ</td>\n",
       "      <td>plant_all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>PreROS</td>\n",
       "      <td>O2</td>\n",
       "      <td>O2</td>\n",
       "      <td>O2</td>\n",
       "      <td>use</td>\n",
       "      <td>ROS precursors; oxygen</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>plant_all</td>\n",
       "      <td></td>\n",
       "      <td>CHEBI:25805</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>AG</td>\n",
       "      <td>all</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>O2</td>\n",
       "      <td>O2</td>\n",
       "      <td>O2</td>\n",
       "      <td>O2</td>\n",
       "      <td>use</td>\n",
       "      <td>dioxygen</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>all</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy    Species    NodeType   Family Clade NodeID NodeName ModelStatus  \\\n",
       "529      MZ  plant_all  metabolite  PostROS   H2O    H2O      H2O      ignore   \n",
       "939      AG        all  metabolite      H2O   H2O    H2O      H2O         use   \n",
       "523      MZ  plant_all  metabolite   PreROS    O2     O2       O2         use   \n",
       "940      AG        all  metabolite       O2    O2     O2       O2         use   \n",
       "\n",
       "            NodeDescription AdditionalInfo  ... GMM_OCD1 GMM_OCD  \\\n",
       "529                   water                 ...      NaN     NaN   \n",
       "939                   water                 ...      NaN     NaN   \n",
       "523  ROS precursors; oxygen                 ...      NaN     NaN   \n",
       "940                dioxygen                 ...      NaN     NaN   \n",
       "\n",
       "    GMM_Description GMM_ShortName GMM_Synonyms   NodeLabel observed_species  \\\n",
       "529             NaN           NaN               Metabolite        plant_all   \n",
       "939             NaN           NaN               Metabolite              all   \n",
       "523             NaN           NaN               Metabolite        plant_all   \n",
       "940             NaN           NaN               Metabolite              all   \n",
       "\n",
       "    also_observed_in chebi_identifier pubmed_identifier  \n",
       "529                               NaN               NaN  \n",
       "939                               NaN               NaN  \n",
       "523                       CHEBI:25805               NaN  \n",
       "940                               NaN               NaN  \n",
       "\n",
       "[4 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PlantCoding\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>Species</th>\n",
       "      <th>NodeType</th>\n",
       "      <th>Family</th>\n",
       "      <th>Clade</th>\n",
       "      <th>NodeID</th>\n",
       "      <th>NodeName</th>\n",
       "      <th>ModelStatus</th>\n",
       "      <th>NodeDescription</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>...</th>\n",
       "      <th>GMM_OCD1</th>\n",
       "      <th>GMM_OCD</th>\n",
       "      <th>GMM_Description</th>\n",
       "      <th>GMM_ShortName</th>\n",
       "      <th>GMM_Synonyms</th>\n",
       "      <th>NodeLabel</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>chebi_identifier</th>\n",
       "      <th>pubmed_identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>MAK</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>CPS</td>\n",
       "      <td>CPS</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td>CPS</td>\n",
       "      <td>use</td>\n",
       "      <td>ent-copalyl diphosphate synthase</td>\n",
       "      <td>also known as GA REQUIRING 1 (GA1)</td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_001702</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>Terpenoid cyclases/Protein prenyltransferases ...</td>\n",
       "      <td>GA1</td>\n",
       "      <td>ABC33,ATCPS1,CPS,CPS1,GA1,TPSGA1</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>826</th>\n",
       "      <td>MAK</td>\n",
       "      <td>stu</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>CPS</td>\n",
       "      <td>CPS</td>\n",
       "      <td>Sotub06g034690.1.1</td>\n",
       "      <td>CPS</td>\n",
       "      <td>use</td>\n",
       "      <td>ent-copalyl diphosphate synthase</td>\n",
       "      <td>also known as GA REQUIRING 1 (GA1)</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>808</th>\n",
       "      <td>MAK</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>AT4G25420</td>\n",
       "      <td>GA20ox1</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin 20-oxidase</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_000842</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>2-oxoglutarate (2OG) and Fe(II)-dependent oxyg...</td>\n",
       "      <td>GA20OX1</td>\n",
       "      <td>20ox1,At2301,GA20OX1,GA5</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828</th>\n",
       "      <td>MAK</td>\n",
       "      <td>stu</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>Sotub03g007160.1.1</td>\n",
       "      <td>GA20ox1</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin 20-oxidase</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>MAK</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>AT5G07200</td>\n",
       "      <td>GA20ox3</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin 20-oxidase</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_023006</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>gibberellin 20-oxidase 3</td>\n",
       "      <td>GA20OX3</td>\n",
       "      <td>20ox3,ATGA20OX3,GA20OX3,YAP169</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>MAK</td>\n",
       "      <td>stu</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>Sotub11g029030.1.1</td>\n",
       "      <td>GA20ox3</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin 20-oxidase</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>MAK</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>AT1G60980</td>\n",
       "      <td>GA20ox4</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin 20-oxidase</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_023006</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>gibberellin 20-oxidase 4</td>\n",
       "      <td>GA20OX4</td>\n",
       "      <td>ATGA20OX4,GA20OX4</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>829</th>\n",
       "      <td>MAK</td>\n",
       "      <td>stu</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>GA20ox</td>\n",
       "      <td>Sotub01g031210.1.1</td>\n",
       "      <td>GA20ox4</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin 20-oxidase</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840</th>\n",
       "      <td>MAK</td>\n",
       "      <td>ath</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>MYB</td>\n",
       "      <td>GAMYB</td>\n",
       "      <td>AT5G06100</td>\n",
       "      <td>MYB33</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin induced TF</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>OCD_all_003344</td>\n",
       "      <td>http://www.gomapman.org/ortholog/OCD_all/OCD_a...</td>\n",
       "      <td>myb domain protein 33</td>\n",
       "      <td>MYB33</td>\n",
       "      <td>ATMYB33,MYB33</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>MAK</td>\n",
       "      <td>stu</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>MYB</td>\n",
       "      <td>GAMYB</td>\n",
       "      <td>Sotub06g030530.1.1</td>\n",
       "      <td>MYB33</td>\n",
       "      <td>use</td>\n",
       "      <td>gibberellin induced TF</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>stu</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy Species      NodeType  Family   Clade              NodeID  \\\n",
       "803     MAK     ath  plant_coding     CPS     CPS           AT4G02780   \n",
       "826     MAK     stu  plant_coding     CPS     CPS  Sotub06g034690.1.1   \n",
       "808     MAK     ath  plant_coding  GA20ox  GA20ox           AT4G25420   \n",
       "828     MAK     stu  plant_coding  GA20ox  GA20ox  Sotub03g007160.1.1   \n",
       "810     MAK     ath  plant_coding  GA20ox  GA20ox           AT5G07200   \n",
       "827     MAK     stu  plant_coding  GA20ox  GA20ox  Sotub11g029030.1.1   \n",
       "811     MAK     ath  plant_coding  GA20ox  GA20ox           AT1G60980   \n",
       "829     MAK     stu  plant_coding  GA20ox  GA20ox  Sotub01g031210.1.1   \n",
       "840     MAK     ath  plant_coding     MYB   GAMYB           AT5G06100   \n",
       "846     MAK     stu  plant_coding     MYB   GAMYB  Sotub06g030530.1.1   \n",
       "\n",
       "    NodeName ModelStatus                   NodeDescription  \\\n",
       "803      CPS         use  ent-copalyl diphosphate synthase   \n",
       "826      CPS         use  ent-copalyl diphosphate synthase   \n",
       "808  GA20ox1         use            gibberellin 20-oxidase   \n",
       "828  GA20ox1         use            gibberellin 20-oxidase   \n",
       "810  GA20ox3         use            gibberellin 20-oxidase   \n",
       "827  GA20ox3         use            gibberellin 20-oxidase   \n",
       "811  GA20ox4         use            gibberellin 20-oxidase   \n",
       "829  GA20ox4         use            gibberellin 20-oxidase   \n",
       "840    MYB33         use            gibberellin induced TF   \n",
       "846    MYB33         use            gibberellin induced TF   \n",
       "\n",
       "                         AdditionalInfo  ...        GMM_OCD1  \\\n",
       "803  also known as GA REQUIRING 1 (GA1)  ...  OCD_all_001702   \n",
       "826  also known as GA REQUIRING 1 (GA1)  ...             NaN   \n",
       "808                                      ...  OCD_all_000842   \n",
       "828                                      ...             NaN   \n",
       "810                                      ...  OCD_all_023006   \n",
       "827                                      ...             NaN   \n",
       "811                                      ...  OCD_all_023006   \n",
       "829                                      ...             NaN   \n",
       "840                                      ...  OCD_all_003344   \n",
       "846                                      ...             NaN   \n",
       "\n",
       "                                               GMM_OCD  \\\n",
       "803  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "826                                                NaN   \n",
       "808  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "828                                                NaN   \n",
       "810  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "827                                                NaN   \n",
       "811  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "829                                                NaN   \n",
       "840  http://www.gomapman.org/ortholog/OCD_all/OCD_a...   \n",
       "846                                                NaN   \n",
       "\n",
       "                                       GMM_Description GMM_ShortName  \\\n",
       "803  Terpenoid cyclases/Protein prenyltransferases ...           GA1   \n",
       "826                                                NaN           NaN   \n",
       "808  2-oxoglutarate (2OG) and Fe(II)-dependent oxyg...       GA20OX1   \n",
       "828                                                NaN           NaN   \n",
       "810                           gibberellin 20-oxidase 3       GA20OX3   \n",
       "827                                                NaN           NaN   \n",
       "811                           gibberellin 20-oxidase 4       GA20OX4   \n",
       "829                                                NaN           NaN   \n",
       "840                              myb domain protein 33         MYB33   \n",
       "846                                                NaN           NaN   \n",
       "\n",
       "                         GMM_Synonyms    NodeLabel observed_species  \\\n",
       "803  ABC33,ATCPS1,CPS,CPS1,GA1,TPSGA1  PlantCoding              ath   \n",
       "826                                    PlantCoding              stu   \n",
       "808          20ox1,At2301,GA20OX1,GA5  PlantCoding              ath   \n",
       "828                                    PlantCoding              stu   \n",
       "810    20ox3,ATGA20OX3,GA20OX3,YAP169  PlantCoding              ath   \n",
       "827                                    PlantCoding              stu   \n",
       "811                 ATGA20OX4,GA20OX4  PlantCoding              ath   \n",
       "829                                    PlantCoding              stu   \n",
       "840                     ATMYB33,MYB33  PlantCoding              ath   \n",
       "846                                    PlantCoding              stu   \n",
       "\n",
       "    also_observed_in chebi_identifier pubmed_identifier  \n",
       "803                               NaN               NaN  \n",
       "826                               NaN               NaN  \n",
       "808                               NaN               NaN  \n",
       "828                               NaN               NaN  \n",
       "810                               NaN               NaN  \n",
       "827                               NaN               NaN  \n",
       "811                               NaN               NaN  \n",
       "829                               NaN               NaN  \n",
       "840                               NaN               NaN  \n",
       "846                               NaN               NaN  \n",
       "\n",
       "[10 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for label, subdf in df_components.groupby('NodeLabel'):\n",
    "    dups =  subdf[subdf.duplicated(['NodeName'], keep=False)]\n",
    "    if dups.shape[0] > 0:\n",
    "        print(label)\n",
    "        display(dups.sort_values('NodeName'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove duplicated metabolites\n",
    "df_components.drop([939, 940], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../data/raw/parsed-components.tsv\n"
     ]
    }
   ],
   "source": [
    "path = os.path.join(input_path, 'parsed-components.tsv')\n",
    "df_components.to_csv(path, sep=\"\\t\", index=None)\n",
    "print(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AddedBy\tSpecies\tNodeType\tFamily\tClade\tNodeID\tNodeName\tModelStatus\tNodeDescription\tAdditionalInfo\tExtDBlink\tProcess\tExternalDB\tModelV\tGMM_OCD1\tGMM_OCD\tGMM_Description\tGMM_ShortName\tGMM_Synonyms\tNodeLabel\tobserved_species\talso_observed_in\tchebi_identifier\tpubmed_identifier\r\n",
      "KG\tath\tplant_coding\tSAM\tSAMS\tAT1G02500\tSAM1\tuse\tSAM synthetase [EC:2.5.1.6]\t\tOCD_all_000621\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000621\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000621\tS-adenosylmethionine synthetase 1\tSAM1\tAtSAM1,MAT1,METK1,SAM-1,SAM1\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tSAM\tSAMS\tAT4G01850\tSAM2\tuse\tSAM synthetase [EC:2.5.1.6]\t\tOCD_all_000621\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000621\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000621\tS-adenosylmethionine synthetase 2\tSAM2\tAtSAM2,MAT2,SAM-2,SAM2\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tSAM\tSAMS\tAT2G36880\tSAM3\tuse\tSAM synthetase [EC:2.5.1.6]\t\tOCD_all_000621\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000621\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000621\tmethionine adenosyltransferase 3\tMETK3\tMAT3,METK3\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tSAM\tSAMS\tAT3G17390\tSAM4\tuse\tSAM synthetase [EC:2.5.1.6]\t\tOCD_all_000621\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000621\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000621\tS-adenosylmethionine synthetase family protein\tMETK4\tMAT4,METK4,MTO3,SAMS3\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tACS\tACS1,2,6\tAT3G61510\tACS1\tuse\tACC synthase [EC:4.4.1.14]\t\tOCD_all_000133\tHormone:ET\tGMM-OCDs\tv2.5\tOCD_all_000133\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000133\tACC synthase 1\tACS1\tACC2,ACS1,AT-ACS1\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tACS\tACS1,2,6\tAT1G01480\tACS2\tuse\tACC synthase [EC:4.4.1.14]\t\tOCD_all_000133\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000133\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000133\t1-amino-cyclopropane-1-carboxylate synthase 2\tACS2\tACC1,ACS2,AT-ACC2\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tACS\tACS4,5,8,9,11\tAT2G22810\tACS4\tuse\tACC synthase [EC:4.4.1.14]\t\tOCD_all_000133\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000133\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000133\t1-aminocyclopropane-1-carboxylate synthase 4\tACS4\tACC4,ACS4,ATACS4\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tACS\tACS4,5,8,9,11\tAT5G65800\tACS5\tuse\tACC synthase [EC:4.4.1.14]\t\tOCD_all_000133\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000133\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000133\tACC synthase 5\tACS5\tACC5,ACS5,ATACS5,CIN5,ETO2\tPlantCoding\tath\t\t\t\r\n",
      "KG\tath\tplant_coding\tACS\tACS1,2,6\tAT4G11280\tACS6\tuse\tACC synthase [EC:4.4.1.14]\t\tOCD_all_000133\tHormone:ET\tGMM-OCDs\tv1.0\tOCD_all_000133\thttp://www.gomapman.org/ortholog/OCD_all/OCD_all_000133\t1-aminocyclopropane-1-carboxylic acid (acc) synthase 6\tACS6\tACC6,ACS6,ATACS6\tPlantCoding\tath\t\t\t\r\n"
     ]
    }
   ],
   "source": [
    "!head $path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Components summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = '''MATCH (n) RETURN DISTINCT n.name AS name, n.level AS level'''\n",
    "nodes = graph.run(q).data()\n",
    "all_nodes_in_components = set([(d[\"name\"], d[\"level\"]) for d in nodes])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "245"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_nodes_in_components)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Complex',\n",
       " 'ExternalCoding',\n",
       " 'Metabolite',\n",
       " 'PlantAbstract',\n",
       " 'PlantCoding',\n",
       " 'PlantNonCoding',\n",
       " 'Process']"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(df_components['NodeLabel'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PlantCoding 142\n",
      "PlantNonCoding 9\n",
      "PlantAbstract 7\n",
      "Complex 1\n",
      "ExternalOrganism 3\n",
      "ExternalCoding 14\n",
      "ExternalNonCoding 0\n",
      "Process 6\n",
      "MetaboliteFamily 4\n",
      "Metabolite 59\n"
     ]
    }
   ],
   "source": [
    "node_dict = {}\n",
    "for label in node_labels:\n",
    "    q = '''MATCH (n:%s) RETURN DISTINCT n.name'''%label\n",
    "    s = set([d['n.name'] for d in graph.run(q).data()])\n",
    "    print(label, len(s))\n",
    "    node_dict[label] = s\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reactions sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges_orig = pd.read_excel(input_path, \n",
    "                         sheet_name=\"Reactions\", \n",
    "                         header=[1], \n",
    "                         dtype=str, \n",
    "                         na_values=helpers.empty_strings)\n",
    "df_edges_orig = df_edges_orig[df_edges_orig['Status'].isin([\"forCB\", \"forCB_INVENTED\"])]\n",
    "df_edges_orig.drop(['Status', 'FOXMES', 'Legacy:Process', 'Legacy:ReactionMode'], axis=1, inplace=True)\n",
    "df_edges_orig.columns = pd.Index(['AddedBy', 'ConnID', 'Species', \n",
    "          'input1_ID', 'input1_level', 'input1_localisation', 'input1_type', \n",
    "          'input2_ID', 'input2_level', 'input2_localisation', 'input2_type', \n",
    "          'input3_ID', 'input3_level', 'input3_localisation', 'input3_type', \n",
    "          'ReactionEffect', 'ReactionMode', 'Modifications',\n",
    "          'output1_ID', 'output1_level', 'output1_localisation', 'output1_type', \n",
    "          'TrustLevel', 'Literature', 'AdditionalInfo', 'Comment', 'ModelV', 'kinetics'],\n",
    "      dtype='object')\n",
    "\n",
    "df_edges_new = pd.read_excel(input_path, \n",
    "                         sheet_name=\"Reactions_New\", \n",
    "                         header=[1], \n",
    "                         dtype=str, \n",
    "                         na_values=helpers.empty_strings)\n",
    "df_edges_new = df_edges_new[~df_edges_new['AddedBy'].isin(['-'])]\n",
    "df_edges_new.drop(['Status'], axis=1, inplace=True)\n",
    "df_edges_new.columns = pd.Index(['AddedBy', 'ConnID', 'Species', \n",
    "          'input1_ID', 'input1_level', 'input1_type', \n",
    "          'input2_ID', 'input2_level', 'input2_type', \n",
    "          'input3_ID', 'input3_level', 'input3_type',                                  \n",
    "          'ReactionEffect', 'ReactionMode', \n",
    "          'output1_ID', 'output1_level', 'output1_type', \n",
    "          'TrustLevel', 'Literature', 'AdditionalInfo', 'Comment', 'ModelV'],\n",
    "      dtype='object')\n",
    "\n",
    "\n",
    "df_edges = pd.concat([df_edges_orig, df_edges_new], sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges = df_edges[~df_edges[\"AddedBy\"].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>output1_ID</th>\n",
       "      <th>output1_level</th>\n",
       "      <th>output1_localisation</th>\n",
       "      <th>output1_type</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>LSD1</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.113.225805</td>\n",
       "      <td>LSD1 interacted with all three catalases both ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>SA</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R4] indirect reaction</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>SA decreased CAT2 activity in a dose-dependent...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>sb</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX3</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>x</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "46        sb  Conn310     ath       CAT       family                 NaN   \n",
       "47        sb  Conn310     ath        SA         node                 NaN   \n",
       "48        sb  Conn310     ath      CAT2         node                 NaN   \n",
       "49        sb  Conn310     ath      CAT2         node                 NaN   \n",
       "1737       x        x       x         x            x                 NaN   \n",
       "\n",
       "       input1_type input2_ID input2_level input2_localisation  ... output1_ID  \\\n",
       "46    plant_coding      LSD1         node                 NaN  ...        NaN   \n",
       "47      metabolite      CAT2         node                 NaN  ...        NaN   \n",
       "48    plant_coding      ACX2         node                 NaN  ...        NaN   \n",
       "49    plant_coding      ACX3         node                 NaN  ...        NaN   \n",
       "1737             x       NaN            x                 NaN  ...          x   \n",
       "\n",
       "     output1_level output1_localisation output1_type  \\\n",
       "46             NaN                  NaN          NaN   \n",
       "47             NaN                  NaN          NaN   \n",
       "48             NaN                  NaN          NaN   \n",
       "49             NaN                  NaN          NaN   \n",
       "1737             x                  NaN            x   \n",
       "\n",
       "                                       TrustLevel  \\\n",
       "46    [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "47                         [R4] indirect reaction   \n",
       "48    [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "49    [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1737                                            x   \n",
       "\n",
       "                          Literature  \\\n",
       "46         DOI:10.1104/pp.113.225805   \n",
       "47    DOI:10.1016/j.chom.2017.01.007   \n",
       "48    DOI:10.1016/j.chom.2017.01.007   \n",
       "49    DOI:10.1016/j.chom.2017.01.007   \n",
       "1737                               x   \n",
       "\n",
       "                                         AdditionalInfo Comment ModelV  \\\n",
       "46    LSD1 interacted with all three catalases both ...     NaN    vNA   \n",
       "47    SA decreased CAT2 activity in a dose-dependent...     NaN    vNA   \n",
       "48    CAT2 Promotes the Activityof ACX2/ACX3 ( (test...     NaN    vNA   \n",
       "49    CAT2 Promotes the Activityof ACX2/ACX3 ( (test...     NaN    vNA   \n",
       "1737                                                  x       x      x   \n",
       "\n",
       "     kinetics  \n",
       "46        NaN  \n",
       "47        NaN  \n",
       "48        NaN  \n",
       "49        NaN  \n",
       "1737      NaN  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Int64Index([1737], dtype='int64')\n"
     ]
    }
   ],
   "source": [
    "x = df_edges[df_edges['AddedBy']=='x'].index\n",
    "print(x)\n",
    "df_edges.drop(x, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[R1] targetted experiments (e.g. Y2H, BIFC)',\n",
       "       '[Ry] invented reaction', '[Rx] incomplete/unspecific reaction',\n",
       "       '[R2] high-throughput experiment (e.g. ChIP-seq)',\n",
       "       '[R3] in-silico prediction', '[R4] indirect reaction'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[\"TrustLevel\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges['trust_level']  = df_edges[\"TrustLevel\"].apply(lambda x: re.search( r\"(R[1|2|3|4|x|y]|undefined)\", x).groups()[0])\n",
    "df_edges['observed_species'] = df_edges[\"Species\"].apply(helpers.get_second_item)\n",
    "df_edges['also_observed_in'] = df_edges[\"Species\"].apply(helpers.rest_of_items)\n",
    "df_edges['Comment'] = df_edges['Comment'].fillna(\"\")\n",
    "df_edges['AdditionalInfo'] = df_edges['AdditionalInfo'].fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['KG', 'MZ', 'ZR', 'MPE', 'ACR', 'MAK', 'ŠT', 'SB'], dtype=object)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges['AddedBy'] = df_edges['AddedBy'].apply(lambda x: x.upper())\n",
    "df_edges[\"AddedBy\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['v1.0', 'v2.5', 'v2.7', 'v2.6', 'vNA'], dtype=object)"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.loc[df_edges['ModelV'].isna(), 'ModelV'] = 'vNA'\n",
    "df_edges['ModelV'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def doi_list(x):\n",
    "    x = only_asci(x.lower())\n",
    "    match = re.findall(\"(?:doi)(?:\\:|\\/)\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"doi:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "def pubmed_list(x):\n",
    "    x = only_asci(x.lower())\n",
    "    match = re.findall(\"(?:pmid)\\:\\s*(.+?)(?:\\s|$)\", x)\n",
    "    if not match is None:\n",
    "        return [\"pmid:\" + m.rstrip('.') for m in match]\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "def list_to_string(x):\n",
    "    return \",\".join([str(i) for i in x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no/bad reference Conn040 \n",
      "no/bad reference Conn118 \n",
      "no/bad reference Conn120 \n",
      "no/bad reference Conn122 \n",
      "no/bad reference Conn182 \n",
      "no/bad reference Conn183 \n",
      "no/bad reference Conn199 ? Kg need to find reference\n",
      "no/bad reference Conn273 KEGG \n",
      "no/bad reference Conn274 KEGG \n",
      "no/bad reference Conn275 KEGG \n",
      "no/bad reference Conn276 KEGG \n",
      "no/bad reference Conn276  10.1073/pnas.98.4.2065\n",
      "no/bad reference Conn277 KEGG \n",
      "no/bad reference Conn278 KEGG \n",
      "no/bad reference Conn279 KEGG \n",
      "no/bad reference Conn280 KEGG \n",
      "no/bad reference Conn281 KEGG \n",
      "no/bad reference Conn282 KEGG \n",
      "no/bad reference Conn283 KEGG \n",
      "no/bad reference Conn309 \n"
     ]
    }
   ],
   "source": [
    "# format literature sources\n",
    "df_edges[\"Literature\"].fillna(value=\"\", inplace=True)\n",
    "for i, row in df_edges.iterrows():\n",
    "    s = row['Literature']\n",
    "    source = doi_list(s)\n",
    "    source += pubmed_list(s)\n",
    "    for z in s.split(\"|\"):\n",
    "        key = z.lower()\n",
    "        if \":\" in key:\n",
    "            if \"aracyc\" in key:\n",
    "                aracyc_string = \"aracyc:\" + z.split(\":\")[1].strip()\n",
    "                source.append(aracyc_string)\n",
    "            elif \"kegg\" in key:\n",
    "                kegg_string = \"kegg:\" + z.split(\":\")[1].strip()\n",
    "                source.append(kegg_string)\n",
    "            elif \"doi\" in key:\n",
    "                # already fetched\n",
    "                continue\n",
    "            elif (\"pubmed\" in key) or (\"pmid\" in key):\n",
    "                pmid_string = \"pubmed:\" + z.split(\":\")[1].strip()\n",
    "        elif \"invented\" in key:\n",
    "            source.append(\"invented\")\n",
    "        else:\n",
    "            print(\"no/bad reference\", row[\"ConnID\"], z)\n",
    "            source.append(\"other:\" + only_asci(z.strip()))\n",
    "    if len(source) > 0:\n",
    "        df_edges.loc[i, \"literature_sources\"] = list_to_string(source)\n",
    "    else:\n",
    "        print(row[\"ConnID\"], z)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges[[\"ConnID\", \"Literature\", \"literature_sources\"]].to_csv(\"lit-check.tsv\", sep=\"\\t\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_df = df_edges.copy()\n",
    "#df_edges = save_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn080</td>\n",
       "      <td>ath</td>\n",
       "      <td>Trichome-initiation</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>process</td>\n",
       "      <td>potyvirus</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[Ry] invented reaction</td>\n",
       "      <td>DOI:10.1105/tpc.111.083261 (Qi, 2011: The jasm...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>v2.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ry</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.111.083261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GID1a,b,c</td>\n",
       "      <td>clade/orthologue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA3</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.106.047415</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.106.047415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GID1a,b,c</td>\n",
       "      <td>clade/orthologue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>GA4</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.106.047415</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.106.047415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA1-GID1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA3-GID1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA4-GID1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>SLY1</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SCF</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.010827</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1105/tpc.010827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>osa</td>\n",
       "      <td>GID2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>SCF</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1126/science.1081077</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>osa</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1126/science.1081077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ath</td>\n",
       "      <td>GA1-GID1-DELLA</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>SCF-SLY1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956 | DOI:10.1105/tpc.01...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956,doi:10.1105/tpc.010827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>MAK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>osa</td>\n",
       "      <td>GA1-GID1-SLR1</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_complex</td>\n",
       "      <td>SCF-GID2</td>\n",
       "      <td>family</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1104/pp.112.200956 | https://doi.org/10...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>osa</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1104/pp.112.200956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>SB</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>SA</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R4] indirect reaction</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>SA decreased CAT2 activity in a dose-dependent...</td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R4</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1016/j.chom.2017.01.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>SB</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1016/j.chom.2017.01.007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>SB</td>\n",
       "      <td>Conn310</td>\n",
       "      <td>ath</td>\n",
       "      <td>CAT2</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plant_coding</td>\n",
       "      <td>ACX3</td>\n",
       "      <td>node</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1016/j.chom.2017.01.007</td>\n",
       "      <td>CAT2 Promotes the Activityof ACX2/ACX3 ( (test...</td>\n",
       "      <td></td>\n",
       "      <td>vNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1016/j.chom.2017.01.007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AddedBy   ConnID Species            input1_ID      input1_level  \\\n",
       "74       KG  Conn080     ath  Trichome-initiation            family   \n",
       "260     MAK      NaN     ath            GID1a,b,c  clade/orthologue   \n",
       "261     MAK      NaN     ath            GID1a,b,c  clade/orthologue   \n",
       "268     MAK      NaN     ath             GA1-GID1            family   \n",
       "269     MAK      NaN     ath             GA3-GID1            family   \n",
       "270     MAK      NaN     ath             GA4-GID1            family   \n",
       "271     MAK      NaN     ath                 SLY1              node   \n",
       "272     MAK      NaN     osa                 GID2              node   \n",
       "273     MAK      NaN     ath       GA1-GID1-DELLA            family   \n",
       "274     MAK      NaN     osa        GA1-GID1-SLR1            family   \n",
       "287      SB  Conn310     ath                   SA              node   \n",
       "288      SB  Conn310     ath                 CAT2              node   \n",
       "289      SB  Conn310     ath                 CAT2              node   \n",
       "\n",
       "    input1_localisation    input1_type  input2_ID input2_level  \\\n",
       "74                  NaN        process  potyvirus       family   \n",
       "260                 NaN   plant_coding        GA3       family   \n",
       "261                 NaN   plant_coding        GA4       family   \n",
       "268                 NaN  plant_complex      DELLA       family   \n",
       "269                 NaN  plant_complex      DELLA       family   \n",
       "270                 NaN  plant_complex      DELLA       family   \n",
       "271                 NaN   plant_coding        SCF       family   \n",
       "272                 NaN   plant_coding        SCF       family   \n",
       "273                 NaN  plant_complex   SCF-SLY1       family   \n",
       "274                 NaN  plant_complex   SCF-GID2       family   \n",
       "287                 NaN     metabolite       CAT2         node   \n",
       "288                 NaN   plant_coding       ACX2         node   \n",
       "289                 NaN   plant_coding       ACX3         node   \n",
       "\n",
       "    input2_localisation  ...                                   TrustLevel  \\\n",
       "74                  NaN  ...                       [Ry] invented reaction   \n",
       "260                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "261                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "268                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "269                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "270                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "271                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "272                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "273                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "274                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "287                 NaN  ...                       [R4] indirect reaction   \n",
       "288                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "289                 NaN  ...  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                            Literature  \\\n",
       "74   DOI:10.1105/tpc.111.083261 (Qi, 2011: The jasm...   \n",
       "260                         DOI:10.1105/tpc.106.047415   \n",
       "261                         DOI:10.1105/tpc.106.047415   \n",
       "268                          DOI:10.1104/pp.112.200956   \n",
       "269                          DOI:10.1104/pp.112.200956   \n",
       "270                          DOI:10.1104/pp.112.200956   \n",
       "271                             DOI:10.1105/tpc.010827   \n",
       "272                        DOI:10.1126/science.1081077   \n",
       "273  DOI:10.1104/pp.112.200956 | DOI:10.1105/tpc.01...   \n",
       "274  DOI:10.1104/pp.112.200956 | https://doi.org/10...   \n",
       "287                     DOI:10.1016/j.chom.2017.01.007   \n",
       "288                     DOI:10.1016/j.chom.2017.01.007   \n",
       "289                     DOI:10.1016/j.chom.2017.01.007   \n",
       "\n",
       "                                        AdditionalInfo Comment ModelV  \\\n",
       "74                                                               v2.7   \n",
       "260                                                               vNA   \n",
       "261                                                               vNA   \n",
       "268                                                               vNA   \n",
       "269                                                               vNA   \n",
       "270                                                               vNA   \n",
       "271                                                               vNA   \n",
       "272                                                               vNA   \n",
       "273                                                               vNA   \n",
       "274                                                               vNA   \n",
       "287  SA decreased CAT2 activity in a dose-dependent...            vNA   \n",
       "288  CAT2 Promotes the Activityof ACX2/ACX3 ( (test...            vNA   \n",
       "289  CAT2 Promotes the Activityof ACX2/ACX3 ( (test...            vNA   \n",
       "\n",
       "    kinetics trust_level observed_species also_observed_in  \\\n",
       "74       NaN          Ry              ath                    \n",
       "260      NaN          R1              ath                    \n",
       "261      NaN          R1              ath                    \n",
       "268      NaN          R1              ath                    \n",
       "269      NaN          R1              ath                    \n",
       "270      NaN          R1              ath                    \n",
       "271      NaN          R1              ath                    \n",
       "272      NaN          R1              osa                    \n",
       "273      NaN          R1              ath                    \n",
       "274      NaN          R1              osa                    \n",
       "287      NaN          R4              ath                    \n",
       "288      NaN          R1              ath                    \n",
       "289      NaN          R1              ath                    \n",
       "\n",
       "                                   literature_sources  \n",
       "74                         doi:10.1105/tpc.111.083261  \n",
       "260                        doi:10.1105/tpc.106.047415  \n",
       "261                        doi:10.1105/tpc.106.047415  \n",
       "268                         doi:10.1104/pp.112.200956  \n",
       "269                         doi:10.1104/pp.112.200956  \n",
       "270                         doi:10.1104/pp.112.200956  \n",
       "271                            doi:10.1105/tpc.010827  \n",
       "272                       doi:10.1126/science.1081077  \n",
       "273  doi:10.1104/pp.112.200956,doi:10.1105/tpc.010827  \n",
       "274                         doi:10.1104/pp.112.200956  \n",
       "287                    doi:10.1016/j.chom.2017.01.007  \n",
       "288                    doi:10.1016/j.chom.2017.01.007  \n",
       "289                    doi:10.1016/j.chom.2017.01.007  \n",
       "\n",
       "[13 rows x 32 columns]"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges[df_edges['ConnID'].duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG,other:10.107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "                                    TrustLevel  \\\n",
       "0  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo Comment ModelV kinetics  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.           v1.0      NaN   \n",
       "1                 ACS catalyse ACC to SAMe reaction.           v1.0      NaN   \n",
       "2                   ACO catalyse ACC to ET reaction.           v1.0      NaN   \n",
       "3  Copper gets transported from the cytoplasm to ...           v1.0      NaN   \n",
       "4  Copper activates the membrane bound ethylene r...           v1.0      NaN   \n",
       "\n",
       "  trust_level observed_species also_observed_in  \\\n",
       "0          R1              ath                    \n",
       "1          R1              ath                    \n",
       "2          R1              ath                    \n",
       "3          R1              ath                    \n",
       "4          R1              ath                    \n",
       "\n",
       "                                  literature_sources  \n",
       "0                  doi:10.1042/bj20120245,other:KEGG  \n",
       "1                  doi:10.1042/bj20120245,other:KEGG  \n",
       "2                  doi:10.1042/bj20120245,other:KEGG  \n",
       "3  doi:10.1042/bj20120245,other:KEGG,other:10.107...  \n",
       "4                  doi:10.1042/bj20120245,other:KEGG  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in ['input1', 'input2', 'input3', 'output1']:\n",
    "    df_edges.loc[:, x + \"_ID\"] = df_edges[x + \"_ID\"].apply(helpers.reorder_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>TrustLevel</th>\n",
       "      <th>Literature</th>\n",
       "      <th>AdditionalInfo</th>\n",
       "      <th>Comment</th>\n",
       "      <th>ModelV</th>\n",
       "      <th>kinetics</th>\n",
       "      <th>trust_level</th>\n",
       "      <th>observed_species</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>SAMS catalyse L-Met to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACS catalyse ACC to SAMe reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>AraCyc:ETHYL-PWY</td>\n",
       "      <td>ACO catalyse ACC to ET reaction.</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper gets transported from the cytoplasm to ...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG,other:10.107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td>[R1] targetted experiments (e.g. Y2H, BIFC)</td>\n",
       "      <td>DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...</td>\n",
       "      <td>Copper activates the membrane bound ethylene r...</td>\n",
       "      <td></td>\n",
       "      <td>v1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R1</td>\n",
       "      <td>ath</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "                                    TrustLevel  \\\n",
       "0  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "1  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "2  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "3  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "4  [R1] targetted experiments (e.g. Y2H, BIFC)   \n",
       "\n",
       "                                          Literature  \\\n",
       "0                                   AraCyc:ETHYL-PWY   \n",
       "1                                   AraCyc:ETHYL-PWY   \n",
       "2                                   AraCyc:ETHYL-PWY   \n",
       "3  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "4  DOI:10.1105/tpc.001768 (Ethylene Biosynthesis ...   \n",
       "\n",
       "                                      AdditionalInfo Comment ModelV kinetics  \\\n",
       "0              SAMS catalyse L-Met to SAMe reaction.           v1.0      NaN   \n",
       "1                 ACS catalyse ACC to SAMe reaction.           v1.0      NaN   \n",
       "2                   ACO catalyse ACC to ET reaction.           v1.0      NaN   \n",
       "3  Copper gets transported from the cytoplasm to ...           v1.0      NaN   \n",
       "4  Copper activates the membrane bound ethylene r...           v1.0      NaN   \n",
       "\n",
       "  trust_level observed_species also_observed_in  \\\n",
       "0          R1              ath                    \n",
       "1          R1              ath                    \n",
       "2          R1              ath                    \n",
       "3          R1              ath                    \n",
       "4          R1              ath                    \n",
       "\n",
       "                                  literature_sources  \n",
       "0                  doi:10.1042/bj20120245,other:KEGG  \n",
       "1                  doi:10.1042/bj20120245,other:KEGG  \n",
       "2                  doi:10.1042/bj20120245,other:KEGG  \n",
       "3  doi:10.1042/bj20120245,other:KEGG,other:10.107...  \n",
       "4                  doi:10.1042/bj20120245,other:KEGG  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_node_to_family(x):\n",
    "    # pathogen proteins are listed as proteins, so cannot use dict\n",
    "    id_, type_, level_ = x.values\n",
    "      \n",
    "    if id_ is np.nan:\n",
    "        return np.nan, np.nan\n",
    "\n",
    "    new_label = None\n",
    "    family_id = None\n",
    "    \n",
    "    ########################\n",
    "    # Simple Cases\n",
    "    ########################\n",
    "    if type_ in ['complex', 'complex [active]', 'complex [activated]', 'complex [inactive]', 'plant_complex']:\n",
    "        if not (id_ in node_dict[\"Complex\"]):\n",
    "            complexes_to_add.append(id_)\n",
    "        new_label = 'Complex'\n",
    "        family_id = id_\n",
    "    \n",
    "    elif type_ in [\"metabolite\"]:\n",
    "        for label in [\"Metabolite\", \"MetaboliteFamily\"]:\n",
    "            if id_ in node_dict[label]:\n",
    "                new_label = label\n",
    "                break\n",
    "        if not new_label:\n",
    "            missing_in_components.update([id_])        \n",
    "            print(id_, type_, level_, \" | (label) not a listed metabolite\")\n",
    "        family_id = id_\n",
    "\n",
    "    elif type_ in ['process']:\n",
    "        if (id_ in node_dict[\"Process\"]):\n",
    "            family_id = id_\n",
    "            new_label = \"Process\"\n",
    "        else:\n",
    "            print(id_, type_, level_, \" | (label) process not a listed process\")\n",
    "        \n",
    "    else:\n",
    "        ########################\n",
    "        # family ID\n",
    "        ########################\n",
    "        check_external = False\n",
    "        if level_ == \"family\":    \n",
    "            family_id = id_\n",
    "        elif level_ in [\"clade\", \"clade/orthologue\"]:\n",
    "            try:\n",
    "                family_id = clade_to_family[id_]\n",
    "                replace_w_family.update([id_])\n",
    "            except KeyError:\n",
    "                check_external = True\n",
    "        elif level_ == \"node\":\n",
    "            try:\n",
    "                family_id = node_to_family[id_]\n",
    "                replace_w_family.update([id_])\n",
    "            except KeyError:\n",
    "                check_external = True\n",
    "        \n",
    "        if check_external:\n",
    "            for label in [\"ExternalOrganism\", \"ExternalCoding\", \"ExternalNonCoding\"]:\n",
    "                if id_ in node_dict[label]:\n",
    "                    new_label = label\n",
    "                    family_id = id_\n",
    "                    break\n",
    "                    \n",
    "        if not family_id:\n",
    "            print(id_, type_, level_, \" | (family id) could not convert to family/external\")\n",
    "                \n",
    "        ########################\n",
    "        # Label\n",
    "        ########################\n",
    "        if (family_id) and (not new_label):\n",
    "            id_labels = [] #looping just in case an id occurs mutiple times\n",
    "            for label in node_labels:\n",
    "                if family_id in node_dict[label]:\n",
    "                    id_labels.append(label)\n",
    "\n",
    "            if len(id_labels) == 1:\n",
    "                new_label = id_labels[0]\n",
    "            elif len(id_labels) > 1:\n",
    "                print(id_, type_, level_, \" | (label) many labels fit\")\n",
    "                missing_in_components.update([id_])       \n",
    "\n",
    "            else:\n",
    "                print(id_, type_, level_, \" | (label) could not find label\")\n",
    "                missing_in_components.update([id_])        \n",
    "\n",
    "    return family_id, new_label\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miR6022 plant_ncRNA node  | (family id) could not convert to family/external\n",
      "LSD1 plant_coding node  | (family id) could not convert to family/external\n"
     ]
    }
   ],
   "source": [
    "complexes_to_add = []\n",
    "missing_in_components = set()\n",
    "replace_w_family = set()\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, level_col, new_id, new_label_col  =\\\n",
    "            [prefix + x for x in ('_ID',  '_type',  '_level',  '_newID', '_label')]    \n",
    "    \n",
    "    df_edges[[new_id, new_label_col]] = df_edges[[id_col, type_col, level_col]].apply(convert_node_to_family, axis=1, result_type='expand')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>also_observed_in</th>\n",
       "      <th>literature_sources</th>\n",
       "      <th>input1_newID</th>\n",
       "      <th>input1_label</th>\n",
       "      <th>input2_newID</th>\n",
       "      <th>input2_label</th>\n",
       "      <th>input3_newID</th>\n",
       "      <th>input3_label</th>\n",
       "      <th>output1_newID</th>\n",
       "      <th>output1_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>SAM</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ET</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG,other:10.107...</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>doi:10.1042/bj20120245,other:KEGG</td>\n",
       "      <td>ETR</td>\n",
       "      <td>PlantCoding</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>Metabolite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETR</td>\n",
       "      <td>PlantCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  also_observed_in                                 literature_sources  \\\n",
       "0                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "1                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "2                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "3                   doi:10.1042/bj20120245,other:KEGG,other:10.107...   \n",
       "4                                   doi:10.1042/bj20120245,other:KEGG   \n",
       "\n",
       "  input1_newID input1_label input2_newID input2_label input3_newID  \\\n",
       "0        L-Met   Metabolite          SAM  PlantCoding          NaN   \n",
       "1         SAMe   Metabolite          ACS  PlantCoding          NaN   \n",
       "2          ACC   Metabolite          ACO  PlantCoding          NaN   \n",
       "3         Cu2+   Metabolite          HMA  PlantCoding          NaN   \n",
       "4          ETR  PlantCoding         Cu2+   Metabolite          NaN   \n",
       "\n",
       "  input3_label output1_newID output1_label  \n",
       "0          NaN          SAMe    Metabolite  \n",
       "1          NaN           ACC    Metabolite  \n",
       "2          NaN            ET    Metabolite  \n",
       "3          NaN          Cu2+    Metabolite  \n",
       "4          NaN           ETR   PlantCoding  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_orthologues(x, prefix=\"\"):\n",
    "    id_, level_, label_ = x.values\n",
    "    ########################\n",
    "    # Specie specific homologues\n",
    "    ########################\n",
    "    \n",
    "    return_D = {f\"{prefix}_{specie}_homologues\":\"\" for specie in all_species}\n",
    "    \n",
    "    if label_ in ['PlantCoding', 'PlantNonCoding',  'PlantAbstract']:\n",
    "\n",
    "        for species in all_species:\n",
    "            species = f\"{species}_homologues\"\n",
    "            k = f\"{prefix}_{species}\"\n",
    "            if level_ == 'node':\n",
    "                return_D[k] = node_ids_key[species][id_]\n",
    "            elif level_ == 'clade':\n",
    "                return_D[k] =  clade_ids_key[species][id_]\n",
    "            elif level_ == 'family':\n",
    "                return_D[k] =  family_ids_key[species][id_]\n",
    "\n",
    "    return_D = {x:helpers.list_to_string(list(return_D[x])) for x in return_D}\n",
    "    return return_D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "new_dfs = []\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    print(prefix)\n",
    "    id_col, level_col, new_label_col  =\\\n",
    "                [prefix + x for x in ('_ID',  '_level',  '_label')]        \n",
    "    \n",
    "    new_df = df_edges[[id_col, level_col, new_label_col ]].apply(get_orthologues, axis=1, result_type='expand', prefix=prefix)\n",
    "    new_dfs.append(new_df)\n",
    "    #df_edges = df_edges.join(new_df, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "homologues_df = pd.concat(new_dfs, sort=False, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "input1_ath_homologues                                            \n",
       "input1_osa_homologues                                            \n",
       "input1_stu_homologues                                            \n",
       "input2_ath_homologues     AT2G36880,AT1G02500,AT4G01850,AT3G17390\n",
       "input2_osa_homologues                                            \n",
       "input2_stu_homologues                                            \n",
       "input3_ath_homologues                                            \n",
       "input3_osa_homologues                                            \n",
       "input3_stu_homologues                                            \n",
       "output1_ath_homologues                                           \n",
       "output1_osa_homologues                                           \n",
       "output1_stu_homologues                                           \n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homologues_df.loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges = df_edges.join(homologues_df, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>input1_stu_homologues</th>\n",
       "      <th>input2_ath_homologues</th>\n",
       "      <th>input2_osa_homologues</th>\n",
       "      <th>input2_stu_homologues</th>\n",
       "      <th>input3_ath_homologues</th>\n",
       "      <th>input3_osa_homologues</th>\n",
       "      <th>input3_stu_homologues</th>\n",
       "      <th>output1_ath_homologues</th>\n",
       "      <th>output1_osa_homologues</th>\n",
       "      <th>output1_stu_homologues</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT2G36880,AT1G02500,AT4G01850,AT3G17390</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT5G65800,AT5G51690,AT4G26200,AT3G49700,AT1G01...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT1G62380,AT1G12010,AT1G77330,AT1G05010,AT2G19590</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>AT5G44790,AT5G21930,AT4G33520,AT1G63440</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  input1_stu_homologues                              input2_ath_homologues  \\\n",
       "0                                  AT2G36880,AT1G02500,AT4G01850,AT3G17390   \n",
       "1                        AT5G65800,AT5G51690,AT4G26200,AT3G49700,AT1G01...   \n",
       "2                        AT1G62380,AT1G12010,AT1G77330,AT1G05010,AT2G19590   \n",
       "3                                  AT5G44790,AT5G21930,AT4G33520,AT1G63440   \n",
       "4                                                                            \n",
       "\n",
       "  input2_osa_homologues input2_stu_homologues input3_ath_homologues  \\\n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "\n",
       "  input3_osa_homologues input3_stu_homologues  \\\n",
       "0                                               \n",
       "1                                               \n",
       "2                                               \n",
       "3                                               \n",
       "4                                               \n",
       "\n",
       "                              output1_ath_homologues output1_osa_homologues  \\\n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                                                             \n",
       "4  AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580                          \n",
       "\n",
       "  output1_stu_homologues  \n",
       "0                         \n",
       "1                         \n",
       "2                         \n",
       "3                         \n",
       "4                         \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "input1\n",
      "input2\n",
      "input3\n",
      "output1\n"
     ]
    }
   ],
   "source": [
    "node_type_to_node_form_dict = {\n",
    "    \"gene\":\"gene\",\n",
    "    \n",
    "    \"protein\":\"protein\",\n",
    "    \"protein [activated]\":\"protein_active\",\n",
    "    'protein [active]': \"protein_active\",\n",
    "    \n",
    "    \"ncRNA\":\"ncRNA\",\n",
    "    \"plant_ncRNA\":\"ncRNA\",\n",
    "    'ta-siRNA':\"ta-siRNA\", \n",
    "        \n",
    "    \"complex\":\"complex\", \n",
    "    \"plant_complex\":\"complex\",\n",
    "    'complex [active]': \"complex_active\",\n",
    "    \n",
    "    \"metabolite\":\"metabolite\",\n",
    "    \n",
    "    \"process\":\"process\", \n",
    "    'process [active]':\"process_active\",\n",
    "\n",
    "    np.nan:\"\", \n",
    "    \"plant_coding\":\"unknown\"\n",
    "}\n",
    "\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, new_form_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_form')]\n",
    "    \n",
    "    x = df_edges[[id_col, type_col]].dropna(how='all')\n",
    "    \n",
    "    print(prefix)\n",
    "    df_edges[new_form_col] = df_edges[type_col].apply(lambda x: node_type_to_node_form_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AddedBy</th>\n",
       "      <th>ConnID</th>\n",
       "      <th>Species</th>\n",
       "      <th>input1_ID</th>\n",
       "      <th>input1_level</th>\n",
       "      <th>input1_localisation</th>\n",
       "      <th>input1_type</th>\n",
       "      <th>input2_ID</th>\n",
       "      <th>input2_level</th>\n",
       "      <th>input2_localisation</th>\n",
       "      <th>...</th>\n",
       "      <th>input3_ath_homologues</th>\n",
       "      <th>input3_osa_homologues</th>\n",
       "      <th>input3_stu_homologues</th>\n",
       "      <th>output1_ath_homologues</th>\n",
       "      <th>output1_osa_homologues</th>\n",
       "      <th>output1_stu_homologues</th>\n",
       "      <th>input1_form</th>\n",
       "      <th>input2_form</th>\n",
       "      <th>input3_form</th>\n",
       "      <th>output1_form</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn001</td>\n",
       "      <td>ath</td>\n",
       "      <td>L-Met</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>SAMS</td>\n",
       "      <td>clade</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn002</td>\n",
       "      <td>ath</td>\n",
       "      <td>SAMe</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACS</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn003</td>\n",
       "      <td>ath</td>\n",
       "      <td>ACC</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>ACO</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein_active</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn004</td>\n",
       "      <td>ath</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>metabolite</td>\n",
       "      <td>HMA</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "      <td>protein</td>\n",
       "      <td></td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KG</td>\n",
       "      <td>Conn005</td>\n",
       "      <td>ath</td>\n",
       "      <td>ETR</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>protein</td>\n",
       "      <td>Cu2+</td>\n",
       "      <td>family</td>\n",
       "      <td>ER</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>protein</td>\n",
       "      <td>metabolite</td>\n",
       "      <td></td>\n",
       "      <td>protein_active</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  AddedBy   ConnID Species input1_ID input1_level input1_localisation  \\\n",
       "0      KG  Conn001     ath     L-Met       family                  ER   \n",
       "1      KG  Conn002     ath      SAMe       family                  ER   \n",
       "2      KG  Conn003     ath       ACC       family                  ER   \n",
       "3      KG  Conn004     ath      Cu2+       family           cytoplasm   \n",
       "4      KG  Conn005     ath       ETR       family                  ER   \n",
       "\n",
       "  input1_type input2_ID input2_level input2_localisation  ...  \\\n",
       "0  metabolite      SAMS        clade                  ER  ...   \n",
       "1  metabolite       ACS       family                  ER  ...   \n",
       "2  metabolite       ACO       family                  ER  ...   \n",
       "3  metabolite       HMA       family                  ER  ...   \n",
       "4     protein      Cu2+       family                  ER  ...   \n",
       "\n",
       "  input3_ath_homologues input3_osa_homologues input3_stu_homologues  \\\n",
       "0                                                                     \n",
       "1                                                                     \n",
       "2                                                                     \n",
       "3                                                                     \n",
       "4                                                                     \n",
       "\n",
       "                              output1_ath_homologues output1_osa_homologues  \\\n",
       "0                                                                             \n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                                                             \n",
       "4  AT2G40940,AT1G66340,AT3G23150,AT1G04310,AT3G04580                          \n",
       "\n",
       "  output1_stu_homologues input1_form     input2_form input3_form  \\\n",
       "0                         metabolite  protein_active               \n",
       "1                         metabolite  protein_active               \n",
       "2                         metabolite  protein_active               \n",
       "3                         metabolite         protein               \n",
       "4                            protein      metabolite               \n",
       "\n",
       "     output1_form  \n",
       "0      metabolite  \n",
       "1      metabolite  \n",
       "2      metabolite  \n",
       "3      metabolite  \n",
       "4  protein_active  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_edges.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_localisation_dict = {\n",
    "    'nuc':'nucleus',\n",
    "    'er':'endoplasmic reticulum',\n",
    "    'golgi':'golgi apparatus'\n",
    "}\n",
    "\n",
    "\n",
    "good_localisations = set([\n",
    "     'nucleus',\n",
    "     'nucleolus',\n",
    "     'cytoplasm',\n",
    "     'vacuole',\n",
    "     'endoplasmic reticulum',\n",
    "     'chloroplast',\n",
    "     'mitochondrion',\n",
    "     'golgi apparatus',\n",
    "     'peroxisome',\n",
    "     'apoplast',\n",
    "     'extracellular'\n",
    "])\n",
    "\n",
    "\n",
    "def node_localisation_std(x):\n",
    "    if not type(x) == str:\n",
    "        return \"\"\n",
    "    \n",
    "    x = x.lower()\n",
    "    if x in node_localisation_dict.keys():\n",
    "        x = node_localisation_dict[x]\n",
    "    \n",
    "    if x in good_localisations:\n",
    "        return x\n",
    "    else:\n",
    "        return \"\"\n",
    "\n",
    "node_localisations = set()\n",
    "new_localisation = set()\n",
    "for prefix in ['input1', 'input2', 'input3', 'output1']:\n",
    "    id_col, type_col, localisation_col  =\\\n",
    "        [prefix + x for x in ('_ID',  '_type',  '_localisation')]\n",
    "    \n",
    "    x = df_edges[['ConnID', id_col, localisation_col]].dropna(how='all', subset=[ id_col, localisation_col])\n",
    "    \n",
    "    #for _, y in x.iterrows():\n",
    "    #    if y[localisation_col] in ['mitochondria?', np.nan]:\n",
    "    #        print(y['ConnID'], \"\\t\", y[id_col], \"\\t\", y[localisation_col])\n",
    "    \n",
    "    \n",
    "    node_localisations.update(x[localisation_col])\n",
    "    \n",
    "    #print(prefix)\n",
    "    df_edges[localisation_col] = df_edges[localisation_col].apply(node_localisation_std)\n",
    "    \n",
    "    new_localisation.update(df_edges[localisation_col])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ER',\n",
       " 'Golgi',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'extracellular',\n",
       " 'mitochondria?',\n",
       " nan,\n",
       " 'nuc',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_localisations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " 'chloroplast',\n",
       " 'cytoplasm',\n",
       " 'endoplasmic reticulum',\n",
       " 'extracellular',\n",
       " 'golgi apparatus',\n",
       " 'nucleus',\n",
       " 'peroxisome',\n",
       " 'vacuole'}"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_localisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_edges.to_csv(os.path.join(\"..\", \"data\", \"raw\", \"edges-sheet.tsv\"), sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "homologue_cols = [f\"{x}_homologues\" for x in all_species]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ath', 'stu', 'osa']"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(\"..\", \"data\", \"raw\", \"complexes_to_add.tsv\"), \"w\") as out:\n",
    "    for c in complexes_to_add:\n",
    "        out.write(f\"{c}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
